# Next Generation Sequencing Data Acquisition

## DNA Stuff

We are primarily going to be interested in representing DNA molecules
in text format as series of nucleotide bases (A, C, G, T).  But, in order
to do this in a consistent fashion that lets us correctly and precisely compare
one sequence of DNA to another we need to know how DNA "works" and we require some
conventions.  

Talk about the fact that it is double-stranded, and hence there is a reverse complement.

Genome references and forward vs reverse strand stuff.


## Sequencing platforms
Basically talk a little about how the technologies work.

Go into paired-end vs single end reads, etc.

Strand direction.  There is a good overview on a blog at:
[http://www.cureffi.org/2012/12/19/forward-and-reverse-reads-in-paired-end-sequencing/](http://www.cureffi.org/2012/12/19/forward-and-reverse-reads-in-paired-end-sequencing/),
to start getting one's head around it.  It ends up being super 
important if you are looking for inversion
breakpoints, etc [@Corbett-DetigSequenceBasedDetectionBreakpoint2012].





## Library Prep Protocols

Gotta mention here about how barcodes work.

How you prepare your libraries dictates what type of data you get.

### WGS

### RAD-Seq methods

### Amplicon Sequencing

### Capture arrays, RAPTURE, etc.