<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 27 An example full workflow from sequences to variants | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics</title>
  <meta name="description" content="A book example for a Chapman &amp; Hall book." />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 27 An example full workflow from sequences to variants | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A book example for a Chapman &amp; Hall book." />
  <meta name="github-repo" content="yihui/bookdown-crc" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 27 An example full workflow from sequences to variants | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics" />
  
  <meta name="twitter:description" content="A book example for a Chapman &amp; Hall book." />
  

<meta name="author" content="Eric C. Anderson" />


<meta name="date" content="2022-03-08" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="whole-genome-alignment-strategies.html"/>
<link rel="next" href="bioinformatic-analysis-on-variant-data.html"/>
<script src="libs/header-attrs/header-attrs.js"></script>
<script src="libs/jquery/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook/css/style.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ECA's Bioinformatics Handbook</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="erics-notes-of-what-he-might-do.html"><a href="erics-notes-of-what-he-might-do.html"><i class="fa fa-check"></i><b>2</b> Eric’s Notes of what he might do</a>
<ul>
<li class="chapter" data-level="2.1" data-path="erics-notes-of-what-he-might-do.html"><a href="erics-notes-of-what-he-might-do.html#table-of-topics"><i class="fa fa-check"></i><b>2.1</b> Table of topics</a></li>
</ul></li>
<li class="part"><span><b>I Part I: Essential Computing Skills</b></span></li>
<li class="chapter" data-level="3" data-path="overview-of-essential-computing-skills.html"><a href="overview-of-essential-computing-skills.html"><i class="fa fa-check"></i><b>3</b> Overview of Essential Computing Skills</a></li>
<li class="chapter" data-level="4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html"><i class="fa fa-check"></i><b>4</b> Essential Unix/Linux Terminal Knowledge</a>
<ul>
<li class="chapter" data-level="4.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#getting-a-bash-shell-on-your-system"><i class="fa fa-check"></i><b>4.1</b> Getting a bash shell on your system</a></li>
<li class="chapter" data-level="4.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#navigating-the-unix-filesystem"><i class="fa fa-check"></i><b>4.2</b> Navigating the Unix filesystem</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#changing-the-working-directory-with-cd"><i class="fa fa-check"></i><b>4.2.1</b> Changing the working directory with <code>cd</code></a></li>
<li class="chapter" data-level="4.2.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#comm-prompt"><i class="fa fa-check"></i><b>4.2.2</b> Updating your command prompt</a></li>
<li class="chapter" data-level="4.2.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#tab-completion-for-paths"><i class="fa fa-check"></i><b>4.2.3</b> TAB-completion for paths</a></li>
<li class="chapter" data-level="4.2.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#listing-the-contents-of-a-directory-with-ls"><i class="fa fa-check"></i><b>4.2.4</b> Listing the contents of a directory with <code>ls</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#globbing"><i class="fa fa-check"></i><b>4.2.5</b> Globbing</a></li>
<li class="chapter" data-level="4.2.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#what-makes-a-good-file-name"><i class="fa fa-check"></i><b>4.2.6</b> What makes a good file-name?</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#the-anatomy-of-a-unix-command"><i class="fa fa-check"></i><b>4.3</b> The anatomy of a Unix command</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#anatomy-command"><i class="fa fa-check"></i><b>4.3.1</b> The <code>command</code></a></li>
<li class="chapter" data-level="4.3.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#the-options"><i class="fa fa-check"></i><b>4.3.2</b> The <em>options</em></a></li>
<li class="chapter" data-level="4.3.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#arguments"><i class="fa fa-check"></i><b>4.3.3</b> Arguments</a></li>
<li class="chapter" data-level="4.3.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#getting-information-about-unix-commands"><i class="fa fa-check"></i><b>4.3.4</b> Getting information about Unix commands</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#handling-manipulating-and-viewing-files-and-streams"><i class="fa fa-check"></i><b>4.4</b> Handling, Manipulating, and Viewing files and streams</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#creating-new-directories"><i class="fa fa-check"></i><b>4.4.1</b> Creating new directories</a></li>
<li class="chapter" data-level="4.4.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#fundamental-file-handling-commands"><i class="fa fa-check"></i><b>4.4.2</b> Fundamental file-handling commands</a></li>
<li class="chapter" data-level="4.4.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#viewing-files"><i class="fa fa-check"></i><b>4.4.3</b> “Viewing” Files</a></li>
<li class="chapter" data-level="4.4.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#redirecting-standard-output-and"><i class="fa fa-check"></i><b>4.4.4</b> Redirecting standard output: <code>&gt;</code> and <code>&gt;&gt;</code></a></li>
<li class="chapter" data-level="4.4.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#stdin-and"><i class="fa fa-check"></i><b>4.4.5</b> stdin, <code>&lt;</code> and <code>|</code></a></li>
<li class="chapter" data-level="4.4.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#stderr"><i class="fa fa-check"></i><b>4.4.6</b> stderr</a></li>
<li class="chapter" data-level="4.4.7" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#symbolic-links"><i class="fa fa-check"></i><b>4.4.7</b> Symbolic links</a></li>
<li class="chapter" data-level="4.4.8" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#file-perm"><i class="fa fa-check"></i><b>4.4.8</b> File Permissions</a></li>
<li class="chapter" data-level="4.4.9" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#editing-text-files-at-the-terminal"><i class="fa fa-check"></i><b>4.4.9</b> Editing text files at the terminal</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#unix-env"><i class="fa fa-check"></i><b>4.5</b> Customizing your Environment</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#appearances-matter"><i class="fa fa-check"></i><b>4.5.1</b> Appearances matter</a></li>
<li class="chapter" data-level="4.5.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#where-are-my-programscommands-at"><i class="fa fa-check"></i><b>4.5.2</b> Where are my programs/commands at?!</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#a-few-more-important-keystrokes"><i class="fa fa-check"></i><b>4.6</b> A Few More Important Keystrokes</a></li>
<li class="chapter" data-level="4.7" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#a-short-list-of-additional-useful-commands."><i class="fa fa-check"></i><b>4.7</b> A short list of additional useful commands.</a></li>
<li class="chapter" data-level="4.8" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#two-important-computing-concepts"><i class="fa fa-check"></i><b>4.8</b> Two important computing concepts</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#compression"><i class="fa fa-check"></i><b>4.8.1</b> Compression</a></li>
<li class="chapter" data-level="4.8.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#hashing"><i class="fa fa-check"></i><b>4.8.2</b> Hashing</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#unix-quick-study-guide"><i class="fa fa-check"></i><b>4.9</b> Unix: Quick Study Guide</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="shell-programming.html"><a href="shell-programming.html"><i class="fa fa-check"></i><b>5</b> Shell programming</a>
<ul>
<li class="chapter" data-level="5.1" data-path="shell-programming.html"><a href="shell-programming.html#an-example-script"><i class="fa fa-check"></i><b>5.1</b> An example script</a></li>
<li class="chapter" data-level="5.2" data-path="shell-programming.html"><a href="shell-programming.html#the-structure-of-a-bash-script"><i class="fa fa-check"></i><b>5.2</b> The Structure of a Bash Script</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="shell-programming.html"><a href="shell-programming.html#paragraph-before-shebang"><i class="fa fa-check"></i><b>5.2.1</b> A bit more on <code>;</code> and <code>&amp;</code></a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="shell-programming.html"><a href="shell-programming.html#variables"><i class="fa fa-check"></i><b>5.3</b> Variables</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="shell-programming.html"><a href="shell-programming.html#assigning-values-to-variables"><i class="fa fa-check"></i><b>5.3.1</b> Assigning values to variables</a></li>
<li class="chapter" data-level="5.3.2" data-path="shell-programming.html"><a href="shell-programming.html#accessing-values-from-variables"><i class="fa fa-check"></i><b>5.3.2</b> Accessing values from variables</a></li>
<li class="chapter" data-level="5.3.3" data-path="shell-programming.html"><a href="shell-programming.html#what-does-the-shell-do-with-the-value-substituted-for-a-variable"><i class="fa fa-check"></i><b>5.3.3</b> What does the shell do with the value substituted for a variable?</a></li>
<li class="chapter" data-level="5.3.4" data-path="shell-programming.html"><a href="shell-programming.html#quotes-and-var-subs"><i class="fa fa-check"></i><b>5.3.4</b> Double and Single Quotation Marks and Variable Substitution</a></li>
<li class="chapter" data-level="5.3.5" data-path="shell-programming.html"><a href="shell-programming.html#one-useful-fancy-variable-substitution-method"><i class="fa fa-check"></i><b>5.3.5</b> One useful, fancy, variable-substitution method</a></li>
<li class="chapter" data-level="5.3.6" data-path="shell-programming.html"><a href="shell-programming.html#bash-arithmetic"><i class="fa fa-check"></i><b>5.3.6</b> Integer Arithmetic with Shell Variables</a></li>
<li class="chapter" data-level="5.3.7" data-path="shell-programming.html"><a href="shell-programming.html#variable-arrays"><i class="fa fa-check"></i><b>5.3.7</b> Variable arrays</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="shell-programming.html"><a href="shell-programming.html#evaluate-a-command-and-substitute-the-result-on-the-command-line"><i class="fa fa-check"></i><b>5.4</b> Evaluate a command and substitute the result on the command line</a></li>
<li class="chapter" data-level="5.5" data-path="shell-programming.html"><a href="shell-programming.html#groupingcollecting-output-from-multiple-commands-commands-and-commands"><i class="fa fa-check"></i><b>5.5</b> Grouping/Collecting output from multiple commands: <code>(commands)</code> and <code>{ commands; }</code></a></li>
<li class="chapter" data-level="5.6" data-path="shell-programming.html"><a href="shell-programming.html#exit-status"><i class="fa fa-check"></i><b>5.6</b> Exit Status</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="shell-programming.html"><a href="shell-programming.html#combinations-of-exit-statuses"><i class="fa fa-check"></i><b>5.6.1</b> Combinations of exit statuses</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="shell-programming.html"><a href="shell-programming.html#unix-for-loops"><i class="fa fa-check"></i><b>5.7</b> Loops and repetition</a></li>
<li class="chapter" data-level="5.8" data-path="shell-programming.html"><a href="shell-programming.html#more-conditional-evaluation-if-then-else-and-friends"><i class="fa fa-check"></i><b>5.8</b> More Conditional Evaluation: <code>if</code>, <code>then</code>, <code>else</code>, and friends</a></li>
<li class="chapter" data-level="5.9" data-path="shell-programming.html"><a href="shell-programming.html#finallypositional-parameters"><i class="fa fa-check"></i><b>5.9</b> Finally…positional parameters</a></li>
<li class="chapter" data-level="5.10" data-path="shell-programming.html"><a href="shell-programming.html#basename-and-dirname-two-useful-little-utilities"><i class="fa fa-check"></i><b>5.10</b> <code>basename</code> and <code>dirname</code> two useful little utilities</a></li>
<li class="chapter" data-level="5.11" data-path="shell-programming.html"><a href="shell-programming.html#bash-functions"><i class="fa fa-check"></i><b>5.11</b> <code>bash</code> functions</a></li>
<li class="chapter" data-level="5.12" data-path="shell-programming.html"><a href="shell-programming.html#reading-files-line-by-line"><i class="fa fa-check"></i><b>5.12</b> reading files line by line</a></li>
<li class="chapter" data-level="5.13" data-path="shell-programming.html"><a href="shell-programming.html#further-reading"><i class="fa fa-check"></i><b>5.13</b> Further reading</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html"><i class="fa fa-check"></i><b>6</b> Sed, awk, and regular expressions</a>
<ul>
<li class="chapter" data-level="6.1" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#awk"><i class="fa fa-check"></i><b>6.1</b> awk</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#line-cycling-tests-and-actions"><i class="fa fa-check"></i><b>6.1.1</b> Line-cycling, tests and actions</a></li>
<li class="chapter" data-level="6.1.2" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#column-splitting-fields--f-nf-print-ofs-and-begin"><i class="fa fa-check"></i><b>6.1.2</b> Column splitting, fields, <code>-F</code>, <code>$</code>, <code>NF</code>, <code>print</code>, <code>OFS</code> and <code>BEGIN</code></a></li>
<li class="chapter" data-level="6.1.3" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#a-brief-introduction-to-regular-expressions"><i class="fa fa-check"></i><b>6.1.3</b> A brief introduction to regular expressions</a></li>
<li class="chapter" data-level="6.1.4" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#a-variety-of-tests"><i class="fa fa-check"></i><b>6.1.4</b> A variety of tests</a></li>
<li class="chapter" data-level="6.1.5" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#code-in-the-action-blocks"><i class="fa fa-check"></i><b>6.1.5</b> Code in the action blocks</a></li>
<li class="chapter" data-level="6.1.6" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#using-awk-to-assign-to-shell-variables"><i class="fa fa-check"></i><b>6.1.6</b> Using <code>awk</code> to assign to shell variables</a></li>
<li class="chapter" data-level="6.1.7" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#passing-variables-into-awk-with--v"><i class="fa fa-check"></i><b>6.1.7</b> Passing Variables into <code>awk</code> with <code>-v</code></a></li>
<li class="chapter" data-level="6.1.8" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#writing-awk-scripts-in-files"><i class="fa fa-check"></i><b>6.1.8</b> Writing awk scripts in files</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#sed"><i class="fa fa-check"></i><b>6.2</b> sed</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html"><i class="fa fa-check"></i><b>7</b> Working on remote servers</a>
<ul>
<li class="chapter" data-level="7.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#accessing-remote-computers"><i class="fa fa-check"></i><b>7.1</b> Accessing remote computers</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#windows"><i class="fa fa-check"></i><b>7.1.1</b> Windows</a></li>
<li class="chapter" data-level="7.1.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#hummingbird"><i class="fa fa-check"></i><b>7.1.2</b> Hummingbird</a></li>
<li class="chapter" data-level="7.1.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#summit"><i class="fa fa-check"></i><b>7.1.3</b> Summit</a></li>
<li class="chapter" data-level="7.1.4" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#sedna"><i class="fa fa-check"></i><b>7.1.4</b> Sedna</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#transferring-files-to-remote-computers"><i class="fa fa-check"></i><b>7.2</b> Transferring files to remote computers</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#sftp-and-several-systems-that-use-it"><i class="fa fa-check"></i><b>7.2.1</b> <code>sftp</code> and several systems that use it</a></li>
<li class="chapter" data-level="7.2.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#git"><i class="fa fa-check"></i><b>7.2.2</b> git</a></li>
<li class="chapter" data-level="7.2.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#globus"><i class="fa fa-check"></i><b>7.2.3</b> Globus</a></li>
<li class="chapter" data-level="7.2.4" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#interfacing-with-the-cloud"><i class="fa fa-check"></i><b>7.2.4</b> Interfacing with “The Cloud”</a></li>
<li class="chapter" data-level="7.2.5" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#get-seqs"><i class="fa fa-check"></i><b>7.2.5</b> Getting files from a sequencing center</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#tmux"><i class="fa fa-check"></i><b>7.3</b> <code>tmux</code>: the terminal multiplexer</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#an-analogy-for-how-tmux-works"><i class="fa fa-check"></i><b>7.3.1</b> An analogy for how <code>tmux</code> works</a></li>
<li class="chapter" data-level="7.3.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#first-steps-with-tmux"><i class="fa fa-check"></i><b>7.3.2</b> First steps with <code>tmux</code></a></li>
<li class="chapter" data-level="7.3.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#further-steps-with-tmux"><i class="fa fa-check"></i><b>7.3.3</b> Further steps with <code>tmux</code></a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#tmux-for-mac-users"><i class="fa fa-check"></i><b>7.4</b> <code>tmux</code> for Mac users</a></li>
<li class="chapter" data-level="7.5" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#installing-software-on-an-hpcc"><i class="fa fa-check"></i><b>7.5</b> Installing Software on an HPCC</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#modules"><i class="fa fa-check"></i><b>7.5.1</b> Modules</a></li>
<li class="chapter" data-level="7.5.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#miniconda"><i class="fa fa-check"></i><b>7.5.2</b> Miniconda</a></li>
<li class="chapter" data-level="7.5.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#installing-java-programs"><i class="fa fa-check"></i><b>7.5.3</b> Installing Java Programs</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#vim-its-time-to-get-serious-with-text-editing"><i class="fa fa-check"></i><b>7.6</b> <code>vim</code>: it’s time to get serious with text editing</a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#using-neovim-and-nvim-r-and-tmux-to-use-r-well-on-the-cluster"><i class="fa fa-check"></i><b>7.6.1</b> Using neovim and Nvim-R and tmux to use R well on the cluster</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="chap-HPCC.html"><a href="chap-HPCC.html"><i class="fa fa-check"></i><b>8</b> High Performance Computing Clusters (HPCC’s)</a>
<ul>
<li class="chapter" data-level="8.1" data-path="chap-HPCC.html"><a href="chap-HPCC.html#an-oversimplified-but-useful-view-of-a-computing-cluster"><i class="fa fa-check"></i><b>8.1</b> An oversimplified, but useful, view of a computing cluster</a></li>
<li class="chapter" data-level="8.2" data-path="chap-HPCC.html"><a href="chap-HPCC.html#cluster-computing-and-the-job-scheduler"><i class="fa fa-check"></i><b>8.2</b> Cluster computing and the job scheduler</a></li>
<li class="chapter" data-level="8.3" data-path="chap-HPCC.html"><a href="chap-HPCC.html#slurm-info"><i class="fa fa-check"></i><b>8.3</b> Learning about the resources on your HPCC</a></li>
<li class="chapter" data-level="8.4" data-path="chap-HPCC.html"><a href="chap-HPCC.html#getting-compute-resources-allocated-to-your-jobs-on-an-hpcc"><i class="fa fa-check"></i><b>8.4</b> Getting compute resources allocated to your jobs on an HPCC</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="chap-HPCC.html"><a href="chap-HPCC.html#interactive-sessions"><i class="fa fa-check"></i><b>8.4.1</b> Interactive sessions</a></li>
<li class="chapter" data-level="8.4.2" data-path="chap-HPCC.html"><a href="chap-HPCC.html#slurm-batch"><i class="fa fa-check"></i><b>8.4.2</b> Batch jobs</a></li>
<li class="chapter" data-level="8.4.3" data-path="chap-HPCC.html"><a href="chap-HPCC.html#slurm-job-arrays"><i class="fa fa-check"></i><b>8.4.3</b> SLURM Job Arrays</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="chap-HPCC.html"><a href="chap-HPCC.html#last-years-stuff.-now-somewhat-defunct"><i class="fa fa-check"></i><b>8.5</b> Last year’s stuff. Now somewhat defunct</a></li>
<li class="chapter" data-level="8.6" data-path="chap-HPCC.html"><a href="chap-HPCC.html#prepation-interlude-an-in-class-exercise-to-make-sure-everything-is-configured-correctly"><i class="fa fa-check"></i><b>8.6</b> PREPATION INTERLUDE: An in-class exercise to make sure everything is configured correctly</a></li>
<li class="chapter" data-level="8.7" data-path="chap-HPCC.html"><a href="chap-HPCC.html#more-boneyard"><i class="fa fa-check"></i><b>8.7</b> More Boneyard…</a></li>
<li class="chapter" data-level="8.8" data-path="chap-HPCC.html"><a href="chap-HPCC.html#the-queue-slurmsgeuge"><i class="fa fa-check"></i><b>8.8</b> The Queue (SLURM/SGE/UGE)</a></li>
<li class="chapter" data-level="8.9" data-path="chap-HPCC.html"><a href="chap-HPCC.html#modules-package"><i class="fa fa-check"></i><b>8.9</b> Modules package</a></li>
<li class="chapter" data-level="8.10" data-path="chap-HPCC.html"><a href="chap-HPCC.html#compiling-programs-without-admin-privileges"><i class="fa fa-check"></i><b>8.10</b> Compiling programs without admin privileges</a></li>
<li class="chapter" data-level="8.11" data-path="chap-HPCC.html"><a href="chap-HPCC.html#job-arrays"><i class="fa fa-check"></i><b>8.11</b> Job arrays</a></li>
<li class="chapter" data-level="8.12" data-path="chap-HPCC.html"><a href="chap-HPCC.html#writing-stdout-and-stderr-to-files"><i class="fa fa-check"></i><b>8.12</b> Writing stdout and stderr to files</a></li>
<li class="chapter" data-level="8.13" data-path="chap-HPCC.html"><a href="chap-HPCC.html#breaking-stuff-down"><i class="fa fa-check"></i><b>8.13</b> Breaking stuff down</a></li>
</ul></li>
<li class="part"><span><b>II Part II: Reproducible Research Strategies</b></span></li>
<li class="chapter" data-level="9" data-path="introduction-to-reproducible-research.html"><a href="introduction-to-reproducible-research.html"><i class="fa fa-check"></i><b>9</b> Introduction to Reproducible Research</a></li>
<li class="chapter" data-level="10" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html"><i class="fa fa-check"></i><b>10</b> Rstudio and Project-centered Organization</a>
<ul>
<li class="chapter" data-level="10.1" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#organizing-big-projects"><i class="fa fa-check"></i><b>10.1</b> Organizing big projects</a></li>
<li class="chapter" data-level="10.2" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#using-rstudio-in-workflows-with-remote-computers-and-hpccs"><i class="fa fa-check"></i><b>10.2</b> Using RStudio in workflows with remote computers and HPCCs</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#keeping-an-rstudio-project-in-sync-with-github"><i class="fa fa-check"></i><b>10.2.1</b> Keeping an RStudio project “in sync” with GitHub</a></li>
<li class="chapter" data-level="10.2.2" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#evaluating-scripts-line-by-line-on-a-remote-machine-from-within-rstudio"><i class="fa fa-check"></i><b>10.2.2</b> Evaluating scripts line by line on a remote machine from within RStudio</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="version-control.html"><a href="version-control.html"><i class="fa fa-check"></i><b>11</b> Version control</a>
<ul>
<li class="chapter" data-level="11.1" data-path="version-control.html"><a href="version-control.html#why-use-version-control"><i class="fa fa-check"></i><b>11.1</b> Why use version control?</a></li>
<li class="chapter" data-level="11.2" data-path="version-control.html"><a href="version-control.html#git-workings"><i class="fa fa-check"></i><b>11.2</b> How git works</a></li>
<li class="chapter" data-level="11.3" data-path="version-control.html"><a href="version-control.html#git-workflow-patterns"><i class="fa fa-check"></i><b>11.3</b> git workflow patterns</a></li>
<li class="chapter" data-level="11.4" data-path="version-control.html"><a href="version-control.html#using-git-with-rstudio"><i class="fa fa-check"></i><b>11.4</b> using git with Rstudio</a></li>
<li class="chapter" data-level="11.5" data-path="version-control.html"><a href="version-control.html#git-on-the-command-line"><i class="fa fa-check"></i><b>11.5</b> git on the command line</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="a-fast-furious-overview-of-the-tidyverse.html"><a href="a-fast-furious-overview-of-the-tidyverse.html"><i class="fa fa-check"></i><b>12</b> A fast, furious overview of the tidyverse</a></li>
<li class="chapter" data-level="13" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html"><i class="fa fa-check"></i><b>13</b> Authoring reproducibly with Rmarkdown</a>
<ul>
<li class="chapter" data-level="13.1" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#notebooks"><i class="fa fa-check"></i><b>13.1</b> Notebooks</a></li>
<li class="chapter" data-level="13.2" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#references"><i class="fa fa-check"></i><b>13.2</b> References</a>
<ul>
<li class="chapter" data-level="13.2.1" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#zotero-and-rmarkdown"><i class="fa fa-check"></i><b>13.2.1</b> Zotero and Rmarkdown</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#bookdown"><i class="fa fa-check"></i><b>13.3</b> Bookdown</a></li>
<li class="chapter" data-level="13.4" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#google-docs"><i class="fa fa-check"></i><b>13.4</b> Google Docs</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="snakemake-chap.html"><a href="snakemake-chap.html"><i class="fa fa-check"></i><b>14</b> Managing Workflows with Snakemake</a>
<ul>
<li class="chapter" data-level="14.1" data-path="snakemake-chap.html"><a href="snakemake-chap.html#a-simple-snakemake-example"><i class="fa fa-check"></i><b>14.1</b> A Simple Snakemake Example</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="snakemake-chap.html"><a href="snakemake-chap.html#snakefile-1-a-simple-rule-to-trimmomcawesome-the-files-from-001"><i class="fa fa-check"></i><b>14.1.1</b> Snakefile #1: A simple rule to TrimmoMcAwesome the files from 001</a></li>
<li class="chapter" data-level="14.1.2" data-path="snakemake-chap.html"><a href="snakemake-chap.html#snakefile-2-a-single-trimmomcawesome-rule-using-wildcards"><i class="fa fa-check"></i><b>14.1.2</b> Snakefile #2: A single TrimmoMcAwesome rule using wildcards</a></li>
<li class="chapter" data-level="14.1.3" data-path="snakemake-chap.html"><a href="snakemake-chap.html#snakefile-3-add-trim_stupendous-in-there"><i class="fa fa-check"></i><b>14.1.3</b> Snakefile #3: Add <code>trim_stupendous</code> in there</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="snakemake-chap.html"><a href="snakemake-chap.html#ugly-complex-file-namesa-job-for-snakemake-input-functions"><i class="fa fa-check"></i><b>14.2</b> Ugly, complex file names…a job for Snakemake input functions!</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="snakemake-chap.html"><a href="snakemake-chap.html#playing-with-pandas"><i class="fa fa-check"></i><b>14.2.1</b> Playing with Pandas</a></li>
<li class="chapter" data-level="14.2.2" data-path="snakemake-chap.html"><a href="snakemake-chap.html#a-quick-sidebar-spoofing-a-snakemake-input-function"><i class="fa fa-check"></i><b>14.2.2</b> A Quick Sidebar: Spoofing a Snakemake input function</a></li>
<li class="chapter" data-level="14.2.3" data-path="snakemake-chap.html"><a href="snakemake-chap.html#snakefile-4-an-input-function-for-trim_awesome"><i class="fa fa-check"></i><b>14.2.3</b> Snakefile #4: an input function for trim_awesome</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="snakemake-chap.html"><a href="snakemake-chap.html#input-functions-from-python-dictionaries"><i class="fa fa-check"></i><b>14.3</b> Input functions from Python Dictionaries</a></li>
<li class="chapter" data-level="14.4" data-path="snakemake-chap.html"><a href="snakemake-chap.html#generating-lists-of-files-with-expand"><i class="fa fa-check"></i><b>14.4</b> Generating lists of files with expand()</a></li>
<li class="chapter" data-level="14.5" data-path="snakemake-chap.html"><a href="snakemake-chap.html#a-python-primer-for-using-snakemake"><i class="fa fa-check"></i><b>14.5</b> A Python Primer for Using Snakemake</a></li>
<li class="chapter" data-level="14.6" data-path="snakemake-chap.html"><a href="snakemake-chap.html#using-snakemake-on-a-computing-cluster"><i class="fa fa-check"></i><b>14.6</b> Using Snakemake on a Computing Cluster</a>
<ul>
<li class="chapter" data-level="14.6.1" data-path="snakemake-chap.html"><a href="snakemake-chap.html#installing-and-configuring-the-snakemake-slurm-profile"><i class="fa fa-check"></i><b>14.6.1</b> Installing and Configuring the Snakemake SLURM profile</a></li>
<li class="chapter" data-level="14.6.2" data-path="snakemake-chap.html"><a href="snakemake-chap.html#playing-with-our-new-slurm-profile"><i class="fa fa-check"></i><b>14.6.2</b> Playing with our new SLURM profile</a></li>
<li class="chapter" data-level="14.6.3" data-path="snakemake-chap.html"><a href="snakemake-chap.html#managing-cluster-resources-with-snakemake"><i class="fa fa-check"></i><b>14.6.3</b> Managing cluster resources with Snakemake</a></li>
<li class="chapter" data-level="14.6.4" data-path="snakemake-chap.html"><a href="snakemake-chap.html#grouping-short-jobs-to-run-on-a-single-node"><i class="fa fa-check"></i><b>14.6.4</b> Grouping short jobs to run on a single node</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Part III: Bioinformatic Analyses</b></span></li>
<li class="chapter" data-level="15" data-path="overview-of-bioinformatic-analyses.html"><a href="overview-of-bioinformatic-analyses.html"><i class="fa fa-check"></i><b>15</b> Overview of Bioinformatic Analyses</a></li>
<li class="chapter" data-level="16" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html"><i class="fa fa-check"></i><b>16</b> DNA Sequences and Sequencing</a>
<ul>
<li class="chapter" data-level="16.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#dna-stuff"><i class="fa fa-check"></i><b>16.1</b> DNA Stuff</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#dna-replication-with-dna-polymerase"><i class="fa fa-check"></i><b>16.1.1</b> DNA Replication with DNA Polymerase</a></li>
<li class="chapter" data-level="16.1.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#the-importance-of-the-3-hydroxyl"><i class="fa fa-check"></i><b>16.1.2</b> The importance of the 3’ hydroxyl…</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#sanger-sequencing"><i class="fa fa-check"></i><b>16.2</b> Sanger sequencing</a></li>
<li class="chapter" data-level="16.3" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#illumina-sequencing-by-synthesis"><i class="fa fa-check"></i><b>16.3</b> Illumina Sequencing by Synthesis</a></li>
<li class="chapter" data-level="16.4" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#library-prep-protocols"><i class="fa fa-check"></i><b>16.4</b> Library Prep Protocols</a>
<ul>
<li class="chapter" data-level="16.4.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#wgs"><i class="fa fa-check"></i><b>16.4.1</b> WGS</a></li>
<li class="chapter" data-level="16.4.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#rad-seq-methods"><i class="fa fa-check"></i><b>16.4.2</b> RAD-Seq methods</a></li>
<li class="chapter" data-level="16.4.3" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#amplicon-sequencing"><i class="fa fa-check"></i><b>16.4.3</b> Amplicon Sequencing</a></li>
<li class="chapter" data-level="16.4.4" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#capture-arrays-rapture-etc."><i class="fa fa-check"></i><b>16.4.4</b> Capture arrays, RAPTURE, etc.</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html"><i class="fa fa-check"></i><b>17</b> Bioinformatic file formats</a>
<ul>
<li class="chapter" data-level="17.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sequences"><i class="fa fa-check"></i><b>17.1</b> Sequences</a></li>
<li class="chapter" data-level="17.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#fastq"><i class="fa fa-check"></i><b>17.2</b> FASTQ</a>
<ul>
<li class="chapter" data-level="17.2.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#illumina-ids"><i class="fa fa-check"></i><b>17.2.1</b> Line 1: Illumina identifier lines</a></li>
<li class="chapter" data-level="17.2.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#bqscores"><i class="fa fa-check"></i><b>17.2.2</b> Line 4: Base quality scores</a></li>
<li class="chapter" data-level="17.2.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#a-fastq-tidyverse-interlude"><i class="fa fa-check"></i><b>17.2.3</b> A FASTQ ‘tidyverse’ Interlude</a></li>
<li class="chapter" data-level="17.2.4" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#comparing-read-1-to-read-2"><i class="fa fa-check"></i><b>17.2.4</b> Comparing read 1 to read 2</a></li>
</ul></li>
<li class="chapter" data-level="17.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#fasta"><i class="fa fa-check"></i><b>17.3</b> FASTA</a>
<ul>
<li class="chapter" data-level="17.3.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#genomic-ranges"><i class="fa fa-check"></i><b>17.3.1</b> Genomic ranges</a></li>
<li class="chapter" data-level="17.3.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#extracting-genomic-ranges-from-a-fasta-file"><i class="fa fa-check"></i><b>17.3.2</b> Extracting genomic ranges from a FASTA file</a></li>
<li class="chapter" data-level="17.3.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#downloading-reference-genomes-from-ncbi"><i class="fa fa-check"></i><b>17.3.3</b> Downloading reference genomes from NCBI</a></li>
</ul></li>
<li class="chapter" data-level="17.4" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sambamfiles"><i class="fa fa-check"></i><b>17.4</b> Alignments</a>
<ul>
<li class="chapter" data-level="17.4.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#how-might-i-align-to-thee-let-me-count-the-ways"><i class="fa fa-check"></i><b>17.4.1</b> How might I align to thee? Let me count the ways…</a></li>
<li class="chapter" data-level="17.4.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#play-with-simple-alignments"><i class="fa fa-check"></i><b>17.4.2</b> Play with simple alignments</a></li>
<li class="chapter" data-level="17.4.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sam-flags"><i class="fa fa-check"></i><b>17.4.3</b> SAM Flags</a></li>
<li class="chapter" data-level="17.4.4" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#cigar"><i class="fa fa-check"></i><b>17.4.4</b> The CIGAR string</a></li>
<li class="chapter" data-level="17.4.5" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#the-seq-and-qual-columns"><i class="fa fa-check"></i><b>17.4.5</b> The SEQ and QUAL columns</a></li>
<li class="chapter" data-level="17.4.6" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sam-headers"><i class="fa fa-check"></i><b>17.4.6</b> SAM File Headers</a></li>
<li class="chapter" data-level="17.4.7" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#the-bam-format"><i class="fa fa-check"></i><b>17.4.7</b> The BAM format</a></li>
<li class="chapter" data-level="17.4.8" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#quick-self-study"><i class="fa fa-check"></i><b>17.4.8</b> Quick self study</a></li>
</ul></li>
<li class="chapter" data-level="17.5" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#variants"><i class="fa fa-check"></i><b>17.5</b> Variants</a>
<ul>
<li class="chapter" data-level="17.5.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#vcf-format-the-body"><i class="fa fa-check"></i><b>17.5.1</b> VCF Format – The Body</a></li>
<li class="chapter" data-level="17.5.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#vcf-format-the-header"><i class="fa fa-check"></i><b>17.5.2</b> VCF Format – The Header</a></li>
<li class="chapter" data-level="17.5.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#boneyard"><i class="fa fa-check"></i><b>17.5.3</b> Boneyard</a></li>
</ul></li>
<li class="chapter" data-level="17.6" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#segments"><i class="fa fa-check"></i><b>17.6</b> Segments</a></li>
<li class="chapter" data-level="17.7" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#conversionextractions-between-different-formats"><i class="fa fa-check"></i><b>17.7</b> Conversion/Extractions between different formats</a></li>
<li class="chapter" data-level="17.8" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#visualization-of-genomic-data"><i class="fa fa-check"></i><b>17.8</b> Visualization of Genomic Data</a>
<ul>
<li class="chapter" data-level="17.8.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sample-data"><i class="fa fa-check"></i><b>17.8.1</b> Sample Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="18" data-path="genome-assembly.html"><a href="genome-assembly.html"><i class="fa fa-check"></i><b>18</b> Genome Assembly</a></li>
<li class="chapter" data-level="19" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><i class="fa fa-check"></i><b>19</b> Alignment of sequence data to a reference genome (and associated steps)</a>
<ul>
<li class="chapter" data-level="19.1" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#read-journey"><i class="fa fa-check"></i><b>19.1</b> The Journey of each DNA Fragment from Organism to Sequencing Read</a></li>
<li class="chapter" data-level="19.2" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#read-groups"><i class="fa fa-check"></i><b>19.2</b> Read Groups</a></li>
<li class="chapter" data-level="19.3" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#aligning-reads-with-bwa"><i class="fa fa-check"></i><b>19.3</b> Aligning reads with <code>bwa</code></a>
<ul>
<li class="chapter" data-level="19.3.1" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#indexing-the-genome-for-alignment"><i class="fa fa-check"></i><b>19.3.1</b> Indexing the genome for alignment</a></li>
<li class="chapter" data-level="19.3.2" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#mapping-reads-with-bwa-mem"><i class="fa fa-check"></i><b>19.3.2</b> Mapping reads with <code>bwa mem</code></a></li>
<li class="chapter" data-level="19.3.3" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#hold-it-right-there-buddy-what-about-the-read-groups"><i class="fa fa-check"></i><b>19.3.3</b> Hold it Right There, Buddy! What about the Read Groups?</a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#samtools"><i class="fa fa-check"></i><b>19.4</b> Processing alignment output with <code>samtools</code></a>
<ul>
<li class="chapter" data-level="19.4.1" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#samtools-subcommands"><i class="fa fa-check"></i><b>19.4.1</b> <code>samtools</code> subcommands</a></li>
</ul></li>
<li class="chapter" data-level="19.5" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#boneyard-below-here"><i class="fa fa-check"></i><b>19.5</b> BONEYARD BELOW HERE</a></li>
<li class="chapter" data-level="19.6" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#preprocess"><i class="fa fa-check"></i><b>19.6</b> Preprocess ?</a></li>
<li class="chapter" data-level="19.7" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#quick-notes-to-self-on-chaining-things"><i class="fa fa-check"></i><b>19.7</b> Quick notes to self on chaining things:</a></li>
<li class="chapter" data-level="19.8" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#merging-bam-files"><i class="fa fa-check"></i><b>19.8</b> Merging BAM files</a></li>
<li class="chapter" data-level="19.9" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#divide-and-conquer-strategies"><i class="fa fa-check"></i><b>19.9</b> Divide and Conquer Strategies</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="variant-calling.html"><a href="variant-calling.html"><i class="fa fa-check"></i><b>20</b> Variant calling</a>
<ul>
<li class="chapter" data-level="20.1" data-path="variant-calling.html"><a href="variant-calling.html#genotype-likelihoods"><i class="fa fa-check"></i><b>20.1</b> Genotype Likelihoods</a>
<ul>
<li class="chapter" data-level="20.1.1" data-path="variant-calling.html"><a href="variant-calling.html#basic-sketch-of-genotype-likelihood-calculations"><i class="fa fa-check"></i><b>20.1.1</b> Basic Sketch of Genotype Likelihood Calculations</a></li>
<li class="chapter" data-level="20.1.2" data-path="variant-calling.html"><a href="variant-calling.html#specifics-of-different-genotype-likelihoods"><i class="fa fa-check"></i><b>20.1.2</b> Specifics of different genotype likelihoods</a></li>
<li class="chapter" data-level="20.1.3" data-path="variant-calling.html"><a href="variant-calling.html#computing-genotype-likelihoods-with-three-different-softwares"><i class="fa fa-check"></i><b>20.1.3</b> Computing genotype likelihoods with three different softwares</a></li>
<li class="chapter" data-level="20.1.4" data-path="variant-calling.html"><a href="variant-calling.html#a-directed-acyclic-graph-for-genotype-likelihoods"><i class="fa fa-check"></i><b>20.1.4</b> A Directed Acyclic Graph For Genotype Likelihoods</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="21" data-path="boneyard-1.html"><a href="boneyard-1.html"><i class="fa fa-check"></i><b>21</b> Boneyard</a></li>
<li class="chapter" data-level="22" data-path="basic-handling-of-vcf-files.html"><a href="basic-handling-of-vcf-files.html"><i class="fa fa-check"></i><b>22</b> Basic Handling of VCF files</a>
<ul>
<li class="chapter" data-level="22.1" data-path="basic-handling-of-vcf-files.html"><a href="basic-handling-of-vcf-files.html#bcftools"><i class="fa fa-check"></i><b>22.1</b> bcftools</a>
<ul>
<li class="chapter" data-level="22.1.1" data-path="basic-handling-of-vcf-files.html"><a href="basic-handling-of-vcf-files.html#tell-me-about-my-vcf-file"><i class="fa fa-check"></i><b>22.1.1</b> Tell me about my VCF file!</a></li>
<li class="chapter" data-level="22.1.2" data-path="basic-handling-of-vcf-files.html"><a href="basic-handling-of-vcf-files.html#get-fragmentsparts-of-my-vcf-file"><i class="fa fa-check"></i><b>22.1.2</b> Get fragments/parts of my VCF file</a></li>
<li class="chapter" data-level="22.1.3" data-path="basic-handling-of-vcf-files.html"><a href="basic-handling-of-vcf-files.html#combine-vcf-files-in-various-ways"><i class="fa fa-check"></i><b>22.1.3</b> Combine VCF files in various ways</a></li>
<li class="chapter" data-level="22.1.4" data-path="basic-handling-of-vcf-files.html"><a href="basic-handling-of-vcf-files.html#filter-out-variants-for-a-variety-of-reasons"><i class="fa fa-check"></i><b>22.1.4</b> Filter out variants for a variety of reasons</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="23" data-path="bioinformatics-for-rad-seq-data-with-and-without-a-reference-genome.html"><a href="bioinformatics-for-rad-seq-data-with-and-without-a-reference-genome.html"><i class="fa fa-check"></i><b>23</b> Bioinformatics for RAD seq data with and without a reference genome</a></li>
<li class="chapter" data-level="24" data-path="processing-amplicon-sequencing-data.html"><a href="processing-amplicon-sequencing-data.html"><i class="fa fa-check"></i><b>24</b> Processing amplicon sequencing data</a></li>
<li class="chapter" data-level="25" data-path="genome-annotation.html"><a href="genome-annotation.html"><i class="fa fa-check"></i><b>25</b> Genome Annotation</a></li>
<li class="chapter" data-level="26" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html"><i class="fa fa-check"></i><b>26</b> Whole genome alignment strategies</a>
<ul>
<li class="chapter" data-level="26.1" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#mapping-of-scaffolds-to-a-closely-related-genome"><i class="fa fa-check"></i><b>26.1</b> Mapping of scaffolds to a closely related genome</a></li>
<li class="chapter" data-level="26.2" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#obtaining-ancestral-states-from-an-outgroup-genome"><i class="fa fa-check"></i><b>26.2</b> Obtaining Ancestral States from an Outgroup Genome</a>
<ul>
<li class="chapter" data-level="26.2.1" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#using-lastz-to-align-coho-to-the-chinook-genome"><i class="fa fa-check"></i><b>26.2.1</b> Using LASTZ to align coho to the chinook genome</a></li>
<li class="chapter" data-level="26.2.2" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#try-on-the-chinook-chromosomes"><i class="fa fa-check"></i><b>26.2.2</b> Try on the chinook chromosomes</a></li>
<li class="chapter" data-level="26.2.3" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#explore-the-other-parameters-more"><i class="fa fa-check"></i><b>26.2.3</b> Explore the other parameters more</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="27" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html"><i class="fa fa-check"></i><b>27</b> An example full workflow from sequences to variants</a>
<ul>
<li class="chapter" data-level="27.1" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#an-overview-of-the-whole-workflow"><i class="fa fa-check"></i><b>27.1</b> An overview of the whole workflow</a></li>
<li class="chapter" data-level="27.2" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#lets-have-a-look-at-the-reference-genome"><i class="fa fa-check"></i><b>27.2</b> Let’s have a look at the reference genome</a>
<ul>
<li class="chapter" data-level="27.2.1" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#first-off-make-sure-that-you-have-added-mamba-to-your-base-environment"><i class="fa fa-check"></i><b>27.2.1</b> First off, make sure that you have added mamba to your base environment</a></li>
<li class="chapter" data-level="27.2.2" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#indexing-the-reference-genome"><i class="fa fa-check"></i><b>27.2.2</b> Indexing the reference genome</a></li>
</ul></li>
<li class="chapter" data-level="27.3" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#a-word-on-organizing-workflows"><i class="fa fa-check"></i><b>27.3</b> A word on organizing workflows</a></li>
<li class="chapter" data-level="27.4" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step-1a.-quality-checking-the-reads"><i class="fa fa-check"></i><b>27.4</b> Step 1a. Quality Checking the Reads</a>
<ul>
<li class="chapter" data-level="27.4.1" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#a-little-looking-at-the-htmls"><i class="fa fa-check"></i><b>27.4.1</b> A Little Looking at the HTMLs</a></li>
<li class="chapter" data-level="27.4.2" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#summarizing-output-from-a-lot-of-different-files-with-multiqc"><i class="fa fa-check"></i><b>27.4.2</b> Summarizing output from a lot of different files with MultiQC</a></li>
</ul></li>
<li class="chapter" data-level="27.5" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step-1b.-trimming-the-reads"><i class="fa fa-check"></i><b>27.5</b> Step 1b. Trimming the Reads</a>
<ul>
<li class="chapter" data-level="27.5.1" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#cycling-over-all-the-files-and-trimming-them"><i class="fa fa-check"></i><b>27.5.1</b> Cycling over all the files and trimming them</a></li>
<li class="chapter" data-level="27.5.2" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#multiqc-your-trimmomatics-results"><i class="fa fa-check"></i><b>27.5.2</b> Multiqc your trimmomatics results</a></li>
</ul></li>
<li class="chapter" data-level="27.6" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step-2.-mapping-each-unit-to-the-genome"><i class="fa fa-check"></i><b>27.6</b> Step 2. Mapping each unit to the genome</a>
<ul>
<li class="chapter" data-level="27.6.1" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#indexing-the-reference-genome-1"><i class="fa fa-check"></i><b>27.6.1</b> Indexing the reference genome</a></li>
<li class="chapter" data-level="27.6.2" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#a-quick-mapping-run-to-stdout"><i class="fa fa-check"></i><b>27.6.2</b> A quick mapping run to stdout</a></li>
</ul></li>
<li class="chapter" data-level="27.7" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step-3.-merging-bams-from-the-same-sample-library"><i class="fa fa-check"></i><b>27.7</b> Step 3. Merging bams from the same sample + library</a></li>
<li class="chapter" data-level="27.8" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step-4.-marking-duplicates"><i class="fa fa-check"></i><b>27.8</b> Step 4. Marking duplicates</a></li>
<li class="chapter" data-level="27.9" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step-5.-merging-dup-marked-bams-from-the-same-sample"><i class="fa fa-check"></i><b>27.9</b> Step 5. Merging dup-marked bams from the same sample</a></li>
<li class="chapter" data-level="27.10" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step-6.-creating-a-gvcf-file-for-each-sample"><i class="fa fa-check"></i><b>27.10</b> Step 6. Creating a gVCF file for each sample</a></li>
<li class="chapter" data-level="27.11" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step-7.-importing-the-gvcf-files-into-a-genomicsdatabase"><i class="fa fa-check"></i><b>27.11</b> Step 7. Importing the gVCF files into a GenomicsDatabase</a></li>
<li class="chapter" data-level="27.12" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step8.-calling-variants-and-creating-vcf-files-from-the-genomicsdatabase"><i class="fa fa-check"></i><b>27.12</b> Step8. Calling variants and creating VCF files from the GenomicsDatabase</a></li>
<li class="chapter" data-level="27.13" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step9.-filtering-variants"><i class="fa fa-check"></i><b>27.13</b> Step9. Filtering variants</a></li>
</ul></li>
<li class="part"><span><b>IV Part IV: Analysis of Big Variant Data</b></span></li>
<li class="chapter" data-level="28" data-path="bioinformatic-analysis-on-variant-data.html"><a href="bioinformatic-analysis-on-variant-data.html"><i class="fa fa-check"></i><b>28</b> Bioinformatic analysis on variant data</a></li>
<li class="part"><span><b>V Part V: Population Genomics</b></span></li>
<li class="chapter" data-level="29" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html"><i class="fa fa-check"></i><b>29</b> Topics in pop gen</a>
<ul>
<li class="chapter" data-level="29.1" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#coalescent"><i class="fa fa-check"></i><b>29.1</b> Coalescent</a></li>
<li class="chapter" data-level="29.2" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#measures-of-genetic-diversity-and-such"><i class="fa fa-check"></i><b>29.2</b> Measures of genetic diversity and such</a></li>
<li class="chapter" data-level="29.3" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#demographic-inference-with-partial-a-partial-i-and-moments"><i class="fa fa-check"></i><b>29.3</b> Demographic inference with <span class="math inline">\(\partial a \partial i\)</span> and <em>moments</em></a></li>
<li class="chapter" data-level="29.4" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#balls-in-boxes"><i class="fa fa-check"></i><b>29.4</b> Balls in Boxes</a></li>
<li class="chapter" data-level="29.5" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#some-landscape-genetics"><i class="fa fa-check"></i><b>29.5</b> Some landscape genetics</a></li>
<li class="chapter" data-level="29.6" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#relationship-inference"><i class="fa fa-check"></i><b>29.6</b> Relationship Inference</a></li>
<li class="chapter" data-level="29.7" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#tests-for-selection"><i class="fa fa-check"></i><b>29.7</b> Tests for Selection</a></li>
<li class="chapter" data-level="29.8" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#multivariate-associations-gea-etc."><i class="fa fa-check"></i><b>29.8</b> Multivariate Associations, GEA, etc.</a></li>
<li class="chapter" data-level="29.9" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#estimating-heritability-in-the-wild"><i class="fa fa-check"></i><b>29.9</b> Estimating heritability in the wild</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references-1.html"><a href="references-1.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://bookdown.org" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="an-example-full-workflow-from-sequences-to-variants" class="section level1" number="27">
<h1><span class="header-section-number">Chapter 27</span> An example full workflow from sequences to variants</h1>
<p>The purpose of this chapter is to present an accessible example workflow
showing all the steps of taking raw sequences in FASTQ files through to
variants in VCF files. It assumes that some sort of reference genome is
available already (be it a chromosome-level assembly or just a scaffold-level
assembly). The workflow shown here is based on the
GATK “Best Practices,” with some modifications to make the workflow
reasonable for non-model organisms that lack many of the genomic resources
expected by the authors of the GATK “Best Practices” (like files of
known variation in the species, etc.)</p>
<p>As an example data set we have created a
small version of a typical NGS data set.
The original data set included 384 Chinook salmon from the Yukon River
basin, sequenced across 4 lanes of an Illumina NovaSeq machine. To create
the subsetted data set, we extracted just the reads that map to a small part
(about 10 megabases from each of four chromosomes, and then a variety of
shorter scaffolds)
of the latest (as of 2022) reference genome (<a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_018296145.1">Otsh_v2.0</a>).
for Chinook salmon. And we included data from only 8 of the 384
individual fish.</p>
<p>Included in the data set is a reference genome that has been downsized similarly,
to include only those portions of the genome from which we extracted the
read from those 8 fish.</p>
<p>Finally, the reads from different individuals were massaged to make it appear that some
individuals were sequenced on multiple lanes or flow cells and so that some individuals
underwent several different library preps. This is an important feature to ensure that
we learn how to properly deal with sequencing of individual libraries across multiple
lanes when marking PCR duplicates.</p>
<p>The data set is available publicly as a .tar file at this link:
<a href="https://drive.google.com/file/d/1LMK-DCkH1RKFAWTR2OKEJ_K9VOjJIZ1b/view?usp=sharing">https://drive.google.com/file/d/1LMK-DCkH1RKFAWTR2OKEJ_K9VOjJIZ1b/view?usp=sharing</a>. If you view that site on a browser you can choose to download the file.</p>
<p>Downloading the file to a Linux server can be done using <code>wget</code> on the command line. Because the file
is large, a little bit of extra trickery is required to download it. This (rather complex) command line
should work:</p>
<div class="sourceCode" id="cb656"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb656-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb656-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">--load-cookies</span> /tmp/cookies.txt <span class="st">&quot;https://docs.google.com/uc?export=download&amp;confirm=</span><span class="va">$(</span><span class="fu">wget</span> <span class="at">--quiet</span> <span class="at">--save-cookies</span> /tmp/cookies.txt <span class="at">--keep-session-cookies</span> <span class="at">--no-check-certificate</span> <span class="st">&#39;https://docs.google.com/uc?export=download&amp;id=1LMK-DCkH1RKFAWTR2OKEJ_K9VOjJIZ1b&#39;</span> <span class="at">-O-</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="at">-rn</span> <span class="st">&#39;s/.*confirm=([0-9A-Za-z_]+).*/\1\n/p&#39;</span><span class="va">)</span><span class="st">&amp;id=1LMK-DCkH1RKFAWTR2OKEJ_K9VOjJIZ1b&quot;</span> <span class="at">-O</span> non-model-wgs-example-data.tar <span class="kw">&amp;&amp;</span> <span class="fu">rm</span> <span class="at">-rf</span> /tmp/cookies.txt</span></code></pre></div>
<p>Once the file <code>non-model-wgs-example-data.tar</code> is downloaded, that .tar archive can be untarred with</p>
<div class="sourceCode" id="cb657"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb657-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb657-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xvf</span> non-model-wgs-example-data.tar</span></code></pre></div>
<p>That will create the directory <code>non-model-wgs-example-data</code>, which contains all the files.</p>
<p>The contents of that directory look like this (in a <code>tree</code> view):</p>
<pre><code>non-model-wgs-example-data
├── README.Rmd
├── README.nb.html
├── data
│   └── messy_names
│       └── fastqs
│           ├── T199967_T2087_HY75HDSX2_L001_R1_001.fastq.gz
│           ├── T199967_T2087_HY75HDSX2_L001_R2_001.fastq.gz
│           ├── T199968_T2087_HY75HDSX2_L001_R1_001.fastq.gz
│           ├── T199968_T2087_HY75HDSX2_L001_R2_001.fastq.gz
│           ├── T199968_T2087_HY75HDSX2_L002_R1_001.fastq.gz
│           ├── T199968_T2087_HY75HDSX2_L002_R2_001.fastq.gz
│           ├── T199969_T2087_HTYYCBBXX_L002_R1_001.fastq.gz
│           ├── T199969_T2087_HTYYCBBXX_L002_R2_001.fastq.gz
│           ├── T199969_T2087_HY75HDSX2_L002_R1_001.fastq.gz
│           ├── T199969_T2087_HY75HDSX2_L002_R2_001.fastq.gz
│           ├── T199969_T2087_HY75HDSX2_L003_R1_001.fastq.gz
│           ├── T199969_T2087_HY75HDSX2_L003_R2_001.fastq.gz
│           ├── T199970_T2087_HY75HDSX2_L003_R1_001.fastq.gz
│           ├── T199970_T2087_HY75HDSX2_L003_R2_001.fastq.gz
│           ├── T199970_T2094_HY75HDSX2_L004_R1_001.fastq.gz
│           ├── T199970_T2094_HY75HDSX2_L004_R2_001.fastq.gz
│           ├── T199971_T2087_HY75HDSX2_L002_R1_001.fastq.gz
│           ├── T199971_T2087_HY75HDSX2_L002_R2_001.fastq.gz
│           ├── T199971_T2087_HY75HDSX2_L004_R1_001.fastq.gz
│           ├── T199971_T2087_HY75HDSX2_L004_R2_001.fastq.gz
│           ├── T199971_T2099_HTYYCBBXX_L002_R1_001.fastq.gz
│           ├── T199971_T2099_HTYYCBBXX_L002_R2_001.fastq.gz
│           ├── T199972_T2087_HTYYCBBXX_L003_R1_001.fastq.gz
│           ├── T199972_T2087_HTYYCBBXX_L003_R2_001.fastq.gz
│           ├── T199972_T2087_HY75HDSX2_L001_R1_001.fastq.gz
│           ├── T199972_T2087_HY75HDSX2_L001_R2_001.fastq.gz
│           ├── T199972_T2094_HTYYCBBXX_L004_R1_001.fastq.gz
│           ├── T199972_T2094_HTYYCBBXX_L004_R2_001.fastq.gz
│           ├── T199972_T2094_HY75HDSX2_L002_R1_001.fastq.gz
│           ├── T199972_T2094_HY75HDSX2_L002_R2_001.fastq.gz
│           ├── T199973_T2087_HY75HDSX2_L002_R1_001.fastq.gz
│           ├── T199973_T2087_HY75HDSX2_L002_R2_001.fastq.gz
│           ├── T199973_T2087_HY75HDSX2_L003_R1_001.fastq.gz
│           ├── T199973_T2087_HY75HDSX2_L003_R2_001.fastq.gz
│           ├── T199973_T2094_HY75HDSX2_L002_R1_001.fastq.gz
│           ├── T199973_T2094_HY75HDSX2_L002_R2_001.fastq.gz
│           ├── T199973_T2094_HY75HDSX2_L003_R1_001.fastq.gz
│           ├── T199973_T2094_HY75HDSX2_L003_R2_001.fastq.gz
│           ├── T199974_T2087_HY75HDSX2_L001_R1_001.fastq.gz
│           ├── T199974_T2087_HY75HDSX2_L001_R2_001.fastq.gz
│           ├── T199974_T2094_HY75HDSX2_L001_R1_001.fastq.gz
│           ├── T199974_T2094_HY75HDSX2_L001_R2_001.fastq.gz
│           ├── T199974_T2099_HY75HDSX2_L001_R1_001.fastq.gz
│           └── T199974_T2099_HY75HDSX2_L001_R2_001.fastq.gz
├── fastq
│   ├── s001---1_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199967_T2087_HY75HDSX2_L001_R1_001.fastq.gz
│   ├── s001---1_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199967_T2087_HY75HDSX2_L001_R2_001.fastq.gz
│   ├── s002---1_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199968_T2087_HY75HDSX2_L001_R1_001.fastq.gz
│   ├── s002---1_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199968_T2087_HY75HDSX2_L001_R2_001.fastq.gz
│   ├── s002---2_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199968_T2087_HY75HDSX2_L002_R1_001.fastq.gz
│   ├── s002---2_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199968_T2087_HY75HDSX2_L002_R2_001.fastq.gz
│   ├── s003---1_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199969_T2087_HY75HDSX2_L002_R1_001.fastq.gz
│   ├── s003---1_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199969_T2087_HY75HDSX2_L002_R2_001.fastq.gz
│   ├── s003---2_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199969_T2087_HY75HDSX2_L003_R1_001.fastq.gz
│   ├── s003---2_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199969_T2087_HY75HDSX2_L003_R2_001.fastq.gz
│   ├── s003---3_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199969_T2087_HTYYCBBXX_L002_R1_001.fastq.gz
│   ├── s003---3_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199969_T2087_HTYYCBBXX_L002_R2_001.fastq.gz
│   ├── s004---1_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199970_T2087_HY75HDSX2_L003_R1_001.fastq.gz
│   ├── s004---1_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199970_T2087_HY75HDSX2_L003_R2_001.fastq.gz
│   ├── s004---2_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199970_T2094_HY75HDSX2_L004_R1_001.fastq.gz
│   ├── s004---2_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199970_T2094_HY75HDSX2_L004_R2_001.fastq.gz
│   ├── s005---1_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199971_T2087_HY75HDSX2_L002_R1_001.fastq.gz
│   ├── s005---1_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199971_T2087_HY75HDSX2_L002_R2_001.fastq.gz
│   ├── s005---2_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199971_T2087_HY75HDSX2_L004_R1_001.fastq.gz
│   ├── s005---2_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199971_T2087_HY75HDSX2_L004_R2_001.fastq.gz
│   ├── s005---3_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199971_T2099_HTYYCBBXX_L002_R1_001.fastq.gz
│   ├── s005---3_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199971_T2099_HTYYCBBXX_L002_R2_001.fastq.gz
│   ├── s006---1_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199972_T2087_HY75HDSX2_L001_R1_001.fastq.gz
│   ├── s006---1_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199972_T2087_HY75HDSX2_L001_R2_001.fastq.gz
│   ├── s006---2_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199972_T2087_HTYYCBBXX_L003_R1_001.fastq.gz
│   ├── s006---2_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199972_T2087_HTYYCBBXX_L003_R2_001.fastq.gz
│   ├── s006---3_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199972_T2094_HY75HDSX2_L002_R1_001.fastq.gz
│   ├── s006---3_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199972_T2094_HY75HDSX2_L002_R2_001.fastq.gz
│   ├── s006---4_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199972_T2094_HTYYCBBXX_L004_R1_001.fastq.gz
│   ├── s006---4_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199972_T2094_HTYYCBBXX_L004_R2_001.fastq.gz
│   ├── s007---1_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199973_T2087_HY75HDSX2_L002_R1_001.fastq.gz
│   ├── s007---1_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199973_T2087_HY75HDSX2_L002_R2_001.fastq.gz
│   ├── s007---2_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199973_T2087_HY75HDSX2_L003_R1_001.fastq.gz
│   ├── s007---2_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199973_T2087_HY75HDSX2_L003_R2_001.fastq.gz
│   ├── s007---3_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199973_T2094_HY75HDSX2_L002_R1_001.fastq.gz
│   ├── s007---3_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199973_T2094_HY75HDSX2_L002_R2_001.fastq.gz
│   ├── s007---4_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199973_T2094_HY75HDSX2_L003_R1_001.fastq.gz
│   ├── s007---4_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199973_T2094_HY75HDSX2_L003_R2_001.fastq.gz
│   ├── s008---1_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199974_T2087_HY75HDSX2_L001_R1_001.fastq.gz
│   ├── s008---1_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199974_T2087_HY75HDSX2_L001_R2_001.fastq.gz
│   ├── s008---2_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199974_T2094_HY75HDSX2_L001_R1_001.fastq.gz
│   ├── s008---2_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199974_T2094_HY75HDSX2_L001_R2_001.fastq.gz
│   ├── s008---3_R1.fq.gz -&gt; ../data/messy_names/fastqs/T199974_T2099_HY75HDSX2_L001_R1_001.fastq.gz
│   └── s008---3_R2.fq.gz -&gt; ../data/messy_names/fastqs/T199974_T2099_HY75HDSX2_L001_R2_001.fastq.gz
├── resources
│   └── genome.fasta
├── samples.tsv
└── units.tsv</code></pre>
<ul>
<li>The README.* files give some background on how this data set was whittled out of the
larger, complete data set.</li>
<li>The directory <code>data/messy_names/fastqs</code>, contains all the FASTQ files as they might appear
after downloading from the sequencing center. Characteristic of such files, they all
have rather daunting names.</li>
<li>The directory <code>fastq</code> at the top level of the directory contains symbolic
links to the big files with messy names. These are there simply to provide
students a simpler variety of names of files to deal with when doing the initial
steps of a sequencing exercise. But, ultimately, we will explore ways of directly
addressing the files with more complex names.</li>
<li>The file <code>samples.tsv</code> just contains all the unique “tidy” sample names
(like <code>s001</code>, <code>s002</code>, etc.) in a single column named <code>sample</code>.</li>
<li>The file <code>units.tsv</code> contains crucial information about all the different pairs
of FASTQ files. Each pair is considered a separate “unit” of one of the samples.<br />
Some of the information in the file pertains to which flowcell and lane the sample-unit was
sequenced on, as well as which library the sample-unit was prepared in.
All of this relates to the “journey” that was taken by each little piece of DNA
that was sequenced (Section <a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#read-journey">19.1</a>). It is worth having a look at the file.</li>
<li>The file <code>resources/genome.fasta</code> contains the “genome” for this exercise.</li>
</ul>
<p>For information about FASTQ and FASTA files see Sections <a href="bioinformatic-file-formats.html#fastq">17.2</a> and <a href="bioinformatic-file-formats.html#fasta">17.3</a>.</p>
<div id="an-overview-of-the-whole-workflow" class="section level2" number="27.1">
<h2><span class="header-section-number">27.1</span> An overview of the whole workflow</h2>
We provide here a simple(ish) flow diagram that shows the different steps/jobs in this
workflow that have to be done. The diagram is made from a Snakemake workflow (see Chapter
@ref(#snakemake-chap)) that Eric has implemented for whole genome sequencing in
non-model organisms. The acyclic directed graph (DAG) output from Snakemake has been
condensed with the function,<code>condense_dag()</code>, from one of Eric’s R packages
(<a href="https://github.com/eriqande/SnakemakeDagR">SnakemakeDagR</a>). Each step/job is a separate
box, and the arrows between the boxes show the dependence of output from the parent box
(at the tail end of the arrow) as input for the child box (at the head end of the arrow).
The numbers in the boxes and along the arrows show how many independent instances of each job
must be run (for different samples, units, chromosomes, etc.)
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:full-dag"></span>
<img src="figs/snakemake-dags/full-dag.svg" alt="A DAG showing all the separate steps or jobs (called &quot;rules&quot; in Snakemake terminology) that we will be exploring while working through the example WGS workflow in this chapter." width="100%" />
<p class="caption">
FIGURE 27.1: A DAG showing all the separate steps or jobs (called “rules” in Snakemake terminology) that we will be exploring while working through the example WGS workflow in this chapter.
</p>
</div>
<p>Burn this DAG into your memory. We will return to versions of it in each section below
to give an indication of which part of the workflow we are tackling.</p>
</div>
<div id="lets-have-a-look-at-the-reference-genome" class="section level2" number="27.2">
<h2><span class="header-section-number">27.2</span> Let’s have a look at the reference genome</h2>
<p>A number of people in the class had problems installing samtools on SUMMIT last
week. The reason for this appears to be that the conda package management system
migrated to a new version of the SSH library, but samtools looks for an older version
of the libcrypto library (perhaps because it uses some of the hashing functions
from that library). This was a known problem when conda upgraded its SSH dependencies,
but seems to have resurfaced with later versions of samtools. However, it can be fixed by
installing version 1.11 of samtools.</p>
<div id="first-off-make-sure-that-you-have-added-mamba-to-your-base-environment" class="section level3" number="27.2.1">
<h3><span class="header-section-number">27.2.1</span> First off, make sure that you have added mamba to your base environment</h3>
<p>I was playing with a lot of things over the weekend, and it was clear that mamba was
orders of magnitude faster than conda for installing environments. (It almost felt
like conda merely stalled on some environments—that wouldn’t make for fun classtime…just
waiting for software to be installed).</p>
<p>So, I am going to require that people use <code>mamba</code> from here on out. First, test to see
if you have mamba in your base conda environment: activate your base conda environment
and see if you have mamba:</p>
<div class="sourceCode" id="cb659"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb659-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb659-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate base</span>
<span id="cb659-2"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb659-2" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span></span></code></pre></div>
<p>If this gives you a screen full of syntax about how to use mamba, like:</p>
<div class="sourceCode" id="cb660"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb660-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-1" aria-hidden="true" tabindex="-1"></a><span class="ex">usage:</span> mamba [-h] [-V] command ...</span>
<span id="cb660-2"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb660-3"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> is a tool for managing and deploying applications, environments and packages.</span>
<span id="cb660-4"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb660-5"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-5" aria-hidden="true" tabindex="-1"></a><span class="ex">Options:</span></span>
<span id="cb660-6"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb660-7"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-7" aria-hidden="true" tabindex="-1"></a><span class="ex">positional</span> arguments:</span>
<span id="cb660-8"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-8" aria-hidden="true" tabindex="-1"></a>  <span class="bu">command</span></span>
<span id="cb660-9"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-9" aria-hidden="true" tabindex="-1"></a>    <span class="ex">clean</span>        Remove unused packages and caches.</span>
<span id="cb660-10"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-10" aria-hidden="true" tabindex="-1"></a>    <span class="ex">compare</span>      Compare packages between conda environments.</span>
<span id="cb660-11"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-11" aria-hidden="true" tabindex="-1"></a>    <span class="ex">config</span>       Modify configuration values in .condarc. This is modeled after the git config command. Writes to the user .condarc file <span class="er">(</span><span class="ex">/home/eriq@colostate.edu/.condarc</span><span class="kw">)</span> <span class="ex">by</span></span>
<span id="cb660-12"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-12" aria-hidden="true" tabindex="-1"></a>                 <span class="ex">default.</span></span>
<span id="cb660-13"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-13" aria-hidden="true" tabindex="-1"></a>    <span class="ex">create</span>       Create a new conda environment from a list of specified packages.</span>
<span id="cb660-14"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-14" aria-hidden="true" tabindex="-1"></a>    <span class="bu">help</span>         Displays a list of available conda commands and their help strings.</span>
<span id="cb660-15"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-15" aria-hidden="true" tabindex="-1"></a>    <span class="ex">info</span>         Display information about current conda install.</span>
<span id="cb660-16"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-16" aria-hidden="true" tabindex="-1"></a>    <span class="ex">init</span>         Initialize conda for shell interaction. [Experimental]</span>
<span id="cb660-17"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install</span>      Installs a list of packages into a specified conda environment.</span>
<span id="cb660-18"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-18" aria-hidden="true" tabindex="-1"></a>    <span class="ex">list</span>         List linked packages in a conda environment.</span>
<span id="cb660-19"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-19" aria-hidden="true" tabindex="-1"></a>    <span class="ex">package</span>      Low-level conda package utility. <span class="er">(</span><span class="ex">EXPERIMENTAL</span><span class="kw">)</span></span>
<span id="cb660-20"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-20" aria-hidden="true" tabindex="-1"></a>    <span class="ex">remove</span>       Remove a list of packages from a specified conda environment.</span>
<span id="cb660-21"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-21" aria-hidden="true" tabindex="-1"></a>    <span class="ex">uninstall</span>    Alias for conda remove.</span>
<span id="cb660-22"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-22" aria-hidden="true" tabindex="-1"></a>    <span class="ex">run</span>          Run an executable in a conda environment. [Experimental]</span>
<span id="cb660-23"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-23" aria-hidden="true" tabindex="-1"></a>    <span class="ex">search</span>       Search for packages and display associated information. The input is a MatchSpec, a query language for conda packages. See examples below.</span>
<span id="cb660-24"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-24" aria-hidden="true" tabindex="-1"></a>    <span class="ex">update</span>       Updates conda packages to the latest compatible version.</span>
<span id="cb660-25"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-25" aria-hidden="true" tabindex="-1"></a>    <span class="ex">upgrade</span>      Alias for conda update.</span>
<span id="cb660-26"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb660-26" aria-hidden="true" tabindex="-1"></a>    <span class="ex">repoquery</span>    Query repositories using mamba.</span></code></pre></div>
<p>then you have mamba.</p>
<p>If, instead, it tells you it doesn’t know anything about the mamba command, then you
need to install it. Make sure that you install it into your base conda environment,
using this command:</p>
<div class="sourceCode" id="cb661"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb661-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb661-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install mamba <span class="at">-n</span> base <span class="at">-c</span> conda-forge</span></code></pre></div>
<p>Now, if you need to get a working samtools (version 1.11) into an environment called
samtools, you can do this:</p>
<div class="sourceCode" id="cb662"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb662-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb662-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> create <span class="at">-n</span> samtools <span class="at">-c</span> bioconda samtools=1.11</span></code></pre></div>
<p>To use samtools you then must activate the environment</p>
<div class="sourceCode" id="cb663"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb663-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb663-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate samtools</span></code></pre></div>
</div>
<div id="indexing-the-reference-genome" class="section level3" number="27.2.2">
<h3><span class="header-section-number">27.2.2</span> Indexing the reference genome</h3>
<p>Samtools has wonderful subcommand, <code>faidx</code> that indexes a fasta file.
We can index our example reference genome by running this command from the
top level of our <code>non-model-wgs-example-data</code> directory.</p>
<div class="sourceCode" id="cb664"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb664-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb664-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> faidx resources/genome.fasta</span></code></pre></div>
<p>This creates the file <code>resources/genome.fasta.fai</code>. The extension
<code>.fai</code> stands for “faidx index.” Look at <code>resources/genome.fasta.fai</code>
using the <code>less</code> viewer. The first few lines look like:</p>
<pre><code>  CM031199.1      14000000        122     60      61
CM031200.1      12000000        14233578        60      61
CM031201.1      10000000        26433700        60      61
CM031202.1      6000000 36600489        60      61
JAEPEU010001976.1       276     42700570        60      61
JAEPEU010009262.1       333     42700932        60      61
JAEPEU010007942.1       367     42701352        60      61
JAEPEU010006675.1       416     42701807        60      61
JAEPEU010003074.1       440     42702311        60      61</code></pre>
<p>Each line gives information about a differet sequence in the referenc fasta
file, listed in the order that the sequences appear in the fasta file.</p>
<p>The five columns of each line hold this information:</p>
<ul>
<li><strong>NAME</strong>: Name of this reference sequence</li>
<li><strong>LENGTH</strong>: Total length of this reference sequence, in bases</li>
<li><strong>OFFSET</strong>: Offset in the FASTA/FASTQ file of this sequence’s first base</li>
<li><strong>LINEBASES</strong>: The number of bases on each line</li>
<li><strong>LINEWIDTH</strong>: The number of bytes in each line, including the newline</li>
</ul>
<p>This information can be used to rapidly find the parts of the file where
any position in the genome is found, for example, if you wanted to find
the 100th base in sequence <code>JAEPEU010001976.1</code> you can see that you have
to go <code>42700570</code> bytes from the start of the file to get to base 1 of
`<code>JAEPEU010001976.1</code>, and then you would have to go forward in the file
for 99 bases. 59 bases forward would get you to the last base on the first
line, then you have an additional byte to eat up (the line ending) and then
40 more bytes would put you at position 100. So, you could reach
the desired point in the fasta file by just going
<code>42700570 + 59 + 1 + 40</code> bytes forward in the file. This is the sort
of thing that can be done very quickly by a computer.</p>
<p>Fortunately you don’t have to do this arithmetic in your head to
find sequences at certain places in the reference genome. Rather, once
you have indexed your FASTA file, you can use <code>samtools faidx</code> with any
number of optional <em>genomic coordinates</em>, to extract those seqences.</p>
<p>A genomic coordinate (or “range”) of a sequence has the form:</p>
<pre><code>chromosome_or_sequence_name:start_position-end_positions</code></pre>
<p>For example:</p>
<pre><code>CM031199.1:50151-50300</code></pre>
<p>gives the genomic coordinates (or the “genome address”) of the 150 base pair
sequence whose first base is at position 50,151 in the genome, and whose
last is at position 50,300. When using <code>samtools faidx</code> bases are numbered
so that the first one is <code>1</code>. This is called <em>base-1</em> indexing. Other
ways of addressing regions of genomes (like those used in BED files, use a
different <em>base-0</em> scheme.)</p>
<p>You can think of <em>base-1</em> schemes as those that give the address of each
base starting from 1. To understand a base-0 scheme in that light, it makes
sense to think about addressing each SNP according to individual “pickets” that
separate each base. (Probably should draw a picture, but you can see it
at this <a href="https://genome-blog.soe.ucsc.edu/blog/2016/12/12/the-ucsc-genome-browser-coordinate-counting-systems/">UCSC genome browser blog site</a>).</p>
<p>At any rate, if you want to get the 500 bases starting from 1,000,001 on the
first two chromosomes in our example reference genome you would do:</p>
<div class="sourceCode" id="cb668"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb668-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb668-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> faidx resources/genome.fasta CM031199.1:1000001-1000500 CM031200.1:1000001-1000500</span></code></pre></div>
<p>Note that if you find it difficult to keep track of all those zeroes, you are
free to put commas into the numbers anywhere that is helpful, for example:</p>
<div class="sourceCode" id="cb669"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb669-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb669-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> faidx resources/genome.fasta CM031199.1:1,000,001-1,000,500 CM031200.1:1,000,001-1,000,500</span></code></pre></div>
<p>So, if you ever need to extract a certain sequence from a genome, this is one way
you can do it!</p>
</div>
</div>
<div id="a-word-on-organizing-workflows" class="section level2" number="27.3">
<h2><span class="header-section-number">27.3</span> A word on organizing workflows</h2>
<p>Now, before we start rolling along with our example bioinformatic workflow,
let us take a moment and appreciate how clean and tidy our <code>non-model-wgs-example-data</code>
directory is. If I do ls on it I see this:</p>
<div class="sourceCode" id="cb670"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb670-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb670-1" aria-hidden="true" tabindex="-1"></a><span class="ex">%</span> ls</span>
<span id="cb670-2"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb670-2" aria-hidden="true" tabindex="-1"></a><span class="ex">data</span>  fastq  README.nb.html  README.Rmd  resources  samples.tsv  units.tsv</span></code></pre></div>
<p>That is all right! There are only seven files or directories there and it
is not a mess.</p>
<p>As we proceed with a bioinformatic pipeline/workflow, it will behoove us to
keep things super nice and clean. In fact I am going to recommend that nearly
everything we do in our workflow should go inside a single directory called
<code>results</code>.</p>
<p>In general, your workflow will involve several types of large files. Some of those
are just things like genomes (.fasta files), genomic features (.gff) files, or known
variants, etc. In general these sorts of things that you might download from a
repository like GenBank should be placed into the <code>resources</code> directory, possibly
in their own subdirectory within <code>resources</code>. The other types of files are those that
get produced when you run bioinformatic software to analyze your data—for example, trimmed
versions of your fastqs after you have trimmed off the low quality bases, etc. These latter
types of files should all go into separate folders within a directory called <code>results</code>.
The directory they go in should reflect what type of step in your bioinformatic process
produced them, for example: <code>trimmed</code>, <code>mapped</code>, <code>variants</code>, etc.</p>
<p>Additionally, there should be a directory, <code>results/logs</code> where you should redirect <code>stderr</code> for
each sample that is run in a certain process. All this will become clear as we start
working through our example.</p>
</div>
<div id="step-1a.-quality-checking-the-reads" class="section level2" number="27.4">
<h2><span class="header-section-number">27.4</span> Step 1a. Quality Checking the Reads</h2>
<p>The first thing we are going to do is Quality Check our fastq files. The
goal of this step is to get a sense for how good (or bad) the data that we got
back from the sequencing center is, in terms of number of reads per sample,
base quality scores, distribution of A, C, G, and T bases, the number of missing
bases, and the occurrence of overrepresented sequences (that could indicate
contamination or untrimmed Illumina adapters, etc.)</p>
The following figure shows where this fits into our overall workflow. We will be
using FASTQC to profile each FASTQ file, and then we will summarize all those results
with MULTIQC.
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:fastqc-multiqc-dag"></span>
<img src="figs/snakemake-dags/fastqc-multiqc.svg" alt="FASTQC steps and the associated MULTIQC step of this section shown in color." width="100%" />
<p class="caption">
FIGURE 27.2: FASTQC steps and the associated MULTIQC step of this section shown in color.
</p>
</div>
<p>There is a standard
piece of Java software called FASTQC that does a reasonably comprehensive job of this.
We will focus on running it interactively for now (we will talk about dispatching
jobs via SLURM later). So, on the cluster, we will want to get a shell on a compute
node. If you want to, on SUMMIT, you can get an interactive shell with:</p>
<div class="sourceCode" id="cb671"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb671-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb671-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sinteractive</span></span></code></pre></div>
<p>But, to be honest, <code>sinteractive</code> is a somewhat old-school shell function that uses
screen to connect you to the interactive shell. This causes some problems for TMUX. So,
I find a better way to get a shell from the oversubscriped interactive partition on
SUMMIT to be:</p>
<div class="sourceCode" id="cb672"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb672-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb672-1" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--partition</span><span class="op">=</span>shas-interactive <span class="at">-t</span> 2:00:00 <span class="at">--pty</span> /bin/bash</span></code></pre></div>
<p>When you do this, and then reconnect to such a session with TMUX, your scrollback shouldn’t
be totally messed up (Thanks Caitlin for reminding me of this <code>sinteractive</code> atrocity!).
I’ve checked with the sys-admins on SUMMIT and they say this is a perfectly acceptable
way of getting an interactive shell on SUMMIT. Also, the <code>-t 2:00:00</code> says
“give me this interactive shell for 2 hours,” which will give us more than
enough time to get through a whole class session.</p>
<p>The first thing we will need to do is get FASTQC using mamba. Put it in its own
environment named <code>fastqc</code>.</p>
<div class="sourceCode" id="cb673"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb673-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb673-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> create <span class="at">-n</span> fastqc <span class="at">-c</span> bioconda fastqc</span></code></pre></div>
<p>Hooray for the wicked speed of mamba!</p>
<p>Once this is set up, we can get the instructions for using fastqc like this:</p>
<div class="sourceCode" id="cb674"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb674-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb674-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate fastqc</span>
<span id="cb674-2"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb674-2" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">--help</span></span></code></pre></div>
<p>That is a fairly lengthy statement about the syntax of this program,
so I won’t copy it into the handbook, here, But read through it and you
shall find that it looks like if we want to just
put all the fastq files on the command line
then <code>fastqc</code> ought to gobble them right up and put the output in files that
are named intelligently. That sounds pretty convenient. Let’s test it out
on a couple of the smallest files that we have (about 15 Mb each):</p>
<div class="sourceCode" id="cb675"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb675-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb675-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make some directories to put the results in</span></span>
<span id="cb675-2"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb675-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/qc/fastqc results/logs/fastqc</span>
<span id="cb675-3"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb675-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb675-4"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb675-4" aria-hidden="true" tabindex="-1"></a><span class="co"># now, try running fastqc on two small files:</span></span>
<span id="cb675-5"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb675-5" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-o</span> results/qc/fastqc <span class="at">--noextract</span> fastq/s006---1_R1.fq.gz fastq/s006---1_R2.fq.gz</span></code></pre></div>
<p>Note that fastqc spits out some output to tell you about the progress that it is making.
It turns out that is going to both <em>stderr</em> (the percent complete lines) and to
<em>stdout</em> (the “complete” lines). So, we probably should capture both of those streams
in the future.</p>
<p>When it is done, look at the output in <code>results/qc/fastqc</code>. See that there are files with
pretty self-explanatory names. For each FASTQ file there is an <code>.html</code> file that has
a report file (with lots of pretty pictures) that can be opened in a web browser. Note that
you have to copy those to your laptop if you want to see them. You could, for example,
do that with Globus. The other output for each FASTQ file is a <code>.zip</code> file of data
that we will use later.</p>
<p>That didn’t take too long. Let’s try the whole schmeer! We can do that easily with
globbing. Note the redirection of <em>stdout</em> to a log file and the <code>2&gt;&amp;1</code> which means, “send
<em>stderr</em> to the same place <em>stdout</em> is going to.”</p>
<div class="sourceCode" id="cb676"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb676-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb676-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-o</span> results/qc/fastqc <span class="at">--noextract</span> fastq/<span class="pp">*</span>.fq.gz  <span class="op">&gt;</span> results/logs/fastqc/fastqc.log <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span></code></pre></div>
<p>Note that if you want to see how much progress is being made on the file
you can look at the bottom of the log file at <code>results/logs/fastqc/fastqc.log</code>
using the <code>tail</code> command in another shell (opened for example with tmux, before you started
the last command). In fact, you can do one better than that by using <code>tail -f</code>.</p>
<p>Since I had another shell going in tmux, I can do this with it:</p>
<div class="sourceCode" id="cb677"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb677-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb677-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span> <span class="at">-f</span> results/logs/fastqc/fastqc.log</span></code></pre></div>
<p>this continuously spits new output to the <code>fastqc.log</code> file to <em>stdout</em>.
The whole thing should take about 5 to 10 minutes on an interactive shell
on SUMMIT.</p>
<div id="a-little-looking-at-the-htmls" class="section level3" number="27.4.1">
<h3><span class="header-section-number">27.4.1</span> A Little Looking at the HTMLs</h3>
<p>Let’s just go through a few of the different pieces of information
in here. It is relatively self-explanatory. I do want to call out the
adapter content pane, and show that it found some instances of the
Illumina Universal Adapter. These are refer to Illumina’s TruSeq Version 3
adapters. So, we see that some of the reads include sequence from the
adapters. Cool. We will come back to that.</p>
</div>
<div id="summarizing-output-from-a-lot-of-different-files-with-multiqc" class="section level3" number="27.4.2">
<h3><span class="header-section-number">27.4.2</span> Summarizing output from a lot of different files with MultiQC</h3>
<p>As you probably noticed if you tried to open all the html reports for every
FASTQ file, it is a bit laborious to look at every report…and this is just
for a small number of files. Imagine if you had to do that for 768 different
FASTQ files!</p>
<p>A good solution for summarizing such outputs is the Java-based program, MultiQC.
This program offers an insightful system for summarizing log files and outputs from
a wide range of different bioinformatics programs. It is capable of summarizing much
more than just fastqc files, but that is how we will first use it.</p>
<p>Let’s get it and put it in a conda environment called multiqc.</p>
<div class="sourceCode" id="cb678"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb678-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb678-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> create <span class="at">-n</span> multiqc <span class="at">-c</span> bioconda multiqc</span></code></pre></div>
<p>Once that is done (and if you use <code>conda</code> instead of <code>mamba</code> it might
never be done) you can activate the environment and get some information
about how to use multiqc with:</p>
<div class="sourceCode" id="cb679"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb679-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb679-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate multiqc</span>
<span id="cb679-2"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb679-2" aria-hidden="true" tabindex="-1"></a><span class="ex">multiqc</span> <span class="at">--help</span> </span></code></pre></div>
<p>Reading the program syntax, it seems like we should be able to
point it to the <code>results/qc/fastqc</code> directory and tell it to put its output
into a <code>results/qc/multiqc</code> directory like this:</p>
<div class="sourceCode" id="cb680"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb680-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb680-1" aria-hidden="true" tabindex="-1"></a><span class="ex">multiqc</span> <span class="at">-v</span> <span class="at">-o</span> results/qc/multiqc results/qc/fastqc</span></code></pre></div>
<p>Note that the command line above is using the <code>-v</code> option, which gives us
<em>verbose</em> output. This is kind of fun because it shows us all the different kinds of
output file types that MultiQC can parse—a whole boatload of them. All we have
now our fastqc files. But we can see the summary report by downloading
<code>results/qc/multiqc/multiqc_report.html</code> and opening it in your browser. Super cool!</p>
</div>
</div>
<div id="step-1b.-trimming-the-reads" class="section level2" number="27.5">
<h2><span class="header-section-number">27.5</span> Step 1b. Trimming the Reads</h2>
<p>The fastqc reports showed that most of the sequences we have are of pretty good quality,
but, as expected, the quality drops off a bit later in the read. Also, we saw that there
were some adapter sequences stuck into 1 to 3% of the reads. We can use the
tool TrimmoMatic to:</p>
<ol style="list-style-type: decimal">
<li>Trim off parts of reads that are of low quality.</li>
<li>Chuck out entire reads that are of low quality.</li>
<li>Remove adapter sequences that are improperly inserted into our reads.</li>
</ol>
This step corresponds to the colored node in our now-familiar workflow DAG:
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:trimmo-dag"></span>
<img src="figs/snakemake-dags/trimmo.svg" alt="Running Trimmomatic to clean up our sequencing reads by removing adapter cruft and discarding parts of the reads with low quality." width="100%" />
<p class="caption">
FIGURE 27.3: Running Trimmomatic to clean up our sequencing reads by removing adapter cruft and discarding parts of the reads with low quality.
</p>
</div>
<p>To use Trimmomatic, we will have to get it. If you don’t have it in a conda environment
yet, you know the drill:</p>
<div class="sourceCode" id="cb681"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb681-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb681-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> create <span class="at">-n</span> trimmo <span class="at">-c</span> bioconda trimmomatic</span></code></pre></div>
<p>Once that is installed you can get a terse and somewhat cryptic message about how
to run the program like this:</p>
<div class="sourceCode" id="cb682"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb682-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb682-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate trimmo</span>
<span id="cb682-2"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb682-2" aria-hidden="true" tabindex="-1"></a><span class="ex">trimmomatic</span></span></code></pre></div>
<p>Which tells us this:</p>
<pre><code>Usage:
       PE [-version] \
          [-threads &lt;threads&gt;] \
          [-phred33|-phred64] \
          [-trimlog &lt;trimLogFile&gt;] \
          [-summary &lt;statsSummaryFile&gt;] \
          [-quiet] \
          [-validatePairs] \
          [-basein &lt;inputBase&gt; | &lt;inputFile1&gt; &lt;inputFile2&gt;] \
          [-baseout &lt;outputBase&gt; | &lt;outputFile1P&gt; &lt;outputFile1U&gt; &lt;outputFile2P&gt; &lt;outputFile2U&gt;] \
          &lt;trimmer1&gt;...
   or:
       SE [-version] \
          [-threads &lt;threads&gt;] \
          [-phred33|-phred64] \
          [-trimlog &lt;trimLogFile&gt;] \
          [-summary &lt;statsSummaryFile&gt;] \
          [-quiet] \
          &lt;inputFile&gt; &lt;outputFile&gt; \
          &lt;trimmer1&gt;...
   or:
       -version</code></pre>
<p>Where I have added backslash-line-endings to make it a little easier to read.</p>
<p>Aha! We can run this dude in paired end mode (PE) or in single end mode (SE).
Since we have paired-end data, we will run it in paired-end mode.</p>
<p>Full details of the syntax and the methodology can be found in the PDF Trimmomatic
manual at <a href="http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf">http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf</a>. This link
seems to not work very well sometimes…</p>
<p>We will be running it on a single thread for now so the really relevant syntax is:</p>
<div class="sourceCode" id="cb684"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb684-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb684-1" aria-hidden="true" tabindex="-1"></a>       <span class="ex">PE</span> <span class="dt">\</span></span>
<span id="cb684-2"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb684-2" aria-hidden="true" tabindex="-1"></a>          [-basein <span class="op">&lt;</span>inputBase<span class="op">&gt;</span> <span class="kw">|</span> <span class="op">&lt;</span>inputFile1<span class="op">&gt;</span> <span class="op">&lt;</span>inputFile2<span class="op">&gt;</span>] <span class="dt">\</span></span>
<span id="cb684-3"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb684-3" aria-hidden="true" tabindex="-1"></a>          [-baseout <span class="op">&lt;</span>outputBase<span class="op">&gt;</span> <span class="kw">|</span> <span class="op">&lt;</span>outputFile1P<span class="op">&gt;</span> <span class="op">&lt;</span>outputFile1U<span class="op">&gt;</span> <span class="op">&lt;</span>outputFile2P<span class="op">&gt;</span> <span class="op">&lt;</span>outputFile2U<span class="op">&gt;</span>] <span class="dt">\</span></span>
<span id="cb684-4"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb684-4" aria-hidden="true" tabindex="-1"></a>          <span class="op">&lt;</span>trimmer1<span class="op">&gt;</span>...</span></code></pre></div>
<p>This syntax message signifies that you can either give it a base file prefix
for the input and output files, like <code>-basein fastq/s001-1  -baseout results/trimmed/s001-1</code>,
or you can explicitly give it the names of the input and output files. I tend to think that
the latter is a better way to do it (always good to be explicit).</p>
<p>So, for example, with our tidy named files in the fastq directory, the first
two lines there might look like:</p>
<div class="sourceCode" id="cb685"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb685-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb685-1" aria-hidden="true" tabindex="-1"></a><span class="ex">trimmomatic</span> PE <span class="dt">\</span></span>
<span id="cb685-2"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb685-2" aria-hidden="true" tabindex="-1"></a>  fastq/s001---1_R1.fq.gz  fastq/s001---1_R2.fq.gz   <span class="dt">\</span></span>
<span id="cb685-3"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb685-3" aria-hidden="true" tabindex="-1"></a>  results/trimmed/s001---1_R1.fq.gz  results/trimmed/s001---1_R1_unpaired.fq.gz  <span class="dt">\</span></span>
<span id="cb685-4"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb685-4" aria-hidden="true" tabindex="-1"></a>  results/trimmed/s001---1_R2.fq.gz  results/trimmed/s001---1_R2_unpaired.fq.gz </span></code></pre></div>
<p>So, the two input files are read 1 and read2, and the four output files are in this order:</p>
<ol style="list-style-type: decimal">
<li>Read 1 successfully trimmed.</li>
<li>Read 1 that is no longer paired (because, for example, its mate was tossed out)</li>
<li>Read 2 successfully trimmed</li>
<li>Read 2 that is no longer paired.</li>
</ol>
<p>The reads that you will use downstream, for mapping, will be the successfully
trimmed ones.</p>
<p>Now, the more complex part of the command line syntax is the innocouous looking
<code>&lt;trimmer1&gt;...</code>. That seemingly simple word belies the opportunity to specify all the
different types of read trimming modules and their parameters
(and the order in which they will be applied) that you want to use. Once again, the
PDF manual is the place to go, but a good place to start will be:</p>
<div class="sourceCode" id="cb686"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb686-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get path to trimmomatic to find the path to the adapter files.</span></span>
<span id="cb686-2"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-2" aria-hidden="true" tabindex="-1"></a><span class="co"># It seems to look for them in a standard /usr/local/share location,</span></span>
<span id="cb686-3"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-3" aria-hidden="true" tabindex="-1"></a><span class="co"># which is not where they are if conda installed.</span></span>
<span id="cb686-4"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-4" aria-hidden="true" tabindex="-1"></a><span class="va">trim_path=$(</span><span class="fu">which</span> trimmomatic<span class="va">)</span></span>
<span id="cb686-5"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-5" aria-hidden="true" tabindex="-1"></a><span class="va">trim_dir=$(</span><span class="fu">dirname</span> <span class="va">$trim_path)</span></span>
<span id="cb686-6"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-6" aria-hidden="true" tabindex="-1"></a><span class="va">adapter_fasta=$trim_dir</span>/../share/trimmomatic/adapters/TruSeq3-PE-2.fa</span>
<span id="cb686-7"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb686-8"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-8" aria-hidden="true" tabindex="-1"></a><span class="co"># make a results/trimmed directory to put the outputs into</span></span>
<span id="cb686-9"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-9" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/trimmed</span>
<span id="cb686-10"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb686-11"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-11" aria-hidden="true" tabindex="-1"></a><span class="ex">trimmomatic</span> PE <span class="dt">\</span></span>
<span id="cb686-12"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-12" aria-hidden="true" tabindex="-1"></a>  fastq/s001---1_R1.fq.gz  fastq/s001---1_R2.fq.gz   <span class="dt">\</span></span>
<span id="cb686-13"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-13" aria-hidden="true" tabindex="-1"></a>  results/trimmed/s001---1_R1.fq.gz  results/trimmed/s001---1_R1_unpaired.fq.gz <span class="dt">\</span></span>
<span id="cb686-14"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-14" aria-hidden="true" tabindex="-1"></a>  results/trimmed/s001---1_R2.fq.gz  results/trimmed/s001---1_R2_unpaired.fq.gz  <span class="dt">\</span></span>
<span id="cb686-15"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-15" aria-hidden="true" tabindex="-1"></a>  ILLUMINACLIP:<span class="va">${adapter_fasta}</span>:2:30:10 <span class="dt">\</span></span>
<span id="cb686-16"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-16" aria-hidden="true" tabindex="-1"></a>  LEADING:3 <span class="dt">\</span></span>
<span id="cb686-17"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-17" aria-hidden="true" tabindex="-1"></a>  TRAILING:3 <span class="dt">\</span></span>
<span id="cb686-18"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-18" aria-hidden="true" tabindex="-1"></a>  SLIDINGWINDOW:4:15 <span class="dt">\</span></span>
<span id="cb686-19"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb686-19" aria-hidden="true" tabindex="-1"></a>  MINLEN:36</span></code></pre></div>
<p>You can go ahead and evaluate those commands and see what happens when
trimmomatic runs. And while it is running, we can try to unpack those last few lines:</p>
<ul>
<li><code>ILLUMINACLIP:${adapter_fasta}:2:30:10</code>: This means remove Illumina adapters that match the
sequences in the fasta file <code>TruSeq3-PE-2.fa</code> that is included with the Trimmomatic distribution.
That file looks like this:</li>
</ul>
<pre><code>&gt;PrefixPE/1
TACACTCTTTCCCTACACGACGCTCTTCCGATCT
&gt;PrefixPE/2
GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT
&gt;PE1
TACACTCTTTCCCTACACGACGCTCTTCCGATCT
&gt;PE1_rc
AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA
&gt;PE2
GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT
&gt;PE2_rc
AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC</code></pre>
<p>Those are the sequences (and the reverse complements) for the TruSeq version3 adapters.
It is pretty fun to look around in your fastq files to find there sequences…(We could
play around with that if you want). This particular fasta file is set up to allow
Trimmomatic to find spurious adapter sequences when they arrived in your reads by
a few different things (see the Palindrome clipping section in the manual).</p>
<ul>
<li><code>LEADING:3</code>: Remove bases at the beginning of the read that have a
Phred-scaled quality score less than 3.</li>
<li><code>TRAILING:3</code>: Remove bases at the end of the read that have a
Phred-scaled quality score less than 3.</li>
<li><code>SLIDINGWINDOW:4:15</code>: Remove any bases within a sliding window of 4 bases that have an
average Phred-scaled quality score of less than 15.</li>
<li><code>MINLEN:36</code>: Toss out any reads that are only 36 bp long after the bad bases have been trimmed.</li>
</ul>
<div id="cycling-over-all-the-files-and-trimming-them" class="section level3" number="27.5.1">
<h3><span class="header-section-number">27.5.1</span> Cycling over all the files and trimming them</h3>
<p>Ultimately, we are going to have to cycle over the fastq files in some way, so,
note that we could have written the above command line by defining a shell variable for the
file prefix. Lets clean this whole thing up with some shell variables, using
<code>s002---1</code> as an example</p>
<div class="sourceCode" id="cb688"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb688-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-1" aria-hidden="true" tabindex="-1"></a><span class="co"># here is a variable to store all the trimmer directives:</span></span>
<span id="cb688-2"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-2" aria-hidden="true" tabindex="-1"></a><span class="va">TRIMMER=</span><span class="st">&quot;ILLUMINACLIP:</span><span class="va">${adapter_fasta}</span><span class="st">:2:30:10 </span><span class="dt">\</span></span>
<span id="cb688-3"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-3" aria-hidden="true" tabindex="-1"></a><span class="st">  LEADING:3 </span><span class="dt">\</span></span>
<span id="cb688-4"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-4" aria-hidden="true" tabindex="-1"></a><span class="st">  TRAILING:3 </span><span class="dt">\</span></span>
<span id="cb688-5"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-5" aria-hidden="true" tabindex="-1"></a><span class="st">  SLIDINGWINDOW:4:15 </span><span class="dt">\</span></span>
<span id="cb688-6"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-6" aria-hidden="true" tabindex="-1"></a><span class="st">  MINLEN:36&quot;</span></span>
<span id="cb688-7"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb688-8"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb688-9"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-9" aria-hidden="true" tabindex="-1"></a><span class="co"># define a file prefix for s002---1</span></span>
<span id="cb688-10"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-10" aria-hidden="true" tabindex="-1"></a><span class="va">FP=</span>s002---1</span>
<span id="cb688-11"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb688-12"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-12" aria-hidden="true" tabindex="-1"></a><span class="co"># now get the names of all the input files and output files</span></span>
<span id="cb688-13"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-13" aria-hidden="true" tabindex="-1"></a><span class="co"># and store them in a shell variable.  Not the fun shell-fu here.</span></span>
<span id="cb688-14"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-14" aria-hidden="true" tabindex="-1"></a><span class="va">INFILES=$(</span><span class="bu">echo</span> fastq/<span class="va">${FP}</span>_<span class="dt">{R1</span><span class="op">,</span><span class="dt">R2}</span>.fq.gz<span class="va">)</span></span>
<span id="cb688-15"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-15" aria-hidden="true" tabindex="-1"></a><span class="va">OUTFILES=$(</span><span class="bu">echo</span> results/trimmed/<span class="va">${FP}</span>_<span class="dt">{R1</span><span class="op">,</span><span class="dt">R2}{</span><span class="op">,</span><span class="dt">_unpaired}</span>.fq.gz<span class="va">)</span></span>
<span id="cb688-16"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb688-17"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-17" aria-hidden="true" tabindex="-1"></a><span class="co"># make a Log file too</span></span>
<span id="cb688-18"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-18" aria-hidden="true" tabindex="-1"></a><span class="va">LOG=</span>results/logs/trimmomatic/<span class="va">$FP</span>.log</span>
<span id="cb688-19"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb688-20"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-20" aria-hidden="true" tabindex="-1"></a><span class="co"># you might want to try &quot;echo $INFILES&quot; and &quot;echo $OUTFILES&quot; </span></span>
<span id="cb688-21"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-21" aria-hidden="true" tabindex="-1"></a><span class="co"># and &quot;echo $LOG&quot; to see what those look like.</span></span>
<span id="cb688-22"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb688-23"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-23" aria-hidden="true" tabindex="-1"></a><span class="co"># We need to make sure that we have directories to put all those</span></span>
<span id="cb688-24"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-24" aria-hidden="true" tabindex="-1"></a><span class="co"># output and log files into.  We can do that with:</span></span>
<span id="cb688-25"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-25" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$(</span><span class="fu">dirname</span> <span class="va">$LOG</span> <span class="va">$OUTFILES)</span></span>
<span id="cb688-26"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb688-27"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-27" aria-hidden="true" tabindex="-1"></a><span class="co"># now that we have defined those variables, we can write our </span></span>
<span id="cb688-28"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-28" aria-hidden="true" tabindex="-1"></a><span class="co"># command line like this:</span></span>
<span id="cb688-29"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb688-30"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb688-30" aria-hidden="true" tabindex="-1"></a><span class="ex">trimmomatic</span> PE <span class="va">$INFILES</span> <span class="va">$OUTFILES</span> <span class="va">$TRIMMER</span> <span class="op">&gt;</span> <span class="va">$LOG</span> <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span></code></pre></div>
<p>Once we have figured out that whole collection of command lines,
and have verified that it works for
our “trial” value of <code>FP</code> (i.e, <code>s002---1</code>), it is time to cycle over all the
different possible values of <code>FP</code>.</p>
<p>We can use globbing and the <code>basename</code> utility to get all the file prefixes. Check this
out:</p>
<div class="sourceCode" id="cb689"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb689-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb689-1" aria-hidden="true" tabindex="-1"></a><span class="va">FILE_PREFIXES=$(</span><span class="fu">basename</span> <span class="at">-a</span> <span class="at">-s</span> _R1.fq.gz fastq/<span class="pp">*</span>R1.fq.gz<span class="va">)</span></span></code></pre></div>
<p>Evaluate that and then run <code>echo $FILE_PREFIXES</code> to see what you have.</p>
<p>With that, we can then define a for loop to run over all 22 pairs of files:</p>
<div class="sourceCode" id="cb690"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb690-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This stays outside of the for loop:</span></span>
<span id="cb690-2"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-2" aria-hidden="true" tabindex="-1"></a><span class="va">TRIMMER=</span><span class="st">&quot;ILLUMINACLIP:</span><span class="va">${adapter_fasta}</span><span class="st">:2:30:10 </span><span class="dt">\</span></span>
<span id="cb690-3"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-3" aria-hidden="true" tabindex="-1"></a><span class="st">  LEADING:3 </span><span class="dt">\</span></span>
<span id="cb690-4"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-4" aria-hidden="true" tabindex="-1"></a><span class="st">  TRAILING:3 </span><span class="dt">\</span></span>
<span id="cb690-5"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-5" aria-hidden="true" tabindex="-1"></a><span class="st">  SLIDINGWINDOW:4:15 </span><span class="dt">\</span></span>
<span id="cb690-6"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-6" aria-hidden="true" tabindex="-1"></a><span class="st">  MINLEN:36&quot;</span></span>
<span id="cb690-7"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb690-8"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-8" aria-hidden="true" tabindex="-1"></a><span class="co"># get list of file prefixes</span></span>
<span id="cb690-9"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-9" aria-hidden="true" tabindex="-1"></a><span class="va">FILE_PREFIXES=$(</span><span class="fu">basename</span> <span class="at">-a</span> <span class="at">-s</span> _R1.fq.gz fastq/<span class="pp">*</span>R1.fq.gz<span class="va">)</span></span>
<span id="cb690-10"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb690-11"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> FP <span class="kw">in</span> <span class="va">$FILE_PREFIXES</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb690-12"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-12" aria-hidden="true" tabindex="-1"></a>  <span class="va">INFILES=$(</span><span class="bu">echo</span> fastq/<span class="va">${FP}</span>_<span class="dt">{R1</span><span class="op">,</span><span class="dt">R2}</span>.fq.gz<span class="va">)</span></span>
<span id="cb690-13"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-13" aria-hidden="true" tabindex="-1"></a>  <span class="va">OUTFILES=$(</span><span class="bu">echo</span> results/trimmed/<span class="va">${FP}</span>_<span class="dt">{R1</span><span class="op">,</span><span class="dt">R2}{</span><span class="op">,</span><span class="dt">_unpaired}</span>.fq.gz<span class="va">)</span></span>
<span id="cb690-14"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-14" aria-hidden="true" tabindex="-1"></a>  <span class="va">LOG=</span>results/logs/trimmomatic/<span class="va">$FP</span>.log</span>
<span id="cb690-15"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb690-16"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$(</span><span class="fu">dirname</span> <span class="va">$LOG</span> <span class="va">$OUTFILES)</span> <span class="co"># this could be done outside the for loop</span></span>
<span id="cb690-17"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb690-18"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-18" aria-hidden="true" tabindex="-1"></a>  <span class="ex">trimmomatic</span> PE <span class="va">$INFILES</span> <span class="va">$OUTFILES</span> <span class="va">$TRIMMER</span> <span class="op">&gt;</span> <span class="va">$LOG</span> <span class="dv">2</span><span class="op">&gt;&amp;</span><span class="dv">1</span></span>
<span id="cb690-19"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb690-20"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb690-20" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code></pre></div>
<p>You can run those, and they will likely take 15 to 20 minutues to get through.
To finish them all, you would typically want to be in a tmux session or something similar
so that the job is not killed when you logout.</p>
<p>Note that for most big and long jobs you wouldn’t run this in an interactive shell,
but, rather, you would run it through SLURM’s <code>sbatch</code>. We will be covering that next
week when we start to talk about mapping. But for now, just want to do some
interactive computing to get a flavor of how one can write and test scripts
on an interactive shell.</p>
</div>
<div id="multiqc-your-trimmomatics-results" class="section level3" number="27.5.2">
<h3><span class="header-section-number">27.5.2</span> Multiqc your trimmomatics results</h3>
<p>It turns out that Multiqc, which so nicely gobbled up and summarized our
fastqc results, also knows how to make nice summaries of the log (stderr/stdout) output
of Trimmomatic. So, once our Trimmomatic loop is done, we can run Multiqc and
direct it to <em>both</em> the directory holding the fastqc output <em>and</em> the directory holding the
Trimmomatic output, so as to get an HTML report that includes both. Witness:</p>
<div class="sourceCode" id="cb691"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb691-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb691-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate multiqc</span>
<span id="cb691-2"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb691-2" aria-hidden="true" tabindex="-1"></a><span class="ex">multiqc</span> <span class="at">-v</span> <span class="at">-f</span> <span class="at">-o</span> results/qc/multiqc results/qc/fastqc results/logs/trimmomatic</span></code></pre></div>
<p>Note that the above command includes the <code>-f</code> option to <em>force</em> it to overwrite existing
output.</p>
<p>As before, you get the results in <code>results/logs/trimmomatic/multiqc_report.html</code>, but you
have to get them on your laptop to view them easily.</p>
</div>
</div>
<div id="step-2.-mapping-each-unit-to-the-genome" class="section level2" number="27.6">
<h2><span class="header-section-number">27.6</span> Step 2. Mapping each unit to the genome</h2>
<p>Mapping or aligning reads to a genome is one of the more computationally
demanding steps in a typical WGS workflow. However, I find it somewhat
more straightforward (less prone to errors and failures and bugs in the software)
than some of the later computational steps (like creating a gVCF file from
HaplotypeCaller).</p>
<p>There are several different software programs available for alignment. I am a fan
of the <code>bwa mem</code> algorithm, and it is commonly used in many production pipelines,
so that is what we will be using.</p>
<p>The output of <code>bwa mem</code> can often be manipulated as it comes streaming out of the program
onto <em>stdout</em>. The standard tool for manipulation such streams is <code>samtools</code>. For this
reason, when we define a conda environment for mapping steps, we will typically
include both the <code>bwa</code> program and the <code>samtools</code> program in it. So, let’s make a
conda environment called <code>bwasam</code> that has both tools in it:</p>
<div class="sourceCode" id="cb692"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb692-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb692-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> create <span class="at">-n</span> bwasam <span class="at">-c</span> bioconda bwa=0.7.17 samtools=1.11</span></code></pre></div>
<p>Once we have that environment, we can activate it and then look at the
syntax for <code>bwa mem</code> by typing <code>bwa mem</code>:</p>
<div class="sourceCode" id="cb693"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb693-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb693-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate bwasam</span>
<span id="cb693-2"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb693-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bwa</span> mem</span></code></pre></div>
<p>The output of that is quite long. Thankfully, the default program values
are often a good place to start, and typically work well. So, for our purposes,
the most germane things to note are the main syntax statement:</p>
<div class="sourceCode" id="cb694"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb694-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb694-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Usage:</span> bwa mem [options] <span class="op">&lt;</span>idxbase<span class="op">&gt;</span> <span class="op">&lt;</span>in1.fq<span class="op">&gt;</span> [in2.fq]</span></code></pre></div>
<p>This means you can give it options, and then you have to pass it
the path to the genome (which is also the prefix used for the bwa indexes,
as we shall see). And then you pass it a fastq file with read 1, and if you have
paired end reads you pass it the corresponding read 2 fastq file.</p>
<p>The other option that we will use is described as:</p>
<pre><code>-R STR        read group header line such as &#39;@RG\tID:foo\tSM:bar&#39; [null]</code></pre>
<p>This lets us add information to the output file about the origin of the
collection of reads in the fastq files (like what sample and library prep they
are from, what type of platform they were sequenced on, and the machine
and lane).</p>
<p>To start out with, we simply want to make a SAM file that we can use
for further explorations of what a SAM/BAM file is, and what
it means for DNA to align to a reference genome. So, make sure that
you are in your <code>non-model-wgs-example-data</code> directory, you have an
interactive compute shell, and that you have the <code>bwasam</code> conda environment
activated. Here we go!</p>
<div id="indexing-the-reference-genome-1" class="section level3" number="27.6.1">
<h3><span class="header-section-number">27.6.1</span> Indexing the reference genome</h3>
<p>The reference genome we are working with is in <code>resources/genome.fasta</code>.
In order to quickly and efficiently find places in the genome to which reads align,
it is necessary to <em>index</em> the reference genome. The bwa subcommand <code>index</code> will
index the reference genome for <code>bwa mem</code> to use.</p>
<p>To index the genome we do:</p>
<div class="sourceCode" id="cb696"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb696-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb696-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bwa</span> index resources/genome.fasta</span></code></pre></div>
<p>This takes about 40 seconds and creates the files:</p>
<pre><code>resources/genome.fasta.amb  resources/genome.fasta.bwt  resources/genome.fasta.sa
resources/genome.fasta.ann  resources/genome.fasta.pac</code></pre>
<p>The particulars of these files are not too important. Just know that they
are required for <code>bwa mem</code> to work. Note that the <code>&lt;idxbase&gt;</code> in <code>bwa mem</code>’s
syntax statement is just prefix for these 5 files: <code>resources/genome.fasta</code>.</p>
</div>
<div id="a-quick-mapping-run-to-stdout" class="section level3" number="27.6.2">
<h3><span class="header-section-number">27.6.2</span> A quick mapping run to stdout</h3>
<p>Now that the genome is indexed let’s go ahead and map the first pair of files
that we trimmed:</p>
<pre><code>results/trimmed/s001---1_R1.fq.gz  results/trimmed/s001---1_R2.fq.gz</code></pre>
<p>and we will just let the output go streaming to our screen:</p>
<div class="sourceCode" id="cb699"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb699-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb699-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bwa</span> mem resources/genome.fasta <span class="dt">\</span></span>
<span id="cb699-2"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb699-2" aria-hidden="true" tabindex="-1"></a>   results/trimmed/s001---1_R1.fq.gz  <span class="dt">\</span></span>
<span id="cb699-3"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb699-3" aria-hidden="true" tabindex="-1"></a>   results/trimmed/s001---1_R2.fq.gz</span></code></pre></div>
<p>You should see a whole lot of lines starting with @ come zooming by first,
then you will see some things that look like progress messages. Eventually,
a bunch of sequences start zooming by. You can hit cntrl-C to stop that.</p>
<p>OK! That was fun. Now, we are actually going to redirect that standard output
into a SAM file, but before we do that, we are going to want to add the <code>-R</code> option
to define the read group for the sequences in this file (we will talk a lot more
about read groups in a moment).</p>
<p>So,</p>
<div class="sourceCode" id="cb700"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb700-1"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb700-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/sam  <span class="co"># a directory in which to put the output from stdout</span></span>
<span id="cb700-2"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb700-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/logs/bwa  <span class="co"># a directory in which to put the output from stderr</span></span>
<span id="cb700-3"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb700-3" aria-hidden="true" tabindex="-1"></a><span class="ex">bwa</span> mem <span class="dt">\</span></span>
<span id="cb700-4"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb700-4" aria-hidden="true" tabindex="-1"></a>-R <span class="st">&#39;@RG\tID:s001_T199967_Lib-1_HY75HDSX2_1_AAGACCGT+CAATCGAC\tSM:T199967\tPL:ILLUMINA\tLB:Lib-1\tPU:HY75HDSX2.1.AAGACCGT+CAATCGAC&#39;</span> <span class="dt">\</span></span>
<span id="cb700-5"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb700-5" aria-hidden="true" tabindex="-1"></a>resources/genome.fasta <span class="dt">\</span></span>
<span id="cb700-6"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb700-6" aria-hidden="true" tabindex="-1"></a>results/trimmed/s001---1_R1.fq.gz  results/trimmed/s001---1_R2.fq.gz <span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb700-7"><a href="an-example-full-workflow-from-sequences-to-variants.html#cb700-7" aria-hidden="true" tabindex="-1"></a>results/sam/s001---1.sam <span class="dv">2</span><span class="op">&gt;</span> results/logs/bwa/s001---1.log</span></code></pre></div>
<p>While that is running (and it may run for a pretty good long while) let’s just note that the read-group string in there is a horrifying
bunch of textual gobbledygook. The information that goes into it is in the
file <code>units.tsv</code>. More on read groups later.</p>
<p>For now, let’s start exploring how sequences can align, and the meaning of all the
lines in the SAM file. This is documented in Section <a href="bioinformatic-file-formats.html#sambamfiles">17.4</a>.</p>
</div>
</div>
<div id="step-3.-merging-bams-from-the-same-sample-library" class="section level2" number="27.7">
<h2><span class="header-section-number">27.7</span> Step 3. Merging bams from the same sample + library</h2>
</div>
<div id="step-4.-marking-duplicates" class="section level2" number="27.8">
<h2><span class="header-section-number">27.8</span> Step 4. Marking duplicates</h2>
</div>
<div id="step-5.-merging-dup-marked-bams-from-the-same-sample" class="section level2" number="27.9">
<h2><span class="header-section-number">27.9</span> Step 5. Merging dup-marked bams from the same sample</h2>
</div>
<div id="step-6.-creating-a-gvcf-file-for-each-sample" class="section level2" number="27.10">
<h2><span class="header-section-number">27.10</span> Step 6. Creating a gVCF file for each sample</h2>
</div>
<div id="step-7.-importing-the-gvcf-files-into-a-genomicsdatabase" class="section level2" number="27.11">
<h2><span class="header-section-number">27.11</span> Step 7. Importing the gVCF files into a GenomicsDatabase</h2>
</div>
<div id="step8.-calling-variants-and-creating-vcf-files-from-the-genomicsdatabase" class="section level2" number="27.12">
<h2><span class="header-section-number">27.12</span> Step8. Calling variants and creating VCF files from the GenomicsDatabase</h2>
</div>
<div id="step9.-filtering-variants" class="section level2" number="27.13">
<h2><span class="header-section-number">27.13</span> Step9. Filtering variants</h2>

</div>
</div>



            </section>

          </div>
        </div>
      </div>
<a href="whole-genome-alignment-strategies.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="bioinformatic-analysis-on-variant-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook/js/app.min.js"></script>
<script src="libs/gitbook/js/clipboard.min.js"></script>
<script src="libs/gitbook/js/plugin-search.js"></script>
<script src="libs/gitbook/js/plugin-sharing.js"></script>
<script src="libs/gitbook/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook/js/plugin-bookdown.js"></script>
<script src="libs/gitbook/js/jquery.highlight.js"></script>
<script src="libs/gitbook/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/eriqande/eca-bioinf-handbook/edit/master/3.12-an-example-full-WGS-workflow.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/eriqande/eca-bioinf-handbook/commits/master/3.12-an-example-full-WGS-workflow.Rmd",
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["eca-bioinf-handbook.pdf", "eca-bioinf-handbook.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
