<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 22 Basic Handling of VCF files | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics</title>
  <meta name="description" content="A book example for a Chapman &amp; Hall book." />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 22 Basic Handling of VCF files | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A book example for a Chapman &amp; Hall book." />
  <meta name="github-repo" content="yihui/bookdown-crc" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 22 Basic Handling of VCF files | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics" />
  
  <meta name="twitter:description" content="A book example for a Chapman &amp; Hall book." />
  

<meta name="author" content="Eric C. Anderson" />


<meta name="date" content="2022-05-04" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="boneyard-1.html"/>
<link rel="next" href="bioinformatics-for-rad-seq-data-with-and-without-a-reference-genome.html"/>
<script src="libs/header-attrs/header-attrs.js"></script>
<script src="libs/jquery/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook/css/style.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ECA's Bioinformatics Handbook</a></li>

<li class="divider"></li>
<li><a href="index.html#preface">Preface<span></span></a></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction<span></span></a></li>
<li class="chapter" data-level="2" data-path="erics-notes-of-what-he-might-do.html"><a href="erics-notes-of-what-he-might-do.html"><i class="fa fa-check"></i><b>2</b> Eric’s Notes of what he might do<span></span></a>
<ul>
<li class="chapter" data-level="2.1" data-path="erics-notes-of-what-he-might-do.html"><a href="erics-notes-of-what-he-might-do.html#table-of-topics"><i class="fa fa-check"></i><b>2.1</b> Table of topics<span></span></a></li>
</ul></li>
<li class="part"><span><b>I Part I: Essential Computing Skills<span></span></b></span></li>
<li class="chapter" data-level="3" data-path="overview-of-essential-computing-skills.html"><a href="overview-of-essential-computing-skills.html"><i class="fa fa-check"></i><b>3</b> Overview of Essential Computing Skills<span></span></a></li>
<li class="chapter" data-level="4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html"><i class="fa fa-check"></i><b>4</b> Essential Unix/Linux Terminal Knowledge<span></span></a>
<ul>
<li class="chapter" data-level="4.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#getting-a-bash-shell-on-your-system"><i class="fa fa-check"></i><b>4.1</b> Getting a bash shell on your system<span></span></a></li>
<li class="chapter" data-level="4.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#navigating-the-unix-filesystem"><i class="fa fa-check"></i><b>4.2</b> Navigating the Unix filesystem<span></span></a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#changing-the-working-directory-with-cd"><i class="fa fa-check"></i><b>4.2.1</b> Changing the working directory with <code>cd</code><span></span></a></li>
<li class="chapter" data-level="4.2.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#comm-prompt"><i class="fa fa-check"></i><b>4.2.2</b> Updating your command prompt<span></span></a></li>
<li class="chapter" data-level="4.2.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#tab-completion-for-paths"><i class="fa fa-check"></i><b>4.2.3</b> TAB-completion for paths<span></span></a></li>
<li class="chapter" data-level="4.2.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#listing-the-contents-of-a-directory-with-ls"><i class="fa fa-check"></i><b>4.2.4</b> Listing the contents of a directory with <code>ls</code><span></span></a></li>
<li class="chapter" data-level="4.2.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#globbing"><i class="fa fa-check"></i><b>4.2.5</b> Globbing<span></span></a></li>
<li class="chapter" data-level="4.2.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#what-makes-a-good-file-name"><i class="fa fa-check"></i><b>4.2.6</b> What makes a good file-name?<span></span></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#the-anatomy-of-a-unix-command"><i class="fa fa-check"></i><b>4.3</b> The anatomy of a Unix command<span></span></a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#anatomy-command"><i class="fa fa-check"></i><b>4.3.1</b> The <code>command</code><span></span></a></li>
<li class="chapter" data-level="4.3.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#the-options"><i class="fa fa-check"></i><b>4.3.2</b> The <em>options</em><span></span></a></li>
<li class="chapter" data-level="4.3.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#arguments"><i class="fa fa-check"></i><b>4.3.3</b> Arguments<span></span></a></li>
<li class="chapter" data-level="4.3.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#getting-information-about-unix-commands"><i class="fa fa-check"></i><b>4.3.4</b> Getting information about Unix commands<span></span></a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#handling-manipulating-and-viewing-files-and-streams"><i class="fa fa-check"></i><b>4.4</b> Handling, Manipulating, and Viewing files and streams<span></span></a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#creating-new-directories"><i class="fa fa-check"></i><b>4.4.1</b> Creating new directories<span></span></a></li>
<li class="chapter" data-level="4.4.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#fundamental-file-handling-commands"><i class="fa fa-check"></i><b>4.4.2</b> Fundamental file-handling commands<span></span></a></li>
<li class="chapter" data-level="4.4.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#viewing-files"><i class="fa fa-check"></i><b>4.4.3</b> “Viewing” Files<span></span></a></li>
<li class="chapter" data-level="4.4.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#redirecting-standard-output-and"><i class="fa fa-check"></i><b>4.4.4</b> Redirecting standard output: <code>&gt;</code> and <code>&gt;&gt;</code><span></span></a></li>
<li class="chapter" data-level="4.4.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#stdin-and"><i class="fa fa-check"></i><b>4.4.5</b> stdin, <code>&lt;</code> and <code>|</code><span></span></a></li>
<li class="chapter" data-level="4.4.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#stderr"><i class="fa fa-check"></i><b>4.4.6</b> stderr<span></span></a></li>
<li class="chapter" data-level="4.4.7" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#symbolic-links"><i class="fa fa-check"></i><b>4.4.7</b> Symbolic links<span></span></a></li>
<li class="chapter" data-level="4.4.8" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#file-perm"><i class="fa fa-check"></i><b>4.4.8</b> File Permissions<span></span></a></li>
<li class="chapter" data-level="4.4.9" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#editing-text-files-at-the-terminal"><i class="fa fa-check"></i><b>4.4.9</b> Editing text files at the terminal<span></span></a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#unix-env"><i class="fa fa-check"></i><b>4.5</b> Customizing your Environment<span></span></a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#appearances-matter"><i class="fa fa-check"></i><b>4.5.1</b> Appearances matter<span></span></a></li>
<li class="chapter" data-level="4.5.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#where-are-my-programscommands-at"><i class="fa fa-check"></i><b>4.5.2</b> Where are my programs/commands at?!<span></span></a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#a-few-more-important-keystrokes"><i class="fa fa-check"></i><b>4.6</b> A Few More Important Keystrokes<span></span></a></li>
<li class="chapter" data-level="4.7" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#a-short-list-of-additional-useful-commands."><i class="fa fa-check"></i><b>4.7</b> A short list of additional useful commands.<span></span></a></li>
<li class="chapter" data-level="4.8" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#two-important-computing-concepts"><i class="fa fa-check"></i><b>4.8</b> Two important computing concepts<span></span></a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#compression"><i class="fa fa-check"></i><b>4.8.1</b> Compression<span></span></a></li>
<li class="chapter" data-level="4.8.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#hashing"><i class="fa fa-check"></i><b>4.8.2</b> Hashing<span></span></a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#unix-quick-study-guide"><i class="fa fa-check"></i><b>4.9</b> Unix: Quick Study Guide<span></span></a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="shell-programming.html"><a href="shell-programming.html"><i class="fa fa-check"></i><b>5</b> Shell programming<span></span></a>
<ul>
<li class="chapter" data-level="5.1" data-path="shell-programming.html"><a href="shell-programming.html#an-example-script"><i class="fa fa-check"></i><b>5.1</b> An example script<span></span></a></li>
<li class="chapter" data-level="5.2" data-path="shell-programming.html"><a href="shell-programming.html#the-structure-of-a-bash-script"><i class="fa fa-check"></i><b>5.2</b> The Structure of a Bash Script<span></span></a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="shell-programming.html"><a href="shell-programming.html#paragraph-before-shebang"><i class="fa fa-check"></i><b>5.2.1</b> A bit more on <code>;</code> and <code>&amp;</code><span></span></a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="shell-programming.html"><a href="shell-programming.html#variables"><i class="fa fa-check"></i><b>5.3</b> Variables<span></span></a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="shell-programming.html"><a href="shell-programming.html#assigning-values-to-variables"><i class="fa fa-check"></i><b>5.3.1</b> Assigning values to variables<span></span></a></li>
<li class="chapter" data-level="5.3.2" data-path="shell-programming.html"><a href="shell-programming.html#accessing-values-from-variables"><i class="fa fa-check"></i><b>5.3.2</b> Accessing values from variables<span></span></a></li>
<li class="chapter" data-level="5.3.3" data-path="shell-programming.html"><a href="shell-programming.html#shell-sub-vars"><i class="fa fa-check"></i><b>5.3.3</b> What does the shell do with the value substituted for a variable?<span></span></a></li>
<li class="chapter" data-level="5.3.4" data-path="shell-programming.html"><a href="shell-programming.html#quotes-and-var-subs"><i class="fa fa-check"></i><b>5.3.4</b> Double and Single Quotation Marks and Variable Substitution<span></span></a></li>
<li class="chapter" data-level="5.3.5" data-path="shell-programming.html"><a href="shell-programming.html#one-useful-fancy-variable-substitution-method"><i class="fa fa-check"></i><b>5.3.5</b> One useful, fancy, variable-substitution method<span></span></a></li>
<li class="chapter" data-level="5.3.6" data-path="shell-programming.html"><a href="shell-programming.html#bash-arithmetic"><i class="fa fa-check"></i><b>5.3.6</b> Integer Arithmetic with Shell Variables<span></span></a></li>
<li class="chapter" data-level="5.3.7" data-path="shell-programming.html"><a href="shell-programming.html#variable-arrays"><i class="fa fa-check"></i><b>5.3.7</b> Variable arrays<span></span></a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="shell-programming.html"><a href="shell-programming.html#bash-sub-result-on-comm-line"><i class="fa fa-check"></i><b>5.4</b> Evaluate a command and substitute the result on the command line<span></span></a></li>
<li class="chapter" data-level="5.5" data-path="shell-programming.html"><a href="shell-programming.html#groupingcollecting-output-from-multiple-commands-commands-and-commands"><i class="fa fa-check"></i><b>5.5</b> Grouping/Collecting output from multiple commands: <code>(commands)</code> and <code>{ commands; }</code><span></span></a></li>
<li class="chapter" data-level="5.6" data-path="shell-programming.html"><a href="shell-programming.html#exit-status"><i class="fa fa-check"></i><b>5.6</b> Exit Status<span></span></a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="shell-programming.html"><a href="shell-programming.html#combinations-of-exit-statuses"><i class="fa fa-check"></i><b>5.6.1</b> Combinations of exit statuses<span></span></a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="shell-programming.html"><a href="shell-programming.html#unix-for-loops"><i class="fa fa-check"></i><b>5.7</b> Loops and repetition<span></span></a></li>
<li class="chapter" data-level="5.8" data-path="shell-programming.html"><a href="shell-programming.html#more-conditional-evaluation-if-then-else-and-friends"><i class="fa fa-check"></i><b>5.8</b> More Conditional Evaluation: <code>if</code>, <code>then</code>, <code>else</code>, and friends<span></span></a></li>
<li class="chapter" data-level="5.9" data-path="shell-programming.html"><a href="shell-programming.html#finallypositional-parameters"><i class="fa fa-check"></i><b>5.9</b> Finally…positional parameters<span></span></a></li>
<li class="chapter" data-level="5.10" data-path="shell-programming.html"><a href="shell-programming.html#basename-and-dirname-two-useful-little-utilities"><i class="fa fa-check"></i><b>5.10</b> <code>basename</code> and <code>dirname</code> two useful little utilities<span></span></a></li>
<li class="chapter" data-level="5.11" data-path="shell-programming.html"><a href="shell-programming.html#bash-functions"><i class="fa fa-check"></i><b>5.11</b> <code>bash</code> functions<span></span></a></li>
<li class="chapter" data-level="5.12" data-path="shell-programming.html"><a href="shell-programming.html#reading-files-line-by-line"><i class="fa fa-check"></i><b>5.12</b> reading files line by line<span></span></a></li>
<li class="chapter" data-level="5.13" data-path="shell-programming.html"><a href="shell-programming.html#further-reading"><i class="fa fa-check"></i><b>5.13</b> Further reading<span></span></a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html"><i class="fa fa-check"></i><b>6</b> Sed, awk, and regular expressions<span></span></a>
<ul>
<li class="chapter" data-level="6.1" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#awk"><i class="fa fa-check"></i><b>6.1</b> awk<span></span></a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#line-cycling-tests-and-actions"><i class="fa fa-check"></i><b>6.1.1</b> Line-cycling, tests and actions<span></span></a></li>
<li class="chapter" data-level="6.1.2" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#column-splitting-fields--f-nf-print-ofs-and-begin"><i class="fa fa-check"></i><b>6.1.2</b> Column splitting, fields, <code>-F</code>, <code>$</code>, <code>NF</code>, <code>print</code>, <code>OFS</code> and <code>BEGIN</code><span></span></a></li>
<li class="chapter" data-level="6.1.3" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#a-brief-introduction-to-regular-expressions"><i class="fa fa-check"></i><b>6.1.3</b> A brief introduction to regular expressions<span></span></a></li>
<li class="chapter" data-level="6.1.4" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#a-variety-of-tests"><i class="fa fa-check"></i><b>6.1.4</b> A variety of tests<span></span></a></li>
<li class="chapter" data-level="6.1.5" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#code-in-the-action-blocks"><i class="fa fa-check"></i><b>6.1.5</b> Code in the action blocks<span></span></a></li>
<li class="chapter" data-level="6.1.6" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#using-awk-to-assign-to-shell-variables"><i class="fa fa-check"></i><b>6.1.6</b> Using <code>awk</code> to assign to shell variables<span></span></a></li>
<li class="chapter" data-level="6.1.7" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#passing-variables-into-awk-with--v"><i class="fa fa-check"></i><b>6.1.7</b> Passing Variables into <code>awk</code> with <code>-v</code><span></span></a></li>
<li class="chapter" data-level="6.1.8" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#writing-awk-scripts-in-files"><i class="fa fa-check"></i><b>6.1.8</b> Writing awk scripts in files<span></span></a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#sed"><i class="fa fa-check"></i><b>6.2</b> sed<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html"><i class="fa fa-check"></i><b>7</b> Working on remote servers<span></span></a>
<ul>
<li class="chapter" data-level="7.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#accessing-remote-computers"><i class="fa fa-check"></i><b>7.1</b> Accessing remote computers<span></span></a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#windows"><i class="fa fa-check"></i><b>7.1.1</b> Windows<span></span></a></li>
<li class="chapter" data-level="7.1.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#hummingbird"><i class="fa fa-check"></i><b>7.1.2</b> Hummingbird<span></span></a></li>
<li class="chapter" data-level="7.1.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#summit"><i class="fa fa-check"></i><b>7.1.3</b> Summit<span></span></a></li>
<li class="chapter" data-level="7.1.4" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#sedna"><i class="fa fa-check"></i><b>7.1.4</b> Sedna<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#transferring-files-to-remote-computers"><i class="fa fa-check"></i><b>7.2</b> Transferring files to remote computers<span></span></a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#sftp-and-several-systems-that-use-it"><i class="fa fa-check"></i><b>7.2.1</b> <code>sftp</code> and several systems that use it<span></span></a></li>
<li class="chapter" data-level="7.2.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#git"><i class="fa fa-check"></i><b>7.2.2</b> git<span></span></a></li>
<li class="chapter" data-level="7.2.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#globus"><i class="fa fa-check"></i><b>7.2.3</b> Globus<span></span></a></li>
<li class="chapter" data-level="7.2.4" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#interfacing-with-the-cloud"><i class="fa fa-check"></i><b>7.2.4</b> Interfacing with “The Cloud”<span></span></a></li>
<li class="chapter" data-level="7.2.5" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#get-seqs"><i class="fa fa-check"></i><b>7.2.5</b> Getting files from a sequencing center<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#editing-files-on-a-remote-server-using-a-local-text-editor"><i class="fa fa-check"></i><b>7.3</b> Editing Files on a Remote Server using a Local Text Editor<span></span></a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#rsubl-ssh-config"><i class="fa fa-check"></i><b>7.3.1</b> Step 1: Set up your SSH config file to automatically apply port forwarding to your connections to your server<span></span></a></li>
<li class="chapter" data-level="7.3.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#download-sublimetext-and-add-the-remotesubl-package-to-it"><i class="fa fa-check"></i><b>7.3.2</b> Download SublimeText and add the RemoteSubl package to it<span></span></a></li>
<li class="chapter" data-level="7.3.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#download-the-rmate-shell-script-to-your-server-and-put-it-on-your-path"><i class="fa fa-check"></i><b>7.3.3</b> Download the <code>rmate</code> shell script to your server and put it on your PATH<span></span></a></li>
<li class="chapter" data-level="7.3.4" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#using-rmate"><i class="fa fa-check"></i><b>7.3.4</b> Using rmate<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#tmux"><i class="fa fa-check"></i><b>7.4</b> <code>tmux</code>: the terminal multiplexer<span></span></a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#an-analogy-for-how-tmux-works"><i class="fa fa-check"></i><b>7.4.1</b> An analogy for how <code>tmux</code> works<span></span></a></li>
<li class="chapter" data-level="7.4.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#first-steps-with-tmux"><i class="fa fa-check"></i><b>7.4.2</b> First steps with <code>tmux</code><span></span></a></li>
<li class="chapter" data-level="7.4.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#further-steps-with-tmux"><i class="fa fa-check"></i><b>7.4.3</b> Further steps with <code>tmux</code><span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#tmux-for-mac-users"><i class="fa fa-check"></i><b>7.5</b> <code>tmux</code> for Mac users<span></span></a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#here-are-more-explicit-step-by-step-instructions-for-setting-this-up-with-summit"><i class="fa fa-check"></i><b>7.5.1</b> Here are more explicit, step-by-step instructions for setting this up with SUMMIT<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#installing-software-on-an-hpcc"><i class="fa fa-check"></i><b>7.6</b> Installing Software on an HPCC<span></span></a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#modules"><i class="fa fa-check"></i><b>7.6.1</b> Modules<span></span></a></li>
<li class="chapter" data-level="7.6.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#miniconda"><i class="fa fa-check"></i><b>7.6.2</b> Miniconda<span></span></a></li>
<li class="chapter" data-level="7.6.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#installing-java-programs"><i class="fa fa-check"></i><b>7.6.3</b> Installing Java Programs<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#vim"><i class="fa fa-check"></i><b>7.7</b> <code>vim</code>: it’s time to get serious with text editing<span></span></a>
<ul>
<li class="chapter" data-level="7.7.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#using-neovim-and-nvim-r-and-tmux-to-use-r-well-on-the-cluster"><i class="fa fa-check"></i><b>7.7.1</b> Using neovim and Nvim-R and tmux to use R well on the cluster<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="chap-HPCC.html"><a href="chap-HPCC.html"><i class="fa fa-check"></i><b>8</b> High Performance Computing Clusters (HPCC’s)<span></span></a>
<ul>
<li class="chapter" data-level="8.1" data-path="chap-HPCC.html"><a href="chap-HPCC.html#an-oversimplified-but-useful-view-of-a-computing-cluster"><i class="fa fa-check"></i><b>8.1</b> An oversimplified, but useful, view of a computing cluster<span></span></a></li>
<li class="chapter" data-level="8.2" data-path="chap-HPCC.html"><a href="chap-HPCC.html#cluster-computing-and-the-job-scheduler"><i class="fa fa-check"></i><b>8.2</b> Cluster computing and the job scheduler<span></span></a></li>
<li class="chapter" data-level="8.3" data-path="chap-HPCC.html"><a href="chap-HPCC.html#slurm-info"><i class="fa fa-check"></i><b>8.3</b> Learning about the resources on your HPCC<span></span></a></li>
<li class="chapter" data-level="8.4" data-path="chap-HPCC.html"><a href="chap-HPCC.html#getting-compute-resources-allocated-to-your-jobs-on-an-hpcc"><i class="fa fa-check"></i><b>8.4</b> Getting compute resources allocated to your jobs on an HPCC<span></span></a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="chap-HPCC.html"><a href="chap-HPCC.html#interactive-sessions"><i class="fa fa-check"></i><b>8.4.1</b> Interactive sessions<span></span></a></li>
<li class="chapter" data-level="8.4.2" data-path="chap-HPCC.html"><a href="chap-HPCC.html#slurm-batch"><i class="fa fa-check"></i><b>8.4.2</b> Batch jobs<span></span></a></li>
<li class="chapter" data-level="8.4.3" data-path="chap-HPCC.html"><a href="chap-HPCC.html#slurm-job-arrays"><i class="fa fa-check"></i><b>8.4.3</b> SLURM Job Arrays<span></span></a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="chap-HPCC.html"><a href="chap-HPCC.html#last-years-stuff.-now-somewhat-defunct"><i class="fa fa-check"></i><b>8.5</b> Last year’s stuff. Now somewhat defunct<span></span></a></li>
<li class="chapter" data-level="8.6" data-path="chap-HPCC.html"><a href="chap-HPCC.html#prepation-interlude-an-in-class-exercise-to-make-sure-everything-is-configured-correctly"><i class="fa fa-check"></i><b>8.6</b> PREPATION INTERLUDE: An in-class exercise to make sure everything is configured correctly<span></span></a></li>
<li class="chapter" data-level="8.7" data-path="chap-HPCC.html"><a href="chap-HPCC.html#more-boneyard"><i class="fa fa-check"></i><b>8.7</b> More Boneyard…<span></span></a></li>
<li class="chapter" data-level="8.8" data-path="chap-HPCC.html"><a href="chap-HPCC.html#the-queue-slurmsgeuge"><i class="fa fa-check"></i><b>8.8</b> The Queue (SLURM/SGE/UGE)<span></span></a></li>
<li class="chapter" data-level="8.9" data-path="chap-HPCC.html"><a href="chap-HPCC.html#modules-package"><i class="fa fa-check"></i><b>8.9</b> Modules package<span></span></a></li>
<li class="chapter" data-level="8.10" data-path="chap-HPCC.html"><a href="chap-HPCC.html#compiling-programs-without-admin-privileges"><i class="fa fa-check"></i><b>8.10</b> Compiling programs without admin privileges<span></span></a></li>
<li class="chapter" data-level="8.11" data-path="chap-HPCC.html"><a href="chap-HPCC.html#job-arrays"><i class="fa fa-check"></i><b>8.11</b> Job arrays<span></span></a></li>
<li class="chapter" data-level="8.12" data-path="chap-HPCC.html"><a href="chap-HPCC.html#writing-stdout-and-stderr-to-files"><i class="fa fa-check"></i><b>8.12</b> Writing stdout and stderr to files<span></span></a></li>
<li class="chapter" data-level="8.13" data-path="chap-HPCC.html"><a href="chap-HPCC.html#breaking-stuff-down"><i class="fa fa-check"></i><b>8.13</b> Breaking stuff down<span></span></a></li>
</ul></li>
<li class="part"><span><b>II Part II: Reproducible Research Strategies<span></span></b></span></li>
<li class="chapter" data-level="9" data-path="introduction-to-reproducible-research.html"><a href="introduction-to-reproducible-research.html"><i class="fa fa-check"></i><b>9</b> Introduction to Reproducible Research<span></span></a></li>
<li class="chapter" data-level="10" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html"><i class="fa fa-check"></i><b>10</b> Rstudio and Project-centered Organization<span></span></a>
<ul>
<li class="chapter" data-level="10.1" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#organizing-big-projects"><i class="fa fa-check"></i><b>10.1</b> Organizing big projects<span></span></a></li>
<li class="chapter" data-level="10.2" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#using-rstudio-in-workflows-with-remote-computers-and-hpccs"><i class="fa fa-check"></i><b>10.2</b> Using RStudio in workflows with remote computers and HPCCs<span></span></a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#keeping-an-rstudio-project-in-sync-with-github"><i class="fa fa-check"></i><b>10.2.1</b> Keeping an RStudio project “in sync” with GitHub<span></span></a></li>
<li class="chapter" data-level="10.2.2" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#evaluating-scripts-line-by-line-on-a-remote-machine-from-within-rstudio"><i class="fa fa-check"></i><b>10.2.2</b> Evaluating scripts line by line on a remote machine from within RStudio<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="version-control.html"><a href="version-control.html"><i class="fa fa-check"></i><b>11</b> Version control<span></span></a>
<ul>
<li class="chapter" data-level="11.1" data-path="version-control.html"><a href="version-control.html#why-use-version-control"><i class="fa fa-check"></i><b>11.1</b> Why use version control?<span></span></a></li>
<li class="chapter" data-level="11.2" data-path="version-control.html"><a href="version-control.html#git-workings"><i class="fa fa-check"></i><b>11.2</b> How git works<span></span></a></li>
<li class="chapter" data-level="11.3" data-path="version-control.html"><a href="version-control.html#git-workflow-patterns"><i class="fa fa-check"></i><b>11.3</b> git workflow patterns<span></span></a></li>
<li class="chapter" data-level="11.4" data-path="version-control.html"><a href="version-control.html#using-git-with-rstudio"><i class="fa fa-check"></i><b>11.4</b> using git with Rstudio<span></span></a></li>
<li class="chapter" data-level="11.5" data-path="version-control.html"><a href="version-control.html#git-on-the-command-line"><i class="fa fa-check"></i><b>11.5</b> git on the command line<span></span></a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="a-fast-furious-overview-of-the-tidyverse.html"><a href="a-fast-furious-overview-of-the-tidyverse.html"><i class="fa fa-check"></i><b>12</b> A fast, furious overview of the tidyverse<span></span></a></li>
<li class="chapter" data-level="13" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html"><i class="fa fa-check"></i><b>13</b> Authoring reproducibly with Rmarkdown<span></span></a>
<ul>
<li class="chapter" data-level="13.1" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#notebooks"><i class="fa fa-check"></i><b>13.1</b> Notebooks<span></span></a></li>
<li class="chapter" data-level="13.2" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#references"><i class="fa fa-check"></i><b>13.2</b> References<span></span></a>
<ul>
<li class="chapter" data-level="13.2.1" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#zotero-and-rmarkdown"><i class="fa fa-check"></i><b>13.2.1</b> Zotero and Rmarkdown<span></span></a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#bookdown"><i class="fa fa-check"></i><b>13.3</b> Bookdown<span></span></a></li>
<li class="chapter" data-level="13.4" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#google-docs"><i class="fa fa-check"></i><b>13.4</b> Google Docs<span></span></a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="snakemake-chap.html"><a href="snakemake-chap.html"><i class="fa fa-check"></i><b>14</b> Managing Workflows with Snakemake<span></span></a>
<ul>
<li class="chapter" data-level="14.1" data-path="snakemake-chap.html"><a href="snakemake-chap.html#a-simple-snakemake-example"><i class="fa fa-check"></i><b>14.1</b> A Simple Snakemake Example<span></span></a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="snakemake-chap.html"><a href="snakemake-chap.html#snakefile-1-a-simple-rule-to-trimmomcawesome-the-files-from-001"><i class="fa fa-check"></i><b>14.1.1</b> Snakefile #1: A simple rule to TrimmoMcAwesome the files from 001<span></span></a></li>
<li class="chapter" data-level="14.1.2" data-path="snakemake-chap.html"><a href="snakemake-chap.html#snakefile-2-a-single-trimmomcawesome-rule-using-wildcards"><i class="fa fa-check"></i><b>14.1.2</b> Snakefile #2: A single TrimmoMcAwesome rule using wildcards<span></span></a></li>
<li class="chapter" data-level="14.1.3" data-path="snakemake-chap.html"><a href="snakemake-chap.html#snakefile-3-add-trim_stupendous-in-there"><i class="fa fa-check"></i><b>14.1.3</b> Snakefile #3: Add <code>trim_stupendous</code> in there<span></span></a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="snakemake-chap.html"><a href="snakemake-chap.html#ugly-complex-file-namesa-job-for-snakemake-input-functions"><i class="fa fa-check"></i><b>14.2</b> Ugly, complex file names…a job for Snakemake input functions!<span></span></a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="snakemake-chap.html"><a href="snakemake-chap.html#playing-with-pandas"><i class="fa fa-check"></i><b>14.2.1</b> Playing with Pandas<span></span></a></li>
<li class="chapter" data-level="14.2.2" data-path="snakemake-chap.html"><a href="snakemake-chap.html#a-quick-sidebar-spoofing-a-snakemake-input-function"><i class="fa fa-check"></i><b>14.2.2</b> A Quick Sidebar: Spoofing a Snakemake input function<span></span></a></li>
<li class="chapter" data-level="14.2.3" data-path="snakemake-chap.html"><a href="snakemake-chap.html#snakefile-4-an-input-function-for-trim_awesome"><i class="fa fa-check"></i><b>14.2.3</b> Snakefile #4: an input function for trim_awesome<span></span></a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="snakemake-chap.html"><a href="snakemake-chap.html#input-functions-from-python-dictionaries"><i class="fa fa-check"></i><b>14.3</b> Input functions from Python Dictionaries<span></span></a></li>
<li class="chapter" data-level="14.4" data-path="snakemake-chap.html"><a href="snakemake-chap.html#generating-lists-of-files-with-expand"><i class="fa fa-check"></i><b>14.4</b> Generating lists of files with expand()<span></span></a></li>
<li class="chapter" data-level="14.5" data-path="snakemake-chap.html"><a href="snakemake-chap.html#a-python-primer-for-using-snakemake"><i class="fa fa-check"></i><b>14.5</b> A Python Primer for Using Snakemake<span></span></a></li>
<li class="chapter" data-level="14.6" data-path="snakemake-chap.html"><a href="snakemake-chap.html#using-snakemake-on-a-computing-cluster"><i class="fa fa-check"></i><b>14.6</b> Using Snakemake on a Computing Cluster<span></span></a>
<ul>
<li class="chapter" data-level="14.6.1" data-path="snakemake-chap.html"><a href="snakemake-chap.html#installing-and-configuring-the-snakemake-slurm-profile"><i class="fa fa-check"></i><b>14.6.1</b> Installing and Configuring the Snakemake SLURM profile<span></span></a></li>
<li class="chapter" data-level="14.6.2" data-path="snakemake-chap.html"><a href="snakemake-chap.html#playing-with-our-new-slurm-profile"><i class="fa fa-check"></i><b>14.6.2</b> Playing with our new SLURM profile<span></span></a></li>
<li class="chapter" data-level="14.6.3" data-path="snakemake-chap.html"><a href="snakemake-chap.html#managing-cluster-resources-with-snakemake"><i class="fa fa-check"></i><b>14.6.3</b> Managing cluster resources with Snakemake<span></span></a></li>
<li class="chapter" data-level="14.6.4" data-path="snakemake-chap.html"><a href="snakemake-chap.html#grouping-short-jobs-to-run-on-a-single-node"><i class="fa fa-check"></i><b>14.6.4</b> Grouping short jobs to run on a single node<span></span></a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Part III: Bioinformatic Analyses<span></span></b></span></li>
<li class="chapter" data-level="15" data-path="overview-of-bioinformatic-analyses.html"><a href="overview-of-bioinformatic-analyses.html"><i class="fa fa-check"></i><b>15</b> Overview of Bioinformatic Analyses<span></span></a></li>
<li class="chapter" data-level="16" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html"><i class="fa fa-check"></i><b>16</b> DNA Sequences and Sequencing<span></span></a>
<ul>
<li class="chapter" data-level="16.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#dna-stuff"><i class="fa fa-check"></i><b>16.1</b> DNA Stuff<span></span></a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#dna-replication-with-dna-polymerase"><i class="fa fa-check"></i><b>16.1.1</b> DNA Replication with DNA Polymerase<span></span></a></li>
<li class="chapter" data-level="16.1.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#the-importance-of-the-3-hydroxyl"><i class="fa fa-check"></i><b>16.1.2</b> The importance of the 3’ hydroxyl…<span></span></a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#sanger-sequencing"><i class="fa fa-check"></i><b>16.2</b> Sanger sequencing<span></span></a></li>
<li class="chapter" data-level="16.3" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#illumina-sequencing-by-synthesis"><i class="fa fa-check"></i><b>16.3</b> Illumina Sequencing by Synthesis<span></span></a></li>
<li class="chapter" data-level="16.4" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#library-prep-protocols"><i class="fa fa-check"></i><b>16.4</b> Library Prep Protocols<span></span></a>
<ul>
<li class="chapter" data-level="16.4.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#wgs"><i class="fa fa-check"></i><b>16.4.1</b> WGS<span></span></a></li>
<li class="chapter" data-level="16.4.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#rad-seq-methods"><i class="fa fa-check"></i><b>16.4.2</b> RAD-Seq methods<span></span></a></li>
<li class="chapter" data-level="16.4.3" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#amplicon-sequencing"><i class="fa fa-check"></i><b>16.4.3</b> Amplicon Sequencing<span></span></a></li>
<li class="chapter" data-level="16.4.4" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#capture-arrays-rapture-etc."><i class="fa fa-check"></i><b>16.4.4</b> Capture arrays, RAPTURE, etc.<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html"><i class="fa fa-check"></i><b>17</b> Bioinformatic file formats<span></span></a>
<ul>
<li class="chapter" data-level="17.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sequences"><i class="fa fa-check"></i><b>17.1</b> Sequences<span></span></a></li>
<li class="chapter" data-level="17.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#fastq"><i class="fa fa-check"></i><b>17.2</b> FASTQ<span></span></a>
<ul>
<li class="chapter" data-level="17.2.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#illumina-ids"><i class="fa fa-check"></i><b>17.2.1</b> Line 1: Illumina identifier lines<span></span></a></li>
<li class="chapter" data-level="17.2.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#bqscores"><i class="fa fa-check"></i><b>17.2.2</b> Line 4: Base quality scores<span></span></a></li>
<li class="chapter" data-level="17.2.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#a-fastq-tidyverse-interlude"><i class="fa fa-check"></i><b>17.2.3</b> A FASTQ ‘tidyverse’ Interlude<span></span></a></li>
<li class="chapter" data-level="17.2.4" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#comparing-read-1-to-read-2"><i class="fa fa-check"></i><b>17.2.4</b> Comparing read 1 to read 2<span></span></a></li>
</ul></li>
<li class="chapter" data-level="17.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#fasta"><i class="fa fa-check"></i><b>17.3</b> FASTA<span></span></a>
<ul>
<li class="chapter" data-level="17.3.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#genomic-ranges"><i class="fa fa-check"></i><b>17.3.1</b> Genomic ranges<span></span></a></li>
<li class="chapter" data-level="17.3.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#extracting-genomic-ranges-from-a-fasta-file"><i class="fa fa-check"></i><b>17.3.2</b> Extracting genomic ranges from a FASTA file<span></span></a></li>
<li class="chapter" data-level="17.3.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#downloading-reference-genomes-from-ncbi"><i class="fa fa-check"></i><b>17.3.3</b> Downloading reference genomes from NCBI<span></span></a></li>
</ul></li>
<li class="chapter" data-level="17.4" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sambamfiles"><i class="fa fa-check"></i><b>17.4</b> Alignments<span></span></a>
<ul>
<li class="chapter" data-level="17.4.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#how-might-i-align-to-thee-let-me-count-the-ways"><i class="fa fa-check"></i><b>17.4.1</b> How might I align to thee? Let me count the ways…<span></span></a></li>
<li class="chapter" data-level="17.4.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#play-with-simple-alignments"><i class="fa fa-check"></i><b>17.4.2</b> Play with simple alignments<span></span></a></li>
<li class="chapter" data-level="17.4.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sam-flags"><i class="fa fa-check"></i><b>17.4.3</b> SAM Flags<span></span></a></li>
<li class="chapter" data-level="17.4.4" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#cigar"><i class="fa fa-check"></i><b>17.4.4</b> The CIGAR string<span></span></a></li>
<li class="chapter" data-level="17.4.5" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#the-seq-and-qual-columns"><i class="fa fa-check"></i><b>17.4.5</b> The SEQ and QUAL columns<span></span></a></li>
<li class="chapter" data-level="17.4.6" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sam-headers"><i class="fa fa-check"></i><b>17.4.6</b> SAM File Headers<span></span></a></li>
<li class="chapter" data-level="17.4.7" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#the-bam-format"><i class="fa fa-check"></i><b>17.4.7</b> The BAM format<span></span></a></li>
<li class="chapter" data-level="17.4.8" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#quick-self-study"><i class="fa fa-check"></i><b>17.4.8</b> Quick self study<span></span></a></li>
</ul></li>
<li class="chapter" data-level="17.5" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#variants"><i class="fa fa-check"></i><b>17.5</b> Variants<span></span></a>
<ul>
<li class="chapter" data-level="17.5.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#vcf-format-the-body"><i class="fa fa-check"></i><b>17.5.1</b> VCF Format – The Body<span></span></a></li>
<li class="chapter" data-level="17.5.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#vcf-format-the-header"><i class="fa fa-check"></i><b>17.5.2</b> VCF Format – The Header<span></span></a></li>
<li class="chapter" data-level="17.5.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#boneyard"><i class="fa fa-check"></i><b>17.5.3</b> Boneyard<span></span></a></li>
</ul></li>
<li class="chapter" data-level="17.6" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#segments"><i class="fa fa-check"></i><b>17.6</b> Segments<span></span></a></li>
<li class="chapter" data-level="17.7" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#conversionextractions-between-different-formats"><i class="fa fa-check"></i><b>17.7</b> Conversion/Extractions between different formats<span></span></a></li>
<li class="chapter" data-level="17.8" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#visualization-of-genomic-data"><i class="fa fa-check"></i><b>17.8</b> Visualization of Genomic Data<span></span></a>
<ul>
<li class="chapter" data-level="17.8.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sample-data"><i class="fa fa-check"></i><b>17.8.1</b> Sample Data<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="18" data-path="genome-assembly.html"><a href="genome-assembly.html"><i class="fa fa-check"></i><b>18</b> Genome Assembly<span></span></a></li>
<li class="chapter" data-level="19" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><i class="fa fa-check"></i><b>19</b> Alignment of sequence data to a reference genome (and associated steps)<span></span></a>
<ul>
<li class="chapter" data-level="19.1" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#read-journey"><i class="fa fa-check"></i><b>19.1</b> The Journey of each DNA Fragment from Organism to Sequencing Read<span></span></a></li>
<li class="chapter" data-level="19.2" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#read-groups"><i class="fa fa-check"></i><b>19.2</b> Read Groups<span></span></a></li>
<li class="chapter" data-level="19.3" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#aligning-reads-with-bwa"><i class="fa fa-check"></i><b>19.3</b> Aligning reads with <code>bwa</code><span></span></a>
<ul>
<li class="chapter" data-level="19.3.1" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#indexing-the-genome-for-alignment"><i class="fa fa-check"></i><b>19.3.1</b> Indexing the genome for alignment<span></span></a></li>
<li class="chapter" data-level="19.3.2" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#mapping-reads-with-bwa-mem"><i class="fa fa-check"></i><b>19.3.2</b> Mapping reads with <code>bwa mem</code><span></span></a></li>
<li class="chapter" data-level="19.3.3" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#hold-it-right-there-buddy-what-about-the-read-groups"><i class="fa fa-check"></i><b>19.3.3</b> Hold it Right There, Buddy! What about the Read Groups?<span></span></a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#samtools"><i class="fa fa-check"></i><b>19.4</b> Processing alignment output with <code>samtools</code><span></span></a>
<ul>
<li class="chapter" data-level="19.4.1" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#samtools-subcommands"><i class="fa fa-check"></i><b>19.4.1</b> <code>samtools</code> subcommands<span></span></a></li>
</ul></li>
<li class="chapter" data-level="19.5" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#boneyard-below-here"><i class="fa fa-check"></i><b>19.5</b> BONEYARD BELOW HERE<span></span></a></li>
<li class="chapter" data-level="19.6" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#preprocess"><i class="fa fa-check"></i><b>19.6</b> Preprocess ?<span></span></a></li>
<li class="chapter" data-level="19.7" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#quick-notes-to-self-on-chaining-things"><i class="fa fa-check"></i><b>19.7</b> Quick notes to self on chaining things:<span></span></a></li>
<li class="chapter" data-level="19.8" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#merging-bam-files"><i class="fa fa-check"></i><b>19.8</b> Merging BAM files<span></span></a></li>
<li class="chapter" data-level="19.9" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#divide-and-conquer-strategies"><i class="fa fa-check"></i><b>19.9</b> Divide and Conquer Strategies<span></span></a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="variant-calling.html"><a href="variant-calling.html"><i class="fa fa-check"></i><b>20</b> Variant calling<span></span></a>
<ul>
<li class="chapter" data-level="20.1" data-path="variant-calling.html"><a href="variant-calling.html#genotype-likelihoods"><i class="fa fa-check"></i><b>20.1</b> Genotype Likelihoods<span></span></a>
<ul>
<li class="chapter" data-level="20.1.1" data-path="variant-calling.html"><a href="variant-calling.html#basic-sketch-of-genotype-likelihood-calculations"><i class="fa fa-check"></i><b>20.1.1</b> Basic Sketch of Genotype Likelihood Calculations<span></span></a></li>
<li class="chapter" data-level="20.1.2" data-path="variant-calling.html"><a href="variant-calling.html#specifics-of-different-genotype-likelihoods"><i class="fa fa-check"></i><b>20.1.2</b> Specifics of different genotype likelihoods<span></span></a></li>
<li class="chapter" data-level="20.1.3" data-path="variant-calling.html"><a href="variant-calling.html#computing-genotype-likelihoods-with-three-different-softwares"><i class="fa fa-check"></i><b>20.1.3</b> Computing genotype likelihoods with three different softwares<span></span></a></li>
<li class="chapter" data-level="20.1.4" data-path="variant-calling.html"><a href="variant-calling.html#a-directed-acyclic-graph-for-genotype-likelihoods"><i class="fa fa-check"></i><b>20.1.4</b> A Directed Acyclic Graph For Genotype Likelihoods<span></span></a></li>
</ul></li>
<li class="chapter" data-level="20.2" data-path="variant-calling.html"><a href="variant-calling.html#gatks-gvcf-workflow"><i class="fa fa-check"></i><b>20.2</b> GATK’s gVCF Workflow<span></span></a>
<ul>
<li class="chapter" data-level="20.2.1" data-path="variant-calling.html"><a href="variant-calling.html#step-1.-running-haplotypecaller-and-getting-one-gvcf-for-each-sample"><i class="fa fa-check"></i><b>20.2.1</b> Step 1. Running HaplotypeCaller and getting one gVCF for each sample<span></span></a></li>
<li class="chapter" data-level="20.2.2" data-path="variant-calling.html"><a href="variant-calling.html#gvcf-format"><i class="fa fa-check"></i><b>20.2.2</b> gVCF files<span></span></a></li>
<li class="chapter" data-level="20.2.3" data-path="variant-calling.html"><a href="variant-calling.html#genomics-db"><i class="fa fa-check"></i><b>20.2.3</b> The Genomics Data Base<span></span></a></li>
<li class="chapter" data-level="20.2.4" data-path="variant-calling.html"><a href="variant-calling.html#gdb-update"><i class="fa fa-check"></i><b>20.2.4</b> Updating the GenomicsDatabase<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="21" data-path="boneyard-1.html"><a href="boneyard-1.html"><i class="fa fa-check"></i><b>21</b> Boneyard<span></span></a></li>
<li class="chapter" data-level="22" data-path="handle-vcf.html"><a href="handle-vcf.html"><i class="fa fa-check"></i><b>22</b> Basic Handling of VCF files<span></span></a>
<ul>
<li class="chapter" data-level="22.1" data-path="handle-vcf.html"><a href="handle-vcf.html#bcftools"><i class="fa fa-check"></i><b>22.1</b> bcftools<span></span></a>
<ul>
<li class="chapter" data-level="22.1.1" data-path="handle-vcf.html"><a href="handle-vcf.html#index-my-vcf-file"><i class="fa fa-check"></i><b>22.1.1</b> Index my VCF file!<span></span></a></li>
<li class="chapter" data-level="22.1.2" data-path="handle-vcf.html"><a href="handle-vcf.html#tell-me-about-my-vcf-file"><i class="fa fa-check"></i><b>22.1.2</b> Tell me about my VCF file!<span></span></a></li>
<li class="chapter" data-level="22.1.3" data-path="handle-vcf.html"><a href="handle-vcf.html#rename-the-samplesindividuals-in-the-file"><i class="fa fa-check"></i><b>22.1.3</b> Rename the samples/individuals in the file<span></span></a></li>
<li class="chapter" data-level="22.1.4" data-path="handle-vcf.html"><a href="handle-vcf.html#get-fragmentsparts-of-my-vcf-file"><i class="fa fa-check"></i><b>22.1.4</b> Get fragments/parts of my VCF file<span></span></a></li>
<li class="chapter" data-level="22.1.5" data-path="handle-vcf.html"><a href="handle-vcf.html#combine-vcf-files-in-various-ways"><i class="fa fa-check"></i><b>22.1.5</b> Combine VCF files in various ways<span></span></a></li>
<li class="chapter" data-level="22.1.6" data-path="handle-vcf.html"><a href="handle-vcf.html#filter-out-variants-for-a-variety-of-reasons"><i class="fa fa-check"></i><b>22.1.6</b> Filter out variants for a variety of reasons<span></span></a></li>
</ul></li>
<li class="chapter" data-level="22.2" data-path="handle-vcf.html"><a href="handle-vcf.html#how-about-some-more-arcane-filtering-tasks-lets-dream-some-up"><i class="fa fa-check"></i><b>22.2</b> How about some more arcane filtering tasks? Let’s dream some up!<span></span></a></li>
<li class="chapter" data-level="22.3" data-path="handle-vcf.html"><a href="handle-vcf.html#reconstituting-a-vcf-file-from-genotype-data"><i class="fa fa-check"></i><b>22.3</b> Reconstituting a VCF file from genotype data<span></span></a>
<ul>
<li class="chapter" data-level="22.3.1" data-path="handle-vcf.html"><a href="handle-vcf.html#why-create-a-vcf-file"><i class="fa fa-check"></i><b>22.3.1</b> Why create a VCF file?<span></span></a></li>
<li class="chapter" data-level="22.3.2" data-path="handle-vcf.html"><a href="handle-vcf.html#converting-the-.genotypes-file-to-vcf"><i class="fa fa-check"></i><b>22.3.2</b> Converting the <code>.genotypes</code> file to VCF<span></span></a></li>
<li class="chapter" data-level="22.3.3" data-path="handle-vcf.html"><a href="handle-vcf.html#reheader-the-vcf"><i class="fa fa-check"></i><b>22.3.3</b> Reheader the VCF<span></span></a></li>
<li class="chapter" data-level="22.3.4" data-path="handle-vcf.html"><a href="handle-vcf.html#where-to-from-here"><i class="fa fa-check"></i><b>22.3.4</b> Where to from here?<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="23" data-path="bioinformatics-for-rad-seq-data-with-and-without-a-reference-genome.html"><a href="bioinformatics-for-rad-seq-data-with-and-without-a-reference-genome.html"><i class="fa fa-check"></i><b>23</b> Bioinformatics for RAD seq data with and without a reference genome<span></span></a></li>
<li class="chapter" data-level="24" data-path="processing-amplicon-sequencing-data.html"><a href="processing-amplicon-sequencing-data.html"><i class="fa fa-check"></i><b>24</b> Processing amplicon sequencing data<span></span></a></li>
<li class="chapter" data-level="25" data-path="variant-annotation.html"><a href="variant-annotation.html"><i class="fa fa-check"></i><b>25</b> Variant Annotation<span></span></a>
<ul>
<li class="chapter" data-level="25.0.1" data-path="variant-annotation.html"><a href="variant-annotation.html#getting-a-gff-file"><i class="fa fa-check"></i><b>25.0.1</b> Getting a GFF file<span></span></a></li>
<li class="chapter" data-level="25.0.2" data-path="variant-annotation.html"><a href="variant-annotation.html#lets-do-the-pearl-millet"><i class="fa fa-check"></i><b>25.0.2</b> Let’s do the pearl millet<span></span></a></li>
<li class="chapter" data-level="25.1" data-path="variant-annotation.html"><a href="variant-annotation.html#boneyard-2"><i class="fa fa-check"></i><b>25.1</b> Boneyard<span></span></a></li>
</ul></li>
<li class="chapter" data-level="26" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html"><i class="fa fa-check"></i><b>26</b> Whole genome alignment strategies<span></span></a>
<ul>
<li class="chapter" data-level="26.1" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#mapping-of-scaffolds-to-a-closely-related-genome"><i class="fa fa-check"></i><b>26.1</b> Mapping of scaffolds to a closely related genome<span></span></a></li>
<li class="chapter" data-level="26.2" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#obtaining-ancestral-states-from-an-outgroup-genome"><i class="fa fa-check"></i><b>26.2</b> Obtaining Ancestral States from an Outgroup Genome<span></span></a>
<ul>
<li class="chapter" data-level="26.2.1" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#using-lastz-to-align-coho-to-the-chinook-genome"><i class="fa fa-check"></i><b>26.2.1</b> Using LASTZ to align coho to the chinook genome<span></span></a></li>
<li class="chapter" data-level="26.2.2" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#try-on-the-chinook-chromosomes"><i class="fa fa-check"></i><b>26.2.2</b> Try on the chinook chromosomes<span></span></a></li>
<li class="chapter" data-level="26.2.3" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#explore-the-other-parameters-more"><i class="fa fa-check"></i><b>26.2.3</b> Explore the other parameters more<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="27" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html"><i class="fa fa-check"></i><b>27</b> An example full workflow from sequences to variants<span></span></a>
<ul>
<li class="chapter" data-level="27.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#an-overview-of-the-whole-workflow"><i class="fa fa-check"></i><b>27.1</b> An overview of the whole workflow<span></span></a></li>
<li class="chapter" data-level="27.2" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#lets-have-a-look-at-the-reference-genome"><i class="fa fa-check"></i><b>27.2</b> Let’s have a look at the reference genome<span></span></a>
<ul>
<li class="chapter" data-level="27.2.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#first-off-make-sure-that-you-have-added-mamba-to-your-base-environment"><i class="fa fa-check"></i><b>27.2.1</b> First off, make sure that you have added mamba to your base environment<span></span></a></li>
<li class="chapter" data-level="27.2.2" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#indexing-the-reference-genome"><i class="fa fa-check"></i><b>27.2.2</b> Indexing the reference genome<span></span></a></li>
</ul></li>
<li class="chapter" data-level="27.3" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#a-word-on-organizing-workflows"><i class="fa fa-check"></i><b>27.3</b> A word on organizing workflows<span></span></a></li>
<li class="chapter" data-level="27.4" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#step-1a.-quality-checking-the-reads"><i class="fa fa-check"></i><b>27.4</b> Step 1a. Quality Checking the Reads<span></span></a>
<ul>
<li class="chapter" data-level="27.4.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#a-little-looking-at-the-htmls"><i class="fa fa-check"></i><b>27.4.1</b> A Little Looking at the HTMLs<span></span></a></li>
<li class="chapter" data-level="27.4.2" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#summarizing-output-from-a-lot-of-different-files-with-multiqc"><i class="fa fa-check"></i><b>27.4.2</b> Summarizing output from a lot of different files with MultiQC<span></span></a></li>
</ul></li>
<li class="chapter" data-level="27.5" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#step-1b.-trimming-the-reads"><i class="fa fa-check"></i><b>27.5</b> Step 1b. Trimming the Reads<span></span></a>
<ul>
<li class="chapter" data-level="27.5.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#cycling-over-all-the-files-and-trimming-them"><i class="fa fa-check"></i><b>27.5.1</b> Cycling over all the files and trimming them<span></span></a></li>
<li class="chapter" data-level="27.5.2" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#multiqc-your-trimmomatics-results"><i class="fa fa-check"></i><b>27.5.2</b> Multiqc your trimmomatics results<span></span></a></li>
</ul></li>
<li class="chapter" data-level="27.6" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#step-2.-mapping-each-unit-to-the-genome"><i class="fa fa-check"></i><b>27.6</b> Step 2. Mapping each unit to the genome<span></span></a>
<ul>
<li class="chapter" data-level="27.6.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#indexing-the-reference-genome-1"><i class="fa fa-check"></i><b>27.6.1</b> Indexing the reference genome<span></span></a></li>
<li class="chapter" data-level="27.6.2" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#a-quick-mapping-run-to-stdout"><i class="fa fa-check"></i><b>27.6.2</b> A quick mapping run to stdout<span></span></a></li>
<li class="chapter" data-level="27.6.3" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#map-with-array"><i class="fa fa-check"></i><b>27.6.3</b> Mapping all 22 pairs of trimmed FASTQ files using a SLURM job array<span></span></a></li>
</ul></li>
<li class="chapter" data-level="27.7" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#step-3.-mark-duplicates-in-bam-files-from-each-individual-and-run-samtools-stats-on-the-results"><i class="fa fa-check"></i><b>27.7</b> Step 3. Mark duplicates in BAM files from each individual, and run <code>samtools stats</code> on the results<span></span></a>
<ul>
<li class="chapter" data-level="27.7.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#run-samtools-stats-on-the-results"><i class="fa fa-check"></i><b>27.7.1</b> Run samtools stats on the results<span></span></a></li>
</ul></li>
<li class="chapter" data-level="27.8" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#make-gvcfs"><i class="fa fa-check"></i><b>27.8</b> Step 4. Creating a gVCF file for each sample<span></span></a>
<ul>
<li class="chapter" data-level="27.8.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#thats-a-dict-thing-to-do"><i class="fa fa-check"></i><b>27.8.1</b> That’s a dict thing to do!<span></span></a></li>
<li class="chapter" data-level="27.8.2" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#setting-up-a-slurm-job-array-for-this"><i class="fa fa-check"></i><b>27.8.2</b> Setting up a SLURM job array for this<span></span></a></li>
<li class="chapter" data-level="27.8.3" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#running-interactively-to-watch-the-output"><i class="fa fa-check"></i><b>27.8.3</b> Running interactively to watch the output<span></span></a></li>
<li class="chapter" data-level="27.8.4" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#launch-the-first-3-jobs-of-our-array"><i class="fa fa-check"></i><b>27.8.4</b> Launch the first 3 jobs of our array<span></span></a></li>
</ul></li>
<li class="chapter" data-level="27.9" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#use-genomicsdbimport"><i class="fa fa-check"></i><b>27.9</b> Step 5. Importing the gVCF files into a GenomicsDatabase<span></span></a>
<ul>
<li class="chapter" data-level="27.9.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#job-array-script-for-the-chromosomes"><i class="fa fa-check"></i><b>27.9.1</b> Job array script for the chromosomes<span></span></a></li>
<li class="chapter" data-level="27.9.2" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#job-array-script-for-the-scaffold-groups"><i class="fa fa-check"></i><b>27.9.2</b> Job array script for the scaffold groups<span></span></a></li>
<li class="chapter" data-level="27.9.3" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#hey-lets-try-updating-the-genomic-data-base"><i class="fa fa-check"></i><b>27.9.3</b> Hey! Let’s try updating the genomic data base<span></span></a></li>
</ul></li>
<li class="chapter" data-level="27.10" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#step-6.-calling-variants-and-creating-vcf-files-from-the-genomicsdatabase"><i class="fa fa-check"></i><b>27.10</b> Step 6. Calling variants and creating VCF files from the GenomicsDatabase<span></span></a>
<ul>
<li class="chapter" data-level="27.10.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#array-script-for-the-four-chromosomes"><i class="fa fa-check"></i><b>27.10.1</b> Array script for the four chromosomes<span></span></a></li>
<li class="chapter" data-level="27.10.2" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#array-script-for-the-two-scaffold-groups"><i class="fa fa-check"></i><b>27.10.2</b> Array script for the two scaffold groups<span></span></a></li>
<li class="chapter" data-level="27.10.3" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#have-a-look-at-the-products"><i class="fa fa-check"></i><b>27.10.3</b> Have a look at the products<span></span></a></li>
<li class="chapter" data-level="27.10.4" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#homework"><i class="fa fa-check"></i><b>27.10.4</b> Homework<span></span></a></li>
<li class="chapter" data-level="27.10.5" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#catenate-all-the-vcf.gz-files-into-a-single-vcf-file-and-index-it"><i class="fa fa-check"></i><b>27.10.5</b> Catenate all the vcf.gz files into a single VCF file and index it<span></span></a></li>
</ul></li>
<li class="chapter" data-level="27.11" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#step-7.-filtering-variants"><i class="fa fa-check"></i><b>27.11</b> Step 7. Filtering variants<span></span></a></li>
<li class="chapter" data-level="27.12" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#step-8.-base-quality-score-recalibration"><i class="fa fa-check"></i><b>27.12</b> Step 8. Base Quality Score Recalibration<span></span></a></li>
</ul></li>
<li class="part"><span><b>IV Part IV: Analysis of Big Variant Data<span></span></b></span></li>
<li class="chapter" data-level="28" data-path="bioinformatic-analysis-on-variant-data.html"><a href="bioinformatic-analysis-on-variant-data.html"><i class="fa fa-check"></i><b>28</b> Bioinformatic analysis on variant data<span></span></a></li>
<li class="part"><span><b>V Part V: Population Genomics<span></span></b></span></li>
<li class="chapter" data-level="29" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html"><i class="fa fa-check"></i><b>29</b> Topics in pop gen<span></span></a>
<ul>
<li class="chapter" data-level="29.1" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#coalescent"><i class="fa fa-check"></i><b>29.1</b> Coalescent<span></span></a></li>
<li class="chapter" data-level="29.2" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#measures-of-genetic-diversity-and-such"><i class="fa fa-check"></i><b>29.2</b> Measures of genetic diversity and such<span></span></a></li>
<li class="chapter" data-level="29.3" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#demographic-inference-with-partial-a-partial-i-and-moments"><i class="fa fa-check"></i><b>29.3</b> Demographic inference with <span class="math inline">\(\partial a \partial i\)</span> and <em>moments</em><span></span></a></li>
<li class="chapter" data-level="29.4" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#balls-in-boxes"><i class="fa fa-check"></i><b>29.4</b> Balls in Boxes<span></span></a></li>
<li class="chapter" data-level="29.5" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#some-landscape-genetics"><i class="fa fa-check"></i><b>29.5</b> Some landscape genetics<span></span></a></li>
<li class="chapter" data-level="29.6" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#relationship-inference"><i class="fa fa-check"></i><b>29.6</b> Relationship Inference<span></span></a></li>
<li class="chapter" data-level="29.7" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#tests-for-selection"><i class="fa fa-check"></i><b>29.7</b> Tests for Selection<span></span></a></li>
<li class="chapter" data-level="29.8" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#multivariate-associations-gea-etc."><i class="fa fa-check"></i><b>29.8</b> Multivariate Associations, GEA, etc.<span></span></a></li>
<li class="chapter" data-level="29.9" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#estimating-heritability-in-the-wild"><i class="fa fa-check"></i><b>29.9</b> Estimating heritability in the wild<span></span></a></li>
</ul></li>
<li><a href="references-1.html#references-1">References<span></span></a></li>
<li class="divider"></li>
<li><a href="https://bookdown.org" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="handle-vcf" class="section level1 hasAnchor" number="22">
<h1><span class="header-section-number">Chapter 22</span> Basic Handling of VCF files<a href="handle-vcf.html#handle-vcf" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>As we saw in the section on bioinformatic formats, VCF files can be
large and unwieldy. The format specification is also such that fields
might have different numbers of subfields, depending, for example, on the
number of alleles found at a variant. Both of these features make it
hard to directly read a VCF file into a, say, R, or some other program
that may wish to treat it purely as tabular data.</p>
<p>This is not to say that you couldn’t just read a VCF file into R directly
as a TAB delimited text file, and then start splitting fields up on it.
However, there are specialized tools for doing operations of VCF files,
and becoming familiar with them can relieve a lot
of the pain of dealing with VCF files.</p>
<p>To have an example VCF file to play with, you can
download one to your Unix workstation with the following commands.
You should put these files in your <code>scratch</code> directory somewhere, perhaps
creating a directory called <code>bcftools-play</code> to put them into. All the following
commands assume that the two example files are in the current working directory.</p>
<p>First we download the <code>vcf.gz</code> version of the VCF file, called <code>all.vcf.gz</code>. This
is the unfiltered VCF file created by our example WGS workflow in Section <a href="example-wgs-flow.html#example-wgs-flow">27</a>:</p>
<div class="sourceCode" id="cb640"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb640-1"><a href="handle-vcf.html#cb640-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">--load-cookies</span> /tmp/cookies.txt <span class="st">&quot;https://docs.google.com/uc?export=download&amp;confirm=</span><span class="va">$(</span><span class="fu">wget</span> <span class="at">--quiet</span> <span class="at">--save-cookies</span> /tmp/cookies.txt <span class="at">--keep-session-cookies</span> <span class="at">--no-check-certificate</span> <span class="st">&#39;https://docs.google.com/uc?export=download&amp;id=1zgFyYbfWU85O4JzmOX7-MOKQs4oXv4fm&#39;</span> <span class="at">-O-</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="at">-rn</span> <span class="st">&#39;s/.*confirm=([0-9A-Za-z_]+).*/\1\n/p&#39;</span><span class="va">)</span><span class="st">&amp;id=1zgFyYbfWU85O4JzmOX7-MOKQs4oXv4fm&quot;</span> <span class="at">-O</span> all.vcf.gz <span class="kw">&amp;&amp;</span> <span class="fu">rm</span> <span class="at">-rf</span> /tmp/cookies.txt</span></code></pre></div>
<p>After that, we also download the BCF version of the same file, just so that everyone
gets familiar with the fact that BCF files can be treated equivalently
to VCF files with <code>bcftools</code>.</p>
<div class="sourceCode" id="cb641"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb641-1"><a href="handle-vcf.html#cb641-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">--load-cookies</span> /tmp/cookies.txt <span class="st">&quot;https://docs.google.com/uc?export=download&amp;confirm=</span><span class="va">$(</span><span class="fu">wget</span> <span class="at">--quiet</span> <span class="at">--save-cookies</span> /tmp/cookies.txt <span class="at">--keep-session-cookies</span> <span class="at">--no-check-certificate</span> <span class="st">&#39;https://docs.google.com/uc?export=download&amp;id=1iUP_UstnmuLSGuBIfb0EYXs1Gw1z-z0o&#39;</span> <span class="at">-O-</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="at">-rn</span> <span class="st">&#39;s/.*confirm=([0-9A-Za-z_]+).*/\1\n/p&#39;</span><span class="va">)</span><span class="st">&amp;id=1iUP_UstnmuLSGuBIfb0EYXs1Gw1z-z0o&quot;</span> <span class="at">-O</span> all.bcf <span class="kw">&amp;&amp;</span> <span class="fu">rm</span> <span class="at">-rf</span> /tmp/cookies.txt</span></code></pre></div>
<p>To repeat, BCF and vcf.gz are two formats serving similar purposes. I will be using both
<code>all.bcf</code> and <code>all.vcf.gz</code> interchangeably in the following examples.</p>
<p>There are two main, well-known programs for handling VCF files: <code>vcftools</code> and <code>bcftools</code>. Both
of these grew out of the 1000 Genomes effort starting about a decade ago. It seems
that <code>vcftools</code> may have been developed first, but, currently, <code>bcftools</code> is being
more actively developed, with new versions and new features being added to it
regularly. <code>vcftools</code> provides some very specific commands for particular analyses
or operations, some of which are not available from <code>bcftools</code>. On the other hand,
<code>bcftools</code> provides a more general interface for operations on VCF files. By this
interface, a great number of the operations done in <code>vcftools</code> are available, but
a little extra knowledge is required to implement them. That said, the range of
possible operations seems much larger in <code>bcftools</code> than in <code>vcftools</code>.</p>
<p>Further, <code>bcftools</code> behaves like a typical Unix utility, allowing data to be streamed
to stdout, and data can be streamed <em>into</em> <code>bcftools</code> (by using the <code>-</code> as the
input file name) from stdin. This lets you pipe output into it the way you
can with most Unix tools. This makes it far more versatile than <code>vcftools</code>.</p>
<div id="bcftools" class="section level2 hasAnchor" number="22.1">
<h2><span class="header-section-number">22.1</span> bcftools<a href="handle-vcf.html#bcftools" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>If you don’t already have <code>bcftools</code>, you can use <code>mamba</code> to create a conda
environment called <code>bcftools</code> that has it:</p>
<div class="sourceCode" id="cb642"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb642-1"><a href="handle-vcf.html#cb642-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> create <span class="at">-n</span> bcftools <span class="at">-c</span> conda-forge <span class="at">-c</span> bioconda bcftools=1.15.1</span>
<span id="cb642-2"><a href="handle-vcf.html#cb642-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate bcftools</span></code></pre></div>
<p>(Note the use of <code>-c conda-forge</code> in the above. This is present because
some of the dependencies for <code>bcftools</code> are not found on the bioconda channel.
Rather they are on the conda-forge channel. If you conda/mamba environment is not
set up to search conda-forge by default, then the <code>-c conda-forge</code> is required to
get all the dependencies.)</p>
<p>Here, we just want to give everyone the chance to run through a few
operations with <code>bcftools</code>, to start to get familiar with its interface.
The first thing to note is that, like <code>samtools</code> (which is maintained by
the same group of people), <code>bcftools</code> possesses a number of different
<em>subcommands</em>. So, the syntax is always like:</p>
<ul>
<li><code>bcftools</code> <em>subcommand</em> options file(s)</li>
</ul>
<p>Also like <code>samtools</code>, <code>bcftools</code> will take input from <em>stdin</em> rather than
from a file—you just pass it <code>-</code> instead of a file name.</p>
<p>The full documentation/manual for <code>bcftools</code> is maintained at:
<a href="http://samtools.github.io/bcftools/bcftools.html">http://samtools.github.io/bcftools/bcftools.html</a>. It is well worth reading through this whole documentation, though it
can be quite terse and intimidating. A friendlier “tutorial” introduction to the
software can be found at <a href="https://samtools.github.io/bcftools/howtos/index.html">https://samtools.github.io/bcftools/howtos/index.html</a>.</p>
<p>Here we are going to get our feet with with a few operations.</p>
<p>First, we will look at the “onboard” documentation. By just entering <code>bcftools</code> you get a
list of all the subcommands that are available:</p>
<div class="sourceCode" id="cb643"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb643-1"><a href="handle-vcf.html#cb643-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span></span></code></pre></div>
<p>If you want the onboard documentation for any of the particular subcommands, you can
just give a naked <code>bctools subcommand</code> command, like:</p>
<div class="sourceCode" id="cb644"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb644-1"><a href="handle-vcf.html#cb644-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> index</span></code></pre></div>
<p>or, for a more daunting set of documentation:</p>
<div class="sourceCode" id="cb645"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb645-1"><a href="handle-vcf.html#cb645-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> roh</span></code></pre></div>
<div id="index-my-vcf-file" class="section level3 hasAnchor" number="22.1.1">
<h3><span class="header-section-number">22.1.1</span> Index my VCF file!<a href="handle-vcf.html#index-my-vcf-file" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The first thing we are going to do is index our VCF files.
We create the default index, a <em>coordinate sorted index</em> which
has the <code>.csi</code> extension. If your VCF file is not in coordinate-sorted
order, you might have to sort it before you do this. However,
<code>all.vcf.gz</code> and <code>all.bcf</code> are already sorted. So:</p>
<div class="sourceCode" id="cb646"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb646-1"><a href="handle-vcf.html#cb646-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> index all.vcf.gz</span>
<span id="cb646-2"><a href="handle-vcf.html#cb646-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> index all.bcf</span></code></pre></div>
<p>Use <code>ls</code> to see the new files created by that operation.</p>
<p>The index allows for rapid access to different parts of the files that
correspond to specific locations in the genome.</p>
</div>
<div id="tell-me-about-my-vcf-file" class="section level3 hasAnchor" number="22.1.2">
<h3><span class="header-section-number">22.1.2</span> Tell me about my VCF file!<a href="handle-vcf.html#tell-me-about-my-vcf-file" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>VCF files are a little daunting. Especially when they are gzipped
(or are BCF files!) they
can seem particularly opaque—learning anything about them in the traditional
manner of uncompressing them and then searching for lines within them
or counting up the number of records can be time consuming. Here are some
<code>bcftools</code> solutions to a few different questions you might have.</p>
<p><strong>Who is in this file?</strong> You can always try to find the last header line
in a VCF file using grep or awk and parse the individuals out yourself, but
it turns out to be faster and safer to use the <code>query</code> subcommand from <code>bcftools</code>
with the <code>-l</code> option. Do it here:</p>
<div class="sourceCode" id="cb647"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb647-1"><a href="handle-vcf.html#cb647-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-l</span> all.vcf.gz</span>
<span id="cb647-2"><a href="handle-vcf.html#cb647-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb647-3"><a href="handle-vcf.html#cb647-3" aria-hidden="true" tabindex="-1"></a><span class="co"># And, of course, you can do the same with the BCF file</span></span>
<span id="cb647-4"><a href="handle-vcf.html#cb647-4" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-l</span> all.bcf</span></code></pre></div>
<p>Then read about it on the manual page. Find the part that describes it.</p>
<p><strong>How many variants are in this file?</strong> This question can be answered quickly with
<code>bcftools stats</code>, which also returns to you a plethora of information about the variants.</p>
<div class="sourceCode" id="cb648"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb648-1"><a href="handle-vcf.html#cb648-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> stats all.vcf.gz <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>The top part of the output tells you how many SNPs and indels (and other types
of variants) there are. Then it tells you about Ts/Tv ratios, then it essentially
gives histogram summaries for allele frequencies,
variant quality scores (QUAL), insertion-deletion sizes, substitution types, read depths, etc.</p>
<p><strong>Where are these variants?</strong> There are several ways to answer this question.
One might be simply to print the CHROM and the POS for each row in the VCF file:</p>
<div class="sourceCode" id="cb649"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb649-1"><a href="handle-vcf.html#cb649-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%CHROM\t%POS\n&#39;</span> all.vcf.gz</span></code></pre></div>
<p>If you want to see where it starts and where it finishes you can do:</p>
<div class="sourceCode" id="cb650"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb650-1"><a href="handle-vcf.html#cb650-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%CHROM\t%POS\n&#39;</span> all.vcf.gz <span class="kw">|</span> <span class="fu">head</span> </span>
<span id="cb650-2"><a href="handle-vcf.html#cb650-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%CHROM\t%POS\n&#39;</span> all.vcf.gz <span class="kw">|</span> <span class="fu">tail</span></span></code></pre></div>
<p>If we wanted to quickly see how many variants were on each of the
chromosomes/scaffolds, sorted by number of variants, we could do:</p>
<div class="sourceCode" id="cb651"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb651-1"><a href="handle-vcf.html#cb651-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%CHROM\t%POS\n&#39;</span> all.vcf.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;{n[$1]++} END {for(i in n) print i, n[i]}&#39;</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-nbr</span> <span class="at">-k</span> 2 <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>This shows one use of the subcommand <code>query</code>, which is quite useful.
Even though it is named <code>query</code> its main purpose is simply extracting fields
of information from a VCF file and spitting them out in a new, user-specified,
typically tabular format.</p>
<p><strong>Give me a glimpse of the file</strong> You can use <code>bcftools view</code> for a number
of things, but at its simplest, you can merely look at the file in VCF format.
(In this manner, it behaves much like <code>samtools view</code> for VCF files).</p>
<div class="sourceCode" id="cb652"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb652-1"><a href="handle-vcf.html#cb652-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show the whole file from the top</span></span>
<span id="cb652-2"><a href="handle-vcf.html#cb652-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view all.bcf <span class="kw">|</span> <span class="fu">less</span></span>
<span id="cb652-3"><a href="handle-vcf.html#cb652-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb652-4"><a href="handle-vcf.html#cb652-4" aria-hidden="true" tabindex="-1"></a><span class="co"># of course, this works with either bcf or vcf or vcf.gz</span></span>
<span id="cb652-5"><a href="handle-vcf.html#cb652-5" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view all.vcf.gz <span class="kw">|</span> <span class="fu">less</span></span>
<span id="cb652-6"><a href="handle-vcf.html#cb652-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb652-7"><a href="handle-vcf.html#cb652-7" aria-hidden="true" tabindex="-1"></a><span class="co"># show just the header with -h.  Here we look at just the last 10 lines of the header</span></span>
<span id="cb652-8"><a href="handle-vcf.html#cb652-8" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-h</span> all.bcf  <span class="kw">|</span> <span class="fu">tail</span></span>
<span id="cb652-9"><a href="handle-vcf.html#cb652-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb652-10"><a href="handle-vcf.html#cb652-10" aria-hidden="true" tabindex="-1"></a><span class="co"># show the variants themselves (no header) with -H</span></span>
<span id="cb652-11"><a href="handle-vcf.html#cb652-11" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-H</span> all.vcf.gz <span class="kw">|</span> <span class="fu">head</span></span></code></pre></div>
<p>Just like you can with <code>samtools view</code> you can convert formats with <code>bcftools view</code>.
Pipe a VCF into it and then use the <code>-O</code> (Big-O, not a zero) option:</p>
<ul>
<li><code>-O z</code>: bgzipped VCF (vcf.gz)</li>
<li><code>-O v</code>: uncompressed VCF (the default)</li>
<li><code>-O u</code>: uncompressed BCF</li>
<li><code>-O b</code>: compressed BCF</li>
</ul>
</div>
<div id="rename-the-samplesindividuals-in-the-file" class="section level3 hasAnchor" number="22.1.3">
<h3><span class="header-section-number">22.1.3</span> Rename the samples/individuals in the file<a href="handle-vcf.html#rename-the-samplesindividuals-in-the-file" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We saw above that the names of the samples in the file are like <code>s001</code>, <code>s002</code>.
This was not actually what I had intended! The names in here are set by the
<code>SM</code> field of the read groups in the BAM files from which variants are called.
In Section <a href="example-wgs-flow.html#map-with-array">27.6.3</a>, where those BAM files were made, I
screwed up and used the wrong column from the <code>numbered-units.tsv</code> file to
set the <code>SM</code> value in the read groups. I wanted to use <code>sample_id</code> not <code>sample</code>.
However, all is not lost. We don’t have to go all the way back to the beginning
and remap everything and call variants. We simply rename the samples in the file.
For this we can use <code>bcftools reheader</code>. First, look at the documentation for that,
both on the web, and with:</p>
<div class="sourceCode" id="cb653"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb653-1"><a href="handle-vcf.html#cb653-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> reheader</span></code></pre></div>
<p>Aha! we see that the web-based documentation is a little more complete, and it tells
us what format to use for a sample-renaming file for the <code>-s</code> option.
Copy the following contents (using <code>nano</code>, for example) into a file called <code>sample-renames.txt</code></p>
<pre><code>s001    T199967
s002    T199968
s003    T199969
s004    T199970
s005    T199971
s006    T199972
s007    T199973
s008    T199974</code></pre>
<p>Then we can make a renamed <code>bcf</code> file with:</p>
<div class="sourceCode" id="cb655"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb655-1"><a href="handle-vcf.html#cb655-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> reheader <span class="at">-s</span> sample-renames.txt all.bcf  <span class="op">&gt;</span> renamed-all.bcf</span></code></pre></div>
<p>and a renamed <code>vcf.gz</code> file with:</p>
<div class="sourceCode" id="cb656"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb656-1"><a href="handle-vcf.html#cb656-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> reheader <span class="at">-s</span> sample-renames.txt all.vcf.gz  <span class="op">&gt;</span> renamed-all.vcf.gz</span></code></pre></div>
<p>In this case, the type of output file (bcf or vcf.gz) is the same
as the type of the input file.</p>
<p><strong>Exercise</strong> Use <code>bcftools view</code> and <code>tail</code> to see that the names have really been
changed. Then use <code>bcftools query</code> to do the same.</p>
</div>
<div id="get-fragmentsparts-of-my-vcf-file" class="section level3 hasAnchor" number="22.1.4">
<h3><span class="header-section-number">22.1.4</span> Get fragments/parts of my VCF file<a href="handle-vcf.html#get-fragmentsparts-of-my-vcf-file" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>There are lots of ways to extract desired bits of information
from a VCF file into a more manageable format.</p>
<p><strong>Extract keyed values from the INFO field</strong> When we did this:</p>
<div class="sourceCode" id="cb657"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb657-1"><a href="handle-vcf.html#cb657-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-H</span> <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>we saw that there is a lot of information in the INFO
field. What if we wanted to extract that? It looks like it could be
hard to parse because the fields are in semi-colon-separated
key-value pairs.</p>
<p>This is another job for <code>bcftools query</code>.
You pass a <em>format string</em> to the <code>-f</code> option that tells the program which fields you want
to extract and how you want to format it. In general, the values are preceded by
a <code>%</code> and subfields of the INFO column can be named described
like <code>%INFO/SUBFIELD</code>. You can ask for TABs between fields with
<code>\t</code> and for line endings with <code>\n</code>. In general you need to wrap all of these format
specifications in single quotes so that the shell does not get confused by them.</p>
<p>Check out some examples:</p>
<div class="sourceCode" id="cb658"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb658-1"><a href="handle-vcf.html#cb658-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract CHROM POS and BaseQRankSum, separated by TABs</span></span>
<span id="cb658-2"><a href="handle-vcf.html#cb658-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%CHROM\t%POS\t%INFO/BaseQRankSum\n&#39;</span> all.vcf.gz <span class="kw">|</span> <span class="fu">less</span></span>
<span id="cb658-3"><a href="handle-vcf.html#cb658-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb658-4"><a href="handle-vcf.html#cb658-4" aria-hidden="true" tabindex="-1"></a><span class="co"># extract CHROM POS and total read depth DP</span></span>
<span id="cb658-5"><a href="handle-vcf.html#cb658-5" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%CHROM\t%POS\t%INFO/DP\n&#39;</span> all.bcf <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>You can even extract information from each of the genotype columns. If you want to print
CHROM and POS and then all of the PHRED-scaled genotype likelihoods for all the samples, separated
by TABs, you can do:</p>
<div class="sourceCode" id="cb659"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb659-1"><a href="handle-vcf.html#cb659-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%CHROM\t%POS\t[%PL\t]\n&#39;</span> all.bcf <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>Note that FORMAT values (i.e., those in the genotype columns) must be wrapped in
<code>[ ]</code> to get all the values to be printed out.</p>
<p><strong>EXERCISE</strong> Extract the CHROM, POS, Maximum Likelihood-estimated Allele Frequency (MLEAF in the INFO
column) for each variant, along with the allele depths (AD) of each individual, all TAB separated, from the
file <code>all.vcf.gz</code>.</p>
<p><strong>View data from specified regions</strong></p>
<p>What if we want to look at variants only in two 10 Kb regions, like
<code>CM031199.1:1-10000</code> and <code>CM031200.1:1000000-1005000</code>? Pass those,
separated by commas, to the <code>-r</code> option (which is an option that applies to many
of the subcommands):</p>
<div class="sourceCode" id="cb660"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb660-1"><a href="handle-vcf.html#cb660-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-H</span> <span class="at">-r</span> CM031199.1:1-10000,CM031200.1:1000000-1005000 all.vcf.gz <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>You can also specify those regions in a file with the <code>-R</code> option.</p>
<p><strong>View data from specified individuals</strong></p>
<p>You can give the sample names (comma separated) to the <code>-s</code> option:</p>
<div class="sourceCode" id="cb661"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb661-1"><a href="handle-vcf.html#cb661-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-H</span> <span class="at">-s</span> s001,s002,s003 all.vcf.gz <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>Or, if you wanted to view all <em>but</em> those two individuals, precede them with a <code>^</code>:</p>
<div class="sourceCode" id="cb662"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb662-1"><a href="handle-vcf.html#cb662-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-H</span> <span class="at">-s</span> ^s001,s002,s003 all.vcf.gz <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>You can also supply a text file with sample names (one-per-line) to the capital letter <code>-S</code> option.</p>
<p>You can combine options, like <code>-r</code> and <code>-s</code>, as well.</p>
</div>
<div id="combine-vcf-files-in-various-ways" class="section level3 hasAnchor" number="22.1.5">
<h3><span class="header-section-number">22.1.5</span> Combine VCF files in various ways<a href="handle-vcf.html#combine-vcf-files-in-various-ways" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Catenate VCF files</strong></p>
<p>If you have VCF files called from the same reference genome filled with
the same samples, it is easy to catenate them together with <code>bcftools concat</code>:</p>
<div class="sourceCode" id="cb663"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb663-1"><a href="handle-vcf.html#cb663-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make two files from different regions</span></span>
<span id="cb663-2"><a href="handle-vcf.html#cb663-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-O</span> z <span class="at">-r</span> CM031199.1:1-10000 all.vcf.gz  <span class="op">&gt;</span> A.vcf.gz</span>
<span id="cb663-3"><a href="handle-vcf.html#cb663-3" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-O</span> z <span class="at">-r</span> CM031200.1:1000000-1005000 all.vcf.gz  <span class="op">&gt;</span> B.vcf.gz</span>
<span id="cb663-4"><a href="handle-vcf.html#cb663-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb663-5"><a href="handle-vcf.html#cb663-5" aria-hidden="true" tabindex="-1"></a><span class="co"># how many variants in each of those?</span></span>
<span id="cb663-6"><a href="handle-vcf.html#cb663-6" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> stats A.vcf.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;/^SN/&#39;</span></span>
<span id="cb663-7"><a href="handle-vcf.html#cb663-7" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> stats B.vcf.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;/^SN/&#39;</span></span>
<span id="cb663-8"><a href="handle-vcf.html#cb663-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb663-9"><a href="handle-vcf.html#cb663-9" aria-hidden="true" tabindex="-1"></a><span class="co"># catenate the back together</span></span>
<span id="cb663-10"><a href="handle-vcf.html#cb663-10" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> concat <span class="at">-Oz</span>  A.vcf.gz B.vcf.gz <span class="op">&gt;</span> CAT.vcf.gz</span>
<span id="cb663-11"><a href="handle-vcf.html#cb663-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb663-12"><a href="handle-vcf.html#cb663-12" aria-hidden="true" tabindex="-1"></a><span class="co"># how many variants in that?</span></span>
<span id="cb663-13"><a href="handle-vcf.html#cb663-13" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> stats CAT.vcf.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;/^SN/&#39;</span></span></code></pre></div>
<p>Note that when using the <code>-O</code> (capital “o”) option to specify the output
type: <code>v</code> = VCF, <code>b</code> = BCF, <code>u</code> = uncompressed BCF, <code>z</code> = bgzipped VCF, you
don’t need a space after the <code>-O</code>.</p>
<p><strong>Merge VCF files</strong></p>
<p>If you have files with different samples in them you
can easily combine them:</p>
<div class="sourceCode" id="cb664"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb664-1"><a href="handle-vcf.html#cb664-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make file with first three samples</span></span>
<span id="cb664-2"><a href="handle-vcf.html#cb664-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-Oz</span> <span class="at">-s</span> s001,s002,s003 all.vcf.gz <span class="op">&gt;</span> first3.vcf.gz</span>
<span id="cb664-3"><a href="handle-vcf.html#cb664-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb664-4"><a href="handle-vcf.html#cb664-4" aria-hidden="true" tabindex="-1"></a><span class="co"># make another with the last three samples</span></span>
<span id="cb664-5"><a href="handle-vcf.html#cb664-5" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-Oz</span> <span class="at">-s</span> s006,s007,s008 all.bcf <span class="op">&gt;</span> last3.vcf.gz</span>
<span id="cb664-6"><a href="handle-vcf.html#cb664-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb664-7"><a href="handle-vcf.html#cb664-7" aria-hidden="true" tabindex="-1"></a><span class="co"># merging requires that the files be indexed</span></span>
<span id="cb664-8"><a href="handle-vcf.html#cb664-8" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> index first3.vcf.gz</span>
<span id="cb664-9"><a href="handle-vcf.html#cb664-9" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> index last3.vcf.gz</span>
<span id="cb664-10"><a href="handle-vcf.html#cb664-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb664-11"><a href="handle-vcf.html#cb664-11" aria-hidden="true" tabindex="-1"></a><span class="co"># merge those into a file with 6 samples</span></span>
<span id="cb664-12"><a href="handle-vcf.html#cb664-12" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> merge <span class="at">-Oz</span> first3.vcf.gz last3.vcf.gz <span class="op">&gt;</span> 6-samples.vcf.gz</span></code></pre></div>
</div>
<div id="filter-out-variants-for-a-variety-of-reasons" class="section level3 hasAnchor" number="22.1.6">
<h3><span class="header-section-number">22.1.6</span> Filter out variants for a variety of reasons<a href="handle-vcf.html#filter-out-variants-for-a-variety-of-reasons" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>There are a lot of ways to filter out variants. <code>bcftools</code> leaves things
very general here, and so just about anything is possible. Some simple ones appear below.
Remember, we are piping the result to <code>bcftools stats</code> just so that we can
see the result. If you really wanted to make a filtered file, you would typically
just redirect it to a file.</p>
<p><strong>Just the biallelic SNPs please</strong> Get things with no more than 2 alleles and
no fewer than two alleles, and of a type = SNP:</p>
<div class="sourceCode" id="cb665"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb665-1"><a href="handle-vcf.html#cb665-1" aria-hidden="true" tabindex="-1"></a><span class="co"># do it and summarize the result to look at it, all in one line:</span></span>
<span id="cb665-2"><a href="handle-vcf.html#cb665-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-Ou</span> <span class="at">-m</span> 2 <span class="at">-M</span> 2 <span class="at">--types</span><span class="op">=</span>snps all.bcf <span class="kw">|</span> <span class="ex">bcftools</span> stats <span class="at">-</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;/^SN/&#39;</span></span></code></pre></div>
<p><strong>Just the biallelic indels please</strong></p>
<div class="sourceCode" id="cb666"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb666-1"><a href="handle-vcf.html#cb666-1" aria-hidden="true" tabindex="-1"></a><span class="co"># do it and see the result all in one line:</span></span>
<span id="cb666-2"><a href="handle-vcf.html#cb666-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-Ou</span> <span class="at">-m</span> 2 <span class="at">-M</span> 2 <span class="at">--types</span><span class="op">=</span>indels all.vcf.gz <span class="kw">|</span> <span class="ex">bcftools</span> stats <span class="at">-</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;/^SN/&#39;</span></span></code></pre></div>
<p>Note the use of <code>-Ou</code> in order to pipe uncompressed BCF output directly into
<code>bcftools stats</code> using the <code>-</code> for a filename.</p>
<p><strong>Fraction of missing sites less than X</strong></p>
<p>If you want to make sure that 60% of your individuals have at least
one read at the genotype, you can do this:</p>
<div class="sourceCode" id="cb667"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb667-1"><a href="handle-vcf.html#cb667-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-i</span> <span class="st">&#39;F_MISSING &lt; 0.4&#39;</span> all.vcf.gz <span class="kw">|</span> <span class="ex">bcftools</span> stats <span class="at">-</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;/^SN/&#39;</span></span></code></pre></div>
<p>Play with setting the <code>F_MISSING</code> to different values and see how that affects
the number of variants retained. (Not much with this example data set, it turns
out, because there is not much missing data.</p>
<p><strong>Exclude based on various features of the data</strong></p>
<p>You can use the <code>-e</code> option to <code>bcftools view</code> or <code>bcftools filter</code> to <em>exclude</em> sites that
meet certain criteria. (You can use <code>-i</code> to <em>include</em> those sites and no others).</p>
<p>For a terrifyingly terse and dense description of what sorts of expressions can be used to
describe the criteria, see the web manual section on expressions: <a href="http://samtools.github.io/bcftools/bcftools.html#expressions">http://samtools.github.io/bcftools/bcftools.html#expressions</a>.</p>
<p>For example, to only keep things with a maximum-likelihood-estimated allele frequency between 0.4 and 0.6:</p>
<div class="sourceCode" id="cb668"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb668-1"><a href="handle-vcf.html#cb668-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-i</span> <span class="st">&#39;INFO/MLEAF &gt;= 0.4 &amp;&amp; INFO/MLEAF &lt;= 0.6&#39;</span> all.bcf <span class="kw">|</span> <span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%INFO/MLEAF\n&#39;</span> <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>Note we are piping the result to <code>bcftools query</code> in order to see what the actual MLEAFs are
after filtering. For the most part, this has worked, except for cases in which there are more than
two allele freqencies. If we wanted to filter those out, we could filter to only biallelic sites,
or, for the sake of illustration, we could retain only those sites at which the MLEAF value for the
<em>first</em> alternate allele is between 0.4 and 0.6:</p>
<div class="sourceCode" id="cb669"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb669-1"><a href="handle-vcf.html#cb669-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-i</span> <span class="st">&#39;INFO/MLEAF[0] &gt;= 0.4 &amp;&amp; INFO/MLEAF[0] &lt;= 0.6&#39;</span> all.bcf <span class="kw">|</span> <span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%INFO/MLEAF\n&#39;</span> <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>Cool!</p>
<p>How about excluding those sites in which any individual had a DP less than 5. We can
test each of the DP columns in the FORMAT columns. We name these <code>FMT/DP</code>.
Note that each test (from each sample)
is combined with an OR by default, so:</p>
<div class="sourceCode" id="cb670"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb670-1"><a href="handle-vcf.html#cb670-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-H</span> <span class="at">-e</span> <span class="st">&#39;FMT/DP &lt; 5&#39;</span> all.bcf <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>To make it easier to see what the DPs are there, let’s print them:</p>
<div class="sourceCode" id="cb671"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb671-1"><a href="handle-vcf.html#cb671-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-e</span> <span class="st">&#39;FMT/DP &lt; 5&#39;</span> all.bcf <span class="kw">|</span> <span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%CHROM\t%POS\t[%DP\t]\n&#39;</span> <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p><strong>More playing</strong></p>
<p>Here is one that might be interesting. Suppose that we want to
filter our data set down to variant sites at which at least
two of the eight individuals had at least 10 reads of the first alternate allele?</p>
<div class="sourceCode" id="cb672"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb672-1"><a href="handle-vcf.html#cb672-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-i</span> <span class="st">&#39;COUNT(FMT/AD[:1] &gt; 10) &gt; 2&#39;</span> all.bcf <span class="kw">|</span> <span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%CHROM\t%POS\t[%AD\t]\n&#39;</span> <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>Whoa! Cool.</p>
</div>
</div>
<div id="how-about-some-more-arcane-filtering-tasks-lets-dream-some-up" class="section level2 hasAnchor" number="22.2">
<h2><span class="header-section-number">22.2</span> How about some more arcane filtering tasks? Let’s dream some up!<a href="handle-vcf.html#how-about-some-more-arcane-filtering-tasks-lets-dream-some-up" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This has only scratched the surface of what is possible with bcftools.</p>
</div>
<div id="reconstituting-a-vcf-file-from-genotype-data" class="section level2 hasAnchor" number="22.3">
<h2><span class="header-section-number">22.3</span> Reconstituting a VCF file from genotype data<a href="handle-vcf.html#reconstituting-a-vcf-file-from-genotype-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>There are many other formats for storing genotype data from high-throughput sequencing,
although few, if any, of them allow for the richness of variant types that can be
encoded in a VCF file, and most of them also don’t allow for all the auxiliary information
(like read depths, allele depths, genotype likelihoods, etc.) to also be easily stored,
indexed, and accessed.</p>
<p>Some of the different formats are specialized to certain types of data. For example,
the program PLINK 2.0 has a binary format that is incredibly efficient for storing
biallelic SNP data. PLINK also has facilities in the program for converting to and
from plink format to VCF. Some formats, however, don’t seem to have any associated
programs for converting to VCF.</p>
<p>There can be times when you receive variant data from other people or sources that are
not in VCF format. Data may be archived or placed in public repositories
in a variety of formats. For example, some of the variant data from a recent pearl
millet sequencing project has been made public at
<a href="https://cegresources.icrisat.org/data_public/PM_SNPs/SNP_calls/">https://cegresources.icrisat.org/data_public/PM_SNPs/SNP_calls/</a>, in the form of <code>.genotypes</code> files.</p>
<p>Here, you can download the first 10000 lines of the file <code>WP.pgchr1.genotype.gz</code>
at this link:
<a href="https://eriqande.github.io/eca-bioinf-handbook/downloads/WP.pgchr1_10K.genotype.gz">https://eriqande.github.io/eca-bioinf-handbook/downloads/WP.pgchr1_10K.genotype.gz</a>, and you
can get the explanation of the columns in the file (including the sample names)
from this link:
<a href="https://eriqande.github.io/eca-bioinf-handbook/downloads/genotype.readme">https://eriqande.github.io/eca-bioinf-handbook/downloads/genotype.readme</a></p>
<p>or on your
cluster, you could use:</p>
<div class="sourceCode" id="cb673"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb673-1"><a href="handle-vcf.html#cb673-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://eriqande.github.io/eca-bioinf-handbook/downloads/WP.pgchr1_10K.genotype.gz</span>
<span id="cb673-2"><a href="handle-vcf.html#cb673-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://eriqande.github.io/eca-bioinf-handbook/downloads/genotype.readme</span></code></pre></div>
<p>These files give the genotypes of individuals using
<a href="https://www.bioinformatics.org/sms/iupac.html">IUPAC nucleotide codes</a> for the bases.
Each diploid sample gets a column, and a single letter gives the types of the two
bases at that site in a sample.</p>
<p>The meaning of each of the columns is given in the <code>genotype.readme</code> file:</p>
<pre><code>Column  Description/SampleID
1   Chromosome
2   Position
3   Reference genotype
4   PE00838
5   PE01458
6   PE05720
7   PE05722
8   PE05724
9   PE08084
...</code></pre>
<p>So, we have the equivalent of a CHROM, and and POS and a REF column, then, we get
one column for each sample.</p>
<p>The <code>.genotypes</code> files look something like this, showing only the first few columns:</p>
<div class="sourceCode" id="cb675"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb675-1"><a href="handle-vcf.html#cb675-1" aria-hidden="true" tabindex="-1"></a><span class="ex">chr1</span>    48      A       <span class="at">-</span> <span class="at">-</span> G G R G <span class="at">-</span> <span class="at">-</span> G <span class="at">-</span> <span class="at">-</span> G G G G G G <span class="at">-</span> R G <span class="at">-</span> G R <span class="at">-</span> G A G G G <span class="at">-</span></span>
<span id="cb675-2"><a href="handle-vcf.html#cb675-2" aria-hidden="true" tabindex="-1"></a><span class="ex">chr1</span>    79      T       T T T T T T <span class="at">-</span> T T <span class="at">-</span> <span class="at">-</span> T T T T T T T T T <span class="at">-</span> T T T T T T T T T</span>
<span id="cb675-3"><a href="handle-vcf.html#cb675-3" aria-hidden="true" tabindex="-1"></a><span class="ex">chr1</span>    83      C       C C C C C C <span class="at">-</span> C C <span class="at">-</span> <span class="at">-</span> C C C C C C C C C <span class="at">-</span> C C C C C Y C C C</span>
<span id="cb675-4"><a href="handle-vcf.html#cb675-4" aria-hidden="true" tabindex="-1"></a><span class="ex">chr1</span>    95      C       C Y Y T C C <span class="at">-</span> C C <span class="at">-</span> <span class="at">-</span> C C C C C Y C C C <span class="at">-</span> C C C C C C C C C</span>
<span id="cb675-5"><a href="handle-vcf.html#cb675-5" aria-hidden="true" tabindex="-1"></a><span class="ex">chr1</span>    118     C       C C C C C C <span class="at">-</span> C C C C C C C C C C C C C C C C C C C C C C C</span>
<span id="cb675-6"><a href="handle-vcf.html#cb675-6" aria-hidden="true" tabindex="-1"></a><span class="ex">chr1</span>    119     C       C M M M C C <span class="at">-</span> C C C C C C C C C M M C C C C C C M C C C C C</span></code></pre></div>
<p>Importantly, here we see that missing data is denoted by a <code>-</code>.</p>
<p>Some of these sites are multiallelic. We can see that by looking at the distinct genotypes
in each row. Here we can use <code>awk</code> to look at the distinct genotypes in a few of the
multiallelic positions found in the <code>WP.pgchr1_10K.genotype.gz</code> we just downloaded:</p>
<div class="sourceCode" id="cb676"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb676-1"><a href="handle-vcf.html#cb676-1" aria-hidden="true" tabindex="-1"></a><span class="ex">gzcat</span> WP.pgchr1_10K.genotype.gz <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb676-2"><a href="handle-vcf.html#cb676-2" aria-hidden="true" tabindex="-1"></a>  awk <span class="st">&#39;</span></span>
<span id="cb676-3"><a href="handle-vcf.html#cb676-3" aria-hidden="true" tabindex="-1"></a><span class="st">    {</span></span>
<span id="cb676-4"><a href="handle-vcf.html#cb676-4" aria-hidden="true" tabindex="-1"></a><span class="st">      ref=$3; </span></span>
<span id="cb676-5"><a href="handle-vcf.html#cb676-5" aria-hidden="true" tabindex="-1"></a><span class="st">      for(i=4;i&lt;=NF;i++) {</span></span>
<span id="cb676-6"><a href="handle-vcf.html#cb676-6" aria-hidden="true" tabindex="-1"></a><span class="st">         if($i!=&quot;-&quot; &amp;&amp; $i!=ref &amp;&amp; !($i in n)) n[$i]++</span></span>
<span id="cb676-7"><a href="handle-vcf.html#cb676-7" aria-hidden="true" tabindex="-1"></a><span class="st">      } </span></span>
<span id="cb676-8"><a href="handle-vcf.html#cb676-8" aria-hidden="true" tabindex="-1"></a><span class="st">      printf(&quot;%s %s %s&quot;, $1, $2, ref); </span></span>
<span id="cb676-9"><a href="handle-vcf.html#cb676-9" aria-hidden="true" tabindex="-1"></a><span class="st">      for(i in n) printf(&quot; %s&quot;, i); </span></span>
<span id="cb676-10"><a href="handle-vcf.html#cb676-10" aria-hidden="true" tabindex="-1"></a><span class="st">      printf(&quot;\n&quot;);  </span></span>
<span id="cb676-11"><a href="handle-vcf.html#cb676-11" aria-hidden="true" tabindex="-1"></a><span class="st">      delete n</span></span>
<span id="cb676-12"><a href="handle-vcf.html#cb676-12" aria-hidden="true" tabindex="-1"></a><span class="st">    }</span></span>
<span id="cb676-13"><a href="handle-vcf.html#cb676-13" aria-hidden="true" tabindex="-1"></a><span class="st">  &#39;</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;NF&gt;5&#39;</span></span></code></pre></div>
<p>That gives use some output that looks like:</p>
<pre><code>chr1 2118 G A R S
chr1 4023 T A C W Y
chr1 4404 G A C R S
chr1 22272 C A M T W
chr1 22843 C A M T
chr1 50267 C A M T
chr1 52108 C A M T Y
chr1 52610 C A M T Y
chr1 54107 A C G M R
chr1 54192 G A C M R S
...</code></pre>
<p>This is good to know! If we want to convert this to a VCF, we will
have to pay attention to the fact that some sites have multiple alternate alleles.</p>
<div id="why-create-a-vcf-file" class="section level3 hasAnchor" number="22.3.1">
<h3><span class="header-section-number">22.3.1</span> Why create a VCF file?<a href="handle-vcf.html#why-create-a-vcf-file" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>You may ask, why do you need a VCF file? Well, perhaps you wish to annotate these variants,
attaching information about their possible effects, given information about coding sequences
that they might occur in. This can be done by identifying where each position sits with
respect to genomic features like mRNA transcript boundaries, exons, and coding sequence.</p>
<p>Information about those sorts of genomic features can be found in a <code>.gff</code> file. The
extension <code>.gff</code> stands for “Genomics Features File.” The pearl millet GFF for the
genome that the variants in the <code>.genotypes</code> were mapped against can
be found at:
<a href="https://cegresources.icrisat.org/data_public/PearlMillet_Genome/v1.0/pm_assembly_v1.0.gff">https://cegresources.icrisat.org/data_public/PearlMillet_Genome/v1.0/pm_assembly_v1.0.gff</a>.</p>
<p>The first few lines of that file look like:</p>
<pre><code>chr1    GLEAN   mRNA    29490538    29492233    0.670065    +   .   ID=Pgl_GLEAN_10008614;
chr1    GLEAN   CDS 29490538    29490605    .   +   0   Parent=Pgl_GLEAN_10008614;
chr1    GLEAN   CDS 29491108    29491390    .   +   1   Parent=Pgl_GLEAN_10008614;
chr1    GLEAN   CDS 29491994    29492233    .   +   0   Parent=Pgl_GLEAN_10008614;
chr1    GLEAN   mRNA    244026666   244030597   0.57488 +   .   ID=Pgl_GLEAN_10020091;
chr1    GLEAN   CDS 244026666   244027096   .   +   0   Parent=Pgl_GLEAN_10020091;
chr1    GLEAN   CDS 244027627   244028075   .   +   1   Parent=Pgl_GLEAN_10020091;
chr1    GLEAN   CDS 244030287   244030597   .   +   2   Parent=Pgl_GLEAN_10020091;
chr1    GLEAN   mRNA    224685509   224689330   0.653208    +   .   ID=Pgl_GLEAN_10021721;
chr1    GLEAN   CDS 224685509   224685562   .   +   0   Parent=Pgl_GLEAN_10021721;
chr1    GLEAN   CDS 224687184   224687617   .   +   0   Parent=Pgl_GLEAN_10021721;
chr1    GLEAN   CDS 224687878   224688208   .   +   1   Parent=Pgl_GLEAN_10021721;
chr1    GLEAN   CDS 224688317   224688388   .   +   0   Parent=Pgl_GLEAN_10021721;
chr1    GLEAN   CDS 224688463   224688664   .   +   0   Parent=Pgl_GLEAN_10021721;
chr1    GLEAN   CDS 224688789   224689330   .   +   2   Parent=Pgl_GLEAN_10021721;</code></pre>
<p>Each row shows the start and stop position of some feature in the genome.
Each of these features may themelves be sub-features of some other parent
feature. For example, the first feature is an mRNA transcript
with the ID <code>Pgl_GLEAN_10008614</code>. Next there are several coding sequences
(CDS) that are subfeatures of the <code>Pgl_GLEAN_10008614</code> transcript, which
can be seen because they have <code>Pgl_GLEAN_10008614</code> as their parent.</p>
<p>Trying to figure out whether a variant occurs in a position within a coding
sequence that will create an amino acid change can
be done by figuring out whether the variant is in a coding sequence and whether the
sequence change causes a non-synonymous change in the codon that it is in, etc. Doing so
by writing your own scripts is hard and subject to a lot of error. (I know, because
I spent a good several days doing that some years ago, and then found I had made a lot
of mistakes when I went to check the interesting sites I found on the
GenBank genome browser!) There is
a lot to figure out because the coding sequences are translated from either the plus (+)
or the minus (-) strand relative to the reference genome, etc.</p>
<p>It is much easier to use a program like <a href="http://pcingola.github.io/SnpEff/">snpEff</a>
to do that for you. <em>However</em>, snpEff takes VCF as the input format. Hence,
one might want to convert a <code>.genotypes</code> file to VCF.</p>
</div>
<div id="converting-the-.genotypes-file-to-vcf" class="section level3 hasAnchor" number="22.3.2">
<h3><span class="header-section-number">22.3.2</span> Converting the <code>.genotypes</code> file to VCF<a href="handle-vcf.html#converting-the-.genotypes-file-to-vcf" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>There are lots of ways we could do this. Today we are going to use awk, because
it is fast and can process the file one line at a time, so it has a very
small memory footprint.</p>
<p>Obviously, the <code>.genotypes</code> file really only has information about the
genotypes themselves, so when we make a VCF file out of it, it is going to
be pretty stripped down. In fact, this is really going to become a
minimal VCF file. There is so little information in it, that the
header for it will be just the minimal header:</p>
<pre><code>##fileformat=VCFv4.2
##FORMAT=&lt;ID=GT,Number=1,Type=String,Description=&quot;Genotype&quot;&gt;</code></pre>
<p>and then it will have the #CHROM header line:</p>
<pre><code>#CHROM  POS ID  REF ALT QUAL    FILTER  INFO    FORMAT  PE00838 PE01458 PE05720 PE05722 ...</code></pre>
<p>And then we will start with the variant lines that will look like this:</p>
<pre><code>chr1    48  .   A   G   100 PASS    .   GT  ./. ./. 1/1 1/1 0/1 ...</code></pre>
<p>For each variant, we have a CHROM and a POS. We don’t have an ID, so
those will all be set to <code>.</code>. We have a REF and an ALT, and for the QUAL column
we will just arbitrarily put in a <code>100</code>, and in the FILTER column we will give each
variant a <code>PASS</code>. There is no other information about these variants so
we list the INFO with <code>.</code> (missing data). Then, the FORMAT column lists only the single
field, <code>GT</code>, which is described in our minimal header, and then the actual genotypes
get listed.</p>
<p>To make such a VCF file, we write an awk script that first prints
the header lines, and then, for each variant, it does two passes through
the information in the <code>.genotypes</code> file:</p>
<ol style="list-style-type: decimal">
<li>A first pass to figure out what the reference and all the different
alternate alleles are. Numbers are assigned to the different alternate
alleles at this step.</li>
<li>A second pass to convert each column to something that looks like <code>0/1</code> or <code>1/1</code>, etc.</li>
</ol>
<p>Before we proceed, let’s confirm that only the 2-nucleotide IUPAC codes are used, which makes sense, because
these are diploids.</p>
<div class="sourceCode" id="cb682"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb682-1"><a href="handle-vcf.html#cb682-1" aria-hidden="true" tabindex="-1"></a><span class="ex">gzcat</span> WP.pgchr1.genotype_10K.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;{for(i=3;i&lt;=NF;i++) n[$i]++} NR&gt;100000 {for(i in n) print i, n[i]; exit}&#39;</span></span>
<span id="cb682-2"><a href="handle-vcf.html#cb682-2" aria-hidden="true" tabindex="-1"></a><span class="ex">A</span> 5240464</span>
<span id="cb682-3"><a href="handle-vcf.html#cb682-3" aria-hidden="true" tabindex="-1"></a><span class="ex">C</span> 7949796</span>
<span id="cb682-4"><a href="handle-vcf.html#cb682-4" aria-hidden="true" tabindex="-1"></a><span class="ex">G</span> 7852659</span>
<span id="cb682-5"><a href="handle-vcf.html#cb682-5" aria-hidden="true" tabindex="-1"></a><span class="ex">K</span> 85767</span>
<span id="cb682-6"><a href="handle-vcf.html#cb682-6" aria-hidden="true" tabindex="-1"></a><span class="ex">M</span> 83639</span>
<span id="cb682-7"><a href="handle-vcf.html#cb682-7" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> 263234</span>
<span id="cb682-8"><a href="handle-vcf.html#cb682-8" aria-hidden="true" tabindex="-1"></a><span class="ex">S</span> 78251</span>
<span id="cb682-9"><a href="handle-vcf.html#cb682-9" aria-hidden="true" tabindex="-1"></a><span class="ex">T</span> 5274252</span>
<span id="cb682-10"><a href="handle-vcf.html#cb682-10" aria-hidden="true" tabindex="-1"></a><span class="ex">W</span> 79988</span>
<span id="cb682-11"><a href="handle-vcf.html#cb682-11" aria-hidden="true" tabindex="-1"></a><span class="ex">Y</span> 265687</span>
<span id="cb682-12"><a href="handle-vcf.html#cb682-12" aria-hidden="true" tabindex="-1"></a><span class="ex">-</span> 10526640</span></code></pre></div>
<p>So, we need a way of converting IUPAC codes to alleles.
Below, for each IUPAC-code-genotype given in the first column, the
first and second alleles are given in the second and third columns:</p>
<div class="sourceCode" id="cb683"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb683-1"><a href="handle-vcf.html#cb683-1" aria-hidden="true" tabindex="-1"></a><span class="ex">A</span> A A</span>
<span id="cb683-2"><a href="handle-vcf.html#cb683-2" aria-hidden="true" tabindex="-1"></a><span class="ex">C</span> C C</span>
<span id="cb683-3"><a href="handle-vcf.html#cb683-3" aria-hidden="true" tabindex="-1"></a><span class="ex">G</span> G G</span>
<span id="cb683-4"><a href="handle-vcf.html#cb683-4" aria-hidden="true" tabindex="-1"></a><span class="ex">T</span> T T</span>
<span id="cb683-5"><a href="handle-vcf.html#cb683-5" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> A G</span>
<span id="cb683-6"><a href="handle-vcf.html#cb683-6" aria-hidden="true" tabindex="-1"></a><span class="ex">Y</span> C T</span>
<span id="cb683-7"><a href="handle-vcf.html#cb683-7" aria-hidden="true" tabindex="-1"></a><span class="ex">S</span> G C</span>
<span id="cb683-8"><a href="handle-vcf.html#cb683-8" aria-hidden="true" tabindex="-1"></a><span class="ex">W</span> A T</span>
<span id="cb683-9"><a href="handle-vcf.html#cb683-9" aria-hidden="true" tabindex="-1"></a><span class="ex">K</span> G T</span>
<span id="cb683-10"><a href="handle-vcf.html#cb683-10" aria-hidden="true" tabindex="-1"></a><span class="ex">M</span> A C</span></code></pre></div>
<p>If we had such a table stored in a file <code>extra/iupac.txt</code>, we could
use <code>awk</code> to write a block of <code>awk</code>
code that defines some awk arrays giving the values of the first and second
allele for each. That would look like this:</p>
<div class="sourceCode" id="cb684"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb684-1"><a href="handle-vcf.html#cb684-1" aria-hidden="true" tabindex="-1"></a> <span class="fu">awk</span> <span class="st">&#39;{printf(&quot;iupac[\&quot;%s\&quot;,1] = \&quot;%s\&quot;;\n&quot;, $1, $2); printf(&quot;iupac[\&quot;%s\&quot;,2] = \&quot;%s\&quot;;\n\n&quot;, $1, $3);}&#39;</span> extra/iupac.txt</span>
<span id="cb684-2"><a href="handle-vcf.html#cb684-2" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;A&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;A&quot;</span><span class="kw">;</span></span>
<span id="cb684-3"><a href="handle-vcf.html#cb684-3" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;A&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;A&quot;</span><span class="kw">;</span></span>
<span id="cb684-4"><a href="handle-vcf.html#cb684-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb684-5"><a href="handle-vcf.html#cb684-5" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;C&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;C&quot;</span><span class="kw">;</span></span>
<span id="cb684-6"><a href="handle-vcf.html#cb684-6" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;C&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;C&quot;</span><span class="kw">;</span></span>
<span id="cb684-7"><a href="handle-vcf.html#cb684-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb684-8"><a href="handle-vcf.html#cb684-8" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;G&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;G&quot;</span><span class="kw">;</span></span>
<span id="cb684-9"><a href="handle-vcf.html#cb684-9" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;G&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;G&quot;</span><span class="kw">;</span></span>
<span id="cb684-10"><a href="handle-vcf.html#cb684-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb684-11"><a href="handle-vcf.html#cb684-11" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;T&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;T&quot;</span><span class="kw">;</span></span>
<span id="cb684-12"><a href="handle-vcf.html#cb684-12" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;T&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;T&quot;</span><span class="kw">;</span></span>
<span id="cb684-13"><a href="handle-vcf.html#cb684-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb684-14"><a href="handle-vcf.html#cb684-14" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;R&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;A&quot;</span><span class="kw">;</span></span>
<span id="cb684-15"><a href="handle-vcf.html#cb684-15" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;R&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;G&quot;</span><span class="kw">;</span></span>
<span id="cb684-16"><a href="handle-vcf.html#cb684-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb684-17"><a href="handle-vcf.html#cb684-17" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;Y&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;C&quot;</span><span class="kw">;</span></span>
<span id="cb684-18"><a href="handle-vcf.html#cb684-18" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;Y&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;T&quot;</span><span class="kw">;</span></span>
<span id="cb684-19"><a href="handle-vcf.html#cb684-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb684-20"><a href="handle-vcf.html#cb684-20" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;S&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;G&quot;</span><span class="kw">;</span></span>
<span id="cb684-21"><a href="handle-vcf.html#cb684-21" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;S&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;C&quot;</span><span class="kw">;</span></span>
<span id="cb684-22"><a href="handle-vcf.html#cb684-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb684-23"><a href="handle-vcf.html#cb684-23" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;W&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;A&quot;</span><span class="kw">;</span></span>
<span id="cb684-24"><a href="handle-vcf.html#cb684-24" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;W&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;T&quot;</span><span class="kw">;</span></span>
<span id="cb684-25"><a href="handle-vcf.html#cb684-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb684-26"><a href="handle-vcf.html#cb684-26" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;K&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;G&quot;</span><span class="kw">;</span></span>
<span id="cb684-27"><a href="handle-vcf.html#cb684-27" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;K&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;T&quot;</span><span class="kw">;</span></span>
<span id="cb684-28"><a href="handle-vcf.html#cb684-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb684-29"><a href="handle-vcf.html#cb684-29" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;M&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;A&quot;</span><span class="kw">;</span></span>
<span id="cb684-30"><a href="handle-vcf.html#cb684-30" aria-hidden="true" tabindex="-1"></a><span class="va">iupac</span><span class="op">[</span><span class="st">&quot;M&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;C&quot;</span><span class="kw">;</span></span></code></pre></div>
<p>Then we can put that code into a little awk script that we will call
<code>genotypes2vcf.awk</code>. You can copy the code in the following
block and put it into a file called <code>script/genotypes2vcf.awk</code> (i.e., it
is called <code>genotypes2vcf.awk</code> inside a directory you made called <code>script</code>):</p>
<div class="sourceCode" id="cb685"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb685-1"><a href="handle-vcf.html#cb685-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we begin by creating arrays that tell us the first and second allele</span></span>
<span id="cb685-2"><a href="handle-vcf.html#cb685-2" aria-hidden="true" tabindex="-1"></a><span class="co"># of each IUPAC-coded genotype</span></span>
<span id="cb685-3"><a href="handle-vcf.html#cb685-3" aria-hidden="true" tabindex="-1"></a><span class="ex">BEGIN</span> {</span>
<span id="cb685-4"><a href="handle-vcf.html#cb685-4" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;A&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;A&quot;</span><span class="kw">;</span></span>
<span id="cb685-5"><a href="handle-vcf.html#cb685-5" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;A&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;A&quot;</span><span class="kw">;</span></span>
<span id="cb685-6"><a href="handle-vcf.html#cb685-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-7"><a href="handle-vcf.html#cb685-7" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;C&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;C&quot;</span><span class="kw">;</span></span>
<span id="cb685-8"><a href="handle-vcf.html#cb685-8" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;C&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;C&quot;</span><span class="kw">;</span></span>
<span id="cb685-9"><a href="handle-vcf.html#cb685-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-10"><a href="handle-vcf.html#cb685-10" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;G&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;G&quot;</span><span class="kw">;</span></span>
<span id="cb685-11"><a href="handle-vcf.html#cb685-11" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;G&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;G&quot;</span><span class="kw">;</span></span>
<span id="cb685-12"><a href="handle-vcf.html#cb685-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-13"><a href="handle-vcf.html#cb685-13" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;T&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;T&quot;</span><span class="kw">;</span></span>
<span id="cb685-14"><a href="handle-vcf.html#cb685-14" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;T&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;T&quot;</span><span class="kw">;</span></span>
<span id="cb685-15"><a href="handle-vcf.html#cb685-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-16"><a href="handle-vcf.html#cb685-16" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;R&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;A&quot;</span><span class="kw">;</span></span>
<span id="cb685-17"><a href="handle-vcf.html#cb685-17" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;R&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;G&quot;</span><span class="kw">;</span></span>
<span id="cb685-18"><a href="handle-vcf.html#cb685-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-19"><a href="handle-vcf.html#cb685-19" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;Y&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;C&quot;</span><span class="kw">;</span></span>
<span id="cb685-20"><a href="handle-vcf.html#cb685-20" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;Y&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;T&quot;</span><span class="kw">;</span></span>
<span id="cb685-21"><a href="handle-vcf.html#cb685-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-22"><a href="handle-vcf.html#cb685-22" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;S&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;G&quot;</span><span class="kw">;</span></span>
<span id="cb685-23"><a href="handle-vcf.html#cb685-23" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;S&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;C&quot;</span><span class="kw">;</span></span>
<span id="cb685-24"><a href="handle-vcf.html#cb685-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-25"><a href="handle-vcf.html#cb685-25" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;W&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;A&quot;</span><span class="kw">;</span></span>
<span id="cb685-26"><a href="handle-vcf.html#cb685-26" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;W&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;T&quot;</span><span class="kw">;</span></span>
<span id="cb685-27"><a href="handle-vcf.html#cb685-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-28"><a href="handle-vcf.html#cb685-28" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;K&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;G&quot;</span><span class="kw">;</span></span>
<span id="cb685-29"><a href="handle-vcf.html#cb685-29" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;K&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;T&quot;</span><span class="kw">;</span></span>
<span id="cb685-30"><a href="handle-vcf.html#cb685-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-31"><a href="handle-vcf.html#cb685-31" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;M&quot;</span>,1<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;A&quot;</span><span class="kw">;</span></span>
<span id="cb685-32"><a href="handle-vcf.html#cb685-32" aria-hidden="true" tabindex="-1"></a>  <span class="va">iupac</span><span class="op">[</span><span class="st">&quot;M&quot;</span>,2<span class="op">]</span> <span class="ex">=</span> <span class="st">&quot;C&quot;</span><span class="kw">;</span></span>
<span id="cb685-33"><a href="handle-vcf.html#cb685-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-34"><a href="handle-vcf.html#cb685-34" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Now, print the minimal header lines</span></span>
<span id="cb685-35"><a href="handle-vcf.html#cb685-35" aria-hidden="true" tabindex="-1"></a>  <span class="bu">printf</span><span class="er">(</span><span class="st">&quot;##fileformat=VCFv4.2\n&quot;</span><span class="kw">)</span></span>
<span id="cb685-36"><a href="handle-vcf.html#cb685-36" aria-hidden="true" tabindex="-1"></a>  <span class="bu">printf</span><span class="er">(</span><span class="st">&quot;##FORMAT=&lt;ID=GT,Number=1,Type=String,Description=</span><span class="dt">\&quot;</span><span class="st">Genotype</span><span class="dt">\&quot;</span><span class="st">&gt;\n&quot;</span><span class="kw">)</span></span>
<span id="cb685-37"><a href="handle-vcf.html#cb685-37" aria-hidden="true" tabindex="-1"></a>  <span class="bu">printf</span><span class="er">(</span><span class="st">&quot;#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT&quot;</span><span class="kw">)</span></span>
<span id="cb685-38"><a href="handle-vcf.html#cb685-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-39"><a href="handle-vcf.html#cb685-39" aria-hidden="true" tabindex="-1"></a><span class="er">}</span></span>
<span id="cb685-40"><a href="handle-vcf.html#cb685-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-41"><a href="handle-vcf.html#cb685-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-42"><a href="handle-vcf.html#cb685-42" aria-hidden="true" tabindex="-1"></a><span class="co"># we catenate the sample names, each on a line starting with SAMPLE</span></span>
<span id="cb685-43"><a href="handle-vcf.html#cb685-43" aria-hidden="true" tabindex="-1"></a><span class="co"># to the input, and then print them all here</span></span>
<span id="cb685-44"><a href="handle-vcf.html#cb685-44" aria-hidden="true" tabindex="-1"></a><span class="va">$1</span><span class="ex">==</span><span class="st">&quot;SAMPLE&quot;</span> {</span>
<span id="cb685-45"><a href="handle-vcf.html#cb685-45" aria-hidden="true" tabindex="-1"></a>  <span class="bu">printf</span><span class="er">(</span><span class="st">&quot;\t%s&quot;</span><span class="ex">,</span> <span class="va">$NF</span><span class="kw">);</span></span>
<span id="cb685-46"><a href="handle-vcf.html#cb685-46" aria-hidden="true" tabindex="-1"></a>  <span class="ex">next</span></span>
<span id="cb685-47"><a href="handle-vcf.html#cb685-47" aria-hidden="true" tabindex="-1"></a><span class="er">}</span></span>
<span id="cb685-48"><a href="handle-vcf.html#cb685-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-49"><a href="handle-vcf.html#cb685-49" aria-hidden="true" tabindex="-1"></a><span class="ex">/xxxxxxxxxxxxxx/</span> {printf<span class="er">(</span><span class="st">&quot;\n&quot;</span><span class="kw">);</span> <span class="ex">next}</span></span>
<span id="cb685-50"><a href="handle-vcf.html#cb685-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-51"><a href="handle-vcf.html#cb685-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-52"><a href="handle-vcf.html#cb685-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-53"><a href="handle-vcf.html#cb685-53" aria-hidden="true" tabindex="-1"></a><span class="co"># then, for each row, we make a first pass and define integers for the</span></span>
<span id="cb685-54"><a href="handle-vcf.html#cb685-54" aria-hidden="true" tabindex="-1"></a><span class="co"># different alternate alleles. Then we make a second pass and print</span></span>
<span id="cb685-55"><a href="handle-vcf.html#cb685-55" aria-hidden="true" tabindex="-1"></a><span class="co"># the VCF line</span></span>
<span id="cb685-56"><a href="handle-vcf.html#cb685-56" aria-hidden="true" tabindex="-1"></a><span class="kw">{</span></span>
<span id="cb685-57"><a href="handle-vcf.html#cb685-57" aria-hidden="true" tabindex="-1"></a>  <span class="va">CHROM=$1</span><span class="kw">;</span></span>
<span id="cb685-58"><a href="handle-vcf.html#cb685-58" aria-hidden="true" tabindex="-1"></a>  <span class="va">POS=$2</span></span>
<span id="cb685-59"><a href="handle-vcf.html#cb685-59" aria-hidden="true" tabindex="-1"></a>  <span class="ex">delete</span> a_ints<span class="kw">;</span> <span class="co"># clear these arrays</span></span>
<span id="cb685-60"><a href="handle-vcf.html#cb685-60" aria-hidden="true" tabindex="-1"></a>  <span class="ex">delete</span> alleles<span class="kw">;</span></span>
<span id="cb685-61"><a href="handle-vcf.html#cb685-61" aria-hidden="true" tabindex="-1"></a>  <span class="ex">a_idx</span> = 0<span class="kw">;</span>  <span class="co"># start the allele index at 0 (the REF)</span></span>
<span id="cb685-62"><a href="handle-vcf.html#cb685-62" aria-hidden="true" tabindex="-1"></a>  <span class="va">alleles</span><span class="op">[</span><span class="va">$3</span><span class="op">]</span> <span class="ex">=</span> a_idx<span class="kw">;</span> <span class="co"># reference allele gets a 0</span></span>
<span id="cb685-63"><a href="handle-vcf.html#cb685-63" aria-hidden="true" tabindex="-1"></a>  <span class="va">a_ints</span><span class="op">[</span>a_idx<span class="op">]</span> <span class="ex">=</span> <span class="va">$3</span><span class="kw">;</span>  <span class="co"># keep this array to have alleles in sorted order,</span></span>
<span id="cb685-64"><a href="handle-vcf.html#cb685-64" aria-hidden="true" tabindex="-1"></a>                       <span class="co"># indexed from 0 to the number of alleles - 1</span></span>
<span id="cb685-65"><a href="handle-vcf.html#cb685-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-66"><a href="handle-vcf.html#cb685-66" aria-hidden="true" tabindex="-1"></a>  <span class="co"># cycle over all the columns, and the two alleles within each genotype</span></span>
<span id="cb685-67"><a href="handle-vcf.html#cb685-67" aria-hidden="true" tabindex="-1"></a>  <span class="co"># in that column, and add any new alleles found to the alleles hash array</span></span>
<span id="cb685-68"><a href="handle-vcf.html#cb685-68" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span><span class="kw">(</span><span class="va">i=</span>4<span class="kw">;</span><span class="ex">i</span><span class="op">&lt;</span>=NF<span class="kw">;</span><span class="ex">i++</span><span class="kw">)</span> <span class="kw">{</span></span>
<span id="cb685-69"><a href="handle-vcf.html#cb685-69" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span><span class="kw">(</span><span class="va">a=</span>1<span class="kw">;</span><span class="ex">a</span><span class="op">&lt;</span>=2<span class="kw">;</span><span class="ex">a++</span><span class="kw">)</span> <span class="cf">if</span><span class="kw">(</span><span class="va">$i</span> != <span class="st">&quot;-&quot;</span><span class="kw">)</span> <span class="kw">{</span></span>
<span id="cb685-70"><a href="handle-vcf.html#cb685-70" aria-hidden="true" tabindex="-1"></a>      <span class="ex">alle</span> = iupac[<span class="va">$i</span>, a]<span class="kw">;</span></span>
<span id="cb685-71"><a href="handle-vcf.html#cb685-71" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span><span class="kw">(</span><span class="ex">!</span><span class="er">(</span><span class="ex">alle</span> in alleles<span class="kw">))</span> <span class="kw">{</span>  <span class="co"># if we have not seen this allele before</span></span>
<span id="cb685-72"><a href="handle-vcf.html#cb685-72" aria-hidden="true" tabindex="-1"></a>        <span class="va">alleles</span><span class="op">[</span>alle<span class="op">]</span> <span class="ex">=</span> ++a_idx<span class="kw">;</span></span>
<span id="cb685-73"><a href="handle-vcf.html#cb685-73" aria-hidden="true" tabindex="-1"></a>        <span class="va">a_ints</span><span class="op">[</span>a_idx<span class="op">]</span> <span class="ex">=</span> alle</span>
<span id="cb685-74"><a href="handle-vcf.html#cb685-74" aria-hidden="true" tabindex="-1"></a>      <span class="kw">}</span></span>
<span id="cb685-75"><a href="handle-vcf.html#cb685-75" aria-hidden="true" tabindex="-1"></a>    <span class="kw">}</span></span>
<span id="cb685-76"><a href="handle-vcf.html#cb685-76" aria-hidden="true" tabindex="-1"></a>  <span class="kw">}</span></span>
<span id="cb685-77"><a href="handle-vcf.html#cb685-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-78"><a href="handle-vcf.html#cb685-78" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Now we can print the VCF line</span></span>
<span id="cb685-79"><a href="handle-vcf.html#cb685-79" aria-hidden="true" tabindex="-1"></a>  <span class="co"># print CHROM, POS, ID, and REF columns</span></span>
<span id="cb685-80"><a href="handle-vcf.html#cb685-80" aria-hidden="true" tabindex="-1"></a>  <span class="bu">printf</span><span class="er">(</span><span class="st">&quot;%s\t%s\t.\t%s&quot;</span><span class="ex">,</span> <span class="va">$1</span>, <span class="va">$2</span>, <span class="va">$3</span><span class="kw">)</span> </span>
<span id="cb685-81"><a href="handle-vcf.html#cb685-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-82"><a href="handle-vcf.html#cb685-82" aria-hidden="true" tabindex="-1"></a>  <span class="co"># print the ALT field, including comma-separated alleles if multiallelic</span></span>
<span id="cb685-83"><a href="handle-vcf.html#cb685-83" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span><span class="kw">(</span><span class="ex">a_idx</span> == 0<span class="kw">)</span> <span class="kw">{</span> <span class="co"># if there are no alternate alleles ALT gets a .</span></span>
<span id="cb685-84"><a href="handle-vcf.html#cb685-84" aria-hidden="true" tabindex="-1"></a>    <span class="bu">printf</span><span class="er">(</span><span class="st">&quot;\t.&quot;</span><span class="kw">)</span></span>
<span id="cb685-85"><a href="handle-vcf.html#cb685-85" aria-hidden="true" tabindex="-1"></a>  <span class="kw">}</span> <span class="cf">else</span> <span class="kw">{</span></span>
<span id="cb685-86"><a href="handle-vcf.html#cb685-86" aria-hidden="true" tabindex="-1"></a>    <span class="bu">printf</span><span class="er">(</span><span class="st">&quot;\t%s&quot;</span><span class="ex">,</span> a_ints[1]<span class="kw">)</span></span>
<span id="cb685-87"><a href="handle-vcf.html#cb685-87" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span><span class="kw">(</span><span class="va">a=</span>2<span class="kw">;</span><span class="ex">a</span><span class="op">&lt;</span>=a_idx<span class="kw">;</span><span class="ex">a++</span><span class="kw">)</span></span>
<span id="cb685-88"><a href="handle-vcf.html#cb685-88" aria-hidden="true" tabindex="-1"></a>      <span class="bu">printf</span><span class="er">(</span><span class="st">&quot;,%s&quot;</span><span class="ex">,</span> a_ints[a]<span class="kw">)</span></span>
<span id="cb685-89"><a href="handle-vcf.html#cb685-89" aria-hidden="true" tabindex="-1"></a>  <span class="kw">}</span></span>
<span id="cb685-90"><a href="handle-vcf.html#cb685-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-91"><a href="handle-vcf.html#cb685-91" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Set all the QUALs to 100, the FILTERs to PASS and the INFO to .</span></span>
<span id="cb685-92"><a href="handle-vcf.html#cb685-92" aria-hidden="true" tabindex="-1"></a>  <span class="bu">printf</span><span class="er">(</span><span class="st">&quot;\t100\tPASS\t.&quot;</span><span class="kw">)</span></span>
<span id="cb685-93"><a href="handle-vcf.html#cb685-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-94"><a href="handle-vcf.html#cb685-94" aria-hidden="true" tabindex="-1"></a>  <span class="co"># make the FORMAT column.  It just has GT</span></span>
<span id="cb685-95"><a href="handle-vcf.html#cb685-95" aria-hidden="true" tabindex="-1"></a>  <span class="bu">printf</span><span class="er">(</span><span class="st">&quot;\tGT&quot;</span><span class="kw">)</span></span>
<span id="cb685-96"><a href="handle-vcf.html#cb685-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb685-97"><a href="handle-vcf.html#cb685-97" aria-hidden="true" tabindex="-1"></a>  <span class="co"># now, cycle over the individuals and print their genotypes</span></span>
<span id="cb685-98"><a href="handle-vcf.html#cb685-98" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span><span class="kw">(</span><span class="va">i=</span>4<span class="kw">;</span><span class="ex">i</span><span class="op">&lt;</span>=NF<span class="kw">;</span><span class="ex">i++</span><span class="kw">)</span> <span class="kw">{</span></span>
<span id="cb685-99"><a href="handle-vcf.html#cb685-99" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span><span class="kw">(</span><span class="va">$i</span><span class="ex">==</span><span class="st">&quot;-&quot;</span><span class="kw">)</span> <span class="kw">{</span></span>
<span id="cb685-100"><a href="handle-vcf.html#cb685-100" aria-hidden="true" tabindex="-1"></a>      <span class="bu">printf</span><span class="er">(</span><span class="st">&quot;\t./.&quot;</span><span class="kw">);</span></span>
<span id="cb685-101"><a href="handle-vcf.html#cb685-101" aria-hidden="true" tabindex="-1"></a>    <span class="kw">}</span> <span class="cf">else</span> <span class="kw">{</span></span>
<span id="cb685-102"><a href="handle-vcf.html#cb685-102" aria-hidden="true" tabindex="-1"></a>      <span class="ex">a</span> = iupac[<span class="va">$i</span>, 1]<span class="kw">;</span></span>
<span id="cb685-103"><a href="handle-vcf.html#cb685-103" aria-hidden="true" tabindex="-1"></a>      <span class="ex">b</span> = iupac[<span class="va">$i</span>, 2]<span class="kw">;</span></span>
<span id="cb685-104"><a href="handle-vcf.html#cb685-104" aria-hidden="true" tabindex="-1"></a>      <span class="bu">printf</span><span class="er">(</span><span class="st">&quot;\t%s/%s&quot;</span><span class="ex">,</span> alleles[a], alleles[b]<span class="kw">)</span></span>
<span id="cb685-105"><a href="handle-vcf.html#cb685-105" aria-hidden="true" tabindex="-1"></a>    <span class="kw">}</span></span>
<span id="cb685-106"><a href="handle-vcf.html#cb685-106" aria-hidden="true" tabindex="-1"></a>  <span class="kw">}</span></span>
<span id="cb685-107"><a href="handle-vcf.html#cb685-107" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb685-108"><a href="handle-vcf.html#cb685-108" aria-hidden="true" tabindex="-1"></a>  <span class="bu">printf</span><span class="er">(</span><span class="st">&quot;\n&quot;</span><span class="kw">)</span></span>
<span id="cb685-109"><a href="handle-vcf.html#cb685-109" aria-hidden="true" tabindex="-1"></a><span class="kw">}</span></span></code></pre></div>
<p>In order to use that <code>awk</code> script we will use <code>awk</code>’s <code>-f</code> option
to read the script, and we will also
pass to <code>awk</code> via <em>stdin</em> the genotypes file with the sample names on top of it followed
by a line with an <code>xxxxxxxxxxxxxxxxx</code></p>
<div class="sourceCode" id="cb686"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb686-1"><a href="handle-vcf.html#cb686-1" aria-hidden="true" tabindex="-1"></a><span class="kw">(</span></span>
<span id="cb686-2"><a href="handle-vcf.html#cb686-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">awk</span> <span class="st">&#39;NR&gt;=5 {printf(&quot;SAMPLE %s\n&quot;, $2);}&#39;</span> genotype.readme<span class="kw">;</span> </span>
<span id="cb686-3"><a href="handle-vcf.html#cb686-3" aria-hidden="true" tabindex="-1"></a>  <span class="bu">echo</span> xxxxxxxxxxxxxxxxx<span class="kw">;</span> </span>
<span id="cb686-4"><a href="handle-vcf.html#cb686-4" aria-hidden="true" tabindex="-1"></a>  <span class="ex">gzcat</span> WP.pgchr1_10K.genotype.gz</span>
<span id="cb686-5"><a href="handle-vcf.html#cb686-5" aria-hidden="true" tabindex="-1"></a><span class="kw">)</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="at">-f</span> script/genotypes2vcf.awk <span class="op">&gt;</span> WP.pgchr1_10K.vcf</span></code></pre></div>
<p>That creates the VCF file, <code>WP.pgchr1_10K.vcf</code>. Woo-hoo! On my old laptop, that takes about 10 seconds. So, about 1 second per 1000 variants.
That means that if you have 3.6 million variants, it will take about an hour to
convert them all to this VCF format. But once they are in VCF, you can
compress and index them
and manipulate them quite quickly.</p>
</div>
<div id="reheader-the-vcf" class="section level3 hasAnchor" number="22.3.3">
<h3><span class="header-section-number">22.3.3</span> Reheader the VCF<a href="handle-vcf.html#reheader-the-vcf" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>It is good practice to have the reference genome information in the VCF header.
We can do that with the <code>bcftools reheader</code> subcommand. To do so, you need to have
the <code>.fai</code> file made by running <code>samtools faidx</code> on the reference genome. You can
download the <code>.fai</code> file from:
<a href="https://eriqande.github.io/eca-bioinf-handbook/downloads/pearl_millet_v1.1.fa.fai">https://eriqande.github.io/eca-bioinf-handbook/downloads/pearl_millet_v1.1.fa.fai</a>
or, on your cluster, use:</p>
<div class="sourceCode" id="cb687"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb687-1"><a href="handle-vcf.html#cb687-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://eriqande.github.io/eca-bioinf-handbook/downloads/pearl_millet_v1.1.fa.fai</span></code></pre></div>
<p>Once we have that, we can reheader the VCF file like this (it is piped to <code>less</code>
just so you can view the result without it bombing your screen).</p>
<div class="sourceCode" id="cb688"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb688-1"><a href="handle-vcf.html#cb688-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> reheader <span class="at">-f</span> pearl_millet_v1.1.fa.fai WP.pgchr1_10K.vcf <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>While we are at it, however, we are going to go one step further and add some
fields to the INFO column. We are going to calculate the alternate allele count (AC)
the total number of alleles of all types (AN) the alternate allele
frequency (AF) and the minor allele
frequency (MAF) for each variant. To add these INFO tags we will use the
<code>fill-tags</code> plug-in for <code>bcftools</code>. This plug-in comes standard with <code>bcftools</code>.
You can see the help information for it like this:</p>
<div class="sourceCode" id="cb689"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb689-1"><a href="handle-vcf.html#cb689-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> +fill-tags <span class="at">-h</span></span></code></pre></div>
<p>For a detailed list of the available tags to fill you can do:</p>
<div class="sourceCode" id="cb690"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb690-1"><a href="handle-vcf.html#cb690-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> +fill-tags <span class="at">--</span> <span class="at">-l</span></span></code></pre></div>
<p>So, we can add the genome to the header and add AN, AC, AF, and MAF fields to the INFO, <em>and</em> compress the file into a vcf.gz, all in one fell swoop with the following, and then
we can index it.</p>
<div class="sourceCode" id="cb691"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb691-1"><a href="handle-vcf.html#cb691-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> reheader <span class="at">-f</span> pearl_millet_v1.1.fa.fai WP.pgchr1_10K.vcf <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb691-2"><a href="handle-vcf.html#cb691-2" aria-hidden="true" tabindex="-1"></a>    bcftools +fill-tags <span class="at">-</span> <span class="at">-Oz</span> <span class="at">-o</span> WP.pgchr1_10K.vcf.gz <span class="at">--</span> <span class="at">-t</span> AN,AC,AF,MAF</span>
<span id="cb691-3"><a href="handle-vcf.html#cb691-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb691-4"><a href="handle-vcf.html#cb691-4" aria-hidden="true" tabindex="-1"></a><span class="co"># then we could index twice to get both a .tbi and a .csi index</span></span>
<span id="cb691-5"><a href="handle-vcf.html#cb691-5" aria-hidden="true" tabindex="-1"></a><span class="co"># like this</span></span>
<span id="cb691-6"><a href="handle-vcf.html#cb691-6" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> index WP.pgchr1_10K.vcf.gz  <span class="co"># makes .csi</span></span>
<span id="cb691-7"><a href="handle-vcf.html#cb691-7" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> index <span class="at">-t</span> WP.pgchr1_10K.vcf.gz  <span class="co"># makes .tbi</span></span></code></pre></div>
<p>And then you could look it at like this:</p>
<div class="sourceCode" id="cb692"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb692-1"><a href="handle-vcf.html#cb692-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view WP.pgchr1_10K.vcf.gz <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>Or you could compute statistics for it like this:</p>
<div class="sourceCode" id="cb693"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb693-1"><a href="handle-vcf.html#cb693-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> stats WP.pgchr1_10K.vcf.gz <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>And, if you wanted to look at only the multiallelic sites you could do this:</p>
<div class="sourceCode" id="cb694"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb694-1"><a href="handle-vcf.html#cb694-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-m</span> 3 WP.pgchr1_10K.vcf.gz <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>Wow! Playing around with this makes it clear that the VCF format is pretty
awesome, and that if you get good at using <code>bcftools</code>, then there are so many
wonderful possibilities.</p>
</div>
<div id="where-to-from-here" class="section level3 hasAnchor" number="22.3.4">
<h3><span class="header-section-number">22.3.4</span> Where to from here?<a href="handle-vcf.html#where-to-from-here" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>A good exercise, (especially for anyone studying pearl millet!), would be
to write a script that cycled over the different chromosomes and made a VCF
for each one, then use <code>bcftools concat</code> to concatenate them into a single VCF file and add
a few info tags. Once that is done, you would have a VCF file ready
to annotate some variants with <code>snpEff</code>.</p>
<p>To automate this all on all the cluster it is worth noting that the <code>.genotype</code>
files can be downloaded like so:</p>
<div class="sourceCode" id="cb695"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb695-1"><a href="handle-vcf.html#cb695-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">--no-check-certificate</span> https://cegresources.icrisat.org/data_public/PM_SNPs/SNP_calls/WP.pgchr7.genotype.gz</span></code></pre></div>
<p>Perhaps on Thursday we can review a Snakefile that would do all this.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="boneyard-1.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="bioinformatics-for-rad-seq-data-with-and-without-a-reference-genome.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook/js/app.min.js"></script>
<script src="libs/gitbook/js/clipboard.min.js"></script>
<script src="libs/gitbook/js/plugin-search.js"></script>
<script src="libs/gitbook/js/plugin-sharing.js"></script>
<script src="libs/gitbook/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook/js/plugin-bookdown.js"></script>
<script src="libs/gitbook/js/jquery.highlight.js"></script>
<script src="libs/gitbook/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/eriqande/eca-bioinf-handbook/edit/master/3.05.5-basic-handling-of-VCF-files.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/eriqande/eca-bioinf-handbook/commits/master/3.05.5-basic-handling-of-VCF-files.Rmd",
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["eca-bioinf-handbook.pdf", "eca-bioinf-handbook.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
