<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 22 Basic Handling of VCF files | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics</title>
  <meta name="description" content="A book example for a Chapman &amp; Hall book." />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 22 Basic Handling of VCF files | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A book example for a Chapman &amp; Hall book." />
  <meta name="github-repo" content="yihui/bookdown-crc" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 22 Basic Handling of VCF files | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics" />
  
  <meta name="twitter:description" content="A book example for a Chapman &amp; Hall book." />
  

<meta name="author" content="Eric C. Anderson" />


<meta name="date" content="2022-04-18" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="boneyard-1.html"/>
<link rel="next" href="bioinformatics-for-rad-seq-data-with-and-without-a-reference-genome.html"/>
<script src="libs/header-attrs/header-attrs.js"></script>
<script src="libs/jquery/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook/css/style.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ECA's Bioinformatics Handbook</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="erics-notes-of-what-he-might-do.html"><a href="erics-notes-of-what-he-might-do.html"><i class="fa fa-check"></i><b>2</b> Eric’s Notes of what he might do</a>
<ul>
<li class="chapter" data-level="2.1" data-path="erics-notes-of-what-he-might-do.html"><a href="erics-notes-of-what-he-might-do.html#table-of-topics"><i class="fa fa-check"></i><b>2.1</b> Table of topics</a></li>
</ul></li>
<li class="part"><span><b>I Part I: Essential Computing Skills</b></span></li>
<li class="chapter" data-level="3" data-path="overview-of-essential-computing-skills.html"><a href="overview-of-essential-computing-skills.html"><i class="fa fa-check"></i><b>3</b> Overview of Essential Computing Skills</a></li>
<li class="chapter" data-level="4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html"><i class="fa fa-check"></i><b>4</b> Essential Unix/Linux Terminal Knowledge</a>
<ul>
<li class="chapter" data-level="4.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#getting-a-bash-shell-on-your-system"><i class="fa fa-check"></i><b>4.1</b> Getting a bash shell on your system</a></li>
<li class="chapter" data-level="4.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#navigating-the-unix-filesystem"><i class="fa fa-check"></i><b>4.2</b> Navigating the Unix filesystem</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#changing-the-working-directory-with-cd"><i class="fa fa-check"></i><b>4.2.1</b> Changing the working directory with <code>cd</code></a></li>
<li class="chapter" data-level="4.2.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#comm-prompt"><i class="fa fa-check"></i><b>4.2.2</b> Updating your command prompt</a></li>
<li class="chapter" data-level="4.2.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#tab-completion-for-paths"><i class="fa fa-check"></i><b>4.2.3</b> TAB-completion for paths</a></li>
<li class="chapter" data-level="4.2.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#listing-the-contents-of-a-directory-with-ls"><i class="fa fa-check"></i><b>4.2.4</b> Listing the contents of a directory with <code>ls</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#globbing"><i class="fa fa-check"></i><b>4.2.5</b> Globbing</a></li>
<li class="chapter" data-level="4.2.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#what-makes-a-good-file-name"><i class="fa fa-check"></i><b>4.2.6</b> What makes a good file-name?</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#the-anatomy-of-a-unix-command"><i class="fa fa-check"></i><b>4.3</b> The anatomy of a Unix command</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#anatomy-command"><i class="fa fa-check"></i><b>4.3.1</b> The <code>command</code></a></li>
<li class="chapter" data-level="4.3.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#the-options"><i class="fa fa-check"></i><b>4.3.2</b> The <em>options</em></a></li>
<li class="chapter" data-level="4.3.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#arguments"><i class="fa fa-check"></i><b>4.3.3</b> Arguments</a></li>
<li class="chapter" data-level="4.3.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#getting-information-about-unix-commands"><i class="fa fa-check"></i><b>4.3.4</b> Getting information about Unix commands</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#handling-manipulating-and-viewing-files-and-streams"><i class="fa fa-check"></i><b>4.4</b> Handling, Manipulating, and Viewing files and streams</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#creating-new-directories"><i class="fa fa-check"></i><b>4.4.1</b> Creating new directories</a></li>
<li class="chapter" data-level="4.4.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#fundamental-file-handling-commands"><i class="fa fa-check"></i><b>4.4.2</b> Fundamental file-handling commands</a></li>
<li class="chapter" data-level="4.4.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#viewing-files"><i class="fa fa-check"></i><b>4.4.3</b> “Viewing” Files</a></li>
<li class="chapter" data-level="4.4.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#redirecting-standard-output-and"><i class="fa fa-check"></i><b>4.4.4</b> Redirecting standard output: <code>&gt;</code> and <code>&gt;&gt;</code></a></li>
<li class="chapter" data-level="4.4.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#stdin-and"><i class="fa fa-check"></i><b>4.4.5</b> stdin, <code>&lt;</code> and <code>|</code></a></li>
<li class="chapter" data-level="4.4.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#stderr"><i class="fa fa-check"></i><b>4.4.6</b> stderr</a></li>
<li class="chapter" data-level="4.4.7" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#symbolic-links"><i class="fa fa-check"></i><b>4.4.7</b> Symbolic links</a></li>
<li class="chapter" data-level="4.4.8" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#file-perm"><i class="fa fa-check"></i><b>4.4.8</b> File Permissions</a></li>
<li class="chapter" data-level="4.4.9" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#editing-text-files-at-the-terminal"><i class="fa fa-check"></i><b>4.4.9</b> Editing text files at the terminal</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#unix-env"><i class="fa fa-check"></i><b>4.5</b> Customizing your Environment</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#appearances-matter"><i class="fa fa-check"></i><b>4.5.1</b> Appearances matter</a></li>
<li class="chapter" data-level="4.5.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#where-are-my-programscommands-at"><i class="fa fa-check"></i><b>4.5.2</b> Where are my programs/commands at?!</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#a-few-more-important-keystrokes"><i class="fa fa-check"></i><b>4.6</b> A Few More Important Keystrokes</a></li>
<li class="chapter" data-level="4.7" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#a-short-list-of-additional-useful-commands."><i class="fa fa-check"></i><b>4.7</b> A short list of additional useful commands.</a></li>
<li class="chapter" data-level="4.8" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#two-important-computing-concepts"><i class="fa fa-check"></i><b>4.8</b> Two important computing concepts</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#compression"><i class="fa fa-check"></i><b>4.8.1</b> Compression</a></li>
<li class="chapter" data-level="4.8.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#hashing"><i class="fa fa-check"></i><b>4.8.2</b> Hashing</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#unix-quick-study-guide"><i class="fa fa-check"></i><b>4.9</b> Unix: Quick Study Guide</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="shell-programming.html"><a href="shell-programming.html"><i class="fa fa-check"></i><b>5</b> Shell programming</a>
<ul>
<li class="chapter" data-level="5.1" data-path="shell-programming.html"><a href="shell-programming.html#an-example-script"><i class="fa fa-check"></i><b>5.1</b> An example script</a></li>
<li class="chapter" data-level="5.2" data-path="shell-programming.html"><a href="shell-programming.html#the-structure-of-a-bash-script"><i class="fa fa-check"></i><b>5.2</b> The Structure of a Bash Script</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="shell-programming.html"><a href="shell-programming.html#paragraph-before-shebang"><i class="fa fa-check"></i><b>5.2.1</b> A bit more on <code>;</code> and <code>&amp;</code></a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="shell-programming.html"><a href="shell-programming.html#variables"><i class="fa fa-check"></i><b>5.3</b> Variables</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="shell-programming.html"><a href="shell-programming.html#assigning-values-to-variables"><i class="fa fa-check"></i><b>5.3.1</b> Assigning values to variables</a></li>
<li class="chapter" data-level="5.3.2" data-path="shell-programming.html"><a href="shell-programming.html#accessing-values-from-variables"><i class="fa fa-check"></i><b>5.3.2</b> Accessing values from variables</a></li>
<li class="chapter" data-level="5.3.3" data-path="shell-programming.html"><a href="shell-programming.html#shell-sub-vars"><i class="fa fa-check"></i><b>5.3.3</b> What does the shell do with the value substituted for a variable?</a></li>
<li class="chapter" data-level="5.3.4" data-path="shell-programming.html"><a href="shell-programming.html#quotes-and-var-subs"><i class="fa fa-check"></i><b>5.3.4</b> Double and Single Quotation Marks and Variable Substitution</a></li>
<li class="chapter" data-level="5.3.5" data-path="shell-programming.html"><a href="shell-programming.html#one-useful-fancy-variable-substitution-method"><i class="fa fa-check"></i><b>5.3.5</b> One useful, fancy, variable-substitution method</a></li>
<li class="chapter" data-level="5.3.6" data-path="shell-programming.html"><a href="shell-programming.html#bash-arithmetic"><i class="fa fa-check"></i><b>5.3.6</b> Integer Arithmetic with Shell Variables</a></li>
<li class="chapter" data-level="5.3.7" data-path="shell-programming.html"><a href="shell-programming.html#variable-arrays"><i class="fa fa-check"></i><b>5.3.7</b> Variable arrays</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="shell-programming.html"><a href="shell-programming.html#bash-sub-result-on-comm-line"><i class="fa fa-check"></i><b>5.4</b> Evaluate a command and substitute the result on the command line</a></li>
<li class="chapter" data-level="5.5" data-path="shell-programming.html"><a href="shell-programming.html#groupingcollecting-output-from-multiple-commands-commands-and-commands"><i class="fa fa-check"></i><b>5.5</b> Grouping/Collecting output from multiple commands: <code>(commands)</code> and <code>{ commands; }</code></a></li>
<li class="chapter" data-level="5.6" data-path="shell-programming.html"><a href="shell-programming.html#exit-status"><i class="fa fa-check"></i><b>5.6</b> Exit Status</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="shell-programming.html"><a href="shell-programming.html#combinations-of-exit-statuses"><i class="fa fa-check"></i><b>5.6.1</b> Combinations of exit statuses</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="shell-programming.html"><a href="shell-programming.html#unix-for-loops"><i class="fa fa-check"></i><b>5.7</b> Loops and repetition</a></li>
<li class="chapter" data-level="5.8" data-path="shell-programming.html"><a href="shell-programming.html#more-conditional-evaluation-if-then-else-and-friends"><i class="fa fa-check"></i><b>5.8</b> More Conditional Evaluation: <code>if</code>, <code>then</code>, <code>else</code>, and friends</a></li>
<li class="chapter" data-level="5.9" data-path="shell-programming.html"><a href="shell-programming.html#finallypositional-parameters"><i class="fa fa-check"></i><b>5.9</b> Finally…positional parameters</a></li>
<li class="chapter" data-level="5.10" data-path="shell-programming.html"><a href="shell-programming.html#basename-and-dirname-two-useful-little-utilities"><i class="fa fa-check"></i><b>5.10</b> <code>basename</code> and <code>dirname</code> two useful little utilities</a></li>
<li class="chapter" data-level="5.11" data-path="shell-programming.html"><a href="shell-programming.html#bash-functions"><i class="fa fa-check"></i><b>5.11</b> <code>bash</code> functions</a></li>
<li class="chapter" data-level="5.12" data-path="shell-programming.html"><a href="shell-programming.html#reading-files-line-by-line"><i class="fa fa-check"></i><b>5.12</b> reading files line by line</a></li>
<li class="chapter" data-level="5.13" data-path="shell-programming.html"><a href="shell-programming.html#further-reading"><i class="fa fa-check"></i><b>5.13</b> Further reading</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html"><i class="fa fa-check"></i><b>6</b> Sed, awk, and regular expressions</a>
<ul>
<li class="chapter" data-level="6.1" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#awk"><i class="fa fa-check"></i><b>6.1</b> awk</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#line-cycling-tests-and-actions"><i class="fa fa-check"></i><b>6.1.1</b> Line-cycling, tests and actions</a></li>
<li class="chapter" data-level="6.1.2" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#column-splitting-fields--f-nf-print-ofs-and-begin"><i class="fa fa-check"></i><b>6.1.2</b> Column splitting, fields, <code>-F</code>, <code>$</code>, <code>NF</code>, <code>print</code>, <code>OFS</code> and <code>BEGIN</code></a></li>
<li class="chapter" data-level="6.1.3" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#a-brief-introduction-to-regular-expressions"><i class="fa fa-check"></i><b>6.1.3</b> A brief introduction to regular expressions</a></li>
<li class="chapter" data-level="6.1.4" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#a-variety-of-tests"><i class="fa fa-check"></i><b>6.1.4</b> A variety of tests</a></li>
<li class="chapter" data-level="6.1.5" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#code-in-the-action-blocks"><i class="fa fa-check"></i><b>6.1.5</b> Code in the action blocks</a></li>
<li class="chapter" data-level="6.1.6" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#using-awk-to-assign-to-shell-variables"><i class="fa fa-check"></i><b>6.1.6</b> Using <code>awk</code> to assign to shell variables</a></li>
<li class="chapter" data-level="6.1.7" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#passing-variables-into-awk-with--v"><i class="fa fa-check"></i><b>6.1.7</b> Passing Variables into <code>awk</code> with <code>-v</code></a></li>
<li class="chapter" data-level="6.1.8" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#writing-awk-scripts-in-files"><i class="fa fa-check"></i><b>6.1.8</b> Writing awk scripts in files</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#sed"><i class="fa fa-check"></i><b>6.2</b> sed</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html"><i class="fa fa-check"></i><b>7</b> Working on remote servers</a>
<ul>
<li class="chapter" data-level="7.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#accessing-remote-computers"><i class="fa fa-check"></i><b>7.1</b> Accessing remote computers</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#windows"><i class="fa fa-check"></i><b>7.1.1</b> Windows</a></li>
<li class="chapter" data-level="7.1.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#hummingbird"><i class="fa fa-check"></i><b>7.1.2</b> Hummingbird</a></li>
<li class="chapter" data-level="7.1.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#summit"><i class="fa fa-check"></i><b>7.1.3</b> Summit</a></li>
<li class="chapter" data-level="7.1.4" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#sedna"><i class="fa fa-check"></i><b>7.1.4</b> Sedna</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#transferring-files-to-remote-computers"><i class="fa fa-check"></i><b>7.2</b> Transferring files to remote computers</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#sftp-and-several-systems-that-use-it"><i class="fa fa-check"></i><b>7.2.1</b> <code>sftp</code> and several systems that use it</a></li>
<li class="chapter" data-level="7.2.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#git"><i class="fa fa-check"></i><b>7.2.2</b> git</a></li>
<li class="chapter" data-level="7.2.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#globus"><i class="fa fa-check"></i><b>7.2.3</b> Globus</a></li>
<li class="chapter" data-level="7.2.4" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#interfacing-with-the-cloud"><i class="fa fa-check"></i><b>7.2.4</b> Interfacing with “The Cloud”</a></li>
<li class="chapter" data-level="7.2.5" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#get-seqs"><i class="fa fa-check"></i><b>7.2.5</b> Getting files from a sequencing center</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#tmux"><i class="fa fa-check"></i><b>7.3</b> <code>tmux</code>: the terminal multiplexer</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#an-analogy-for-how-tmux-works"><i class="fa fa-check"></i><b>7.3.1</b> An analogy for how <code>tmux</code> works</a></li>
<li class="chapter" data-level="7.3.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#first-steps-with-tmux"><i class="fa fa-check"></i><b>7.3.2</b> First steps with <code>tmux</code></a></li>
<li class="chapter" data-level="7.3.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#further-steps-with-tmux"><i class="fa fa-check"></i><b>7.3.3</b> Further steps with <code>tmux</code></a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#tmux-for-mac-users"><i class="fa fa-check"></i><b>7.4</b> <code>tmux</code> for Mac users</a></li>
<li class="chapter" data-level="7.5" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#installing-software-on-an-hpcc"><i class="fa fa-check"></i><b>7.5</b> Installing Software on an HPCC</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#modules"><i class="fa fa-check"></i><b>7.5.1</b> Modules</a></li>
<li class="chapter" data-level="7.5.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#miniconda"><i class="fa fa-check"></i><b>7.5.2</b> Miniconda</a></li>
<li class="chapter" data-level="7.5.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#installing-java-programs"><i class="fa fa-check"></i><b>7.5.3</b> Installing Java Programs</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#vim-its-time-to-get-serious-with-text-editing"><i class="fa fa-check"></i><b>7.6</b> <code>vim</code>: it’s time to get serious with text editing</a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#using-neovim-and-nvim-r-and-tmux-to-use-r-well-on-the-cluster"><i class="fa fa-check"></i><b>7.6.1</b> Using neovim and Nvim-R and tmux to use R well on the cluster</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="chap-HPCC.html"><a href="chap-HPCC.html"><i class="fa fa-check"></i><b>8</b> High Performance Computing Clusters (HPCC’s)</a>
<ul>
<li class="chapter" data-level="8.1" data-path="chap-HPCC.html"><a href="chap-HPCC.html#an-oversimplified-but-useful-view-of-a-computing-cluster"><i class="fa fa-check"></i><b>8.1</b> An oversimplified, but useful, view of a computing cluster</a></li>
<li class="chapter" data-level="8.2" data-path="chap-HPCC.html"><a href="chap-HPCC.html#cluster-computing-and-the-job-scheduler"><i class="fa fa-check"></i><b>8.2</b> Cluster computing and the job scheduler</a></li>
<li class="chapter" data-level="8.3" data-path="chap-HPCC.html"><a href="chap-HPCC.html#slurm-info"><i class="fa fa-check"></i><b>8.3</b> Learning about the resources on your HPCC</a></li>
<li class="chapter" data-level="8.4" data-path="chap-HPCC.html"><a href="chap-HPCC.html#getting-compute-resources-allocated-to-your-jobs-on-an-hpcc"><i class="fa fa-check"></i><b>8.4</b> Getting compute resources allocated to your jobs on an HPCC</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="chap-HPCC.html"><a href="chap-HPCC.html#interactive-sessions"><i class="fa fa-check"></i><b>8.4.1</b> Interactive sessions</a></li>
<li class="chapter" data-level="8.4.2" data-path="chap-HPCC.html"><a href="chap-HPCC.html#slurm-batch"><i class="fa fa-check"></i><b>8.4.2</b> Batch jobs</a></li>
<li class="chapter" data-level="8.4.3" data-path="chap-HPCC.html"><a href="chap-HPCC.html#slurm-job-arrays"><i class="fa fa-check"></i><b>8.4.3</b> SLURM Job Arrays</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="chap-HPCC.html"><a href="chap-HPCC.html#last-years-stuff.-now-somewhat-defunct"><i class="fa fa-check"></i><b>8.5</b> Last year’s stuff. Now somewhat defunct</a></li>
<li class="chapter" data-level="8.6" data-path="chap-HPCC.html"><a href="chap-HPCC.html#prepation-interlude-an-in-class-exercise-to-make-sure-everything-is-configured-correctly"><i class="fa fa-check"></i><b>8.6</b> PREPATION INTERLUDE: An in-class exercise to make sure everything is configured correctly</a></li>
<li class="chapter" data-level="8.7" data-path="chap-HPCC.html"><a href="chap-HPCC.html#more-boneyard"><i class="fa fa-check"></i><b>8.7</b> More Boneyard…</a></li>
<li class="chapter" data-level="8.8" data-path="chap-HPCC.html"><a href="chap-HPCC.html#the-queue-slurmsgeuge"><i class="fa fa-check"></i><b>8.8</b> The Queue (SLURM/SGE/UGE)</a></li>
<li class="chapter" data-level="8.9" data-path="chap-HPCC.html"><a href="chap-HPCC.html#modules-package"><i class="fa fa-check"></i><b>8.9</b> Modules package</a></li>
<li class="chapter" data-level="8.10" data-path="chap-HPCC.html"><a href="chap-HPCC.html#compiling-programs-without-admin-privileges"><i class="fa fa-check"></i><b>8.10</b> Compiling programs without admin privileges</a></li>
<li class="chapter" data-level="8.11" data-path="chap-HPCC.html"><a href="chap-HPCC.html#job-arrays"><i class="fa fa-check"></i><b>8.11</b> Job arrays</a></li>
<li class="chapter" data-level="8.12" data-path="chap-HPCC.html"><a href="chap-HPCC.html#writing-stdout-and-stderr-to-files"><i class="fa fa-check"></i><b>8.12</b> Writing stdout and stderr to files</a></li>
<li class="chapter" data-level="8.13" data-path="chap-HPCC.html"><a href="chap-HPCC.html#breaking-stuff-down"><i class="fa fa-check"></i><b>8.13</b> Breaking stuff down</a></li>
</ul></li>
<li class="part"><span><b>II Part II: Reproducible Research Strategies</b></span></li>
<li class="chapter" data-level="9" data-path="introduction-to-reproducible-research.html"><a href="introduction-to-reproducible-research.html"><i class="fa fa-check"></i><b>9</b> Introduction to Reproducible Research</a></li>
<li class="chapter" data-level="10" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html"><i class="fa fa-check"></i><b>10</b> Rstudio and Project-centered Organization</a>
<ul>
<li class="chapter" data-level="10.1" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#organizing-big-projects"><i class="fa fa-check"></i><b>10.1</b> Organizing big projects</a></li>
<li class="chapter" data-level="10.2" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#using-rstudio-in-workflows-with-remote-computers-and-hpccs"><i class="fa fa-check"></i><b>10.2</b> Using RStudio in workflows with remote computers and HPCCs</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#keeping-an-rstudio-project-in-sync-with-github"><i class="fa fa-check"></i><b>10.2.1</b> Keeping an RStudio project “in sync” with GitHub</a></li>
<li class="chapter" data-level="10.2.2" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#evaluating-scripts-line-by-line-on-a-remote-machine-from-within-rstudio"><i class="fa fa-check"></i><b>10.2.2</b> Evaluating scripts line by line on a remote machine from within RStudio</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="version-control.html"><a href="version-control.html"><i class="fa fa-check"></i><b>11</b> Version control</a>
<ul>
<li class="chapter" data-level="11.1" data-path="version-control.html"><a href="version-control.html#why-use-version-control"><i class="fa fa-check"></i><b>11.1</b> Why use version control?</a></li>
<li class="chapter" data-level="11.2" data-path="version-control.html"><a href="version-control.html#git-workings"><i class="fa fa-check"></i><b>11.2</b> How git works</a></li>
<li class="chapter" data-level="11.3" data-path="version-control.html"><a href="version-control.html#git-workflow-patterns"><i class="fa fa-check"></i><b>11.3</b> git workflow patterns</a></li>
<li class="chapter" data-level="11.4" data-path="version-control.html"><a href="version-control.html#using-git-with-rstudio"><i class="fa fa-check"></i><b>11.4</b> using git with Rstudio</a></li>
<li class="chapter" data-level="11.5" data-path="version-control.html"><a href="version-control.html#git-on-the-command-line"><i class="fa fa-check"></i><b>11.5</b> git on the command line</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="a-fast-furious-overview-of-the-tidyverse.html"><a href="a-fast-furious-overview-of-the-tidyverse.html"><i class="fa fa-check"></i><b>12</b> A fast, furious overview of the tidyverse</a></li>
<li class="chapter" data-level="13" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html"><i class="fa fa-check"></i><b>13</b> Authoring reproducibly with Rmarkdown</a>
<ul>
<li class="chapter" data-level="13.1" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#notebooks"><i class="fa fa-check"></i><b>13.1</b> Notebooks</a></li>
<li class="chapter" data-level="13.2" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#references"><i class="fa fa-check"></i><b>13.2</b> References</a>
<ul>
<li class="chapter" data-level="13.2.1" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#zotero-and-rmarkdown"><i class="fa fa-check"></i><b>13.2.1</b> Zotero and Rmarkdown</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#bookdown"><i class="fa fa-check"></i><b>13.3</b> Bookdown</a></li>
<li class="chapter" data-level="13.4" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#google-docs"><i class="fa fa-check"></i><b>13.4</b> Google Docs</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="snakemake-chap.html"><a href="snakemake-chap.html"><i class="fa fa-check"></i><b>14</b> Managing Workflows with Snakemake</a>
<ul>
<li class="chapter" data-level="14.1" data-path="snakemake-chap.html"><a href="snakemake-chap.html#a-simple-snakemake-example"><i class="fa fa-check"></i><b>14.1</b> A Simple Snakemake Example</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="snakemake-chap.html"><a href="snakemake-chap.html#snakefile-1-a-simple-rule-to-trimmomcawesome-the-files-from-001"><i class="fa fa-check"></i><b>14.1.1</b> Snakefile #1: A simple rule to TrimmoMcAwesome the files from 001</a></li>
<li class="chapter" data-level="14.1.2" data-path="snakemake-chap.html"><a href="snakemake-chap.html#snakefile-2-a-single-trimmomcawesome-rule-using-wildcards"><i class="fa fa-check"></i><b>14.1.2</b> Snakefile #2: A single TrimmoMcAwesome rule using wildcards</a></li>
<li class="chapter" data-level="14.1.3" data-path="snakemake-chap.html"><a href="snakemake-chap.html#snakefile-3-add-trim_stupendous-in-there"><i class="fa fa-check"></i><b>14.1.3</b> Snakefile #3: Add <code>trim_stupendous</code> in there</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="snakemake-chap.html"><a href="snakemake-chap.html#ugly-complex-file-namesa-job-for-snakemake-input-functions"><i class="fa fa-check"></i><b>14.2</b> Ugly, complex file names…a job for Snakemake input functions!</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="snakemake-chap.html"><a href="snakemake-chap.html#playing-with-pandas"><i class="fa fa-check"></i><b>14.2.1</b> Playing with Pandas</a></li>
<li class="chapter" data-level="14.2.2" data-path="snakemake-chap.html"><a href="snakemake-chap.html#a-quick-sidebar-spoofing-a-snakemake-input-function"><i class="fa fa-check"></i><b>14.2.2</b> A Quick Sidebar: Spoofing a Snakemake input function</a></li>
<li class="chapter" data-level="14.2.3" data-path="snakemake-chap.html"><a href="snakemake-chap.html#snakefile-4-an-input-function-for-trim_awesome"><i class="fa fa-check"></i><b>14.2.3</b> Snakefile #4: an input function for trim_awesome</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="snakemake-chap.html"><a href="snakemake-chap.html#input-functions-from-python-dictionaries"><i class="fa fa-check"></i><b>14.3</b> Input functions from Python Dictionaries</a></li>
<li class="chapter" data-level="14.4" data-path="snakemake-chap.html"><a href="snakemake-chap.html#generating-lists-of-files-with-expand"><i class="fa fa-check"></i><b>14.4</b> Generating lists of files with expand()</a></li>
<li class="chapter" data-level="14.5" data-path="snakemake-chap.html"><a href="snakemake-chap.html#a-python-primer-for-using-snakemake"><i class="fa fa-check"></i><b>14.5</b> A Python Primer for Using Snakemake</a></li>
<li class="chapter" data-level="14.6" data-path="snakemake-chap.html"><a href="snakemake-chap.html#using-snakemake-on-a-computing-cluster"><i class="fa fa-check"></i><b>14.6</b> Using Snakemake on a Computing Cluster</a>
<ul>
<li class="chapter" data-level="14.6.1" data-path="snakemake-chap.html"><a href="snakemake-chap.html#installing-and-configuring-the-snakemake-slurm-profile"><i class="fa fa-check"></i><b>14.6.1</b> Installing and Configuring the Snakemake SLURM profile</a></li>
<li class="chapter" data-level="14.6.2" data-path="snakemake-chap.html"><a href="snakemake-chap.html#playing-with-our-new-slurm-profile"><i class="fa fa-check"></i><b>14.6.2</b> Playing with our new SLURM profile</a></li>
<li class="chapter" data-level="14.6.3" data-path="snakemake-chap.html"><a href="snakemake-chap.html#managing-cluster-resources-with-snakemake"><i class="fa fa-check"></i><b>14.6.3</b> Managing cluster resources with Snakemake</a></li>
<li class="chapter" data-level="14.6.4" data-path="snakemake-chap.html"><a href="snakemake-chap.html#grouping-short-jobs-to-run-on-a-single-node"><i class="fa fa-check"></i><b>14.6.4</b> Grouping short jobs to run on a single node</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Part III: Bioinformatic Analyses</b></span></li>
<li class="chapter" data-level="15" data-path="overview-of-bioinformatic-analyses.html"><a href="overview-of-bioinformatic-analyses.html"><i class="fa fa-check"></i><b>15</b> Overview of Bioinformatic Analyses</a></li>
<li class="chapter" data-level="16" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html"><i class="fa fa-check"></i><b>16</b> DNA Sequences and Sequencing</a>
<ul>
<li class="chapter" data-level="16.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#dna-stuff"><i class="fa fa-check"></i><b>16.1</b> DNA Stuff</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#dna-replication-with-dna-polymerase"><i class="fa fa-check"></i><b>16.1.1</b> DNA Replication with DNA Polymerase</a></li>
<li class="chapter" data-level="16.1.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#the-importance-of-the-3-hydroxyl"><i class="fa fa-check"></i><b>16.1.2</b> The importance of the 3’ hydroxyl…</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#sanger-sequencing"><i class="fa fa-check"></i><b>16.2</b> Sanger sequencing</a></li>
<li class="chapter" data-level="16.3" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#illumina-sequencing-by-synthesis"><i class="fa fa-check"></i><b>16.3</b> Illumina Sequencing by Synthesis</a></li>
<li class="chapter" data-level="16.4" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#library-prep-protocols"><i class="fa fa-check"></i><b>16.4</b> Library Prep Protocols</a>
<ul>
<li class="chapter" data-level="16.4.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#wgs"><i class="fa fa-check"></i><b>16.4.1</b> WGS</a></li>
<li class="chapter" data-level="16.4.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#rad-seq-methods"><i class="fa fa-check"></i><b>16.4.2</b> RAD-Seq methods</a></li>
<li class="chapter" data-level="16.4.3" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#amplicon-sequencing"><i class="fa fa-check"></i><b>16.4.3</b> Amplicon Sequencing</a></li>
<li class="chapter" data-level="16.4.4" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#capture-arrays-rapture-etc."><i class="fa fa-check"></i><b>16.4.4</b> Capture arrays, RAPTURE, etc.</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html"><i class="fa fa-check"></i><b>17</b> Bioinformatic file formats</a>
<ul>
<li class="chapter" data-level="17.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sequences"><i class="fa fa-check"></i><b>17.1</b> Sequences</a></li>
<li class="chapter" data-level="17.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#fastq"><i class="fa fa-check"></i><b>17.2</b> FASTQ</a>
<ul>
<li class="chapter" data-level="17.2.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#illumina-ids"><i class="fa fa-check"></i><b>17.2.1</b> Line 1: Illumina identifier lines</a></li>
<li class="chapter" data-level="17.2.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#bqscores"><i class="fa fa-check"></i><b>17.2.2</b> Line 4: Base quality scores</a></li>
<li class="chapter" data-level="17.2.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#a-fastq-tidyverse-interlude"><i class="fa fa-check"></i><b>17.2.3</b> A FASTQ ‘tidyverse’ Interlude</a></li>
<li class="chapter" data-level="17.2.4" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#comparing-read-1-to-read-2"><i class="fa fa-check"></i><b>17.2.4</b> Comparing read 1 to read 2</a></li>
</ul></li>
<li class="chapter" data-level="17.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#fasta"><i class="fa fa-check"></i><b>17.3</b> FASTA</a>
<ul>
<li class="chapter" data-level="17.3.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#genomic-ranges"><i class="fa fa-check"></i><b>17.3.1</b> Genomic ranges</a></li>
<li class="chapter" data-level="17.3.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#extracting-genomic-ranges-from-a-fasta-file"><i class="fa fa-check"></i><b>17.3.2</b> Extracting genomic ranges from a FASTA file</a></li>
<li class="chapter" data-level="17.3.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#downloading-reference-genomes-from-ncbi"><i class="fa fa-check"></i><b>17.3.3</b> Downloading reference genomes from NCBI</a></li>
</ul></li>
<li class="chapter" data-level="17.4" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sambamfiles"><i class="fa fa-check"></i><b>17.4</b> Alignments</a>
<ul>
<li class="chapter" data-level="17.4.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#how-might-i-align-to-thee-let-me-count-the-ways"><i class="fa fa-check"></i><b>17.4.1</b> How might I align to thee? Let me count the ways…</a></li>
<li class="chapter" data-level="17.4.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#play-with-simple-alignments"><i class="fa fa-check"></i><b>17.4.2</b> Play with simple alignments</a></li>
<li class="chapter" data-level="17.4.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sam-flags"><i class="fa fa-check"></i><b>17.4.3</b> SAM Flags</a></li>
<li class="chapter" data-level="17.4.4" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#cigar"><i class="fa fa-check"></i><b>17.4.4</b> The CIGAR string</a></li>
<li class="chapter" data-level="17.4.5" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#the-seq-and-qual-columns"><i class="fa fa-check"></i><b>17.4.5</b> The SEQ and QUAL columns</a></li>
<li class="chapter" data-level="17.4.6" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sam-headers"><i class="fa fa-check"></i><b>17.4.6</b> SAM File Headers</a></li>
<li class="chapter" data-level="17.4.7" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#the-bam-format"><i class="fa fa-check"></i><b>17.4.7</b> The BAM format</a></li>
<li class="chapter" data-level="17.4.8" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#quick-self-study"><i class="fa fa-check"></i><b>17.4.8</b> Quick self study</a></li>
</ul></li>
<li class="chapter" data-level="17.5" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#variants"><i class="fa fa-check"></i><b>17.5</b> Variants</a>
<ul>
<li class="chapter" data-level="17.5.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#vcf-format-the-body"><i class="fa fa-check"></i><b>17.5.1</b> VCF Format – The Body</a></li>
<li class="chapter" data-level="17.5.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#vcf-format-the-header"><i class="fa fa-check"></i><b>17.5.2</b> VCF Format – The Header</a></li>
<li class="chapter" data-level="17.5.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#boneyard"><i class="fa fa-check"></i><b>17.5.3</b> Boneyard</a></li>
</ul></li>
<li class="chapter" data-level="17.6" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#segments"><i class="fa fa-check"></i><b>17.6</b> Segments</a></li>
<li class="chapter" data-level="17.7" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#conversionextractions-between-different-formats"><i class="fa fa-check"></i><b>17.7</b> Conversion/Extractions between different formats</a></li>
<li class="chapter" data-level="17.8" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#visualization-of-genomic-data"><i class="fa fa-check"></i><b>17.8</b> Visualization of Genomic Data</a>
<ul>
<li class="chapter" data-level="17.8.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sample-data"><i class="fa fa-check"></i><b>17.8.1</b> Sample Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="18" data-path="genome-assembly.html"><a href="genome-assembly.html"><i class="fa fa-check"></i><b>18</b> Genome Assembly</a></li>
<li class="chapter" data-level="19" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><i class="fa fa-check"></i><b>19</b> Alignment of sequence data to a reference genome (and associated steps)</a>
<ul>
<li class="chapter" data-level="19.1" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#read-journey"><i class="fa fa-check"></i><b>19.1</b> The Journey of each DNA Fragment from Organism to Sequencing Read</a></li>
<li class="chapter" data-level="19.2" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#read-groups"><i class="fa fa-check"></i><b>19.2</b> Read Groups</a></li>
<li class="chapter" data-level="19.3" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#aligning-reads-with-bwa"><i class="fa fa-check"></i><b>19.3</b> Aligning reads with <code>bwa</code></a>
<ul>
<li class="chapter" data-level="19.3.1" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#indexing-the-genome-for-alignment"><i class="fa fa-check"></i><b>19.3.1</b> Indexing the genome for alignment</a></li>
<li class="chapter" data-level="19.3.2" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#mapping-reads-with-bwa-mem"><i class="fa fa-check"></i><b>19.3.2</b> Mapping reads with <code>bwa mem</code></a></li>
<li class="chapter" data-level="19.3.3" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#hold-it-right-there-buddy-what-about-the-read-groups"><i class="fa fa-check"></i><b>19.3.3</b> Hold it Right There, Buddy! What about the Read Groups?</a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#samtools"><i class="fa fa-check"></i><b>19.4</b> Processing alignment output with <code>samtools</code></a>
<ul>
<li class="chapter" data-level="19.4.1" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#samtools-subcommands"><i class="fa fa-check"></i><b>19.4.1</b> <code>samtools</code> subcommands</a></li>
</ul></li>
<li class="chapter" data-level="19.5" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#boneyard-below-here"><i class="fa fa-check"></i><b>19.5</b> BONEYARD BELOW HERE</a></li>
<li class="chapter" data-level="19.6" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#preprocess"><i class="fa fa-check"></i><b>19.6</b> Preprocess ?</a></li>
<li class="chapter" data-level="19.7" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#quick-notes-to-self-on-chaining-things"><i class="fa fa-check"></i><b>19.7</b> Quick notes to self on chaining things:</a></li>
<li class="chapter" data-level="19.8" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#merging-bam-files"><i class="fa fa-check"></i><b>19.8</b> Merging BAM files</a></li>
<li class="chapter" data-level="19.9" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#divide-and-conquer-strategies"><i class="fa fa-check"></i><b>19.9</b> Divide and Conquer Strategies</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="variant-calling.html"><a href="variant-calling.html"><i class="fa fa-check"></i><b>20</b> Variant calling</a>
<ul>
<li class="chapter" data-level="20.1" data-path="variant-calling.html"><a href="variant-calling.html#genotype-likelihoods"><i class="fa fa-check"></i><b>20.1</b> Genotype Likelihoods</a>
<ul>
<li class="chapter" data-level="20.1.1" data-path="variant-calling.html"><a href="variant-calling.html#basic-sketch-of-genotype-likelihood-calculations"><i class="fa fa-check"></i><b>20.1.1</b> Basic Sketch of Genotype Likelihood Calculations</a></li>
<li class="chapter" data-level="20.1.2" data-path="variant-calling.html"><a href="variant-calling.html#specifics-of-different-genotype-likelihoods"><i class="fa fa-check"></i><b>20.1.2</b> Specifics of different genotype likelihoods</a></li>
<li class="chapter" data-level="20.1.3" data-path="variant-calling.html"><a href="variant-calling.html#computing-genotype-likelihoods-with-three-different-softwares"><i class="fa fa-check"></i><b>20.1.3</b> Computing genotype likelihoods with three different softwares</a></li>
<li class="chapter" data-level="20.1.4" data-path="variant-calling.html"><a href="variant-calling.html#a-directed-acyclic-graph-for-genotype-likelihoods"><i class="fa fa-check"></i><b>20.1.4</b> A Directed Acyclic Graph For Genotype Likelihoods</a></li>
</ul></li>
<li class="chapter" data-level="20.2" data-path="variant-calling.html"><a href="variant-calling.html#gatks-gvcf-workflow"><i class="fa fa-check"></i><b>20.2</b> GATK’s gVCF Workflow</a>
<ul>
<li class="chapter" data-level="20.2.1" data-path="variant-calling.html"><a href="variant-calling.html#step-1.-running-haplotypecaller-and-getting-one-gvcf-for-each-sample"><i class="fa fa-check"></i><b>20.2.1</b> Step 1. Running HaplotypeCaller and getting one gVCF for each sample</a></li>
<li class="chapter" data-level="20.2.2" data-path="variant-calling.html"><a href="variant-calling.html#gvcf-format"><i class="fa fa-check"></i><b>20.2.2</b> gVCF files</a></li>
<li class="chapter" data-level="20.2.3" data-path="variant-calling.html"><a href="variant-calling.html#genomics-db"><i class="fa fa-check"></i><b>20.2.3</b> The Genomics Data Base</a></li>
<li class="chapter" data-level="20.2.4" data-path="variant-calling.html"><a href="variant-calling.html#gdb-update"><i class="fa fa-check"></i><b>20.2.4</b> Updating the GenomicsDatabase</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="21" data-path="boneyard-1.html"><a href="boneyard-1.html"><i class="fa fa-check"></i><b>21</b> Boneyard</a></li>
<li class="chapter" data-level="22" data-path="handle-vcf.html"><a href="handle-vcf.html"><i class="fa fa-check"></i><b>22</b> Basic Handling of VCF files</a>
<ul>
<li class="chapter" data-level="22.1" data-path="handle-vcf.html"><a href="handle-vcf.html#bcftools"><i class="fa fa-check"></i><b>22.1</b> bcftools</a>
<ul>
<li class="chapter" data-level="22.1.1" data-path="handle-vcf.html"><a href="handle-vcf.html#index-my-vcf-file"><i class="fa fa-check"></i><b>22.1.1</b> Index my VCF file!</a></li>
<li class="chapter" data-level="22.1.2" data-path="handle-vcf.html"><a href="handle-vcf.html#tell-me-about-my-vcf-file"><i class="fa fa-check"></i><b>22.1.2</b> Tell me about my VCF file!</a></li>
<li class="chapter" data-level="22.1.3" data-path="handle-vcf.html"><a href="handle-vcf.html#rename-the-samplesindividuals-in-the-file"><i class="fa fa-check"></i><b>22.1.3</b> Rename the samples/individuals in the file</a></li>
<li class="chapter" data-level="22.1.4" data-path="handle-vcf.html"><a href="handle-vcf.html#get-fragmentsparts-of-my-vcf-file"><i class="fa fa-check"></i><b>22.1.4</b> Get fragments/parts of my VCF file</a></li>
<li class="chapter" data-level="22.1.5" data-path="handle-vcf.html"><a href="handle-vcf.html#combine-vcf-files-in-various-ways"><i class="fa fa-check"></i><b>22.1.5</b> Combine VCF files in various ways</a></li>
<li class="chapter" data-level="22.1.6" data-path="handle-vcf.html"><a href="handle-vcf.html#filter-out-variants-for-a-variety-of-reasons"><i class="fa fa-check"></i><b>22.1.6</b> Filter out variants for a variety of reasons</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="23" data-path="bioinformatics-for-rad-seq-data-with-and-without-a-reference-genome.html"><a href="bioinformatics-for-rad-seq-data-with-and-without-a-reference-genome.html"><i class="fa fa-check"></i><b>23</b> Bioinformatics for RAD seq data with and without a reference genome</a></li>
<li class="chapter" data-level="24" data-path="processing-amplicon-sequencing-data.html"><a href="processing-amplicon-sequencing-data.html"><i class="fa fa-check"></i><b>24</b> Processing amplicon sequencing data</a></li>
<li class="chapter" data-level="25" data-path="genome-annotation.html"><a href="genome-annotation.html"><i class="fa fa-check"></i><b>25</b> Genome Annotation</a></li>
<li class="chapter" data-level="26" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html"><i class="fa fa-check"></i><b>26</b> Whole genome alignment strategies</a>
<ul>
<li class="chapter" data-level="26.1" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#mapping-of-scaffolds-to-a-closely-related-genome"><i class="fa fa-check"></i><b>26.1</b> Mapping of scaffolds to a closely related genome</a></li>
<li class="chapter" data-level="26.2" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#obtaining-ancestral-states-from-an-outgroup-genome"><i class="fa fa-check"></i><b>26.2</b> Obtaining Ancestral States from an Outgroup Genome</a>
<ul>
<li class="chapter" data-level="26.2.1" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#using-lastz-to-align-coho-to-the-chinook-genome"><i class="fa fa-check"></i><b>26.2.1</b> Using LASTZ to align coho to the chinook genome</a></li>
<li class="chapter" data-level="26.2.2" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#try-on-the-chinook-chromosomes"><i class="fa fa-check"></i><b>26.2.2</b> Try on the chinook chromosomes</a></li>
<li class="chapter" data-level="26.2.3" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#explore-the-other-parameters-more"><i class="fa fa-check"></i><b>26.2.3</b> Explore the other parameters more</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="27" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html"><i class="fa fa-check"></i><b>27</b> An example full workflow from sequences to variants</a>
<ul>
<li class="chapter" data-level="27.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#an-overview-of-the-whole-workflow"><i class="fa fa-check"></i><b>27.1</b> An overview of the whole workflow</a></li>
<li class="chapter" data-level="27.2" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#lets-have-a-look-at-the-reference-genome"><i class="fa fa-check"></i><b>27.2</b> Let’s have a look at the reference genome</a>
<ul>
<li class="chapter" data-level="27.2.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#first-off-make-sure-that-you-have-added-mamba-to-your-base-environment"><i class="fa fa-check"></i><b>27.2.1</b> First off, make sure that you have added mamba to your base environment</a></li>
<li class="chapter" data-level="27.2.2" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#indexing-the-reference-genome"><i class="fa fa-check"></i><b>27.2.2</b> Indexing the reference genome</a></li>
</ul></li>
<li class="chapter" data-level="27.3" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#a-word-on-organizing-workflows"><i class="fa fa-check"></i><b>27.3</b> A word on organizing workflows</a></li>
<li class="chapter" data-level="27.4" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#step-1a.-quality-checking-the-reads"><i class="fa fa-check"></i><b>27.4</b> Step 1a. Quality Checking the Reads</a>
<ul>
<li class="chapter" data-level="27.4.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#a-little-looking-at-the-htmls"><i class="fa fa-check"></i><b>27.4.1</b> A Little Looking at the HTMLs</a></li>
<li class="chapter" data-level="27.4.2" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#summarizing-output-from-a-lot-of-different-files-with-multiqc"><i class="fa fa-check"></i><b>27.4.2</b> Summarizing output from a lot of different files with MultiQC</a></li>
</ul></li>
<li class="chapter" data-level="27.5" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#step-1b.-trimming-the-reads"><i class="fa fa-check"></i><b>27.5</b> Step 1b. Trimming the Reads</a>
<ul>
<li class="chapter" data-level="27.5.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#cycling-over-all-the-files-and-trimming-them"><i class="fa fa-check"></i><b>27.5.1</b> Cycling over all the files and trimming them</a></li>
<li class="chapter" data-level="27.5.2" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#multiqc-your-trimmomatics-results"><i class="fa fa-check"></i><b>27.5.2</b> Multiqc your trimmomatics results</a></li>
</ul></li>
<li class="chapter" data-level="27.6" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#step-2.-mapping-each-unit-to-the-genome"><i class="fa fa-check"></i><b>27.6</b> Step 2. Mapping each unit to the genome</a>
<ul>
<li class="chapter" data-level="27.6.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#indexing-the-reference-genome-1"><i class="fa fa-check"></i><b>27.6.1</b> Indexing the reference genome</a></li>
<li class="chapter" data-level="27.6.2" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#a-quick-mapping-run-to-stdout"><i class="fa fa-check"></i><b>27.6.2</b> A quick mapping run to stdout</a></li>
<li class="chapter" data-level="27.6.3" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#map-with-array"><i class="fa fa-check"></i><b>27.6.3</b> Mapping all 22 pairs of trimmed FASTQ files using a SLURM job array</a></li>
</ul></li>
<li class="chapter" data-level="27.7" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#step-3.-mark-duplicates-in-bam-files-from-each-individual-and-run-samtools-stats-on-the-results"><i class="fa fa-check"></i><b>27.7</b> Step 3. Mark duplicates in BAM files from each individual, and run <code>samtools stats</code> on the results</a>
<ul>
<li class="chapter" data-level="27.7.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#run-samtools-stats-on-the-results"><i class="fa fa-check"></i><b>27.7.1</b> Run samtools stats on the results</a></li>
</ul></li>
<li class="chapter" data-level="27.8" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#make-gvcfs"><i class="fa fa-check"></i><b>27.8</b> Step 4. Creating a gVCF file for each sample</a>
<ul>
<li class="chapter" data-level="27.8.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#thats-a-dict-thing-to-do"><i class="fa fa-check"></i><b>27.8.1</b> That’s a dict thing to do!</a></li>
<li class="chapter" data-level="27.8.2" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#setting-up-a-slurm-job-array-for-this"><i class="fa fa-check"></i><b>27.8.2</b> Setting up a SLURM job array for this</a></li>
<li class="chapter" data-level="27.8.3" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#running-interactively-to-watch-the-output"><i class="fa fa-check"></i><b>27.8.3</b> Running interactively to watch the output</a></li>
<li class="chapter" data-level="27.8.4" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#launch-the-first-3-jobs-of-our-array"><i class="fa fa-check"></i><b>27.8.4</b> Launch the first 3 jobs of our array</a></li>
</ul></li>
<li class="chapter" data-level="27.9" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#use-genomicsdbimport"><i class="fa fa-check"></i><b>27.9</b> Step 5. Importing the gVCF files into a GenomicsDatabase</a>
<ul>
<li class="chapter" data-level="27.9.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#job-array-script-for-the-chromosomes"><i class="fa fa-check"></i><b>27.9.1</b> Job array script for the chromosomes</a></li>
<li class="chapter" data-level="27.9.2" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#job-array-script-for-the-scaffold-groups"><i class="fa fa-check"></i><b>27.9.2</b> Job array script for the scaffold groups</a></li>
<li class="chapter" data-level="27.9.3" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#hey-lets-try-updating-the-genomic-data-base"><i class="fa fa-check"></i><b>27.9.3</b> Hey! Let’s try updating the genomic data base</a></li>
</ul></li>
<li class="chapter" data-level="27.10" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#step-6.-calling-variants-and-creating-vcf-files-from-the-genomicsdatabase"><i class="fa fa-check"></i><b>27.10</b> Step 6. Calling variants and creating VCF files from the GenomicsDatabase</a>
<ul>
<li class="chapter" data-level="27.10.1" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#array-script-for-the-four-chromosomes"><i class="fa fa-check"></i><b>27.10.1</b> Array script for the four chromosomes</a></li>
<li class="chapter" data-level="27.10.2" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#array-script-for-the-two-scaffold-groups"><i class="fa fa-check"></i><b>27.10.2</b> Array script for the two scaffold groups</a></li>
<li class="chapter" data-level="27.10.3" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#have-a-look-at-the-products"><i class="fa fa-check"></i><b>27.10.3</b> Have a look at the products</a></li>
<li class="chapter" data-level="27.10.4" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#homework"><i class="fa fa-check"></i><b>27.10.4</b> Homework</a></li>
<li class="chapter" data-level="27.10.5" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#catenate-all-the-vcf.gz-files-into-a-single-vcf-file-and-index-it"><i class="fa fa-check"></i><b>27.10.5</b> Catenate all the vcf.gz files into a single VCF file and index it</a></li>
</ul></li>
<li class="chapter" data-level="27.11" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#step-7.-filtering-variants"><i class="fa fa-check"></i><b>27.11</b> Step 7. Filtering variants</a></li>
<li class="chapter" data-level="27.12" data-path="example-wgs-flow.html"><a href="example-wgs-flow.html#step-8.-base-quality-score-recalibration"><i class="fa fa-check"></i><b>27.12</b> Step 8. Base Quality Score Recalibration</a></li>
</ul></li>
<li class="part"><span><b>IV Part IV: Analysis of Big Variant Data</b></span></li>
<li class="chapter" data-level="28" data-path="bioinformatic-analysis-on-variant-data.html"><a href="bioinformatic-analysis-on-variant-data.html"><i class="fa fa-check"></i><b>28</b> Bioinformatic analysis on variant data</a></li>
<li class="part"><span><b>V Part V: Population Genomics</b></span></li>
<li class="chapter" data-level="29" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html"><i class="fa fa-check"></i><b>29</b> Topics in pop gen</a>
<ul>
<li class="chapter" data-level="29.1" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#coalescent"><i class="fa fa-check"></i><b>29.1</b> Coalescent</a></li>
<li class="chapter" data-level="29.2" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#measures-of-genetic-diversity-and-such"><i class="fa fa-check"></i><b>29.2</b> Measures of genetic diversity and such</a></li>
<li class="chapter" data-level="29.3" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#demographic-inference-with-partial-a-partial-i-and-moments"><i class="fa fa-check"></i><b>29.3</b> Demographic inference with <span class="math inline">\(\partial a \partial i\)</span> and <em>moments</em></a></li>
<li class="chapter" data-level="29.4" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#balls-in-boxes"><i class="fa fa-check"></i><b>29.4</b> Balls in Boxes</a></li>
<li class="chapter" data-level="29.5" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#some-landscape-genetics"><i class="fa fa-check"></i><b>29.5</b> Some landscape genetics</a></li>
<li class="chapter" data-level="29.6" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#relationship-inference"><i class="fa fa-check"></i><b>29.6</b> Relationship Inference</a></li>
<li class="chapter" data-level="29.7" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#tests-for-selection"><i class="fa fa-check"></i><b>29.7</b> Tests for Selection</a></li>
<li class="chapter" data-level="29.8" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#multivariate-associations-gea-etc."><i class="fa fa-check"></i><b>29.8</b> Multivariate Associations, GEA, etc.</a></li>
<li class="chapter" data-level="29.9" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#estimating-heritability-in-the-wild"><i class="fa fa-check"></i><b>29.9</b> Estimating heritability in the wild</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references-1.html"><a href="references-1.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://bookdown.org" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="handle-vcf" class="section level1" number="22">
<h1><span class="header-section-number">Chapter 22</span> Basic Handling of VCF files</h1>
<p>As we saw in the section on bioinformatic formats, VCF files can be
large and unwieldy. The format specification is also such that fields
might have different numbers of subfields, depending, for example, on the
number of alleles found at a variant. Both of these features make it
hard to directly read a VCF file into a, say, R, or some other program
that may wish to treat it purely as tabular data.</p>
<p>This is not to say that you couldn’t just read a VCF file into R directly
as a TAB delimited text file, and then start splitting fields up on it.
However, there are specialized tools for doing operations of VCF files,
and becoming familiar with them can relieve a lot
of the pain of dealing with VCF files.</p>
<p>To have an example VCF file to play with, you can
download one to your Unix workstation with the following commands.
You should put these files in your <code>scratch</code> directory somewhere, perhaps
creating a directory called <code>bcftools-play</code> to put them into. All the following
commands assume that the two example files are in the current working directory.</p>
<p>First we download the <code>vcf.gz</code> version of the VCF file, called <code>all.vcf.gz</code>. This
is the unfiltered VCF file created by our example WGS workflow in Section <a href="example-wgs-flow.html#example-wgs-flow">27</a>:</p>
<div class="sourceCode" id="cb624"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb624-1"><a href="handle-vcf.html#cb624-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">--load-cookies</span> /tmp/cookies.txt <span class="st">&quot;https://docs.google.com/uc?export=download&amp;confirm=</span><span class="va">$(</span><span class="fu">wget</span> <span class="at">--quiet</span> <span class="at">--save-cookies</span> /tmp/cookies.txt <span class="at">--keep-session-cookies</span> <span class="at">--no-check-certificate</span> <span class="st">&#39;https://docs.google.com/uc?export=download&amp;id=1zgFyYbfWU85O4JzmOX7-MOKQs4oXv4fm&#39;</span> <span class="at">-O-</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="at">-rn</span> <span class="st">&#39;s/.*confirm=([0-9A-Za-z_]+).*/\1\n/p&#39;</span><span class="va">)</span><span class="st">&amp;id=1zgFyYbfWU85O4JzmOX7-MOKQs4oXv4fm&quot;</span> <span class="at">-O</span> all.vcf.gz <span class="kw">&amp;&amp;</span> <span class="fu">rm</span> <span class="at">-rf</span> /tmp/cookies.txt</span></code></pre></div>
<p>After that, we also download the BCF version of the same file, just so that everyone
gets familiar with the fact that BCF files can be treated equivalently
to VCF files with <code>bcftools</code>.</p>
<div class="sourceCode" id="cb625"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb625-1"><a href="handle-vcf.html#cb625-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">--load-cookies</span> /tmp/cookies.txt <span class="st">&quot;https://docs.google.com/uc?export=download&amp;confirm=</span><span class="va">$(</span><span class="fu">wget</span> <span class="at">--quiet</span> <span class="at">--save-cookies</span> /tmp/cookies.txt <span class="at">--keep-session-cookies</span> <span class="at">--no-check-certificate</span> <span class="st">&#39;https://docs.google.com/uc?export=download&amp;id=1iUP_UstnmuLSGuBIfb0EYXs1Gw1z-z0o&#39;</span> <span class="at">-O-</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="at">-rn</span> <span class="st">&#39;s/.*confirm=([0-9A-Za-z_]+).*/\1\n/p&#39;</span><span class="va">)</span><span class="st">&amp;id=1iUP_UstnmuLSGuBIfb0EYXs1Gw1z-z0o&quot;</span> <span class="at">-O</span> all.bcf <span class="kw">&amp;&amp;</span> <span class="fu">rm</span> <span class="at">-rf</span> /tmp/cookies.txt</span></code></pre></div>
<p>To repeat, BCF and vcf.gz are two formats serving similar purposes. I will be using both
<code>all.bcf</code> and <code>all.vcf.gz</code> interchangeably in the following examples.</p>
<p>There are two main, well-known programs for handling VCF files: <code>vcftools</code> and <code>bcftools</code>. Both
of these grew out of the 1000 Genomes effort starting about a decade ago. It seems
that <code>vcftools</code> may have been developed first, but, currently, <code>bcftools</code> is being
more actively developed, with new versions and new features being added to it
regularly. <code>vcftools</code> provides some very specific commands for particular analyses
or operations, some of which are not available from <code>bcftools</code>. On the other hand,
<code>bcftools</code> provides a more general interface for operations on VCF files. By this
interface, a great number of the operations done in <code>vcftools</code> are available, but
a little extra knowledge is required to implement them. That said, the range of
possible operations seems much larger in <code>bcftools</code> than in <code>vcftools</code>.</p>
<p>Further, <code>bcftools</code> behaves like a typical Unix utility, allowing data to be streamed
to stdout, and data can be streamed <em>into</em> <code>bcftools</code> (by using the <code>-</code> as the
input file name) from stdin. This lets you pipe output into it the way you
can with most Unix tools. This makes it far more versatile than <code>vcftools</code>.</p>
<div id="bcftools" class="section level2" number="22.1">
<h2><span class="header-section-number">22.1</span> bcftools</h2>
<p>If you don’t already have <code>bcftools</code>, you can use <code>mamba</code> to create a conda
environment called <code>bcftools</code> that has it:</p>
<div class="sourceCode" id="cb626"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb626-1"><a href="handle-vcf.html#cb626-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> create <span class="at">-n</span> bcftools <span class="at">-c</span> conda-forge <span class="at">-c</span> bioconda bcftools=1.15.1</span>
<span id="cb626-2"><a href="handle-vcf.html#cb626-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate bcftools</span></code></pre></div>
<p>(Note the use of <code>-c conda-forge</code> in the above. This is present because
some of the dependencies for <code>bcftools</code> are not found on the bioconda channel.
Rather they are on the conda-forge channel. If you conda/mamba environment is not
set up to search conda-forge by default, then the <code>-c conda-forge</code> is required to
get all the dependencies.)</p>
<p>Here, we just want to give everyone the chance to run through a few
operations with <code>bcftools</code>, to start to get familiar with its interface.
The first thing to note is that, like <code>samtools</code> (which is maintained by
the same group of people), <code>bcftools</code> possesses a number of different
<em>subcommands</em>. So, the syntax is always like:</p>
<ul>
<li><code>bcftools</code> <em>subcommand</em> options file(s)</li>
</ul>
<p>Also like <code>samtools</code>, <code>bcftools</code> will take input from <em>stdin</em> rather than
from a file—you just pass it <code>-</code> instead of a file name.</p>
<p>The full documentation/manual for <code>bcftools</code> is maintained at:
<a href="http://samtools.github.io/bcftools/bcftools.html">http://samtools.github.io/bcftools/bcftools.html</a>. It is well worth reading through this whole documentation, though it
can be quite terse and intimidating. A friendlier “tutorial” introduction to the
software can be found at <a href="https://samtools.github.io/bcftools/howtos/index.html">https://samtools.github.io/bcftools/howtos/index.html</a>.</p>
<p>Here we are going to get our feet with with a few operations.</p>
<p>First, we will look at the “onboard” documentation. By just entering <code>bcftools</code> you get a
list of all the subcommands that are available:</p>
<div class="sourceCode" id="cb627"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb627-1"><a href="handle-vcf.html#cb627-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span></span></code></pre></div>
<p>If you want the onboard documentation for any of the particular subcommands, you can
just give a naked <code>bctools subcommand</code> command, like:</p>
<div class="sourceCode" id="cb628"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb628-1"><a href="handle-vcf.html#cb628-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> index</span></code></pre></div>
<p>or, for a more daunting set of documentation:</p>
<div class="sourceCode" id="cb629"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb629-1"><a href="handle-vcf.html#cb629-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> roh</span></code></pre></div>
<div id="index-my-vcf-file" class="section level3" number="22.1.1">
<h3><span class="header-section-number">22.1.1</span> Index my VCF file!</h3>
<p>The first thing we are going to do is index our VCF files.
We create the default index, a <em>coordinate sorted index</em> which
has the <code>.csi</code> extension. If your VCF file is not in coordinate-sorted
order, you might have to sort it before you do this. However,
<code>all.vcf.gz</code> and <code>all.bcf</code> are already sorted. So:</p>
<div class="sourceCode" id="cb630"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb630-1"><a href="handle-vcf.html#cb630-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> index all.vcf.gz</span>
<span id="cb630-2"><a href="handle-vcf.html#cb630-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> index all.bcf</span></code></pre></div>
<p>Use <code>ls</code> to see the new files created by that operation.</p>
<p>The index allows for rapid access to different parts of the files that
correspond to specific locations in the genome.</p>
</div>
<div id="tell-me-about-my-vcf-file" class="section level3" number="22.1.2">
<h3><span class="header-section-number">22.1.2</span> Tell me about my VCF file!</h3>
<p>VCF files are a little daunting. Especially when they are gzipped
(or are BCF files!) they
can seem particularly opaque—learning anything about them in the traditional
manner of uncompressing them and then searching for lines within them
or counting up the number of records can be time consuming. Here are some
<code>bcftools</code> solutions to a few different questions you might have.</p>
<p><strong>Who is in this file?</strong> You can always try to find the last header line
in a VCF file using grep or awk and parse the individuals out yourself, but
it turns out to be faster and safer to use the <code>query</code> subcommand from <code>bcftools</code>
with the <code>-l</code> option. Do it here:</p>
<div class="sourceCode" id="cb631"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb631-1"><a href="handle-vcf.html#cb631-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-l</span> all.vcf.gz</span>
<span id="cb631-2"><a href="handle-vcf.html#cb631-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb631-3"><a href="handle-vcf.html#cb631-3" aria-hidden="true" tabindex="-1"></a><span class="co"># And, of course, you can do the same with the BCF file</span></span>
<span id="cb631-4"><a href="handle-vcf.html#cb631-4" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-l</span> all.bcf</span></code></pre></div>
<p>Then read about it on the manual page. Find the part that describes it.</p>
<p><strong>How many variants are in this file?</strong> This question can be answered quickly with
<code>bcftools stats</code>, which also returns to you a plethora of information about the variants.</p>
<div class="sourceCode" id="cb632"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb632-1"><a href="handle-vcf.html#cb632-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> stats all.vcf.gz <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>The top part of the output tells you how many SNPs and indels (and other types
of variants) there are. Then it tells you about Ts/Tv ratios, then it essentially
gives histogram summaries for allele frequencies,
variant quality scores (QUAL), insertion-deletion sizes, substitution types, read depths, etc.</p>
<p><strong>Where are these variants?</strong> There are several ways to answer this question.
One might be simply to print the CHROM and the POS for each row in the VCF file:</p>
<div class="sourceCode" id="cb633"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb633-1"><a href="handle-vcf.html#cb633-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%CHROM\t%POS\n&#39;</span> all.vcf.gz</span></code></pre></div>
<p>If you want to see where it starts and where it finishes you can do:</p>
<div class="sourceCode" id="cb634"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb634-1"><a href="handle-vcf.html#cb634-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%CHROM\t%POS\n&#39;</span> all.vcf.gz <span class="kw">|</span> <span class="fu">head</span> </span>
<span id="cb634-2"><a href="handle-vcf.html#cb634-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%CHROM\t%POS\n&#39;</span> all.vcf.gz <span class="kw">|</span> <span class="fu">tail</span></span></code></pre></div>
<p>If we wanted to quickly see how many variants were on each of the
chromosomes/scaffolds, sorted by number of variants, we could do:</p>
<div class="sourceCode" id="cb635"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb635-1"><a href="handle-vcf.html#cb635-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%CHROM\t%POS\n&#39;</span> all.vcf.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;{n[$1]++} END {for(i in n) print i, n[i]}&#39;</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-nbr</span> <span class="at">-k</span> 2 <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>This shows one use of the subcommand <code>query</code>, which is quite useful.
Even though it is named <code>query</code> its main purpose is simply extracting fields
of information from a VCF file and spitting them out in a new, user-specified,
typically tabular format.</p>
<p><strong>Give me a glimpse of the file</strong> You can use <code>bcftools view</code> for a number
of things, but at its simplest, you can merely look at the file in VCF format.
(In this manner, it behaves much like <code>samtools view</code> for VCF files).</p>
<div class="sourceCode" id="cb636"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb636-1"><a href="handle-vcf.html#cb636-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show the whole file from the top</span></span>
<span id="cb636-2"><a href="handle-vcf.html#cb636-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view all.bcf <span class="kw">|</span> <span class="fu">less</span></span>
<span id="cb636-3"><a href="handle-vcf.html#cb636-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb636-4"><a href="handle-vcf.html#cb636-4" aria-hidden="true" tabindex="-1"></a><span class="co"># of course, this works with either bcf or vcf or vcf.gz</span></span>
<span id="cb636-5"><a href="handle-vcf.html#cb636-5" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view all.vcf.gz <span class="kw">|</span> <span class="fu">less</span></span>
<span id="cb636-6"><a href="handle-vcf.html#cb636-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb636-7"><a href="handle-vcf.html#cb636-7" aria-hidden="true" tabindex="-1"></a><span class="co"># show just the header with -h.  Here we look at just the last 10 lines of the header</span></span>
<span id="cb636-8"><a href="handle-vcf.html#cb636-8" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-h</span> all.bcf  <span class="kw">|</span> <span class="fu">tail</span></span>
<span id="cb636-9"><a href="handle-vcf.html#cb636-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb636-10"><a href="handle-vcf.html#cb636-10" aria-hidden="true" tabindex="-1"></a><span class="co"># show the variants themselves (no header) with -H</span></span>
<span id="cb636-11"><a href="handle-vcf.html#cb636-11" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-H</span> all.vcf.gz <span class="kw">|</span> <span class="fu">head</span></span></code></pre></div>
<p>Just like you can with <code>samtools view</code> you can convert formats with <code>bcftools view</code>.
Pipe a VCF into it and then use the <code>-O</code> (Big-O, not a zero) option:</p>
<ul>
<li><code>-O z</code>: bgzipped VCF (vcf.gz)</li>
<li><code>-O v</code>: uncompressed VCF (the default)</li>
<li><code>-O u</code>: uncompressed BCF</li>
<li><code>-O b</code>: compressed BCF</li>
</ul>
</div>
<div id="rename-the-samplesindividuals-in-the-file" class="section level3" number="22.1.3">
<h3><span class="header-section-number">22.1.3</span> Rename the samples/individuals in the file</h3>
<p>We saw above that the names of the samples in the file are like <code>s001</code>, <code>s002</code>.
This was not actually what I had intended! The names in here are set by the
<code>SM</code> field of the read groups in the BAM files from which variants are called.
In Section <a href="example-wgs-flow.html#map-with-array">27.6.3</a>, where those BAM files were made, I
screwed up and used the wrong column from the <code>numbered-units.tsv</code> file to
set the <code>SM</code> value in the read groups. I wanted to use <code>sample_id</code> not <code>sample</code>.
However, all is not lost. We don’t have to go all the way back to the beginning
and remap everything and call variants. We simply rename the samples in the file.
For this we can use <code>bcftools reheader</code>. First, look at the documentation for that,
both on the web, and with:</p>
<div class="sourceCode" id="cb637"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb637-1"><a href="handle-vcf.html#cb637-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> reheader</span></code></pre></div>
<p>Aha! we see that the web-based documentation is a little more complete, and it tells
us what format to use for a sample-renaming file for the <code>-s</code> option.
Copy the following contents (using <code>nano</code>, for example) into a file called <code>sample-renames.txt</code></p>
<pre><code>s001    T199967
s002    T199968
s003    T199969
s004    T199970
s005    T199971
s006    T199972
s007    T199973
s008    T199974</code></pre>
<p>Then we can make a renamed <code>bcf</code> file with:</p>
<div class="sourceCode" id="cb639"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb639-1"><a href="handle-vcf.html#cb639-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> reheader <span class="at">-s</span> sample-renames.txt all.bcf  <span class="op">&gt;</span> renamed-all.bcf</span></code></pre></div>
<p>and a renamed <code>vcf.gz</code> file with:</p>
<div class="sourceCode" id="cb640"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb640-1"><a href="handle-vcf.html#cb640-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> reheader <span class="at">-s</span> sample-renames.txt all.vcf.gz  <span class="op">&gt;</span> renamed-all.vcf.gz</span></code></pre></div>
<p>In this case, the type of output file (bcf or vcf.gz) is the same
as the type of the input file.</p>
<p><strong>Exercise</strong> Use <code>bcftools view</code> and <code>tail</code> to see that the names have really been
changed. Then use <code>bcftools query</code> to do the same.</p>
</div>
<div id="get-fragmentsparts-of-my-vcf-file" class="section level3" number="22.1.4">
<h3><span class="header-section-number">22.1.4</span> Get fragments/parts of my VCF file</h3>
<p>There are lots of ways to extract desired bits of information
from a VCF file into a more manageable format.</p>
<p><strong>Extract keyed values from the INFO field</strong> When we did this:</p>
<div class="sourceCode" id="cb641"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb641-1"><a href="handle-vcf.html#cb641-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-H</span> <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>we saw that there is a lot of information in the INFO
field. What if we wanted to extract that? It looks like it could be
hard to parse because the fields are in semi-colon-separated
key-value pairs.</p>
<p>This is another job for <code>bcftools query</code>.
You pass a <em>format string</em> to the <code>-f</code> option that tells the program which fields you want
to extract and how you want to format it. In general, the values are preceded by
a <code>%</code> and subfields of the INFO column can be named described
like <code>%INFO/SUBFIELD</code>. You can ask for TABs between fields with
<code>\t</code> and for line endings with <code>\n</code>. In general you need to wrap all of these format
specifications in single quotes so that the shell does not get confused by them.</p>
<p>Check out some examples:</p>
<div class="sourceCode" id="cb642"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb642-1"><a href="handle-vcf.html#cb642-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract CHROM POS and BaseQRankSum, separated by TABs</span></span>
<span id="cb642-2"><a href="handle-vcf.html#cb642-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%CHROM\t%POS\t%INFO/BaseQRankSum\n&#39;</span> all.vcf.gz <span class="kw">|</span> <span class="fu">less</span></span>
<span id="cb642-3"><a href="handle-vcf.html#cb642-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb642-4"><a href="handle-vcf.html#cb642-4" aria-hidden="true" tabindex="-1"></a><span class="co"># extract CHROM POS and total read depth DP</span></span>
<span id="cb642-5"><a href="handle-vcf.html#cb642-5" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%CHROM\t%POS\t%INFO/DP\n&#39;</span> all.bcf <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>You can even extract information from each of the genotype columns. If you want to print
CHROM and POS and then all of the PHRED-scaled genotype likelihoods for all the samples, separated
by TABs, you can do:</p>
<div class="sourceCode" id="cb643"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb643-1"><a href="handle-vcf.html#cb643-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> query <span class="at">-f</span> <span class="st">&#39;%CHROM\t%POS\t[%PL\t]\n&#39;</span> all.bcf <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>Note that FORMAT values (i.e., those in the genotype columns) must be wrapped in
<code>[ ]</code> to get all the values to be printed out.</p>
<p><strong>EXERCISE</strong> Extract the CHROM, POS, Maximum Likelihood-estimated Allele Frequency (MLEAF in the INFO
column) for each variant, along with the allele depths (AD) of each individual, all TAB separated, from the
file <code>all.vcf.gz</code>.</p>
<p><strong>View data from specified regions</strong></p>
<p>What if we want to look at variants only in two 10 Kb regions, like
<code>CM031199.1:1-10000</code> and <code>CM031200.1:1000000-1005000</code>? Pass those,
separated by commas, to the <code>-r</code> option (which is an option that applies to many
of the subcommands):</p>
<div class="sourceCode" id="cb644"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb644-1"><a href="handle-vcf.html#cb644-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-H</span> <span class="at">-r</span> CM031199.1:1-10000,CM031200.1:1000000-1005000 all.vcf.gz <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>You can also specify those regions in a file with the <code>-R</code> option.</p>
<p><strong>View data from specified individuals</strong></p>
<p>You can give the sample names (comma separated) to the <code>-s</code> option:</p>
<div class="sourceCode" id="cb645"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb645-1"><a href="handle-vcf.html#cb645-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-H</span> <span class="at">-s</span> s001,s002,s003 all.vcf.gz <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>Or, if you wanted to view all <em>but</em> those two individuals, precede them with a <code>^</code>:</p>
<div class="sourceCode" id="cb646"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb646-1"><a href="handle-vcf.html#cb646-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-H</span> <span class="at">-s</span> ^s001,s002,s003 all.vcf.gz <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>You can also supply a text file with sample names (one-per-line) to the capital letter <code>-S</code> option.</p>
<p>You can combine options, like <code>-r</code> and <code>-s</code>, as well.</p>
</div>
<div id="combine-vcf-files-in-various-ways" class="section level3" number="22.1.5">
<h3><span class="header-section-number">22.1.5</span> Combine VCF files in various ways</h3>
<p><strong>Catenate VCF files</strong></p>
<p>If you have VCF files called from the same reference genome filled with
the same samples, it is easy to catenate them together with <code>bcftools concat</code>:</p>
<div class="sourceCode" id="cb647"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb647-1"><a href="handle-vcf.html#cb647-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make two files from different regions</span></span>
<span id="cb647-2"><a href="handle-vcf.html#cb647-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-O</span> z <span class="at">-r</span> CM031199.1:1-10000 all.vcf.gz  <span class="op">&gt;</span> A.vcf.gz</span>
<span id="cb647-3"><a href="handle-vcf.html#cb647-3" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-O</span> z <span class="at">-r</span> CM031200.1:1000000-1005000 all.vcf.gz  <span class="op">&gt;</span> B.vcf.gz</span>
<span id="cb647-4"><a href="handle-vcf.html#cb647-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb647-5"><a href="handle-vcf.html#cb647-5" aria-hidden="true" tabindex="-1"></a><span class="co"># how many variants in each of those?</span></span>
<span id="cb647-6"><a href="handle-vcf.html#cb647-6" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> stats A.vcf.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;/^SN/&#39;</span></span>
<span id="cb647-7"><a href="handle-vcf.html#cb647-7" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> stats B.vcf.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;/^SN/&#39;</span></span>
<span id="cb647-8"><a href="handle-vcf.html#cb647-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb647-9"><a href="handle-vcf.html#cb647-9" aria-hidden="true" tabindex="-1"></a><span class="co"># catenate the back together</span></span>
<span id="cb647-10"><a href="handle-vcf.html#cb647-10" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> concat <span class="at">-Oz</span>  A.vcf.gz B.vcf.gz <span class="op">&gt;</span> CAT.vcf.gz</span>
<span id="cb647-11"><a href="handle-vcf.html#cb647-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb647-12"><a href="handle-vcf.html#cb647-12" aria-hidden="true" tabindex="-1"></a><span class="co"># how many variants in that?</span></span>
<span id="cb647-13"><a href="handle-vcf.html#cb647-13" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> stats CAT.vcf.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;/^SN/&#39;</span></span></code></pre></div>
<p>Note that when using the <code>-O</code> (capital “o”) option to specify the output
type: <code>v</code> = VCF, <code>b</code> = BCF, <code>u</code> = uncompressed BCF, <code>z</code> = bgzipped VCF, you
don’t need a space after the <code>-O</code>.</p>
<p><strong>Merge VCF files</strong></p>
<p>If you have files with different samples in them you
can easily combine them:</p>
<div class="sourceCode" id="cb648"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb648-1"><a href="handle-vcf.html#cb648-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make file with first three samples</span></span>
<span id="cb648-2"><a href="handle-vcf.html#cb648-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-Oz</span> <span class="at">-s</span> s001,s002,s003 all.vcf.gz <span class="op">&gt;</span> first3.vcf.gz</span>
<span id="cb648-3"><a href="handle-vcf.html#cb648-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb648-4"><a href="handle-vcf.html#cb648-4" aria-hidden="true" tabindex="-1"></a><span class="co"># make another with the last three samples</span></span>
<span id="cb648-5"><a href="handle-vcf.html#cb648-5" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-Oz</span> <span class="at">-s</span> s006,s007,s008 all.bcf <span class="op">&gt;</span> last3.vcf.gz</span>
<span id="cb648-6"><a href="handle-vcf.html#cb648-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb648-7"><a href="handle-vcf.html#cb648-7" aria-hidden="true" tabindex="-1"></a><span class="co"># merging requires that the files be indexed</span></span>
<span id="cb648-8"><a href="handle-vcf.html#cb648-8" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> index first3.vcf.gz</span>
<span id="cb648-9"><a href="handle-vcf.html#cb648-9" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> index last3.vcf.gz</span>
<span id="cb648-10"><a href="handle-vcf.html#cb648-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb648-11"><a href="handle-vcf.html#cb648-11" aria-hidden="true" tabindex="-1"></a><span class="co"># merge those into a file with 6 samples</span></span>
<span id="cb648-12"><a href="handle-vcf.html#cb648-12" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> merge <span class="at">-Oz</span> first3.vcf.gz last3.vcf.gz <span class="op">&gt;</span> 6-samples.vcf.gz</span></code></pre></div>
</div>
<div id="filter-out-variants-for-a-variety-of-reasons" class="section level3" number="22.1.6">
<h3><span class="header-section-number">22.1.6</span> Filter out variants for a variety of reasons</h3>
<p>There are a lot of ways to filter out variants. <code>bcftools</code> leaves things
very general here, and so just about anything is possible. Some simple ones:</p>
<p><strong>Just the biallelic SNPs please</strong> Get things with no more than 2 alleles and
no fewer than two alleles, and of a type = SNP:</p>
<div class="sourceCode" id="cb649"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb649-1"><a href="handle-vcf.html#cb649-1" aria-hidden="true" tabindex="-1"></a><span class="co"># do it and summarize the result to look at it, all in one line:</span></span>
<span id="cb649-2"><a href="handle-vcf.html#cb649-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-Ou</span> <span class="at">-m</span> 2 <span class="at">-M</span> 2 <span class="at">--types</span><span class="op">=</span>snps all.bcf <span class="kw">|</span> <span class="ex">bcftools</span> stats <span class="at">-</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;/^SN/&#39;</span></span></code></pre></div>
<p><strong>Just the biallelic indels please</strong></p>
<div class="sourceCode" id="cb650"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb650-1"><a href="handle-vcf.html#cb650-1" aria-hidden="true" tabindex="-1"></a><span class="co"># do it and see the result all in one line:</span></span>
<span id="cb650-2"><a href="handle-vcf.html#cb650-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-Ou</span> <span class="at">-m</span> 2 <span class="at">-M</span> 2 <span class="at">--types</span><span class="op">=</span>indels all.vcf.gz <span class="kw">|</span> <span class="ex">bcftools</span> stats <span class="at">-</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;/^SN/&#39;</span></span></code></pre></div>
<p>Note the use of <code>-Ou</code> in order to pipe uncompressed BCF output directly into
<code>bcftools stats</code> using the <code>-</code> for a filename.</p>
<p><strong>Fraction of missing sites less than X</strong></p>
<p>If you want to make sure that 60% of your individuals have at least
one read at the genotype, you can do this:</p>
<div class="sourceCode" id="cb651"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb651-1"><a href="handle-vcf.html#cb651-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-i</span> <span class="st">&#39;F_MISSING &lt; 0.4&#39;</span> all.vcf.gz <span class="kw">|</span> <span class="ex">bcftools</span> stats <span class="at">-</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;/^SN/&#39;</span></span></code></pre></div>
<p>Play with setting the <code>F_MISSING</code> to different values and see how that affects
the number of variants retained. (Not much with this example data set, it turns
out, because there is not much missing data.</p>
<p><strong>Exclude based on various features of the data</strong></p>
<p>You can use the <code>-e</code> option to <code>bcftools view</code> or <code>bcftools filter</code> to <em>exclude</em> sites that
meet certain criteria. (You can use <code>-i</code> to <em>include</em> those sites and no others).</p>
<p>For a terrifying terse and dense description of what sorts of expressions can be used to
describe the criteria, see the web manual section on expressions: <a href="http://samtools.github.io/bcftools/bcftools.html#expressions">http://samtools.github.io/bcftools/bcftools.html#expressions</a>.</p>
<p>For example, to only keep things with a maximum-likelihood-estimated allele frequency between 0.4 and 0.6:</p>
<div class="sourceCode" id="cb652"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb652-1"><a href="handle-vcf.html#cb652-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-H</span> <span class="at">-i</span> <span class="st">&#39;INFO/MLEAF &gt;= 0.4 &amp;&amp; INFO/MLEAF &lt;= 0.6&#39;</span> all.bcf <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>We are using the <code>-H</code> option there just so that we can easily see the result using <code>less</code>.</p>
<p>How about excluding those sites in which any individual had a DP less than 5. We can
test each of the DP columns in the FORMAT columns. We name these <code>FMT/DP</code>.
Note that each test (from each sample)
is combined with an OR by default, so:</p>
<div class="sourceCode" id="cb653"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb653-1"><a href="handle-vcf.html#cb653-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-H</span> <span class="at">-e</span> <span class="st">&#39;FMT/DP &lt; 5&#39;</span> all.bcf <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<hr />
<p>This has only scratched the surface of what is possible with bcftools.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="boneyard-1.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="bioinformatics-for-rad-seq-data-with-and-without-a-reference-genome.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook/js/app.min.js"></script>
<script src="libs/gitbook/js/clipboard.min.js"></script>
<script src="libs/gitbook/js/plugin-search.js"></script>
<script src="libs/gitbook/js/plugin-sharing.js"></script>
<script src="libs/gitbook/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook/js/plugin-bookdown.js"></script>
<script src="libs/gitbook/js/jquery.highlight.js"></script>
<script src="libs/gitbook/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/eriqande/eca-bioinf-handbook/edit/master/3.05.5-basic-handling-of-VCF-files.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/eriqande/eca-bioinf-handbook/commits/master/3.05.5-basic-handling-of-VCF-files.Rmd",
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["eca-bioinf-handbook.pdf", "eca-bioinf-handbook.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
