<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 High Performance Computing (HPC) Environments | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics</title>
  <meta name="description" content="A book example for a Chapman &amp; Hall book." />
  <meta name="generator" content="bookdown 0.16 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 High Performance Computing (HPC) Environments | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A book example for a Chapman &amp; Hall book." />
  <meta name="github-repo" content="yihui/bookdown-crc" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 High Performance Computing (HPC) Environments | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics" />
  
  <meta name="twitter:description" content="A book example for a Chapman &amp; Hall book." />
  

<meta name="author" content="Eric C. Anderson" />


<meta name="date" content="2020-02-05" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="sed-awk-and-regular-expressions.html"/>
<link rel="next" href="introduction-to-reproducible-research.html"/>
<script src="libs/jquery/jquery.min.js"></script>
<link href="libs/gitbook/css/style.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ECA's Bioinformatics Handbook</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="erics-notes-of-what-he-might-do.html"><a href="erics-notes-of-what-he-might-do.html"><i class="fa fa-check"></i><b>2</b> Eric’s Notes of what he might do</a><ul>
<li class="chapter" data-level="2.1" data-path="erics-notes-of-what-he-might-do.html"><a href="erics-notes-of-what-he-might-do.html#table-of-topics"><i class="fa fa-check"></i><b>2.1</b> Table of topics</a></li>
</ul></li>
<li class="part"><span><b>I Part I: Essential Computing Skills</b></span></li>
<li class="chapter" data-level="3" data-path="overview-of-essential-computing-skills.html"><a href="overview-of-essential-computing-skills.html"><i class="fa fa-check"></i><b>3</b> Overview of Essential Computing Skills</a></li>
<li class="chapter" data-level="4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html"><i class="fa fa-check"></i><b>4</b> Essential Unix/Linux Terminal Knowledge</a><ul>
<li class="chapter" data-level="4.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#getting-a-bash-shell-on-your-system"><i class="fa fa-check"></i><b>4.1</b> Getting a bash shell on your system</a></li>
<li class="chapter" data-level="4.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#navigating-the-unix-filesystem"><i class="fa fa-check"></i><b>4.2</b> Navigating the Unix filesystem</a><ul>
<li class="chapter" data-level="4.2.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#changing-the-working-directory-with-cd"><i class="fa fa-check"></i><b>4.2.1</b> Changing the working directory with <code>cd</code></a></li>
<li class="chapter" data-level="4.2.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#updating-your-command-prompt"><i class="fa fa-check"></i><b>4.2.2</b> Updating your command prompt</a></li>
<li class="chapter" data-level="4.2.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#tab-completion-for-paths"><i class="fa fa-check"></i><b>4.2.3</b> TAB-completion for paths</a></li>
<li class="chapter" data-level="4.2.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#listing-the-contents-of-a-directory-with-ls"><i class="fa fa-check"></i><b>4.2.4</b> Listing the contents of a directory with <code>ls</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#globbing"><i class="fa fa-check"></i><b>4.2.5</b> Globbing</a></li>
<li class="chapter" data-level="4.2.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#what-makes-a-good-file-name"><i class="fa fa-check"></i><b>4.2.6</b> What makes a good file-name?</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#the-anatomy-of-a-unix-command"><i class="fa fa-check"></i><b>4.3</b> The anatomy of a Unix command</a><ul>
<li class="chapter" data-level="4.3.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#anatomy-command"><i class="fa fa-check"></i><b>4.3.1</b> The <code>command</code></a></li>
<li class="chapter" data-level="4.3.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#the-options"><i class="fa fa-check"></i><b>4.3.2</b> The <em>options</em></a></li>
<li class="chapter" data-level="4.3.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#arguments"><i class="fa fa-check"></i><b>4.3.3</b> Arguments</a></li>
<li class="chapter" data-level="4.3.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#getting-information-about-unix-commands"><i class="fa fa-check"></i><b>4.3.4</b> Getting information about Unix commands</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#handling-manipulating-and-viewing-files-and-streams"><i class="fa fa-check"></i><b>4.4</b> Handling, Manipulating, and Viewing files and streams</a><ul>
<li class="chapter" data-level="4.4.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#creating-new-directories"><i class="fa fa-check"></i><b>4.4.1</b> Creating new directories</a></li>
<li class="chapter" data-level="4.4.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#fundamental-file-handling-commands"><i class="fa fa-check"></i><b>4.4.2</b> Fundamental file-handling commands</a></li>
<li class="chapter" data-level="4.4.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#viewing-files"><i class="fa fa-check"></i><b>4.4.3</b> “Viewing” Files</a></li>
<li class="chapter" data-level="4.4.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#redirecting-standard-output-and"><i class="fa fa-check"></i><b>4.4.4</b> Redirecting standard output: <code>&gt;</code> and <code>&gt;&gt;</code></a></li>
<li class="chapter" data-level="4.4.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#stdin-and"><i class="fa fa-check"></i><b>4.4.5</b> stdin, <code>&lt;</code> and <code>|</code></a></li>
<li class="chapter" data-level="4.4.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#stderr"><i class="fa fa-check"></i><b>4.4.6</b> stderr</a></li>
<li class="chapter" data-level="4.4.7" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#symbolic-links"><i class="fa fa-check"></i><b>4.4.7</b> Symbolic links</a></li>
<li class="chapter" data-level="4.4.8" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#file-permissions"><i class="fa fa-check"></i><b>4.4.8</b> File Permissions</a></li>
<li class="chapter" data-level="4.4.9" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#editing-text-files-at-the-terminal"><i class="fa fa-check"></i><b>4.4.9</b> Editing text files at the terminal</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#unix-env"><i class="fa fa-check"></i><b>4.5</b> Customizing your Environment</a><ul>
<li class="chapter" data-level="4.5.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#appearances-matter"><i class="fa fa-check"></i><b>4.5.1</b> Appearances matter</a></li>
<li class="chapter" data-level="4.5.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#where-are-my-programscommands-at"><i class="fa fa-check"></i><b>4.5.2</b> Where are my programs/commands at?!</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#a-few-more-important-keystrokes"><i class="fa fa-check"></i><b>4.6</b> A Few More Important Keystrokes</a></li>
<li class="chapter" data-level="4.7" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#a-short-list-of-additional-useful-commands."><i class="fa fa-check"></i><b>4.7</b> A short list of additional useful commands.</a></li>
<li class="chapter" data-level="4.8" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#two-important-computing-concepts"><i class="fa fa-check"></i><b>4.8</b> Two important computing concepts</a><ul>
<li class="chapter" data-level="4.8.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#compression"><i class="fa fa-check"></i><b>4.8.1</b> Compression</a></li>
<li class="chapter" data-level="4.8.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#hashing"><i class="fa fa-check"></i><b>4.8.2</b> Hashing</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#unix-quick-study-guide"><i class="fa fa-check"></i><b>4.9</b> Unix: Quick Study Guide</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="shell-programming.html"><a href="shell-programming.html"><i class="fa fa-check"></i><b>5</b> Shell programming</a><ul>
<li class="chapter" data-level="5.1" data-path="shell-programming.html"><a href="shell-programming.html#advanced-repitition"><i class="fa fa-check"></i><b>5.1</b> Advanced repitition</a></li>
<li class="chapter" data-level="5.2" data-path="shell-programming.html"><a href="shell-programming.html#variables"><i class="fa fa-check"></i><b>5.2</b> Variables</a></li>
<li class="chapter" data-level="5.3" data-path="shell-programming.html"><a href="shell-programming.html#looping"><i class="fa fa-check"></i><b>5.3</b> looping</a></li>
<li class="chapter" data-level="5.4" data-path="shell-programming.html"><a href="shell-programming.html#further-reading"><i class="fa fa-check"></i><b>5.4</b> Further reading</a></li>
<li class="chapter" data-level="5.5" data-path="shell-programming.html"><a href="shell-programming.html#reading-files-line-by-line"><i class="fa fa-check"></i><b>5.5</b> reading files line by line</a></li>
<li class="chapter" data-level="5.6" data-path="shell-programming.html"><a href="shell-programming.html#difference-between-double-and-single-quotes"><i class="fa fa-check"></i><b>5.6</b> Difference between double and single quotes</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html"><i class="fa fa-check"></i><b>6</b> Sed, awk, and regular expressions</a></li>
<li class="chapter" data-level="7" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html"><i class="fa fa-check"></i><b>7</b> High Performance Computing (HPC) Environments</a><ul>
<li class="chapter" data-level="7.1" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#accessing-remote-computers"><i class="fa fa-check"></i><b>7.1</b> Accessing remote computers</a></li>
<li class="chapter" data-level="7.2" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#transferring-files-to-remote-computers"><i class="fa fa-check"></i><b>7.2</b> Transferring files to remote computers</a><ul>
<li class="chapter" data-level="7.2.1" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#scp"><i class="fa fa-check"></i><b>7.2.1</b> scp</a></li>
<li class="chapter" data-level="7.2.2" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#globus"><i class="fa fa-check"></i><b>7.2.2</b> Globus</a></li>
<li class="chapter" data-level="7.2.3" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#interfacing-with-the-cloud"><i class="fa fa-check"></i><b>7.2.3</b> Interfacing with “The Cloud”</a></li>
<li class="chapter" data-level="7.2.4" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#getting-files-from-a-sequencing-center"><i class="fa fa-check"></i><b>7.2.4</b> Getting files from a sequencing center</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#activatinginstalling-software"><i class="fa fa-check"></i><b>7.3</b> Activating/Installing software</a><ul>
<li class="chapter" data-level="7.3.1" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#modules"><i class="fa fa-check"></i><b>7.3.1</b> Modules</a></li>
<li class="chapter" data-level="7.3.2" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#miniconda"><i class="fa fa-check"></i><b>7.3.2</b> Miniconda</a></li>
<li class="chapter" data-level="7.3.3" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#exporting-environments"><i class="fa fa-check"></i><b>7.3.3</b> Exporting environments</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#boneyard"><i class="fa fa-check"></i><b>7.4</b> Boneyard</a></li>
<li class="chapter" data-level="7.5" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#the-queue-slurmsgeuge"><i class="fa fa-check"></i><b>7.5</b> The Queue (SLURM/SGE/UGE)</a></li>
<li class="chapter" data-level="7.6" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#modules-package"><i class="fa fa-check"></i><b>7.6</b> Modules package</a></li>
<li class="chapter" data-level="7.7" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#compiling-programs-without-admin-privileges"><i class="fa fa-check"></i><b>7.7</b> Compiling programs without admin privileges</a></li>
<li class="chapter" data-level="7.8" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#job-arrays"><i class="fa fa-check"></i><b>7.8</b> Job arrays</a></li>
<li class="chapter" data-level="7.9" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#writing-stdout-and-stderr-to-files"><i class="fa fa-check"></i><b>7.9</b> Writing stdout and stderr to files</a></li>
<li class="chapter" data-level="7.10" data-path="high-performance-computing-hpc-environments.html"><a href="high-performance-computing-hpc-environments.html#breaking-stuff-down"><i class="fa fa-check"></i><b>7.10</b> Breaking stuff down</a></li>
</ul></li>
<li class="part"><span><b>II Part II: Reproducible Research Strategies</b></span></li>
<li class="chapter" data-level="8" data-path="introduction-to-reproducible-research.html"><a href="introduction-to-reproducible-research.html"><i class="fa fa-check"></i><b>8</b> Introduction to Reproducible Research</a></li>
<li class="chapter" data-level="9" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html"><i class="fa fa-check"></i><b>9</b> Rstudio and Project-centered Organization</a><ul>
<li class="chapter" data-level="9.1" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#organizing-big-projects"><i class="fa fa-check"></i><b>9.1</b> Organizing big projects</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="version-control.html"><a href="version-control.html"><i class="fa fa-check"></i><b>10</b> Version control</a><ul>
<li class="chapter" data-level="10.1" data-path="version-control.html"><a href="version-control.html#why-use-version-control"><i class="fa fa-check"></i><b>10.1</b> Why use version control?</a></li>
<li class="chapter" data-level="10.2" data-path="version-control.html"><a href="version-control.html#git-workings"><i class="fa fa-check"></i><b>10.2</b> How git works</a></li>
<li class="chapter" data-level="10.3" data-path="version-control.html"><a href="version-control.html#git-workflow-patterns"><i class="fa fa-check"></i><b>10.3</b> git workflow patterns</a></li>
<li class="chapter" data-level="10.4" data-path="version-control.html"><a href="version-control.html#using-git-with-rstudio"><i class="fa fa-check"></i><b>10.4</b> using git with Rstudio</a></li>
<li class="chapter" data-level="10.5" data-path="version-control.html"><a href="version-control.html#git-on-the-command-line"><i class="fa fa-check"></i><b>10.5</b> git on the command line</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="a-fast-furious-overview-of-the-tidyverse.html"><a href="a-fast-furious-overview-of-the-tidyverse.html"><i class="fa fa-check"></i><b>11</b> A fast, furious overview of the tidyverse</a></li>
<li class="chapter" data-level="12" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html"><i class="fa fa-check"></i><b>12</b> Authoring reproducibly with Rmarkdown</a><ul>
<li class="chapter" data-level="12.1" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#notebooks"><i class="fa fa-check"></i><b>12.1</b> Notebooks</a></li>
<li class="chapter" data-level="12.2" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#references"><i class="fa fa-check"></i><b>12.2</b> References</a><ul>
<li class="chapter" data-level="12.2.1" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#zotero-and-rmarkdown"><i class="fa fa-check"></i><b>12.2.1</b> Zotero and Rmarkdown</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#bookdown"><i class="fa fa-check"></i><b>12.3</b> Bookdown</a></li>
<li class="chapter" data-level="12.4" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#google-docs"><i class="fa fa-check"></i><b>12.4</b> Google Docs</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="using-python.html"><a href="using-python.html"><i class="fa fa-check"></i><b>13</b> Using python</a></li>
<li class="part"><span><b>III Part III: Bioinformatic Analyses</b></span></li>
<li class="chapter" data-level="14" data-path="overview-of-bioinformatic-analyses.html"><a href="overview-of-bioinformatic-analyses.html"><i class="fa fa-check"></i><b>14</b> Overview of Bioinformatic Analyses</a></li>
<li class="chapter" data-level="15" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html"><i class="fa fa-check"></i><b>15</b> DNA Sequences and Sequencing</a><ul>
<li class="chapter" data-level="15.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#dna-stuff"><i class="fa fa-check"></i><b>15.1</b> DNA Stuff</a><ul>
<li class="chapter" data-level="15.1.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#dna-replication-with-dna-polymerase"><i class="fa fa-check"></i><b>15.1.1</b> DNA Replication with DNA Polymerase</a></li>
<li class="chapter" data-level="15.1.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#the-importance-of-the-3-hydroxyl"><i class="fa fa-check"></i><b>15.1.2</b> The importance of the 3’ hydroxyl…</a></li>
</ul></li>
<li class="chapter" data-level="15.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#sanger-sequencing"><i class="fa fa-check"></i><b>15.2</b> Sanger sequencing</a></li>
<li class="chapter" data-level="15.3" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#illumina-sequencing-by-synthesis"><i class="fa fa-check"></i><b>15.3</b> Illumina Sequencing by Synthesis</a></li>
<li class="chapter" data-level="15.4" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#library-prep-protocols"><i class="fa fa-check"></i><b>15.4</b> Library Prep Protocols</a><ul>
<li class="chapter" data-level="15.4.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#wgs"><i class="fa fa-check"></i><b>15.4.1</b> WGS</a></li>
<li class="chapter" data-level="15.4.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#rad-seq-methods"><i class="fa fa-check"></i><b>15.4.2</b> RAD-Seq methods</a></li>
<li class="chapter" data-level="15.4.3" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#amplicon-sequencing"><i class="fa fa-check"></i><b>15.4.3</b> Amplicon Sequencing</a></li>
<li class="chapter" data-level="15.4.4" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#capture-arrays-rapture-etc."><i class="fa fa-check"></i><b>15.4.4</b> Capture arrays, RAPTURE, etc.</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="16" data-path="bioinformatic-file-formats-and-associated-tools.html"><a href="bioinformatic-file-formats-and-associated-tools.html"><i class="fa fa-check"></i><b>16</b> Bioinformatic file formats and associated tools</a><ul>
<li class="chapter" data-level="16.1" data-path="bioinformatic-file-formats-and-associated-tools.html"><a href="bioinformatic-file-formats-and-associated-tools.html#sequences"><i class="fa fa-check"></i><b>16.1</b> Sequences</a></li>
<li class="chapter" data-level="16.2" data-path="bioinformatic-file-formats-and-associated-tools.html"><a href="bioinformatic-file-formats-and-associated-tools.html#fasta"><i class="fa fa-check"></i><b>16.2</b> FASTA</a></li>
<li class="chapter" data-level="16.3" data-path="bioinformatic-file-formats-and-associated-tools.html"><a href="bioinformatic-file-formats-and-associated-tools.html#fastq"><i class="fa fa-check"></i><b>16.3</b> FASTQ</a><ul>
<li class="chapter" data-level="16.3.1" data-path="bioinformatic-file-formats-and-associated-tools.html"><a href="bioinformatic-file-formats-and-associated-tools.html#line-1-illumina-identifier-lines"><i class="fa fa-check"></i><b>16.3.1</b> Line 1: Illumina identifier lines</a></li>
<li class="chapter" data-level="16.3.2" data-path="bioinformatic-file-formats-and-associated-tools.html"><a href="bioinformatic-file-formats-and-associated-tools.html#line-4-base-quality-scores"><i class="fa fa-check"></i><b>16.3.2</b> Line 4: Base quality scores</a></li>
<li class="chapter" data-level="16.3.3" data-path="bioinformatic-file-formats-and-associated-tools.html"><a href="bioinformatic-file-formats-and-associated-tools.html#a-fastq-tidyverse-interlude"><i class="fa fa-check"></i><b>16.3.3</b> A FASTQ ‘tidyverse’ Interlude</a></li>
<li class="chapter" data-level="16.3.4" data-path="bioinformatic-file-formats-and-associated-tools.html"><a href="bioinformatic-file-formats-and-associated-tools.html#comparing-read-1-to-read-2"><i class="fa fa-check"></i><b>16.3.4</b> Comparing read 1 to read 2</a></li>
</ul></li>
<li class="chapter" data-level="16.4" data-path="bioinformatic-file-formats-and-associated-tools.html"><a href="bioinformatic-file-formats-and-associated-tools.html#alignments"><i class="fa fa-check"></i><b>16.4</b> Alignments</a><ul>
<li class="chapter" data-level="16.4.1" data-path="bioinformatic-file-formats-and-associated-tools.html"><a href="bioinformatic-file-formats-and-associated-tools.html#samtools"><i class="fa fa-check"></i><b>16.4.1</b> Samtools</a></li>
</ul></li>
<li class="chapter" data-level="16.5" data-path="bioinformatic-file-formats-and-associated-tools.html"><a href="bioinformatic-file-formats-and-associated-tools.html#variants"><i class="fa fa-check"></i><b>16.5</b> Variants</a></li>
<li class="chapter" data-level="16.6" data-path="bioinformatic-file-formats-and-associated-tools.html"><a href="bioinformatic-file-formats-and-associated-tools.html#segments"><i class="fa fa-check"></i><b>16.6</b> Segments</a></li>
<li class="chapter" data-level="16.7" data-path="bioinformatic-file-formats-and-associated-tools.html"><a href="bioinformatic-file-formats-and-associated-tools.html#conversionextractions-between-different-formats"><i class="fa fa-check"></i><b>16.7</b> Conversion/Extractions between different formats</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="genome-assembly.html"><a href="genome-assembly.html"><i class="fa fa-check"></i><b>17</b> Genome Assembly</a></li>
<li class="chapter" data-level="18" data-path="alignment-of-sequence-data-to-a-reference-genome.html"><a href="alignment-of-sequence-data-to-a-reference-genome.html"><i class="fa fa-check"></i><b>18</b> Alignment of sequence data to a reference genome</a><ul>
<li class="chapter" data-level="18.1" data-path="alignment-of-sequence-data-to-a-reference-genome.html"><a href="alignment-of-sequence-data-to-a-reference-genome.html#preprocess"><i class="fa fa-check"></i><b>18.1</b> Preprocess ?</a></li>
<li class="chapter" data-level="18.2" data-path="alignment-of-sequence-data-to-a-reference-genome.html"><a href="alignment-of-sequence-data-to-a-reference-genome.html#read-groups"><i class="fa fa-check"></i><b>18.2</b> Read Groups</a></li>
<li class="chapter" data-level="18.3" data-path="alignment-of-sequence-data-to-a-reference-genome.html"><a href="alignment-of-sequence-data-to-a-reference-genome.html#merging-bam-files"><i class="fa fa-check"></i><b>18.3</b> Merging BAM files</a></li>
<li class="chapter" data-level="18.4" data-path="alignment-of-sequence-data-to-a-reference-genome.html"><a href="alignment-of-sequence-data-to-a-reference-genome.html#divide-and-conquer-strategies"><i class="fa fa-check"></i><b>18.4</b> Divide and Conquer Strategies</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="variant-calling-with-gatk.html"><a href="variant-calling-with-gatk.html"><i class="fa fa-check"></i><b>19</b> Variant calling with GATK</a></li>
<li class="chapter" data-level="20" data-path="bioinformatics-for-rad-seq-data-with-and-without-a-reference-genome.html"><a href="bioinformatics-for-rad-seq-data-with-and-without-a-reference-genome.html"><i class="fa fa-check"></i><b>20</b> Bioinformatics for RAD seq data with and without a reference genome</a></li>
<li class="chapter" data-level="21" data-path="processing-amplicon-sequencing-data.html"><a href="processing-amplicon-sequencing-data.html"><i class="fa fa-check"></i><b>21</b> Processing amplicon sequencing data</a></li>
<li class="chapter" data-level="22" data-path="genome-annotation.html"><a href="genome-annotation.html"><i class="fa fa-check"></i><b>22</b> Genome Annotation</a></li>
<li class="chapter" data-level="23" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html"><i class="fa fa-check"></i><b>23</b> Whole genome alignment strategies</a><ul>
<li class="chapter" data-level="23.1" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#mapping-of-scaffolds-to-a-closely-related-genome"><i class="fa fa-check"></i><b>23.1</b> Mapping of scaffolds to a closely related genome</a></li>
<li class="chapter" data-level="23.2" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#obtaining-ancestral-states-from-an-outgroup-genome"><i class="fa fa-check"></i><b>23.2</b> Obtaining Ancestral States from an Outgroup Genome</a><ul>
<li class="chapter" data-level="23.2.1" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#using-lastz-to-align-coho-to-the-chinook-genome"><i class="fa fa-check"></i><b>23.2.1</b> Using LASTZ to align coho to the chinook genome</a></li>
<li class="chapter" data-level="23.2.2" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#try-on-the-chinook-chromosomes"><i class="fa fa-check"></i><b>23.2.2</b> Try on the chinook chromosomes</a></li>
<li class="chapter" data-level="23.2.3" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#explore-the-other-parameters-more"><i class="fa fa-check"></i><b>23.2.3</b> Explore the other parameters more</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>IV Part IV: Analysis of Big Variant Data</b></span></li>
<li class="chapter" data-level="24" data-path="bioinformatic-analysis-on-variant-data.html"><a href="bioinformatic-analysis-on-variant-data.html"><i class="fa fa-check"></i><b>24</b> Bioinformatic analysis on variant data</a></li>
<li class="part"><span><b>V Part V: Population Genomics</b></span></li>
<li class="chapter" data-level="25" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html"><i class="fa fa-check"></i><b>25</b> Topics in pop gen</a><ul>
<li class="chapter" data-level="25.1" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#coalescent"><i class="fa fa-check"></i><b>25.1</b> Coalescent</a></li>
<li class="chapter" data-level="25.2" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#measures-of-genetic-diversity-and-such"><i class="fa fa-check"></i><b>25.2</b> Measures of genetic diversity and such</a></li>
<li class="chapter" data-level="25.3" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#demographic-inference-with-partial-a-partial-i-and-moments"><i class="fa fa-check"></i><b>25.3</b> Demographic inference with <span class="math inline">\(\partial a \partial i\)</span> and <em>moments</em></a></li>
<li class="chapter" data-level="25.4" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#balls-in-boxes"><i class="fa fa-check"></i><b>25.4</b> Balls in Boxes</a></li>
<li class="chapter" data-level="25.5" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#some-landscape-genetics"><i class="fa fa-check"></i><b>25.5</b> Some landscape genetics</a></li>
<li class="chapter" data-level="25.6" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#relationship-inference"><i class="fa fa-check"></i><b>25.6</b> Relationship Inference</a></li>
<li class="chapter" data-level="25.7" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#tests-for-selection"><i class="fa fa-check"></i><b>25.7</b> Tests for Selection</a></li>
<li class="chapter" data-level="25.8" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#multivariate-associations-gea-etc."><i class="fa fa-check"></i><b>25.8</b> Multivariate Associations, GEA, etc.</a></li>
<li class="chapter" data-level="25.9" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#estimating-heritability-in-the-wild"><i class="fa fa-check"></i><b>25.9</b> Estimating heritability in the wild</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references-1.html"><a href="references-1.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://bookdown.org" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="high-performance-computing-hpc-environments" class="section level1">
<h1><span class="header-section-number">Chapter 7</span> High Performance Computing (HPC) Environments</h1>
<p>Hey Eric! You might consider breaking this into two separate chatpers: 1 = working on remote computers
and 2 = high-performance computing. The first could include all the stuff about scp, globus, rclone,
and google drive.</p>
<p>This is going to be a chapter on using High Performance Computing clusters.</p>
<p>There is a lot to convey here about using queues and things.</p>
<p>I know SGE pretty well at this point, but others might use slurm.</p>
<p>Here is a good page with a comparison: <a href="https://confluence.csiro.au/display/SC/Reference+Guide%3A+Migrating+from+SGE+to+SLURM">https://confluence.csiro.au/display/SC/Reference+Guide%3A+Migrating+from+SGE+to+SLURM</a></p>
<p>And here is a good primer on SGE stuff: <a href="https://confluence.si.edu/display/HPC/Monitoring+your+Jobs">https://confluence.si.edu/display/HPC/Monitoring+your+Jobs</a></p>
<p>I guess I’ll have to see what the CSU students have access to.</p>
<div id="accessing-remote-computers" class="section level2">
<h2><span class="header-section-number">7.1</span> Accessing remote computers</h2>
<p>Start off with some stuff about ssh and scp.</p>
<p>Maybe have a section on public and private keys so you don’t have to put
your password in every time (I would like to get better with that,
as well).</p>
<p>Here is the deal on that. On a mac, you can use</p>
<pre><code>ssh-keygen -t rsa -b 4096</code></pre>
<p>and be sure to give it a password, so that your private key is not unencrypted.
For, if it is, then there is a chance (I do believe) that if someone were to
obtain that file, they could gain access to all the computers you are authorized on.</p>
<p>Note that a lot of tutorials on the web have you generating a private key without
any encryption. That is lame.</p>
<p>Then copy the .ssh/id_rsa.pub key to the .ssh/authorized_keys file on the server
(creating it if it needs to be there). Then ssh to the server and your Mac will pop
up a window asking for a password or will prompt on the command line for one (note that the password stays local!). You put in
the password that you used when you created the private key. That
can be saved in the Mac keychain (on old version of OSX you might
get asked if you want to save it in the Mac keychain). Voila!
Now you have access to the server with no need to type a password in there.</p>
<p>Note, since Sierra, you need to add this to a <code>.ssh/config</code> file to get the password
stored in the keychain:</p>
<pre><code>Host *
   AddKeysToAgent yes
   UseKeychain yes     </code></pre>
<p>And, it turns out that you also need to make sure the permissions on that
file are set appropriately:</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb55-1" data-line-number="1"><span class="fu">chmod</span> 600 ~/.ssh/config</a></code></pre></div>
<p>It is actually pretty darn simple…</p>
<p>Note, to have access from another computer to the server, you probably just create
a keypair for that computer, and add the public key to the authorized_keys.</p>
<p>Then add something like this to your .bashrc:</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb56-1" data-line-number="1"><span class="co"># for quick ssh to hoffman</span></a>
<a class="sourceLine" id="cb56-2" data-line-number="2"><span class="bu">alias</span> hoffy=<span class="st">&#39;ssh eriq@hoffman2.idre.ucla.edu&#39;</span></a>
<a class="sourceLine" id="cb56-3" data-line-number="3"></a>
<a class="sourceLine" id="cb56-4" data-line-number="4"><span class="co"># to be used in scp.  like scp file $hoff:~/</span></a>
<a class="sourceLine" id="cb56-5" data-line-number="5"><span class="va">hoff=</span>eriq<span class="ex">@hoffman2.idre.ucla.edu</span></a></code></pre></div>
<p>Note that this should only be done on a private computer account, not on a shared account on a computer.
Otherwise, everyone using that account will have access.</p>
<p>Also, an interesting thing to investigate might be SSHF/FUSE which just might
let one mount the HPC filesystem on a local directory, so you can work with
files there using RStudio, etc, get them all debugged, and then run them.</p>
<p>Check out some information about that here:
<a href="https://www.digitalocean.com/community/tutorials/how-to-use-sshfs-to-mount-remote-file-systems-over-ssh">https://www.digitalocean.com/community/tutorials/how-to-use-sshfs-to-mount-remote-file-systems-over-ssh</a>.</p>
<p>I went to <a href="https://osxfuse.github.io/">https://osxfuse.github.io/</a> and I downloaded SSHFS 2.5.0 and FUSE for Mac OS X. Then I installed them,
doing a simple default install for each.</p>
<p>Then, you make a mountpoint, which you can do in your home directory if you want:</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb57-1" data-line-number="1"><span class="fu">mkdir</span> ~/hoffman  # this is the mountpoint</a>
<a class="sourceLine" id="cb57-2" data-line-number="2"><span class="ex">sshfs</span> -o allow_other,defer_permissions,IdentityFile=~/.ssh/id_rsa eriq@hoffman2.idre.ucla.edu:/u/home/e/eriq/  ~/hoffman</a>
<a class="sourceLine" id="cb57-3" data-line-number="3"></a>
<a class="sourceLine" id="cb57-4" data-line-number="4"><span class="co"># close it out:</span></a>
<a class="sourceLine" id="cb57-5" data-line-number="5"><span class="fu">umount</span> ~/hoffman/</a></code></pre></div>
<p>Note that the absolute path to my home folder there seemed to be important.</p>
<p>Also, I can’t get to <code>/u/nobackup/kruegg/eriq</code> from here by going through the symlink in my home directory
on hoffman, since that is essentially a different volume. But I should
be able to do this:</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb58-1" data-line-number="1"><span class="ex">sshfs</span> -o allow_other,defer_permissions,IdentityFile=~/.ssh/id_rsa eriq@hoffman2.idre.ucla.edu:/u/nobackup/kruegg/eriq  ~/hoffman</a></code></pre></div>
<p>Yep! that works. And I can open all those files from within RStudio. Cool.</p>
<p>Note, I will have to have a separate mount point for $SCRATCH as well.</p>
<p>Shoot! That works really seamlessly!</p>
<p>I can also add to my .bashrc:</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb59-1" data-line-number="1"><span class="bu">alias</span> hoffuse=<span class="st">&#39;sshfs -o allow_other,defer_permissions,IdentityFile=~/.ssh/id_rsa eriq@hoffman2.idre.ucla.edu:/u/nobackup/kruegg/eriq  ~/hoffman&#39;</span></a></code></pre></div>
<p>Now, the bad news: you can open an Rstudio project from that remote, mounted volume, but it
doesn’t seem to really work. It is incredibly slow, and then fails to quit properly, etc.</p>
<p>After force-quitting it, I am unable to unmount the hoffuse volume. What a mess. OK, I probably
won’t try (running the rstudio project over FUSE). But it might be useful still for editing
small things.</p>
</div>
<div id="transferring-files-to-remote-computers" class="section level2">
<h2><span class="header-section-number">7.2</span> Transferring files to remote computers</h2>
<div id="scp" class="section level3">
<h3><span class="header-section-number">7.2.1</span> scp</h3>
</div>
<div id="globus" class="section level3">
<h3><span class="header-section-number">7.2.2</span> Globus</h3>
</div>
<div id="interfacing-with-the-cloud" class="section level3">
<h3><span class="header-section-number">7.2.3</span> Interfacing with “The Cloud”</h3>
<p>Increasingly, data scientists and tech companies alike are keeping their
data “in the cloud.” This means that they
pay a large tech firm like Amazon, Dropbox, or Google to store their data for them in
a place that can be accessed via the internet. There are many advantages to this
model. For one thing, the company that serves the data often will create multiple copies
of the data for backup and redundancy: a fire in a single data center is not a calamity
because the data are also stored elsewhere, and can often be accessed seamlessly from those
other locations with no apparent disruption of service. For another, companies that are
in the business of storing and serving
data to multiple clients have data centers that are well-networked, so that getting
data onto and off of their storage systems can be done very quickly over the internet
by an end-user with a good internet connection.</p>
<p>Five years ago, the idea of storing next generation sequencing data might have
sounded a little
crazy—it always seemed a laborious task getting the data off of the remote server at the
sequencing center, so why not just keep the data in-house once you have it?
To be sure, keeping a copy of your
data in-house still can make sense for long-term data archiving needs, but, today, cloud
storage for your sequencing data can make a lot of sense. A few reasons are:</p>
<ol style="list-style-type: decimal">
<li>Transferring your data from the cloud to the remote HPC system
that you use to process the data can be very fast.</li>
<li>As above, your data can be redundantly backed up.</li>
<li>If your institution (university, agency, etc.) has an agreement with a cloud storage
service that provides you with unlimited storage and free network access, then storing
your sequencing data in the cloud will cost considerably less than buying a dedicated
large system of hard drives for data backup. (One must wonder if service
agreements might not be at risk of renegotiation if many researchers start using their
unlimited institutional cloud storage space to store and/or archive their
next generation sequencing data sets. My own agency’s contract with Google runs
through 2021…but I have to think that these services are making plenty of money, even
if a handful of researchers store big sequence data in the cloud. Nonetheless, you
should be careful not to put multiple copies of data sets, or intermediate files that
are easily regenerated, up in the cloud.)</li>
<li>If you are a PI with many lab members wishing to access the same data set, or even if
you are just a regular Joe/Joanna researcher but you wish to share your data, it is
possible to effect that using your cloud service’s sharing settings. We will discuss
how to do this with Google Drive.</li>
</ol>
<p>There are clearly advantages to using the cloud, but one small hurdle remains. Most
of the time, working in an HPC environment, we are using Unix, which provides a consistent
set of tools for interfacing with other computers using SSH-based protocols (like <code>scp</code>
for copying files from one remote computer to another). Unfortunately, many common
cloud storage services do not offer an SSH based interface. Rather, they typically process
requests from clients using an HTTPS protocol. This protocol, which effectively runs the
world-wide web, is a natural choice for cloud services that most people will access
using a web browser; however, Unix does not traditionally come with a utility or command
to easily process the types of HTTPS transactions needed to network with
cloud storage. Furthermore, there must be some security when it comes to accessing
your cloud-based storage—you don’t want everyone to be able to access your files, so
your cloud service needs to have some way of authenticating people
(you and your labmates for example) that are authorized to access your data.</p>
<p>These problems have been overcome by a utility called <code>rclone</code>, the product of a
comprehensive open-source software project that brings the functionality of the
<code>rsync</code> utility (a common Unix tool used to synchronize and mirror file systems)
to cloud-based storage. (Note: <code>rclone</code> has nothing to do with the R programming
language, despite its name that looks like an R package.)
Currently <code>rclone</code> provides a consistent interface for accessing
files from over 35 different cloud storage providers, including Box, Dropbox, Google Drive,
and Microsoft OneDrive. Binaries for <code>rclone</code> can be downloaded for your desktop
machine from <a href="https://rclone.org/downloads/">https://rclone.org/downloads/</a>. We will
talk about how to install it on your HPC system later.</p>
<p>Once <code>rclone</code> is installed and in your <code>PATH</code>, you invoke it in your terminal
with the command <code>rclone</code>. Before we get into the details of the various <code>rclone</code> subcommands,
it will be helpful to take a glance at the information <code>rclone</code> records when it
configures itself to talk to your cloud service. To do so, it creates a file called <code>~/.config/rclone/rclone.conf</code>, where it stores information about all the different
connections to cloud services you have set up. For example, that
file on my system looks like this:</p>
<pre><code>[gdrive-rclone]
type = drive
scope = drive
root_folder_id = 1I2EDV465N5732Tx1FFAiLWOqZRJcAzUd
token = {&quot;access_token&quot;:&quot;bs43.94cUFOe6SjjkofZ&quot;,&quot;token_type&quot;:&quot;Bearer&quot;,&quot;refresh_token&quot;:&quot;1/MrtfsRoXhgc&quot;,&quot;expiry&quot;:&quot;2019-04-29T22:51:58.148286-06:00&quot;}
client_id = 2934793-oldk97lhld88dlkh301hd.apps.googleusercontent.com
client_secret = MMq3jdsjdjgKTGH4rNV_y-NbbG</code></pre>
<p>In this configuration:</p>
<ul>
<li><code>gdrive-rclone</code> is the name by which rclone refers to this cloud storage location</li>
<li><code>root_folder_id</code> is the ID of the Google Drive folder that can be thought of as the root directory of <code>gdrive-rclone</code>. This ID is not the simple name of that directory on
your Google Drive, rather it is the unique name given by Google Drive to that directory.
You can see it by navigating in your browser to the directory you want and finding it
after the last slash in the URL. For example, in the above case, the URL is:
<code>https://drive.google.com/drive/u/1/folders/1I2EDV465N5732Tx1FFAiLWOqZRJcAzUd</code></li>
<li><code>client_id</code> and <code>client_secret</code> are like a username and a shared secret that <code>rclone</code> uses
to authenticate the user to Google Drive as who they say they are.</li>
<li><code>token</code> are the credentials used by <code>rclone</code> to make requests of Google Drive on the basis
of the user.</li>
</ul>
<p>Note: the above does not include my
real credentials, as then anyone could use them to access my Google Drive!</p>
<p>To set up your own configuration file to use Google Drive, you will use the <code>rclone config</code>
command, but before you do that, you will want to wrangle a client_id from Google. Follow
the directions at <a href="https://rclone.org/drive/#making-your-own-client-id">https://rclone.org/drive/#making-your-own-client-id</a>. Things are a little different from in their step
by step, but you can muddle through to get to a screen with a client_ID and a client
secret that you can copy onto your clipboard.</p>
<p>Once you have done that, then run <code>rclone config</code> and follow the prompts. A
typical session of <code>rclone config</code> for Google Drive access is given
<a href="https://rclone.org/drive/">here</a>. Don’t choose to do the advanced setup; however
do use “auto config,” which will bounce up a web page and let you authenticate rclone
to your Google account.</p>
<div id="basic-maneuvers" class="section level4">
<h4><span class="header-section-number">7.2.3.1</span> Basic Maneuvers</h4>
<p>The syntax for use is:</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb61-1" data-line-number="1"><span class="ex">rclone</span> [options] subcommand  parameter1 [parameter 2...]</a></code></pre></div>
<p>The “subcommand” part tells <code>rclone</code> what you want to do, like <code>copy</code> or <code>sync</code>, and
the “parameter” part of the above syntax is typically a path
specification to a directory or a file. In using rclone to access the
cloud there is not a root directory, like <code>/</code> in Unix. Instead, each remote
cloud access point is treated as the root directory, and you refer to it
by the name of the configuration followed by a colon. In our example,
<code>gdrive-rclone:</code> is the root, and we don’t need to add a <code>/</code> after it to
start a path with it. Thus <code>gdrive-rclone:this_dir/that_dir</code> is a
valid path for <code>rclone</code> to a location on my Google Drive.</p>
<p>Very often when moving, copying, or syncing files, the parameters
consist of:</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode s"><code class="sourceCode gnuassembler"><a class="sourceLine" id="cb62-1" data-line-number="1">source-directory  destination-directory</a></code></pre></div>
<p>One very important point is that, unlike the Unix commands <code>cp</code> and <code>mv</code>, rclone
likes to operate on directories, not on multiple named files.</p>
<p>A few key subcommands:</p>
<ul>
<li><code>ls</code>, <code>lsd</code>, and <code>lsl</code> are like <code>ls</code>, <code>ls -d</code> and <code>ls -l</code></li>
</ul>
<div class="sourceCode" id="cb63"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb63-1" data-line-number="1"><span class="ex">rclone</span>  lsd gdrive-rclone:</a>
<a class="sourceLine" id="cb63-2" data-line-number="2"><span class="ex">rclone</span>  lsd gdrive-rclone:NOFU</a></code></pre></div>
<ul>
<li><code>copy</code>: copy the <em>contents</em> of a source <em>directory</em> to a destination <em>directory</em>. One super cool
thing about this is that <code>rclone</code> won’t re-copy files that are already on the destination and which
are identical to those in the source directory.</li>
</ul>
<div class="sourceCode" id="cb64"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb64-1" data-line-number="1"><span class="ex">rclone</span> copy bams gdrive-rclone:NOFU/bams</a></code></pre></div>
<p>Note that the destination directory will be created if it does not already exist.<br />
- <code>sync</code>: make the contents of the destination directory look just like the
contents of the source directory. <em>WARNING</em> This will delete files in the destination
directory that do not appear in the source directory.</p>
<p>A few key options:</p>
<ul>
<li><code>--dry-run</code>: don’t actually copy, sync, or move anything. Just tell me what you would have done.</li>
<li><code>-P</code>, <code>-v</code>, <code>-vv</code>: give me progress information, verbose output, or super-verbose output, respectively.</li>
<li><code>--tpslimit 10</code>: don’t make any more than 10 transactions a second with Google Drive (should always be used when transferring files)</li>
<li><code>---fast-list</code>: combine multiple transactions together. Should always be used with Google Drive,
especially when handling lots of files.</li>
<li><code>--drive-shared-with-me</code>: make the “root” directory a directory that shows all
of the Google Drive folders that people have shared with you. This is key for accessing
folders that have been shared with you.</li>
</ul>
<p>For example, try something like:</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb65-1" data-line-number="1"><span class="ex">rclone</span> --drive-shared-with-me lsd gdrive-rclone:</a></code></pre></div>
</div>
<div id="rclone-filter" class="section level4">
<h4><span class="header-section-number">7.2.3.2</span> filtering: Be particular about the files you transfer</h4>
<p><code>rclone</code> works a little differently than the Unix utility <code>cp</code>. In particular,
<code>rclone</code> is not set up very well to copy individual files. While there is a
an <code>rclone</code> command known as <code>copyto</code> that will allow you copy a single file,
you cannot (apparently) specify multiple, individual files that you wish to copy.</p>
<p>In other words, you can’t do:</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb66-1" data-line-number="1"><span class="ex">rclone</span> copyto this_file.txt that_file.txt another_file.bam gdrive-rclone:dest_dir</a></code></pre></div>
<p>In general, you will be better off using <code>rclone</code> to copy the <em>contents</em> of a directory
to the inside of the destination directory. However, there are options in <code>rclone</code> that
can keep you from being totally indiscriminate about the files you transfer. In other words,
you can <em>filter</em> the files that get transferred. You can read about that at
<a href="https://rclone.org/filtering/">https://rclone.org/filtering/</a>.</p>
<p>For a quick example, imagine that you have a directory called <code>Data</code> on you Google Drive
that contains both VCF and BAM files. You want to get only the VCF files (ending with <code>.vcf.gz</code>, say)
onto the current working directory on your cluster. Then something like this works:</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb67-1" data-line-number="1"><span class="ex">rclone</span> copy --include *.vcf.gz gdrive-rclone:Data ./</a></code></pre></div>
<p>Note that, if you are issuing this command on a Unix system in a directory
where the pattern <code>*.vcf.gz</code> will expand (by globbing) to multiple files, you will
get an error. In that case, wrap the pattern in a pair of single quotes to keep
the shell from expanding it, like this:</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb68-1" data-line-number="1"><span class="ex">rclone</span> copy --include <span class="st">&#39;*.vcf.gz&#39;</span> gdrive-rclone:Data ./</a></code></pre></div>
</div>
<div id="feel-free-to-make-lots-of-configurations" class="section level4">
<h4><span class="header-section-number">7.2.3.3</span> Feel free to make lots of configurations</h4>
<p>You might want to configure a remote for each directory-specific project.
You can do that by just editing the configuration file. For example,
if I had a directory deep within my Google Drive, inside a chain of folders that
looked like, say, <code>Projects/NGS/Species/Salmon/Chinook/CentralValley/WinterRun</code>
where I was keeping
all my data on a project concerning winter-run Chinook salmon, then it would be
quite inconvenient to type <code>Projects/NGS/Species/Salmon/Chinook/CentralValley/WinterRun</code>
every time I wanted to copy or sync something within that directory. Instead,
I could add the following
lines to my configuration file, essentially copying the existing configuration and
then modifying the configuration name and the <code>root_folder_id</code> to be the
Google Drive identifier for the folder <code>Projects/NGS/Species/Salmon/Chinook/CentralValley/WinterRun</code> (which
one can find by navigating to that folder in a web browser and pulling the ID from the
end of the URL.) The updated configuration could look like:</p>
<pre><code>[gdrive-winter-run]
type = drive
scope = drive
root_folder_id = 1MjOrclmP1udhxOTvLWDHFBVET1dF6CIn
token = {&quot;access_token&quot;:&quot;bs43.94cUFOe6SjjkofZ&quot;,&quot;token_type&quot;:&quot;Bearer&quot;,&quot;refresh_token&quot;:&quot;1/MrtfsRoXhgc&quot;,&quot;expiry&quot;:&quot;2019-04-29T22:51:58.148286-06:00&quot;}
client_id = 2934793-oldk97lhld88dlkh301hd.apps.googleusercontent.com
client_secret = MMq3jdsjdjgKTGH4rNV_y-NbbG</code></pre>
<p>As long as the directory is still within the same Google Drive account, you can re-use
all the authorization information, and just change the <code>[name]</code> part and the <code>root_folder_id</code>.
Now this:</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb70-1" data-line-number="1"><span class="ex">rclone</span> copy src_dir gdrive-winter-run: </a></code></pre></div>
<p>puts items into <code>Projects/NGS/Species/Salmon/Chinook/CentralValley/WinterRun</code> on the Google Drive
without having to type that God-awful long path name.</p>
</div>
<div id="installing-rclone-on-a-remote-machine-without-sudo-access" class="section level4">
<h4><span class="header-section-number">7.2.3.4</span> Installing rclone on a remote machine without sudo access</h4>
<p>The instructions on the website require root access. You don’t have to have root
access to install rclone locally in your home directory somewhere.
Copy the download link from <a href="https://rclone.org/downloads/">https://rclone.org/downloads/</a> for
the type of operating system your remote machine uses (most likely Linux if it is a cluster).
Then transfer that with <code>wget</code>, unzip it and put the binary in your PATH. It will look
something like this:</p>
<pre><code>wget https://downloads.rclone.org/rclone-current-linux-amd64.zip
unzip rclone-current-linux-amd64.zip
cp rclone-current-linux-amd64/rclone ~/bin</code></pre>
<p>You won’t get manual pages on your system, but you can always find the docs on the web.</p>
</div>
<div id="setting-up-configurations-on-the-remote-machine" class="section level4">
<h4><span class="header-section-number">7.2.3.5</span> Setting up configurations on the remote machine…</h4>
<p>Is as easy as copying your config file to where it should go, which
is easy to find using the command:</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb72-1" data-line-number="1"><span class="ex">rclone</span> config file</a></code></pre></div>
</div>
<div id="encrypting-your-config-file" class="section level4">
<h4><span class="header-section-number">7.2.3.6</span> Encrypting your config file</h4>
<p>This makes sense, and it easy: with <code>rclone config</code> encryption is one of
the options. When it is encrypted, use <code>rclone config show</code> to see what
it looks like in clear text.</p>
</div>
<div id="some-other-usage-tips" class="section level4">
<h4><span class="header-section-number">7.2.3.7</span> Some other usage tips</h4>
<p>Following an email exchange with Ren, I should mention how to do an md5
checksum on the remote server to make sure that everything is correctly there.</p>
</div>
</div>
<div id="getting-files-from-a-sequencing-center" class="section level3">
<h3><span class="header-section-number">7.2.4</span> Getting files from a sequencing center</h3>
<p>Very often sequencing centers will post all the data from a single
run of a machine at a secured (or unsecured) http address. You will
need to download those files to operate on them on your cluster or
local machine. However some of the files available on the server
will likely belong to other researchers and you don’t want to waste time
downloading them.</p>
<p>Let’s take an example. Suppose you are sent an email from the sequencing
center that says something like:</p>
<blockquote>
<p>Your samples are AW_F1 (female) and AW_M1 (male).
You should be able to access the data from this link provided by YCGA:
<a href="http://sysg1.cs.yale.edu:3010/5lnO9bs3zfa8LOhESfsYfq3Dc/061719/" class="uri">http://sysg1.cs.yale.edu:3010/5lnO9bs3zfa8LOhESfsYfq3Dc/061719/</a></p>
</blockquote>
<p>You can easily access this web address using <code>rclone</code>. You could set up a new
remote in your rclone config to point to <code>http://sysg1.cs.yale.edu</code>,
but, since you will only be using this once, to get your data, it makes
more sense to just specify the remote on the command line. This can be
done by passing <code>rclone</code> the URL address via the <code>--http-url</code> option, and
then, after that, telling it what protocol to use by adding <code>:http:</code> to
the command. Here is what you would use to list the directories available
at the sequencing center URL:</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb73-1" data-line-number="1"><span class="co"># here is the command</span></a>
<a class="sourceLine" id="cb73-2" data-line-number="2"><span class="ex">%</span> rclone lsd --http-url http://sysg1.cs.yale.edu:3010/5lnO9bs3zfa8LOhESfsYfq3Dc/061719/ :http:</a>
<a class="sourceLine" id="cb73-3" data-line-number="3"></a>
<a class="sourceLine" id="cb73-4" data-line-number="4"><span class="co"># and here is the output</span></a>
<a class="sourceLine" id="cb73-5" data-line-number="5">          <span class="ex">-1</span> 1969-12-31 16:00:00        -1 sjg73_fqs</a>
<a class="sourceLine" id="cb73-6" data-line-number="6">          <span class="ex">-1</span> 1969-12-31 16:00:00        -1 sjg73_supernova_fqs</a></code></pre></div>
<p>Aha! There are two directories that might hold our sequencing data.
I wonder what is in those diretories? The <code>rclone tree</code> command is the
perfect way to drill down into those diretories and look at their contents:</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb74-1" data-line-number="1"><span class="ex">%</span> rclone tree --http-url http://sysg1.cs.yale.edu:3010/5lnO9bs3zfa8LOhESfsYfq3Dc/061719/ :http:</a>
<a class="sourceLine" id="cb74-2" data-line-number="2"><span class="ex">/</span></a>
<a class="sourceLine" id="cb74-3" data-line-number="3">├── <span class="ex">sjg73_fqs</span></a>
<a class="sourceLine" id="cb74-4" data-line-number="4">│   ├── <span class="ex">AW_F1</span></a>
<a class="sourceLine" id="cb74-5" data-line-number="5">│   │   ├── <span class="ex">AW_F1_S2_L001_I1_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-6" data-line-number="6">│   │   ├── <span class="ex">AW_F1_S2_L001_R1_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-7" data-line-number="7">│   │   └── <span class="ex">AW_F1_S2_L001_R2_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-8" data-line-number="8">│   ├── <span class="ex">AW_M1</span></a>
<a class="sourceLine" id="cb74-9" data-line-number="9">│   │   ├── <span class="ex">AW_M1_S3_L001_I1_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-10" data-line-number="10">│   │   ├── <span class="ex">AW_M1_S3_L001_R1_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-11" data-line-number="11">│   │   └── <span class="ex">AW_M1_S3_L001_R2_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-12" data-line-number="12">│   └── <span class="ex">ESP_A1</span></a>
<a class="sourceLine" id="cb74-13" data-line-number="13">│       ├── <span class="ex">ESP_A1_S1_L001_I1_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-14" data-line-number="14">│       ├── <span class="ex">ESP_A1_S1_L001_R1_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-15" data-line-number="15">│       └── <span class="ex">ESP_A1_S1_L001_R2_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-16" data-line-number="16">└── <span class="ex">sjg73_supernova_fqs</span></a>
<a class="sourceLine" id="cb74-17" data-line-number="17">    ├── <span class="ex">AW_F1</span></a>
<a class="sourceLine" id="cb74-18" data-line-number="18">    │   ├── <span class="ex">AW_F1_S2_L001_I1_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-19" data-line-number="19">    │   ├── <span class="ex">AW_F1_S2_L001_R1_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-20" data-line-number="20">    │   └── <span class="ex">AW_F1_S2_L001_R2_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-21" data-line-number="21">    ├── <span class="ex">AW_M1</span></a>
<a class="sourceLine" id="cb74-22" data-line-number="22">    │   ├── <span class="ex">AW_M1_S3_L001_I1_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-23" data-line-number="23">    │   ├── <span class="ex">AW_M1_S3_L001_R1_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-24" data-line-number="24">    │   └── <span class="ex">AW_M1_S3_L001_R2_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-25" data-line-number="25">    └── <span class="ex">ESP_A1</span></a>
<a class="sourceLine" id="cb74-26" data-line-number="26">        ├── <span class="ex">ESP_A1_S1_L001_I1_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-27" data-line-number="27">        ├── <span class="ex">ESP_A1_S1_L001_R1_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-28" data-line-number="28">        └── <span class="ex">ESP_A1_S1_L001_R2_001.fastq.gz</span></a>
<a class="sourceLine" id="cb74-29" data-line-number="29"></a>
<a class="sourceLine" id="cb74-30" data-line-number="30"><span class="ex">8</span> directories, 18 files</a></code></pre></div>
<p>Whoa! That is pretty cool!. From this output we see that there are
subdirectories named <code>AW_F1</code> and <code>AW_M1</code> that hold the files that
we want. And, of course, the <code>ESP_A1</code> samples must belong to someone
else. It would be great if we could just download the files we wanted,
excluding the ones in the <code>ESP_A1</code> directories. It turns out that there is!
<code>rclone</code> has an <code>--exclude</code> option to exclude paths that match certain
patterns (see Section <a href="high-performance-computing-hpc-environments.html#rclone-filter">7.2.3.2</a>, above). We can
experiment by giving <code>rclone copy</code> the <code>--dry-run</code> command to see which
files will be transferred. If we don’t do any filtering, we see this
when we try to dry-run copy the directories to our local directory <code>Alewife/fastqs</code>:</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb75-1" data-line-number="1"><span class="ex">%</span> rclone copy --dry-run --http-url http://sysg1.cs.yale.edu:3010/5lnO9bs3zfa8LOhESfsYfq3Dc/061719/ :http: Alewife/fastqs/</a>
<a class="sourceLine" id="cb75-2" data-line-number="2"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_fqs/ESP_A1/ESP_A1_S1_L001_I1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-3" data-line-number="3"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_fqs/ESP_A1/ESP_A1_S1_L001_R1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-4" data-line-number="4"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_fqs/ESP_A1/ESP_A1_S1_L001_R2_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-5" data-line-number="5"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_supernova_fqs/AW_M1/AW_M1_S3_L001_I1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-6" data-line-number="6"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_supernova_fqs/AW_M1/AW_M1_S3_L001_R1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-7" data-line-number="7"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_supernova_fqs/AW_M1/AW_M1_S3_L001_R2_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-8" data-line-number="8"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_supernova_fqs/AW_F1/AW_F1_S2_L001_I1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-9" data-line-number="9"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_supernova_fqs/AW_F1/AW_F1_S2_L001_R1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-10" data-line-number="10"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_supernova_fqs/AW_F1/AW_F1_S2_L001_R2_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-11" data-line-number="11"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_supernova_fqs/ESP_A1/ESP_A1_S1_L001_I1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-12" data-line-number="12"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_supernova_fqs/ESP_A1/ESP_A1_S1_L001_R1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-13" data-line-number="13"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_supernova_fqs/ESP_A1/ESP_A1_S1_L001_R2_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-14" data-line-number="14"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_fqs/AW_F1/AW_F1_S2_L001_I1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-15" data-line-number="15"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_fqs/AW_F1/AW_F1_S2_L001_R1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-16" data-line-number="16"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_fqs/AW_F1/AW_F1_S2_L001_R2_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-17" data-line-number="17"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_fqs/AW_M1/AW_M1_S3_L001_I1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-18" data-line-number="18"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_fqs/AW_M1/AW_M1_S3_L001_R1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb75-19" data-line-number="19"><span class="ex">2019/07/11</span> 10:33:43 NOTICE: sjg73_fqs/AW_M1/AW_M1_S3_L001_R2_001.fastq.gz: Not copying as --dry-run</a></code></pre></div>
<p>Since we do not want to copy the <code>ESP_A1</code> files we see if we can exclude
them:</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb76-1" data-line-number="1"><span class="ex">%</span> rclone copy --exclude */ESP_A1/* --dry-run --http-url http://sysg1.cs.yale.edu:3010/5lnO9bs3zfa8LOhESfsYfq3Dc/061719/ :http: Alewife/fastqs/</a>
<a class="sourceLine" id="cb76-2" data-line-number="2"><span class="ex">2019/07/11</span> 10:37:22 NOTICE: sjg73_fqs/AW_F1/AW_F1_S2_L001_I1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb76-3" data-line-number="3"><span class="ex">2019/07/11</span> 10:37:22 NOTICE: sjg73_fqs/AW_F1/AW_F1_S2_L001_R2_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb76-4" data-line-number="4"><span class="ex">2019/07/11</span> 10:37:22 NOTICE: sjg73_fqs/AW_F1/AW_F1_S2_L001_R1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb76-5" data-line-number="5"><span class="ex">2019/07/11</span> 10:37:22 NOTICE: sjg73_fqs/AW_M1/AW_M1_S3_L001_I1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb76-6" data-line-number="6"><span class="ex">2019/07/11</span> 10:37:22 NOTICE: sjg73_fqs/AW_M1/AW_M1_S3_L001_R2_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb76-7" data-line-number="7"><span class="ex">2019/07/11</span> 10:37:22 NOTICE: sjg73_fqs/AW_M1/AW_M1_S3_L001_R1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb76-8" data-line-number="8"><span class="ex">2019/07/11</span> 10:37:22 NOTICE: sjg73_supernova_fqs/AW_F1/AW_F1_S2_L001_I1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb76-9" data-line-number="9"><span class="ex">2019/07/11</span> 10:37:22 NOTICE: sjg73_supernova_fqs/AW_F1/AW_F1_S2_L001_R2_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb76-10" data-line-number="10"><span class="ex">2019/07/11</span> 10:37:22 NOTICE: sjg73_supernova_fqs/AW_F1/AW_F1_S2_L001_R1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb76-11" data-line-number="11"><span class="ex">2019/07/11</span> 10:37:22 NOTICE: sjg73_supernova_fqs/AW_M1/AW_M1_S3_L001_I1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb76-12" data-line-number="12"><span class="ex">2019/07/11</span> 10:37:22 NOTICE: sjg73_supernova_fqs/AW_M1/AW_M1_S3_L001_R1_001.fastq.gz: Not copying as --dry-run</a>
<a class="sourceLine" id="cb76-13" data-line-number="13"><span class="ex">2019/07/11</span> 10:37:22 NOTICE: sjg73_supernova_fqs/AW_M1/AW_M1_S3_L001_R2_001.fastq.gz: Not copying as --dry-run</a></code></pre></div>
<p>Booyah! That gets us just what we want. So, then we remove the <code>--dry-run</code> option,
and maybe add <code>-v -P</code> to give us verbose output and progress information, and copy all of our files:</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb77-1" data-line-number="1"><span class="ex">%</span> rclone copy --exclude */ESP_A1/* -v -P  --http-url http://sysg1.cs.yale.edu:3010/5lnO9bs3zfa8LOhESfsYfq3Dc/061719/ :http: Alewife/fastqs/</a></code></pre></div>
</div>
</div>
<div id="activatinginstalling-software" class="section level2">
<h2><span class="header-section-number">7.3</span> Activating/Installing software</h2>
<div id="modules" class="section level3">
<h3><span class="header-section-number">7.3.1</span> Modules</h3>
<p>This is if your sys admin has made it easy.</p>
</div>
<div id="miniconda" class="section level3">
<h3><span class="header-section-number">7.3.2</span> Miniconda</h3>
<p>This is how one will probably want to do it</p>
<div id="testing-this-on-summit" class="section level4">
<h4><span class="header-section-number">7.3.2.1</span> Testing this on Summit</h4>
<p>I just want to quickly try this:</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb78-1" data-line-number="1"><span class="fu">ssh</span> eriq@colostate.edu@login.rc.colorado.edu</a>
<a class="sourceLine" id="cb78-2" data-line-number="2"></a>
<a class="sourceLine" id="cb78-3" data-line-number="3"><span class="co"># get on compile nodes</span></a>
<a class="sourceLine" id="cb78-4" data-line-number="4"><span class="fu">ssh</span> scompile</a>
<a class="sourceLine" id="cb78-5" data-line-number="5"></a>
<a class="sourceLine" id="cb78-6" data-line-number="6"><span class="co"># I checked modules and found no samtools, bcftools, etc.</span></a>
<a class="sourceLine" id="cb78-7" data-line-number="7"></a>
<a class="sourceLine" id="cb78-8" data-line-number="8"><span class="co"># Now install miniconda</span></a>
<a class="sourceLine" id="cb78-9" data-line-number="9"><span class="fu">mkdir</span> conda_install</a>
<a class="sourceLine" id="cb78-10" data-line-number="10"><span class="bu">cd</span> conda_install/</a>
<a class="sourceLine" id="cb78-11" data-line-number="11"><span class="fu">wget</span> https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh</a>
<a class="sourceLine" id="cb78-12" data-line-number="12"><span class="fu">chmod</span> u+x Miniconda3-latest-Linux-x86_64.sh </a>
<a class="sourceLine" id="cb78-13" data-line-number="13"><span class="ex">./Miniconda3-latest-Linux-x86_64.sh</span> </a>
<a class="sourceLine" id="cb78-14" data-line-number="14"></a>
<a class="sourceLine" id="cb78-15" data-line-number="15"><span class="co"># then you follow the prompts and agree to the license and the </span></a>
<a class="sourceLine" id="cb78-16" data-line-number="16"><span class="co"># default install location.</span></a>
<a class="sourceLine" id="cb78-17" data-line-number="17"></a>
<a class="sourceLine" id="cb78-18" data-line-number="18"><span class="co">## </span><span class="al">NOTE</span><span class="co">: Might want to install to a different location if there</span></a>
<a class="sourceLine" id="cb78-19" data-line-number="19"><span class="co">## are serious caps on hard disk usage in home directory..</span></a>
<a class="sourceLine" id="cb78-20" data-line-number="20"></a>
<a class="sourceLine" id="cb78-21" data-line-number="21"><span class="bu">source</span> ~/.bashrc</a>
<a class="sourceLine" id="cb78-22" data-line-number="22"></a>
<a class="sourceLine" id="cb78-23" data-line-number="23"><span class="co"># after that I have it!</span></a>
<a class="sourceLine" id="cb78-24" data-line-number="24"><span class="kw">(</span><span class="ex">base</span><span class="kw">)</span> [<span class="ex">eriq@colostate.edu@shas0136</span> conda_install]$</a></code></pre></div>
<p>Now, let’s see if we can get samtools. Just google “samtools conda” to
get an idea of how to do it:</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb79-1" data-line-number="1"><span class="ex">conda</span> install -c bioconda samtools</a></code></pre></div>
<p>AFter that, it just works! Cool!</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb80-1" data-line-number="1"><span class="kw">(</span><span class="ex">base</span><span class="kw">)</span> [<span class="ex">eriq@colostate.edu@shas0136</span> ~]$ samtools</a>
<a class="sourceLine" id="cb80-2" data-line-number="2"></a>
<a class="sourceLine" id="cb80-3" data-line-number="3"><span class="ex">Program</span>: samtools (Tools for alignments in the SAM format)</a>
<a class="sourceLine" id="cb80-4" data-line-number="4"><span class="ex">Version</span>: 1.9 (using htslib 1.9)</a>
<a class="sourceLine" id="cb80-5" data-line-number="5"></a>
<a class="sourceLine" id="cb80-6" data-line-number="6"><span class="ex">Usage</span>:   samtools <span class="op">&lt;</span>command<span class="op">&gt;</span> [options]</a>
<a class="sourceLine" id="cb80-7" data-line-number="7"></a>
<a class="sourceLine" id="cb80-8" data-line-number="8"><span class="ex">Commands</span>:</a>
<a class="sourceLine" id="cb80-9" data-line-number="9">  <span class="ex">--</span> Indexing</a>
<a class="sourceLine" id="cb80-10" data-line-number="10">     <span class="ex">dict</span>           create a sequence dictionary file</a>
<a class="sourceLine" id="cb80-11" data-line-number="11">     <span class="ex">faidx</span>          index/extract FASTA</a>
<a class="sourceLine" id="cb80-12" data-line-number="12">     <span class="ex">fqidx</span>          index/extract FASTQ</a>
<a class="sourceLine" id="cb80-13" data-line-number="13">     <span class="ex">index</span>          index alignment</a>
<a class="sourceLine" id="cb80-14" data-line-number="14"></a>
<a class="sourceLine" id="cb80-15" data-line-number="15"><span class="ex">...</span></a></code></pre></div>
<p>All right! That is amazing. Next steps:</p>
<ol style="list-style-type: decimal">
<li>Tell students about establishing different environments.</li>
<li>Learn about how to make a minimal environment for a project and how to
record that and be able to distribute/propagate it.</li>
</ol>
<p>Along those lines, I want to see if, when I install samtools into a new environment,
it re-downloads it or not…</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb81-1" data-line-number="1"><span class="ex">conda</span> create --name quick-test</a>
<a class="sourceLine" id="cb81-2" data-line-number="2"><span class="ex">conda</span> activate quick-test</a>
<a class="sourceLine" id="cb81-3" data-line-number="3"></a>
<a class="sourceLine" id="cb81-4" data-line-number="4"><span class="co"># i also made an environment in a specified directory (you</span></a>
<a class="sourceLine" id="cb81-5" data-line-number="5"><span class="co"># could put these within a project directory)</span></a>
<a class="sourceLine" id="cb81-6" data-line-number="6"><span class="ex">conda</span> create --prefix ./test-envs-dir </a>
<a class="sourceLine" id="cb81-7" data-line-number="7"><span class="ex">conda</span> activate ./test-envs-dir</a>
<a class="sourceLine" id="cb81-8" data-line-number="8"></a>
<a class="sourceLine" id="cb81-9" data-line-number="9"><span class="co"># now, let&#39;s install bcftools there</span></a>
<a class="sourceLine" id="cb81-10" data-line-number="10"><span class="ex">conda</span> install -c bioconda bcftools</a>
<a class="sourceLine" id="cb81-11" data-line-number="11"></a>
<a class="sourceLine" id="cb81-12" data-line-number="12"><span class="co"># note that it doesn&#39;t re-download the dependencies, as far as I can tell.</span></a>
<a class="sourceLine" id="cb81-13" data-line-number="13"><span class="ex">conda</span> activate base</a>
<a class="sourceLine" id="cb81-14" data-line-number="14"><span class="ex">bcftools</span> <span class="co"># not found in environment base</span></a>
<a class="sourceLine" id="cb81-15" data-line-number="15"></a>
<a class="sourceLine" id="cb81-16" data-line-number="16"><span class="ex">conda</span> activate ./test-envs-dir</a>
<a class="sourceLine" id="cb81-17" data-line-number="17"><span class="ex">bcftools</span>  # it is found in this environment.  Cool.</a>
<a class="sourceLine" id="cb81-18" data-line-number="18"></a>
<a class="sourceLine" id="cb81-19" data-line-number="19"><span class="ex">conda</span> activate quick-test</a>
<a class="sourceLine" id="cb81-20" data-line-number="20"><span class="ex">bcftools</span>  # it ain<span class="st">&#39;t here</span></a></code></pre></div>
<p>Now, after that, bcftools is in ./test-envs-dir/bcftools</p>
<p>So, what if we install it into another environment. Does it symlink it?</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb82-1" data-line-number="1"><span class="ex">conda</span> activate base</a>
<a class="sourceLine" id="cb82-2" data-line-number="2"><span class="ex">conda</span> install -c bioconda bcftools</a>
<a class="sourceLine" id="cb82-3" data-line-number="3"><span class="ex">bcftools</span></a>
<a class="sourceLine" id="cb82-4" data-line-number="4"></a>
<a class="sourceLine" id="cb82-5" data-line-number="5"><span class="co"># whoa! That errored out!</span></a>
<a class="sourceLine" id="cb82-6" data-line-number="6"><span class="ex">bcftools</span>: error while loading shared libraries: libcrypto.so.1.0.0: cannot open shared object file: No such file or directory</a>
<a class="sourceLine" id="cb82-7" data-line-number="7"></a>
<a class="sourceLine" id="cb82-8" data-line-number="8"><span class="co"># that is a serious problem.  </span></a>
<a class="sourceLine" id="cb82-9" data-line-number="9"></a>
<a class="sourceLine" id="cb82-10" data-line-number="10"><span class="co"># Can I get it in my other environment?</span></a>
<a class="sourceLine" id="cb82-11" data-line-number="11"><span class="ex">conda</span> activate quick-test</a>
<a class="sourceLine" id="cb82-12" data-line-number="12"><span class="ex">conda</span> install -c bioconda bcftools</a>
<a class="sourceLine" id="cb82-13" data-line-number="13"><span class="ex">bcftools</span></a>
<a class="sourceLine" id="cb82-14" data-line-number="14"></a>
<a class="sourceLine" id="cb82-15" data-line-number="15"><span class="co"># that totally works, but it still doesn&#39;t in base...</span></a>
<a class="sourceLine" id="cb82-16" data-line-number="16"></a>
<a class="sourceLine" id="cb82-17" data-line-number="17"><span class="co"># so, what if we add samtools to our other environments?</span></a>
<a class="sourceLine" id="cb82-18" data-line-number="18"><span class="co"># that works fine.</span></a></code></pre></div>
<p>But, samtools/bcftools dependency issues are a known problem: <a href="https://github.com/sunbeam-labs/sunbeam/issues/181" class="uri">https://github.com/sunbeam-labs/sunbeam/issues/181</a>
Basically they rely on different versions of some ssh libs.</p>
<p>Note that installing bcftools first things work. But what if we make another environment
and install samtools first again?</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb83-1" data-line-number="1"><span class="ex">conda</span> create --name samtools-first</a>
<a class="sourceLine" id="cb83-2" data-line-number="2"><span class="ex">conda</span> activate samtools-first</a>
<a class="sourceLine" id="cb83-3" data-line-number="3"><span class="ex">conda</span> install -c bioconda samtools  # this didn<span class="st">&#39;t download anything new</span></a>
<a class="sourceLine" id="cb83-4" data-line-number="4"><span class="st">conda install -c bioconda bcftools</span></a>
<a class="sourceLine" id="cb83-5" data-line-number="5"></a>
<a class="sourceLine" id="cb83-6" data-line-number="6"><span class="st"># BOOM! THIS CREATES A FAIL.  SO, YOU GOTTA INSTALL BCFTOOLS</span></a>
<a class="sourceLine" id="cb83-7" data-line-number="7"><span class="st"># FIRST.  FAR OUT.</span></a></code></pre></div>
</div>
</div>
<div id="exporting-environments" class="section level3">
<h3><span class="header-section-number">7.3.3</span> Exporting environments</h3>
<p>Looks like we should be able to do this. Let’s do the one that works:</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb84-1" data-line-number="1"><span class="kw">(</span><span class="ex">quick-test</span><span class="kw">)</span> [<span class="ex">~</span>]<span class="ex">--%</span> conda env export <span class="op">&gt;</span> quick-test-env.yml</a>
<a class="sourceLine" id="cb84-2" data-line-number="2"></a>
<a class="sourceLine" id="cb84-3" data-line-number="3"><span class="kw">(</span><span class="ex">quick-test</span><span class="kw">)</span> [<span class="ex">~</span>]<span class="ex">--%</span> cat quick-test-env.yml </a>
<a class="sourceLine" id="cb84-4" data-line-number="4"><span class="ex">name</span>: quick-test</a>
<a class="sourceLine" id="cb84-5" data-line-number="5"><span class="ex">channels</span>:</a>
<a class="sourceLine" id="cb84-6" data-line-number="6">  <span class="ex">-</span> bioconda</a>
<a class="sourceLine" id="cb84-7" data-line-number="7">  <span class="ex">-</span> defaults</a>
<a class="sourceLine" id="cb84-8" data-line-number="8"><span class="ex">dependencies</span>:</a>
<a class="sourceLine" id="cb84-9" data-line-number="9">  <span class="ex">-</span> _libgcc_mutex=0.1=main</a>
<a class="sourceLine" id="cb84-10" data-line-number="10">  <span class="ex">-</span> bcftools=1.9=ha228f0b_4</a>
<a class="sourceLine" id="cb84-11" data-line-number="11">  <span class="ex">-</span> bzip2=1.0.8=h7b6447c_0</a>
<a class="sourceLine" id="cb84-12" data-line-number="12">  <span class="ex">-</span> ca-certificates=2019.5.15=1</a>
<a class="sourceLine" id="cb84-13" data-line-number="13">  <span class="ex">-</span> curl=7.65.3=hbc83047_0</a>
<a class="sourceLine" id="cb84-14" data-line-number="14">  <span class="ex">-</span> htslib=1.9=ha228f0b_7</a>
<a class="sourceLine" id="cb84-15" data-line-number="15">  <span class="ex">-</span> krb5=1.16.1=h173b8e3_7</a>
<a class="sourceLine" id="cb84-16" data-line-number="16">  <span class="ex">-</span> libcurl=7.65.3=h20c2e04_0</a>
<a class="sourceLine" id="cb84-17" data-line-number="17">  <span class="ex">-</span> libdeflate=1.0=h14c3975_1</a>
<a class="sourceLine" id="cb84-18" data-line-number="18">  <span class="ex">-</span> libedit=3.1.20181209=hc058e9b_0</a>
<a class="sourceLine" id="cb84-19" data-line-number="19">  <span class="ex">-</span> libgcc-ng=9.1.0=hdf63c60_0</a>
<a class="sourceLine" id="cb84-20" data-line-number="20">  <span class="ex">-</span> libssh2=1.8.2=h1ba5d50_0</a>
<a class="sourceLine" id="cb84-21" data-line-number="21">  <span class="ex">-</span> libstdcxx-ng=9.1.0=hdf63c60_0</a>
<a class="sourceLine" id="cb84-22" data-line-number="22">  <span class="ex">-</span> ncurses=6.1=he6710b0_1</a>
<a class="sourceLine" id="cb84-23" data-line-number="23">  <span class="ex">-</span> openssl=1.1.1c=h7b6447c_1</a>
<a class="sourceLine" id="cb84-24" data-line-number="24">  <span class="ex">-</span> samtools=1.9=h10a08f8_12</a>
<a class="sourceLine" id="cb84-25" data-line-number="25">  <span class="ex">-</span> tk=8.6.8=hbc83047_0</a>
<a class="sourceLine" id="cb84-26" data-line-number="26">  <span class="ex">-</span> xz=5.2.4=h14c3975_4</a>
<a class="sourceLine" id="cb84-27" data-line-number="27">  <span class="ex">-</span> zlib=1.2.11=h7b6447c_3</a>
<a class="sourceLine" id="cb84-28" data-line-number="28"><span class="ex">prefix</span>: /home/eriq@colostate.edu/miniconda3/envs/quick-test</a>
<a class="sourceLine" id="cb84-29" data-line-number="29"></a>
<a class="sourceLine" id="cb84-30" data-line-number="30"></a>
<a class="sourceLine" id="cb84-31" data-line-number="31"><span class="co"># OK, that is cool.  Now, if we wanted to email that to someone,</span></a>
<a class="sourceLine" id="cb84-32" data-line-number="32"><span class="co"># we could, and then they could do this:</span></a>
<a class="sourceLine" id="cb84-33" data-line-number="33"><span class="ex">conda</span> env create --name dupie-quick -f quick-test-env.yml </a>
<a class="sourceLine" id="cb84-34" data-line-number="34"></a>
<a class="sourceLine" id="cb84-35" data-line-number="35"><span class="co"># that environment then has samtools and bcftools</span></a>
<a class="sourceLine" id="cb84-36" data-line-number="36"></a>
<a class="sourceLine" id="cb84-37" data-line-number="37"><span class="co"># note that it probably would try to name it quick-test if we didn&#39;t</span></a>
<a class="sourceLine" id="cb84-38" data-line-number="38"><span class="co"># pass in the name there...</span></a></code></pre></div>
</div>
</div>
<div id="boneyard" class="section level2">
<h2><span class="header-section-number">7.4</span> Boneyard</h2>
<p>This is a variant of rsync that let’s you sync stuff up to google drive. It might be a better
solution than rcp for getting stuff onto and off of google drive. Here is a link:
<a href="https://rclone.org/">https://rclone.org/</a>. I need to evaluate it. It might also be a good way
to backup some of my workstuff on my laptop to Google Drive (and maybe also for other people to create replicas and have a decent backup if they have unlimited Google Drive storage).</p>
<p>I got this working. It is important to set your own OAuth client ID:
<a href="https://forum.rclone.org/t/very-slow-sync-to-google-drive/6903">https://forum.rclone.org/t/very-slow-sync-to-google-drive/6903</a></p>
<p>After that I did like this:</p>
<pre><code>rclone sync -vv --tpslimit 10 --fast-list Otsh_v1.0_genomic.fna  gdrive-rclone:spoogee-spoogee</code></pre>
<p>which did 2 Gb of fasta into the spoogee-spoogee directory pretty quickly.</p>
<p>But, with something that has lots of files, it took longer:</p>
<pre><code># this is only about 100 Mb but took a long time
rclone copy -P --tpslimit 10 --fast-list  rubias  gdrive-rclone:rubias</code></pre>
<p>However, once that is done, you can sync it and it finds that parts that have changed pretty quickly.</p>
<p>it appears to do that by file modification times:</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb87-1" data-line-number="1"><span class="ex">2019-04-19</span> 23:21 /Otsh_v1.0/--% rclone sync -vv --tpslimit 10 --fast-list Otsh_v1.0_genomic.fna  gdrive-rclone:spoogee-spoogee</a>
<a class="sourceLine" id="cb87-2" data-line-number="2"><span class="ex">2019/04/19</span> 23:21:36 DEBUG : rclone: Version <span class="st">&quot;v1.47.0&quot;</span> starting with parameters [<span class="st">&quot;rclone&quot;</span> <span class="st">&quot;sync&quot;</span> <span class="st">&quot;-vv&quot;</span> <span class="st">&quot;--tpslimit&quot;</span> <span class="st">&quot;10&quot;</span> <span class="st">&quot;--fast-list&quot;</span> <span class="st">&quot;Otsh_v1.0_genomic.fna&quot;</span> <span class="st">&quot;gdrive-rclone:spoogee-spoogee&quot;</span>]</a>
<a class="sourceLine" id="cb87-3" data-line-number="3"><span class="ex">2019/04/19</span> 23:21:36 DEBUG : Using config file from <span class="st">&quot;/Users/eriq/.config/rclone/rclone.conf&quot;</span></a>
<a class="sourceLine" id="cb87-4" data-line-number="4"><span class="ex">2019/04/19</span> 23:21:36 INFO  : Starting HTTP transaction limiter: max 10 transactions/s with burst 1</a>
<a class="sourceLine" id="cb87-5" data-line-number="5"><span class="ex">2019/04/19</span> 23:21:37 DEBUG : GCF_002872995.1_Otsh_v1.0_genomic.gff.gz: Excluded</a>
<a class="sourceLine" id="cb87-6" data-line-number="6"><span class="ex">2019/04/19</span> 23:21:37 DEBUG : Otsh_v1.0_genomic.dict: Excluded</a>
<a class="sourceLine" id="cb87-7" data-line-number="7"><span class="ex">2019/04/19</span> 23:21:37 DEBUG : Otsh_v1.0_genomic.fna.amb: Excluded</a>
<a class="sourceLine" id="cb87-8" data-line-number="8"><span class="ex">2019/04/19</span> 23:21:37 DEBUG : Otsh_v1.0_genomic.fna.ann: Excluded</a>
<a class="sourceLine" id="cb87-9" data-line-number="9"><span class="ex">2019/04/19</span> 23:21:37 DEBUG : Otsh_v1.0_genomic.fna.bwt: Excluded</a>
<a class="sourceLine" id="cb87-10" data-line-number="10"><span class="ex">2019/04/19</span> 23:21:37 DEBUG : Otsh_v1.0_genomic.fna.fai: Excluded</a>
<a class="sourceLine" id="cb87-11" data-line-number="11"><span class="ex">2019/04/19</span> 23:21:37 DEBUG : Otsh_v1.0_genomic.fna.pac: Excluded</a>
<a class="sourceLine" id="cb87-12" data-line-number="12"><span class="ex">2019/04/19</span> 23:21:37 DEBUG : Otsh_v1.0_genomic.fna.sa: Excluded</a>
<a class="sourceLine" id="cb87-13" data-line-number="13"><span class="ex">2019/04/19</span> 23:21:37 INFO  : Google drive root <span class="st">&#39;spoogee-spoogee&#39;</span>: Waiting for checks to finish</a>
<a class="sourceLine" id="cb87-14" data-line-number="14"><span class="ex">2019/04/19</span> 23:21:37 DEBUG : Otsh_v1.0_genomic.fna: Size and modification time the same (differ by 0s, within tolerance 1s)</a>
<a class="sourceLine" id="cb87-15" data-line-number="15"><span class="ex">2019/04/19</span> 23:21:37 DEBUG : Otsh_v1.0_genomic.fna: Unchanged skipping</a>
<a class="sourceLine" id="cb87-16" data-line-number="16"><span class="ex">2019/04/19</span> 23:21:37 INFO  : Google drive root <span class="st">&#39;spoogee-spoogee&#39;</span>: Waiting for transfers to finish</a>
<a class="sourceLine" id="cb87-17" data-line-number="17"><span class="ex">2019/04/19</span> 23:21:37 INFO  : Waiting for deletions to finish</a>
<a class="sourceLine" id="cb87-18" data-line-number="18"><span class="ex">2019/04/19</span> 23:21:37 INFO  : </a>
<a class="sourceLine" id="cb87-19" data-line-number="19"><span class="ex">Transferred</span>:             0 / 0 Bytes, -, 0 Bytes/s, ETA -</a>
<a class="sourceLine" id="cb87-20" data-line-number="20"><span class="ex">Errors</span>:                 0</a>
<a class="sourceLine" id="cb87-21" data-line-number="21"><span class="ex">Checks</span>:                 1 / 1, 100%</a>
<a class="sourceLine" id="cb87-22" data-line-number="22"><span class="ex">Transferred</span>:            0 / 0, -</a>
<a class="sourceLine" id="cb87-23" data-line-number="23"><span class="ex">Elapsed</span> time:        1.3s</a>
<a class="sourceLine" id="cb87-24" data-line-number="24"></a>
<a class="sourceLine" id="cb87-25" data-line-number="25"><span class="ex">2019/04/19</span> 23:21:37 DEBUG : 5 go routines active</a>
<a class="sourceLine" id="cb87-26" data-line-number="26"><span class="ex">2019/04/19</span> 23:21:37 DEBUG : rclone: Version <span class="st">&quot;v1.47.0&quot;</span> finishing with parameters [<span class="st">&quot;rclone&quot;</span> <span class="st">&quot;sync&quot;</span> <span class="st">&quot;-vv&quot;</span> <span class="st">&quot;--tpslimit&quot;</span> <span class="st">&quot;10&quot;</span> <span class="st">&quot;--fast-list&quot;</span> <span class="st">&quot;Otsh_v1.0_genomic.fna&quot;</span> <span class="st">&quot;gdrive-rclone:spoogee-spoogee&quot;</span>]</a>
<a class="sourceLine" id="cb87-27" data-line-number="27"><span class="ex">2</span></a></code></pre></div>
<p>So, for moving big files around that might be a good way forward. I will have to do a test with some big files.</p>
<p>And I need to test it with team drives so that multiple individuals can pull stuff off of the Bird Genoscape drive for example.</p>
<p>It would be nice to have safeguards so people don’t trash stuff accidentally….</p>
<div id="rclone-on-hoffman" class="section level4">
<h4><span class="header-section-number">7.4.0.1</span> rclone on Hoffman</h4>
<p>Their default install script expects sudo access to put it in /usr/local
but I don’t on hoffman, obviously, so I just downloaded the the install script and edited
the section for Linux to look like this at the relevant part</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb88-1" data-line-number="1"><span class="kw">case</span> <span class="va">$OS</span><span class="kw"> in</span></a>
<a class="sourceLine" id="cb88-2" data-line-number="2">  <span class="st">&#39;linux&#39;</span><span class="kw">)</span></a>
<a class="sourceLine" id="cb88-3" data-line-number="3">    <span class="co">#binary</span></a>
<a class="sourceLine" id="cb88-4" data-line-number="4">    <span class="fu">cp</span> rclone ~/bin/rclone.new</a>
<a class="sourceLine" id="cb88-5" data-line-number="5">    <span class="fu">chmod</span> 755 ~/bin/rclone.new</a>
<a class="sourceLine" id="cb88-6" data-line-number="6">    <span class="co">#chown root:root /usr/bin/rclone.new</span></a>
<a class="sourceLine" id="cb88-7" data-line-number="7">    <span class="fu">mv</span> ~/bin/rclone.new ~/bin/rclone</a>
<a class="sourceLine" id="cb88-8" data-line-number="8">    <span class="co">#manuals</span></a>
<a class="sourceLine" id="cb88-9" data-line-number="9">    <span class="co">#mkdir -p /usr/local/share/man/man1</span></a>
<a class="sourceLine" id="cb88-10" data-line-number="10">    <span class="co">#cp rclone.1 /usr/local/share/man/man1/</span></a>
<a class="sourceLine" id="cb88-11" data-line-number="11">    <span class="co">#mandb</span></a>
<a class="sourceLine" id="cb88-12" data-line-number="12">    <span class="kw">;;</span></a></code></pre></div>
<p>I don’t get man pages, but I get it in ~/bin no problem.</p>
<p>To set up the configuration, check where it belongs:</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb89-1" data-line-number="1"><span class="ex">%</span> rclone config file</a>
<a class="sourceLine" id="cb89-2" data-line-number="2"><span class="ex">Configuration</span> file doesn<span class="st">&#39;t exist, but rclone will use this path:</span></a>
<a class="sourceLine" id="cb89-3" data-line-number="3"><span class="st">/u/home/e/eriq/.config/rclone/rclone.conf</span></a></code></pre></div>
<p>And then I just put my config file from my laptop on there. I just pasted the stuff
in whilst emacsing it. Holy cow! That is super easy.</p>
<p>Note that the config file is where you can also set default options like tpslimit and fast-list I think.</p>
<p>So, the OAuth stuff is all stored in that config file. And if you can set it up on one machine you can
go put it on any others that you want. That is awesome.</p>
<p>When it was done, I tested it:</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb90-1" data-line-number="1"><span class="ex">%</span> rclone sync -vv  --drive-shared-with-me  gdrive-rclone:BaselinePaper  BaselinePaper_here</a>
<a class="sourceLine" id="cb90-2" data-line-number="2"><span class="ex">2019/04/29</span> 14:49:24 DEBUG : rclone: Version <span class="st">&quot;v1.47.0&quot;</span> starting with parameters [<span class="st">&quot;rclone&quot;</span> <span class="st">&quot;sync&quot;</span> <span class="st">&quot;-vv&quot;</span> <span class="st">&quot;--drive-shared-with-me&quot;</span> <span class="st">&quot;gdrive-rclone:BaselinePaper&quot;</span> <span class="st">&quot;BaselinePaper_here&quot;</span>]</a>
<a class="sourceLine" id="cb90-3" data-line-number="3"><span class="ex">2019/04/29</span> 14:49:24 DEBUG : Using config file from <span class="st">&quot;/u/home/e/eriq/.config/rclone/rclone.conf&quot;</span></a>
<a class="sourceLine" id="cb90-4" data-line-number="4"><span class="ex">2019/04/29</span> 14:49:25 INFO  : Local file system at /u/home/e/eriq/BaselinePaper_here: Waiting for checks to finish</a>
<a class="sourceLine" id="cb90-5" data-line-number="5"><span class="ex">2019/04/29</span> 14:49:25 INFO  : Local file system at /u/home/e/eriq/BaselinePaper_here: Waiting for transfers to finish</a>
<a class="sourceLine" id="cb90-6" data-line-number="6"><span class="ex">2019/04/29</span> 14:49:26 DEBUG : Local file system at /u/home/e/eriq/BaselinePaper_here: File to upload is small (41922 bytes), <span class="ex">uploading</span> instead of streaming</a>
<a class="sourceLine" id="cb90-7" data-line-number="7"><span class="ex">2019/04/29</span> 14:49:26 DEBUG : BaselinePaper_Body.docx: Failed to pre-allocate: operation not supported</a>
<a class="sourceLine" id="cb90-8" data-line-number="8"><span class="ex">2019/04/29</span> 14:49:26 INFO  : BaselinePaper_Body.docx: Copied (new)</a>
<a class="sourceLine" id="cb90-9" data-line-number="9"><span class="ex">2019/04/29</span> 14:49:26 DEBUG : BaselinePaper_Body.docx: Updating size of doc after download to 41922</a>
<a class="sourceLine" id="cb90-10" data-line-number="10"><span class="ex">2019/04/29</span> 14:49:26 INFO  : BaselinePaper_Body.docx: Copied (Rcat, new)</a>
<a class="sourceLine" id="cb90-11" data-line-number="11"><span class="ex">2019/04/29</span> 14:49:27 DEBUG : Local file system at /u/home/e/eriq/BaselinePaper_here: File to upload is small (57172 bytes), <span class="ex">uploading</span> instead of streaming</a>
<a class="sourceLine" id="cb90-12" data-line-number="12"><span class="ex">2019/04/29</span> 14:49:27 DEBUG : ResponseToReviewers_eca.docx: Failed to pre-allocate: operation not supported</a>
<a class="sourceLine" id="cb90-13" data-line-number="13"><span class="ex">2019/04/29</span> 14:49:27 INFO  : ResponseToReviewers_eca.docx: Copied (new)</a>
<a class="sourceLine" id="cb90-14" data-line-number="14"><span class="ex">2019/04/29</span> 14:49:27 DEBUG : ResponseToReviewers_eca.docx: Updating size of doc after download to 57172</a>
<a class="sourceLine" id="cb90-15" data-line-number="15"><span class="ex">2019/04/29</span> 14:49:27 INFO  : ResponseToReviewers_eca.docx: Copied (Rcat, new)</a>
<a class="sourceLine" id="cb90-16" data-line-number="16"><span class="ex">2019/04/29</span> 14:49:27 INFO  : Waiting for deletions to finish</a>
<a class="sourceLine" id="cb90-17" data-line-number="17"><span class="ex">2019/04/29</span> 14:49:27 INFO  : </a>
<a class="sourceLine" id="cb90-18" data-line-number="18"><span class="ex">Transferred</span>:      193.543k / 193.543 kBytes, 100%, 79.377 kBytes/s, ETA 0s</a>
<a class="sourceLine" id="cb90-19" data-line-number="19"><span class="ex">Errors</span>:                 0</a>
<a class="sourceLine" id="cb90-20" data-line-number="20"><span class="ex">Checks</span>:                 0 / 0, -</a>
<a class="sourceLine" id="cb90-21" data-line-number="21"><span class="ex">Transferred</span>:            4 / 4, 100%</a>
<a class="sourceLine" id="cb90-22" data-line-number="22"><span class="ex">Elapsed</span> time:        2.4s</a>
<a class="sourceLine" id="cb90-23" data-line-number="23"></a>
<a class="sourceLine" id="cb90-24" data-line-number="24"><span class="ex">2019/04/29</span> 14:49:27 DEBUG : 6 go routines active</a>
<a class="sourceLine" id="cb90-25" data-line-number="25"><span class="ex">2019/04/29</span> 14:49:27 DEBUG : rclone: Version <span class="st">&quot;v1.47.0&quot;</span> finishing with parameters [<span class="st">&quot;rclone&quot;</span> <span class="st">&quot;sync&quot;</span> <span class="st">&quot;-vv&quot;</span> <span class="st">&quot;--drive-shared-with-me&quot;</span> <span class="st">&quot;gdrive-rclone:BaselinePaper&quot;</span> <span class="st">&quot;BaselinePaper_here&quot;</span>]</a></code></pre></div>
<p>That was fast and super solid.</p>
</div>
<div id="encrypt-the-config-file" class="section level4">
<h4><span class="header-section-number">7.4.0.2</span> Encrypt the config file</h4>
<p>You can use <code>rclone config edit</code> to set a password for the config file. Then it
encrypts that so no one is able to run wild if they just get that file. You have to
provide your password to do any of the rclone commands. If you want to see the
config file use <code>rclone config show</code>. You could always copy that elsewhere, and then
re-encrypt it.</p>
<p>Here is some nice stuff for summarizing all the information from the different runs from the chinook-wgs project:</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb91-1" data-line-number="1"><span class="ex">qacct</span> -o eriq -b 09271925 -j ml <span class="kw">|</span> <span class="ex">tidy-qacct</span></a></code></pre></div>
<p>Explain scratch space and how clusters are configured with respect to storage, etc.</p>
<p>Strategies—break names up with consistent characters:</p>
<ul>
<li>dashes within population names</li>
<li>underscores for different groups of chromosomes</li>
<li>periods for catenating pairs of pops</li>
</ul>
<p>etc. Basically, it just makes it much easier to split things up
when the time comes.</p>
</div>
</div>
<div id="the-queue-slurmsgeuge" class="section level2">
<h2><span class="header-section-number">7.5</span> The Queue (SLURM/SGE/UGE)</h2>
</div>
<div id="modules-package" class="section level2">
<h2><span class="header-section-number">7.6</span> Modules package</h2>
</div>
<div id="compiling-programs-without-admin-privileges" class="section level2">
<h2><span class="header-section-number">7.7</span> Compiling programs without admin privileges</h2>
<p>Inevitably you will want to use a piece of software that is not available as
a module or is not otherwise installed on they system.</p>
<p>Typically these software programs have a frightful web of dependencies.</p>
<p>Unix/Linux distros typically maintain all these dependencies as libraries or packages
that can be installed using a <code>rpm</code> or <code>yum</code>. However, the simple “plug-and-play” approach
to using these programs requires have administrator privileges so that the software can
be installed in one of the (typically protected) paths in the root (like <code>/usr/bin</code>).</p>
<p>But, you can use these programs to install packages into your home directory. Once you have done
that, you need to let your system know where to look for these packages when it needs them
(i.e., when running a program or <em>linking</em> to it whilst compiling up a program that uses it
as a dependency.</p>
<p>Hoffman2 runs CentOS. Turns out that CentOS uses <code>yum</code> as a package manager.</p>
<p>Let’s see if we can install llvm using yum.</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb92-1" data-line-number="1"><span class="ex">yum</span> search all llvm <span class="co"># &lt;- this got me to devtoolset-7-all.x86_64 : Package shipping all available toolsets.</span></a>
<a class="sourceLine" id="cb92-2" data-line-number="2"></a>
<a class="sourceLine" id="cb92-3" data-line-number="3"><span class="co"># a little web searching made it look like llvm-toolset-7-5.0.1-4.el7.x86_64.rpm or devtoolset-7-llvm-7.0-5.el7.x86_64.rpm</span></a>
<a class="sourceLine" id="cb92-4" data-line-number="4"><span class="co"># might be what we want.  The first is a dependency of the second...</span></a>
<a class="sourceLine" id="cb92-5" data-line-number="5"><span class="fu">mkdir</span> ~/centos</a></code></pre></div>
<p>Was using instructions at <a href="https://stackoverflow.com/questions/36651091/how-to-install-packages-in-linux-centos-without-root-user-with-automatic-depen">https://stackoverflow.com/questions/36651091/how-to-install-packages-in-linux-centos-without-root-user-with-automatic-depen</a></p>
<p>Couldn’t get yum downloader to download any packages. The whole thing looked like it was going to
be a mess, so I thought I would try with miniconda.</p>
<p>I installed miniconda (python 2.7 version) into <code>/u/nobackup/kruegg/eriq/programs/miniconda/</code> and then did this:</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb93-1" data-line-number="1"><span class="co"># probably could have listed them all at once, but wanted to watch them go </span></a>
<a class="sourceLine" id="cb93-2" data-line-number="2"><span class="co"># one at a time...</span></a>
<a class="sourceLine" id="cb93-3" data-line-number="3"><span class="ex">conda</span> install numpy</a>
<a class="sourceLine" id="cb93-4" data-line-number="4"><span class="ex">conda</span> install scipy</a>
<a class="sourceLine" id="cb93-5" data-line-number="5"><span class="ex">conda</span> install pandas</a>
<a class="sourceLine" id="cb93-6" data-line-number="6"><span class="ex">conda</span> install numba</a>
<a class="sourceLine" id="cb93-7" data-line-number="7"></a>
<a class="sourceLine" id="cb93-8" data-line-number="8"><span class="co"># those all ran great.</span></a>
<a class="sourceLine" id="cb93-9" data-line-number="9"></a>
<a class="sourceLine" id="cb93-10" data-line-number="10"><span class="ex">conda</span> install pysnptools</a>
<a class="sourceLine" id="cb93-11" data-line-number="11"></a>
<a class="sourceLine" id="cb93-12" data-line-number="12"><span class="co"># that one didn&#39;t find a match, but I found on the web that I should try:</span></a>
<a class="sourceLine" id="cb93-13" data-line-number="13"><span class="ex">conda</span> install -c bioconda pysnptools </a>
<a class="sourceLine" id="cb93-14" data-line-number="14"></a>
<a class="sourceLine" id="cb93-15" data-line-number="15"><span class="co"># that worked!</span></a></code></pre></div>
<p>Also we want to touch briefly on LD_PATH (linking failures—and note that libraries are often
named libxxx.a) and CPATH (for failure to find xxxx.h), etc.</p>
</div>
<div id="job-arrays" class="section level2">
<h2><span class="header-section-number">7.8</span> Job arrays</h2>
<p>Definitely mention the <code>eval</code> keyword in bash for when you want to print
command lines with redirects.</p>
<p>Show the routine for it, and develop a good approach to efficiently
orchestrating redos. If you know the taskIDs of the ones that failed
then it is pretty easy to write an awk script that picks out the
commands and puts them in a new file. Actually, it is probably
better to just cycle over the numbers and use the -t option
to launch each. Then there is now changing the job-ids file.</p>
<p>In fact, I am starting to think that the -t option is better than
putting it into the file.</p>
<p>Question: if you give something on the command line, does that override
the directive in the header of the file? If so, then you don’t even
need to change the file. Note that using the qsub command line options
instead of the directives really opens up a lot of possibilities for
writing useful scripts that are flexible.</p>
<p>Also use short names for the jobs and have a system for naming the
redos (append numbers so you know which round it is, too)
possibly base the name on the ways things failed the first time. Like,
<code>fsttf1</code> = “Fst run for things that failed due to time limits, 1”. Or
structure things so that redos can just be done by invoking it with -t
and the jobid.</p>
</div>
<div id="writing-stdout-and-stderr-to-files" class="section level2">
<h2><span class="header-section-number">7.9</span> Writing stdout and stderr to files</h2>
<p>This is always good to do. Note that <code>stdbuf</code> is super useful here so that
things don’t get buffered super long. (PCAngsd doesn’t seem to write antyhing till
the end…)</p>
</div>
<div id="breaking-stuff-down" class="section level2">
<h2><span class="header-section-number">7.10</span> Breaking stuff down</h2>
<p>It is probably worth talking about how problems can be broken down into
smaller ones. Maybe give an example, and then say that we will be talking about
this for every step of the way in bioinformatic pipelines.</p>
<p>One thing to note—sometimes processes go awry for one reason or another.
When things are in smaller chunks it is not such a huge investment to
re-run it. (Unlike stuff that runs for two weeks before you realize that
it ain’t working right).</p>

</div>
</div>



            </section>

          </div>
        </div>
      </div>
<a href="sed-awk-and-regular-expressions.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="introduction-to-reproducible-research.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook/js/app.min.js"></script>
<script src="libs/gitbook/js/lunr.js"></script>
<script src="libs/gitbook/js/clipboard.min.js"></script>
<script src="libs/gitbook/js/plugin-search.js"></script>
<script src="libs/gitbook/js/plugin-sharing.js"></script>
<script src="libs/gitbook/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook/js/plugin-bookdown.js"></script>
<script src="libs/gitbook/js/jquery.highlight.js"></script>
<script src="libs/gitbook/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/eriqande/eca-bioinf-handbook/edit/master/1.03-hpc.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/eriqande/eca-bioinf-handbook/commits/master/1.03-hpc.Rmd",
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["eca-bioinf-handbook.pdf", "eca-bioinf-handbook.epub"],
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
