<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 8 High Performance Computing Clusters (HPCC’s) | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics</title>
  <meta name="description" content="A book example for a Chapman &amp; Hall book." />
  <meta name="generator" content="bookdown 0.16 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 8 High Performance Computing Clusters (HPCC’s) | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A book example for a Chapman &amp; Hall book." />
  <meta name="github-repo" content="yihui/bookdown-crc" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 8 High Performance Computing Clusters (HPCC’s) | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics" />
  
  <meta name="twitter:description" content="A book example for a Chapman &amp; Hall book." />
  

<meta name="author" content="Eric C. Anderson" />


<meta name="date" content="2020-03-05" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="working-on-remote-servers.html"/>
<link rel="next" href="introduction-to-reproducible-research.html"/>
<script src="libs/jquery/jquery.min.js"></script>
<link href="libs/gitbook/css/style.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ECA's Bioinformatics Handbook</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="erics-notes-of-what-he-might-do.html"><a href="erics-notes-of-what-he-might-do.html"><i class="fa fa-check"></i><b>2</b> Eric’s Notes of what he might do</a><ul>
<li class="chapter" data-level="2.1" data-path="erics-notes-of-what-he-might-do.html"><a href="erics-notes-of-what-he-might-do.html#table-of-topics"><i class="fa fa-check"></i><b>2.1</b> Table of topics</a></li>
</ul></li>
<li class="part"><span><b>I Part I: Essential Computing Skills</b></span></li>
<li class="chapter" data-level="3" data-path="overview-of-essential-computing-skills.html"><a href="overview-of-essential-computing-skills.html"><i class="fa fa-check"></i><b>3</b> Overview of Essential Computing Skills</a></li>
<li class="chapter" data-level="4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html"><i class="fa fa-check"></i><b>4</b> Essential Unix/Linux Terminal Knowledge</a><ul>
<li class="chapter" data-level="4.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#getting-a-bash-shell-on-your-system"><i class="fa fa-check"></i><b>4.1</b> Getting a bash shell on your system</a></li>
<li class="chapter" data-level="4.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#navigating-the-unix-filesystem"><i class="fa fa-check"></i><b>4.2</b> Navigating the Unix filesystem</a><ul>
<li class="chapter" data-level="4.2.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#changing-the-working-directory-with-cd"><i class="fa fa-check"></i><b>4.2.1</b> Changing the working directory with <code>cd</code></a></li>
<li class="chapter" data-level="4.2.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#updating-your-command-prompt"><i class="fa fa-check"></i><b>4.2.2</b> Updating your command prompt</a></li>
<li class="chapter" data-level="4.2.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#tab-completion-for-paths"><i class="fa fa-check"></i><b>4.2.3</b> TAB-completion for paths</a></li>
<li class="chapter" data-level="4.2.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#listing-the-contents-of-a-directory-with-ls"><i class="fa fa-check"></i><b>4.2.4</b> Listing the contents of a directory with <code>ls</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#globbing"><i class="fa fa-check"></i><b>4.2.5</b> Globbing</a></li>
<li class="chapter" data-level="4.2.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#what-makes-a-good-file-name"><i class="fa fa-check"></i><b>4.2.6</b> What makes a good file-name?</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#the-anatomy-of-a-unix-command"><i class="fa fa-check"></i><b>4.3</b> The anatomy of a Unix command</a><ul>
<li class="chapter" data-level="4.3.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#anatomy-command"><i class="fa fa-check"></i><b>4.3.1</b> The <code>command</code></a></li>
<li class="chapter" data-level="4.3.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#the-options"><i class="fa fa-check"></i><b>4.3.2</b> The <em>options</em></a></li>
<li class="chapter" data-level="4.3.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#arguments"><i class="fa fa-check"></i><b>4.3.3</b> Arguments</a></li>
<li class="chapter" data-level="4.3.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#getting-information-about-unix-commands"><i class="fa fa-check"></i><b>4.3.4</b> Getting information about Unix commands</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#handling-manipulating-and-viewing-files-and-streams"><i class="fa fa-check"></i><b>4.4</b> Handling, Manipulating, and Viewing files and streams</a><ul>
<li class="chapter" data-level="4.4.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#creating-new-directories"><i class="fa fa-check"></i><b>4.4.1</b> Creating new directories</a></li>
<li class="chapter" data-level="4.4.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#fundamental-file-handling-commands"><i class="fa fa-check"></i><b>4.4.2</b> Fundamental file-handling commands</a></li>
<li class="chapter" data-level="4.4.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#viewing-files"><i class="fa fa-check"></i><b>4.4.3</b> “Viewing” Files</a></li>
<li class="chapter" data-level="4.4.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#redirecting-standard-output-and"><i class="fa fa-check"></i><b>4.4.4</b> Redirecting standard output: <code>&gt;</code> and <code>&gt;&gt;</code></a></li>
<li class="chapter" data-level="4.4.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#stdin-and"><i class="fa fa-check"></i><b>4.4.5</b> stdin, <code>&lt;</code> and <code>|</code></a></li>
<li class="chapter" data-level="4.4.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#stderr"><i class="fa fa-check"></i><b>4.4.6</b> stderr</a></li>
<li class="chapter" data-level="4.4.7" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#symbolic-links"><i class="fa fa-check"></i><b>4.4.7</b> Symbolic links</a></li>
<li class="chapter" data-level="4.4.8" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#file-permissions"><i class="fa fa-check"></i><b>4.4.8</b> File Permissions</a></li>
<li class="chapter" data-level="4.4.9" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#editing-text-files-at-the-terminal"><i class="fa fa-check"></i><b>4.4.9</b> Editing text files at the terminal</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#unix-env"><i class="fa fa-check"></i><b>4.5</b> Customizing your Environment</a><ul>
<li class="chapter" data-level="4.5.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#appearances-matter"><i class="fa fa-check"></i><b>4.5.1</b> Appearances matter</a></li>
<li class="chapter" data-level="4.5.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#where-are-my-programscommands-at"><i class="fa fa-check"></i><b>4.5.2</b> Where are my programs/commands at?!</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#a-few-more-important-keystrokes"><i class="fa fa-check"></i><b>4.6</b> A Few More Important Keystrokes</a></li>
<li class="chapter" data-level="4.7" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#a-short-list-of-additional-useful-commands."><i class="fa fa-check"></i><b>4.7</b> A short list of additional useful commands.</a></li>
<li class="chapter" data-level="4.8" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#two-important-computing-concepts"><i class="fa fa-check"></i><b>4.8</b> Two important computing concepts</a><ul>
<li class="chapter" data-level="4.8.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#compression"><i class="fa fa-check"></i><b>4.8.1</b> Compression</a></li>
<li class="chapter" data-level="4.8.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#hashing"><i class="fa fa-check"></i><b>4.8.2</b> Hashing</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#unix-quick-study-guide"><i class="fa fa-check"></i><b>4.9</b> Unix: Quick Study Guide</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="shell-programming.html"><a href="shell-programming.html"><i class="fa fa-check"></i><b>5</b> Shell programming</a><ul>
<li class="chapter" data-level="5.1" data-path="shell-programming.html"><a href="shell-programming.html#an-example-script"><i class="fa fa-check"></i><b>5.1</b> An example script</a></li>
<li class="chapter" data-level="5.2" data-path="shell-programming.html"><a href="shell-programming.html#the-structure-of-a-bash-script"><i class="fa fa-check"></i><b>5.2</b> The Structure of a Bash Script</a><ul>
<li class="chapter" data-level="5.2.1" data-path="shell-programming.html"><a href="shell-programming.html#a-bit-more-on-and"><i class="fa fa-check"></i><b>5.2.1</b> A bit more on <code>;</code> and <code>&amp;</code></a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="shell-programming.html"><a href="shell-programming.html#variables"><i class="fa fa-check"></i><b>5.3</b> Variables</a><ul>
<li class="chapter" data-level="5.3.1" data-path="shell-programming.html"><a href="shell-programming.html#assigning-values-to-variables"><i class="fa fa-check"></i><b>5.3.1</b> Assigning values to variables</a></li>
<li class="chapter" data-level="5.3.2" data-path="shell-programming.html"><a href="shell-programming.html#accessing-values-from-variables"><i class="fa fa-check"></i><b>5.3.2</b> Accessing values from variables</a></li>
<li class="chapter" data-level="5.3.3" data-path="shell-programming.html"><a href="shell-programming.html#what-does-the-shell-do-with-the-value-substituted-for-a-variable"><i class="fa fa-check"></i><b>5.3.3</b> What does the shell do with the value substituted for a variable?</a></li>
<li class="chapter" data-level="5.3.4" data-path="shell-programming.html"><a href="shell-programming.html#double-and-single-quotation-marks-and-variable-substitution"><i class="fa fa-check"></i><b>5.3.4</b> Double and Single Quotation Marks and Variable Substitution</a></li>
<li class="chapter" data-level="5.3.5" data-path="shell-programming.html"><a href="shell-programming.html#one-useful-fancy-variable-substitution-method"><i class="fa fa-check"></i><b>5.3.5</b> One useful, fancy, variable-substitution method</a></li>
<li class="chapter" data-level="5.3.6" data-path="shell-programming.html"><a href="shell-programming.html#variable-arrays"><i class="fa fa-check"></i><b>5.3.6</b> Variable arrays</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="shell-programming.html"><a href="shell-programming.html#evaluate-a-command-and-substitute-the-result-on-the-command-line"><i class="fa fa-check"></i><b>5.4</b> Evaluate a command and substitute the result on the command line</a></li>
<li class="chapter" data-level="5.5" data-path="shell-programming.html"><a href="shell-programming.html#groupingcollecting-output-from-multiple-commands-commands-and-commands"><i class="fa fa-check"></i><b>5.5</b> Grouping/Collecting output from multiple commands: <code>(commands)</code> and <code>{ commands; }</code></a></li>
<li class="chapter" data-level="5.6" data-path="shell-programming.html"><a href="shell-programming.html#exit-status"><i class="fa fa-check"></i><b>5.6</b> Exit Status</a><ul>
<li class="chapter" data-level="5.6.1" data-path="shell-programming.html"><a href="shell-programming.html#combinations-of-exit-statuses"><i class="fa fa-check"></i><b>5.6.1</b> Combinations of exit statuses</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="shell-programming.html"><a href="shell-programming.html#loops-and-repetition"><i class="fa fa-check"></i><b>5.7</b> Loops and repetition</a></li>
<li class="chapter" data-level="5.8" data-path="shell-programming.html"><a href="shell-programming.html#more-conditional-evaluation-if-then-else-and-friends"><i class="fa fa-check"></i><b>5.8</b> More Conditional Evaluation: <code>if</code>, <code>then</code>, <code>else</code>, and friends</a></li>
<li class="chapter" data-level="5.9" data-path="shell-programming.html"><a href="shell-programming.html#finallypositional-parameters"><i class="fa fa-check"></i><b>5.9</b> Finally…positional parameters</a></li>
<li class="chapter" data-level="5.10" data-path="shell-programming.html"><a href="shell-programming.html#basename-and-dirname-two-useful-little-utilities"><i class="fa fa-check"></i><b>5.10</b> <code>basename</code> and <code>dirname</code> two useful little utilities</a></li>
<li class="chapter" data-level="5.11" data-path="shell-programming.html"><a href="shell-programming.html#bash-functions"><i class="fa fa-check"></i><b>5.11</b> <code>bash</code> functions</a></li>
<li class="chapter" data-level="5.12" data-path="shell-programming.html"><a href="shell-programming.html#reading-files-line-by-line"><i class="fa fa-check"></i><b>5.12</b> reading files line by line</a></li>
<li class="chapter" data-level="5.13" data-path="shell-programming.html"><a href="shell-programming.html#further-reading"><i class="fa fa-check"></i><b>5.13</b> Further reading</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html"><i class="fa fa-check"></i><b>6</b> Sed, awk, and regular expressions</a><ul>
<li class="chapter" data-level="6.1" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#awk"><i class="fa fa-check"></i><b>6.1</b> awk</a><ul>
<li class="chapter" data-level="6.1.1" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#line-cycling-tests-and-actions"><i class="fa fa-check"></i><b>6.1.1</b> Line-cycling, tests and actions</a></li>
<li class="chapter" data-level="6.1.2" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#column-splitting-fields--f-nf-print-ofs-and-begin"><i class="fa fa-check"></i><b>6.1.2</b> Column splitting, fields, <code>-F</code>, <code>$</code>, <code>NF</code>, <code>print</code>, <code>OFS</code> and <code>BEGIN</code></a></li>
<li class="chapter" data-level="6.1.3" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#a-brief-introduction-to-regular-expressions"><i class="fa fa-check"></i><b>6.1.3</b> A brief introduction to regular expressions</a></li>
<li class="chapter" data-level="6.1.4" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#a-variety-of-tests"><i class="fa fa-check"></i><b>6.1.4</b> A variety of tests</a></li>
<li class="chapter" data-level="6.1.5" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#code-in-the-action-blocks"><i class="fa fa-check"></i><b>6.1.5</b> Code in the action blocks</a></li>
<li class="chapter" data-level="6.1.6" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#using-awk-to-assign-to-shell-variables"><i class="fa fa-check"></i><b>6.1.6</b> Using <code>awk</code> to assign to shell variables</a></li>
<li class="chapter" data-level="6.1.7" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#passing-variables-into-awk-with--v"><i class="fa fa-check"></i><b>6.1.7</b> Passing Variables into <code>awk</code> with <code>-v</code></a></li>
<li class="chapter" data-level="6.1.8" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#writing-awk-scripts-in-files"><i class="fa fa-check"></i><b>6.1.8</b> Writing awk scripts in files</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#sed"><i class="fa fa-check"></i><b>6.2</b> sed</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html"><i class="fa fa-check"></i><b>7</b> Working on remote servers</a><ul>
<li class="chapter" data-level="7.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#accessing-remote-computers"><i class="fa fa-check"></i><b>7.1</b> Accessing remote computers</a><ul>
<li class="chapter" data-level="7.1.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#windows"><i class="fa fa-check"></i><b>7.1.1</b> Windows</a></li>
<li class="chapter" data-level="7.1.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#hummingbird"><i class="fa fa-check"></i><b>7.1.2</b> Hummingbird</a></li>
<li class="chapter" data-level="7.1.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#summit"><i class="fa fa-check"></i><b>7.1.3</b> Summit</a></li>
<li class="chapter" data-level="7.1.4" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#sedna"><i class="fa fa-check"></i><b>7.1.4</b> Sedna</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#transferring-files-to-remote-computers"><i class="fa fa-check"></i><b>7.2</b> Transferring files to remote computers</a><ul>
<li class="chapter" data-level="7.2.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#sftp-via-lftp"><i class="fa fa-check"></i><b>7.2.1</b> <code>sftp</code> (via <code>lftp</code>)</a></li>
<li class="chapter" data-level="7.2.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#git"><i class="fa fa-check"></i><b>7.2.2</b> git</a></li>
<li class="chapter" data-level="7.2.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#globus"><i class="fa fa-check"></i><b>7.2.3</b> Globus</a></li>
<li class="chapter" data-level="7.2.4" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#interfacing-with-the-cloud"><i class="fa fa-check"></i><b>7.2.4</b> Interfacing with “The Cloud”</a></li>
<li class="chapter" data-level="7.2.5" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#getting-files-from-a-sequencing-center"><i class="fa fa-check"></i><b>7.2.5</b> Getting files from a sequencing center</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#tmux"><i class="fa fa-check"></i><b>7.3</b> <code>tmux</code>: the terminal multiplexer</a><ul>
<li class="chapter" data-level="7.3.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#an-analogy-for-how-tmux-works"><i class="fa fa-check"></i><b>7.3.1</b> An analogy for how <code>tmux</code> works</a></li>
<li class="chapter" data-level="7.3.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#first-steps-with-tmux"><i class="fa fa-check"></i><b>7.3.2</b> First steps with <code>tmux</code></a></li>
<li class="chapter" data-level="7.3.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#further-steps-with-tmux"><i class="fa fa-check"></i><b>7.3.3</b> Further steps with <code>tmux</code></a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#vim-its-time-to-get-serious-with-text-editing"><i class="fa fa-check"></i><b>7.4</b> <code>vim</code>: it’s time to get serious with text editing</a><ul>
<li class="chapter" data-level="7.4.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#using-neovim-and-nvim-r-and-tmux-to-use-r-well-on-the-cluster"><i class="fa fa-check"></i><b>7.4.1</b> Using neovim and Nvim-R and tmux to use R well on the cluster</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="high-performance-computing-clusters-hpccs.html"><a href="high-performance-computing-clusters-hpccs.html"><i class="fa fa-check"></i><b>8</b> High Performance Computing Clusters (HPCC’s)</a><ul>
<li class="chapter" data-level="8.1" data-path="high-performance-computing-clusters-hpccs.html"><a href="high-performance-computing-clusters-hpccs.html#an-oversimplified-but-useful-view-of-a-computing-cluster"><i class="fa fa-check"></i><b>8.1</b> An oversimplified, but useful, view of a computing cluster</a></li>
<li class="chapter" data-level="8.2" data-path="high-performance-computing-clusters-hpccs.html"><a href="high-performance-computing-clusters-hpccs.html#cluster-computing-and-the-job-scheduler"><i class="fa fa-check"></i><b>8.2</b> Cluster computing and the job scheduler</a></li>
<li class="chapter" data-level="8.3" data-path="high-performance-computing-clusters-hpccs.html"><a href="high-performance-computing-clusters-hpccs.html#learning-about-the-resources-on-your-hpcc"><i class="fa fa-check"></i><b>8.3</b> Learning about the resources on your HPCC</a></li>
<li class="chapter" data-level="8.4" data-path="high-performance-computing-clusters-hpccs.html"><a href="high-performance-computing-clusters-hpccs.html#installing-software-on-an-hpcc"><i class="fa fa-check"></i><b>8.4</b> Installing Software on an HPCC</a><ul>
<li class="chapter" data-level="8.4.1" data-path="high-performance-computing-clusters-hpccs.html"><a href="high-performance-computing-clusters-hpccs.html#miniconda"><i class="fa fa-check"></i><b>8.4.1</b> Miniconda</a></li>
<li class="chapter" data-level="8.4.2" data-path="high-performance-computing-clusters-hpccs.html"><a href="high-performance-computing-clusters-hpccs.html#modules"><i class="fa fa-check"></i><b>8.4.2</b> Modules</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="high-performance-computing-clusters-hpccs.html"><a href="high-performance-computing-clusters-hpccs.html#getting-compute-resources-allocated-to-your-jobs-on-an-hpcc"><i class="fa fa-check"></i><b>8.5</b> Getting compute resources allocated to your jobs on an HPCC</a><ul>
<li class="chapter" data-level="8.5.1" data-path="high-performance-computing-clusters-hpccs.html"><a href="high-performance-computing-clusters-hpccs.html#interactive-sessions"><i class="fa fa-check"></i><b>8.5.1</b> Interactive sessions</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="high-performance-computing-clusters-hpccs.html"><a href="high-performance-computing-clusters-hpccs.html#the-queue-slurmsgeuge"><i class="fa fa-check"></i><b>8.6</b> The Queue (SLURM/SGE/UGE)</a></li>
<li class="chapter" data-level="8.7" data-path="high-performance-computing-clusters-hpccs.html"><a href="high-performance-computing-clusters-hpccs.html#modules-package"><i class="fa fa-check"></i><b>8.7</b> Modules package</a></li>
<li class="chapter" data-level="8.8" data-path="high-performance-computing-clusters-hpccs.html"><a href="high-performance-computing-clusters-hpccs.html#compiling-programs-without-admin-privileges"><i class="fa fa-check"></i><b>8.8</b> Compiling programs without admin privileges</a></li>
<li class="chapter" data-level="8.9" data-path="high-performance-computing-clusters-hpccs.html"><a href="high-performance-computing-clusters-hpccs.html#job-arrays"><i class="fa fa-check"></i><b>8.9</b> Job arrays</a></li>
<li class="chapter" data-level="8.10" data-path="high-performance-computing-clusters-hpccs.html"><a href="high-performance-computing-clusters-hpccs.html#writing-stdout-and-stderr-to-files"><i class="fa fa-check"></i><b>8.10</b> Writing stdout and stderr to files</a></li>
<li class="chapter" data-level="8.11" data-path="high-performance-computing-clusters-hpccs.html"><a href="high-performance-computing-clusters-hpccs.html#breaking-stuff-down"><i class="fa fa-check"></i><b>8.11</b> Breaking stuff down</a></li>
</ul></li>
<li class="part"><span><b>II Part II: Reproducible Research Strategies</b></span></li>
<li class="chapter" data-level="9" data-path="introduction-to-reproducible-research.html"><a href="introduction-to-reproducible-research.html"><i class="fa fa-check"></i><b>9</b> Introduction to Reproducible Research</a></li>
<li class="chapter" data-level="10" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html"><i class="fa fa-check"></i><b>10</b> Rstudio and Project-centered Organization</a><ul>
<li class="chapter" data-level="10.1" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#organizing-big-projects"><i class="fa fa-check"></i><b>10.1</b> Organizing big projects</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="version-control.html"><a href="version-control.html"><i class="fa fa-check"></i><b>11</b> Version control</a><ul>
<li class="chapter" data-level="11.1" data-path="version-control.html"><a href="version-control.html#why-use-version-control"><i class="fa fa-check"></i><b>11.1</b> Why use version control?</a></li>
<li class="chapter" data-level="11.2" data-path="version-control.html"><a href="version-control.html#git-workings"><i class="fa fa-check"></i><b>11.2</b> How git works</a></li>
<li class="chapter" data-level="11.3" data-path="version-control.html"><a href="version-control.html#git-workflow-patterns"><i class="fa fa-check"></i><b>11.3</b> git workflow patterns</a></li>
<li class="chapter" data-level="11.4" data-path="version-control.html"><a href="version-control.html#using-git-with-rstudio"><i class="fa fa-check"></i><b>11.4</b> using git with Rstudio</a></li>
<li class="chapter" data-level="11.5" data-path="version-control.html"><a href="version-control.html#git-on-the-command-line"><i class="fa fa-check"></i><b>11.5</b> git on the command line</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="a-fast-furious-overview-of-the-tidyverse.html"><a href="a-fast-furious-overview-of-the-tidyverse.html"><i class="fa fa-check"></i><b>12</b> A fast, furious overview of the tidyverse</a></li>
<li class="chapter" data-level="13" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html"><i class="fa fa-check"></i><b>13</b> Authoring reproducibly with Rmarkdown</a><ul>
<li class="chapter" data-level="13.1" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#notebooks"><i class="fa fa-check"></i><b>13.1</b> Notebooks</a></li>
<li class="chapter" data-level="13.2" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#references"><i class="fa fa-check"></i><b>13.2</b> References</a><ul>
<li class="chapter" data-level="13.2.1" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#zotero-and-rmarkdown"><i class="fa fa-check"></i><b>13.2.1</b> Zotero and Rmarkdown</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#bookdown"><i class="fa fa-check"></i><b>13.3</b> Bookdown</a></li>
<li class="chapter" data-level="13.4" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#google-docs"><i class="fa fa-check"></i><b>13.4</b> Google Docs</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="using-python.html"><a href="using-python.html"><i class="fa fa-check"></i><b>14</b> Using python</a></li>
<li class="part"><span><b>III Part III: Bioinformatic Analyses</b></span></li>
<li class="chapter" data-level="15" data-path="overview-of-bioinformatic-analyses.html"><a href="overview-of-bioinformatic-analyses.html"><i class="fa fa-check"></i><b>15</b> Overview of Bioinformatic Analyses</a></li>
<li class="chapter" data-level="16" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html"><i class="fa fa-check"></i><b>16</b> DNA Sequences and Sequencing</a><ul>
<li class="chapter" data-level="16.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#dna-stuff"><i class="fa fa-check"></i><b>16.1</b> DNA Stuff</a><ul>
<li class="chapter" data-level="16.1.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#dna-replication-with-dna-polymerase"><i class="fa fa-check"></i><b>16.1.1</b> DNA Replication with DNA Polymerase</a></li>
<li class="chapter" data-level="16.1.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#the-importance-of-the-3-hydroxyl"><i class="fa fa-check"></i><b>16.1.2</b> The importance of the 3’ hydroxyl…</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#sanger-sequencing"><i class="fa fa-check"></i><b>16.2</b> Sanger sequencing</a></li>
<li class="chapter" data-level="16.3" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#illumina-sequencing-by-synthesis"><i class="fa fa-check"></i><b>16.3</b> Illumina Sequencing by Synthesis</a></li>
<li class="chapter" data-level="16.4" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#library-prep-protocols"><i class="fa fa-check"></i><b>16.4</b> Library Prep Protocols</a><ul>
<li class="chapter" data-level="16.4.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#wgs"><i class="fa fa-check"></i><b>16.4.1</b> WGS</a></li>
<li class="chapter" data-level="16.4.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#rad-seq-methods"><i class="fa fa-check"></i><b>16.4.2</b> RAD-Seq methods</a></li>
<li class="chapter" data-level="16.4.3" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#amplicon-sequencing"><i class="fa fa-check"></i><b>16.4.3</b> Amplicon Sequencing</a></li>
<li class="chapter" data-level="16.4.4" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#capture-arrays-rapture-etc."><i class="fa fa-check"></i><b>16.4.4</b> Capture arrays, RAPTURE, etc.</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html"><i class="fa fa-check"></i><b>17</b> Bioinformatic file formats</a><ul>
<li class="chapter" data-level="17.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sequences"><i class="fa fa-check"></i><b>17.1</b> Sequences</a></li>
<li class="chapter" data-level="17.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#fastq"><i class="fa fa-check"></i><b>17.2</b> FASTQ</a><ul>
<li class="chapter" data-level="17.2.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#line-1-illumina-identifier-lines"><i class="fa fa-check"></i><b>17.2.1</b> Line 1: Illumina identifier lines</a></li>
<li class="chapter" data-level="17.2.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#line-4-base-quality-scores"><i class="fa fa-check"></i><b>17.2.2</b> Line 4: Base quality scores</a></li>
<li class="chapter" data-level="17.2.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#a-fastq-tidyverse-interlude"><i class="fa fa-check"></i><b>17.2.3</b> A FASTQ ‘tidyverse’ Interlude</a></li>
<li class="chapter" data-level="17.2.4" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#comparing-read-1-to-read-2"><i class="fa fa-check"></i><b>17.2.4</b> Comparing read 1 to read 2</a></li>
</ul></li>
<li class="chapter" data-level="17.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#fasta"><i class="fa fa-check"></i><b>17.3</b> FASTA</a><ul>
<li class="chapter" data-level="17.3.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#genomic-ranges"><i class="fa fa-check"></i><b>17.3.1</b> Genomic ranges</a></li>
<li class="chapter" data-level="17.3.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#extracting-genomic-ranges-from-a-fasta-file"><i class="fa fa-check"></i><b>17.3.2</b> Extracting genomic ranges from a FASTA file</a></li>
<li class="chapter" data-level="17.3.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#downloading-reference-genomes-from-ncbi"><i class="fa fa-check"></i><b>17.3.3</b> Downloading reference genomes from NCBI</a></li>
</ul></li>
<li class="chapter" data-level="17.4" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#alignments"><i class="fa fa-check"></i><b>17.4</b> Alignments</a><ul>
<li class="chapter" data-level="17.4.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#how-might-i-align-to-thee-let-me-count-the-ways"><i class="fa fa-check"></i><b>17.4.1</b> How might I align to thee? Let me count the ways…</a></li>
<li class="chapter" data-level="17.4.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#play-with-simple-alignments"><i class="fa fa-check"></i><b>17.4.2</b> Play with simple alignments</a></li>
<li class="chapter" data-level="17.4.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sam-flags"><i class="fa fa-check"></i><b>17.4.3</b> SAM Flags</a></li>
<li class="chapter" data-level="17.4.4" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#the-cigar-string"><i class="fa fa-check"></i><b>17.4.4</b> The CIGAR string</a></li>
<li class="chapter" data-level="17.4.5" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#the-seq-and-qual-columns"><i class="fa fa-check"></i><b>17.4.5</b> The SEQ and QUAL columns</a></li>
<li class="chapter" data-level="17.4.6" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sam-file-headers"><i class="fa fa-check"></i><b>17.4.6</b> SAM File Headers</a></li>
<li class="chapter" data-level="17.4.7" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#the-bam-format"><i class="fa fa-check"></i><b>17.4.7</b> The BAM format</a></li>
<li class="chapter" data-level="17.4.8" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#quick-self-study"><i class="fa fa-check"></i><b>17.4.8</b> Quick self study</a></li>
</ul></li>
<li class="chapter" data-level="17.5" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#variants"><i class="fa fa-check"></i><b>17.5</b> Variants</a></li>
<li class="chapter" data-level="17.6" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#segments"><i class="fa fa-check"></i><b>17.6</b> Segments</a></li>
<li class="chapter" data-level="17.7" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#conversionextractions-between-different-formats"><i class="fa fa-check"></i><b>17.7</b> Conversion/Extractions between different formats</a></li>
<li class="chapter" data-level="17.8" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#visualization-of-genomic-data"><i class="fa fa-check"></i><b>17.8</b> Visualization of Genomic Data</a><ul>
<li class="chapter" data-level="17.8.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sample-data"><i class="fa fa-check"></i><b>17.8.1</b> Sample Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="18" data-path="genome-assembly.html"><a href="genome-assembly.html"><i class="fa fa-check"></i><b>18</b> Genome Assembly</a></li>
<li class="chapter" data-level="19" data-path="alignment-of-sequence-data-to-a-reference-genome.html"><a href="alignment-of-sequence-data-to-a-reference-genome.html"><i class="fa fa-check"></i><b>19</b> Alignment of sequence data to a reference genome</a><ul>
<li class="chapter" data-level="19.1" data-path="alignment-of-sequence-data-to-a-reference-genome.html"><a href="alignment-of-sequence-data-to-a-reference-genome.html#preprocess"><i class="fa fa-check"></i><b>19.1</b> Preprocess ?</a></li>
<li class="chapter" data-level="19.2" data-path="alignment-of-sequence-data-to-a-reference-genome.html"><a href="alignment-of-sequence-data-to-a-reference-genome.html#read-groups"><i class="fa fa-check"></i><b>19.2</b> Read Groups</a></li>
<li class="chapter" data-level="19.3" data-path="alignment-of-sequence-data-to-a-reference-genome.html"><a href="alignment-of-sequence-data-to-a-reference-genome.html#merging-bam-files"><i class="fa fa-check"></i><b>19.3</b> Merging BAM files</a></li>
<li class="chapter" data-level="19.4" data-path="alignment-of-sequence-data-to-a-reference-genome.html"><a href="alignment-of-sequence-data-to-a-reference-genome.html#divide-and-conquer-strategies"><i class="fa fa-check"></i><b>19.4</b> Divide and Conquer Strategies</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="variant-calling-with-gatk.html"><a href="variant-calling-with-gatk.html"><i class="fa fa-check"></i><b>20</b> Variant calling with GATK</a></li>
<li class="chapter" data-level="21" data-path="bioinformatics-for-rad-seq-data-with-and-without-a-reference-genome.html"><a href="bioinformatics-for-rad-seq-data-with-and-without-a-reference-genome.html"><i class="fa fa-check"></i><b>21</b> Bioinformatics for RAD seq data with and without a reference genome</a></li>
<li class="chapter" data-level="22" data-path="processing-amplicon-sequencing-data.html"><a href="processing-amplicon-sequencing-data.html"><i class="fa fa-check"></i><b>22</b> Processing amplicon sequencing data</a></li>
<li class="chapter" data-level="23" data-path="genome-annotation.html"><a href="genome-annotation.html"><i class="fa fa-check"></i><b>23</b> Genome Annotation</a></li>
<li class="chapter" data-level="24" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html"><i class="fa fa-check"></i><b>24</b> Whole genome alignment strategies</a><ul>
<li class="chapter" data-level="24.1" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#mapping-of-scaffolds-to-a-closely-related-genome"><i class="fa fa-check"></i><b>24.1</b> Mapping of scaffolds to a closely related genome</a></li>
<li class="chapter" data-level="24.2" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#obtaining-ancestral-states-from-an-outgroup-genome"><i class="fa fa-check"></i><b>24.2</b> Obtaining Ancestral States from an Outgroup Genome</a><ul>
<li class="chapter" data-level="24.2.1" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#using-lastz-to-align-coho-to-the-chinook-genome"><i class="fa fa-check"></i><b>24.2.1</b> Using LASTZ to align coho to the chinook genome</a></li>
<li class="chapter" data-level="24.2.2" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#try-on-the-chinook-chromosomes"><i class="fa fa-check"></i><b>24.2.2</b> Try on the chinook chromosomes</a></li>
<li class="chapter" data-level="24.2.3" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#explore-the-other-parameters-more"><i class="fa fa-check"></i><b>24.2.3</b> Explore the other parameters more</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>IV Part IV: Analysis of Big Variant Data</b></span></li>
<li class="chapter" data-level="25" data-path="bioinformatic-analysis-on-variant-data.html"><a href="bioinformatic-analysis-on-variant-data.html"><i class="fa fa-check"></i><b>25</b> Bioinformatic analysis on variant data</a></li>
<li class="part"><span><b>V Part V: Population Genomics</b></span></li>
<li class="chapter" data-level="26" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html"><i class="fa fa-check"></i><b>26</b> Topics in pop gen</a><ul>
<li class="chapter" data-level="26.1" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#coalescent"><i class="fa fa-check"></i><b>26.1</b> Coalescent</a></li>
<li class="chapter" data-level="26.2" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#measures-of-genetic-diversity-and-such"><i class="fa fa-check"></i><b>26.2</b> Measures of genetic diversity and such</a></li>
<li class="chapter" data-level="26.3" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#demographic-inference-with-partial-a-partial-i-and-moments"><i class="fa fa-check"></i><b>26.3</b> Demographic inference with <span class="math inline">\(\partial a \partial i\)</span> and <em>moments</em></a></li>
<li class="chapter" data-level="26.4" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#balls-in-boxes"><i class="fa fa-check"></i><b>26.4</b> Balls in Boxes</a></li>
<li class="chapter" data-level="26.5" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#some-landscape-genetics"><i class="fa fa-check"></i><b>26.5</b> Some landscape genetics</a></li>
<li class="chapter" data-level="26.6" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#relationship-inference"><i class="fa fa-check"></i><b>26.6</b> Relationship Inference</a></li>
<li class="chapter" data-level="26.7" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#tests-for-selection"><i class="fa fa-check"></i><b>26.7</b> Tests for Selection</a></li>
<li class="chapter" data-level="26.8" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#multivariate-associations-gea-etc."><i class="fa fa-check"></i><b>26.8</b> Multivariate Associations, GEA, etc.</a></li>
<li class="chapter" data-level="26.9" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#estimating-heritability-in-the-wild"><i class="fa fa-check"></i><b>26.9</b> Estimating heritability in the wild</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references-1.html"><a href="references-1.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://bookdown.org" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="high-performance-computing-clusters-hpccs" class="section level1">
<h1><span class="header-section-number">Chapter 8</span> High Performance Computing Clusters (HPCC’s)</h1>
<p>One interesting aspect of modern next generation sequencing data is simply its
sheer size: it is not uncommon to receive a half or a full terabyte of data from
a sequencing center. It is impractical to store this much data on your
laptop, let alone analyze it there. Crunching through such a quantity of data on a single
computer or server could take a very long time. Instead, you will likely break
up the analysis of such data into a number of smaller jobs, and send them off to
run on an assortment of different computers in a High Performance Computing Cluster (HPCC).</p>
<p>Thus, even if you have immersed yourself in bioinformatic data file formats and honed
your skills at shell programming and accessing remote computers, sequence data analysis
remains such a formidable foe that there is still one last key area of computing
in which you must be fluent, in order to comfortably do bioinformatics: you must understand
how to submit and manage jobs sent to an HPCC.</p>
<p>My first experience with HPCC’s occurred when I started analyzing
high-throughput sequencer output. I had over 15 years experience in shell
programming at that time, and I was given some example analysis scripts to emulate, but I still
found it took several weeks before I was moderately comfortable in an HPCC environment. Most
HPCC’s have some sort of tutorial web pages that provide a little bit of background
on cluster computing, but I didn’t find the ones available to me, at the time, to
be particularly helpful.</p>
<p>The goal of this chapter is to provide the sort of background I wish that I had
when I started doing cluster computing for bioinformatics. I will not be providing
a comprehensive overview of parallel computation. For example, we will not focus at all
upon the rich tradition of parallel computing applications through “message passing” interfaces which
can maintain a synchronized analysis from a single program executing
on multiple computers at the same time. Rather, we will focus on the manner in which
most bioinformatic problems can be broken down into a series of smaller jobs,
each of which can be run, independently, on its own processor without the
need for maintaining synchrony between multiple processes.</p>
<p>We start with an overview of what an HPCC consists of, defining a few important
terms. Then we provide some background on the fundamental problem of cluster
computing: namely that a lot of people want to use the computing power of the cluster,
but it needs to be allocated to users in an equitable fashion. An understanding of this
forms the basis for our discussion of <em>job scheduling</em> and the methods you must
use to tell the <em>job scheduler</em> what resources you will need, so that those resources
will eventually be allocated to you. We will cover a job scheduler called SLURM, which
stands for the Simple Linux Utility for Resource Management. It is the scheduler used
on the Summit supercomputer in Boulder and the Hummingbird and Sedna clusters deployed
at UCSC and the NWFSC, respectively.</p>
<p>Interspersed with our discussion of SLURM, we will cover
methods for installing software to use for your analyses.</p>
<div id="an-oversimplified-but-useful-view-of-a-computing-cluster" class="section level2">
<h2><span class="header-section-number">8.1</span> An oversimplified, but useful, view of a computing cluster</h2>
<p>At its simplest, an HPCC can be thought of as a whole lot of computers that are all
put in a room somewhere for the purposes of doing a lot of computations. Most of these
computers do not possess all the elements that typically come to mind when you
think of a “computer.” For example, none of them are attached to monitors—each
computer resembles just the “box” part of a desktop computer. Each of these “boxes” is called
a <em>node</em>. The node is the unit in a cluster that corresponds most closely to what you think of
as a “computer.” As is typical of most computers today, each of these nodes has some
amount of Random Access Memory (RAM) that is <em>only accessible by the node itself</em>. RAM
is the space in memory that is used for active computation and calculation.</p>
<p>Each of these
nodes might also have a hard drive used for operating system software. The bulk of the
hard drive space that each node can access, however, is in the form of a large array of hard disks
that are connected to <em>all</em> of the nodes by an interface that allows data to be transferred
back and forth between each node and the hard-drive array at speeds that would be expected of
an internal hard drive (i.e., this array of hard drives is not just plugged in with a USB
cable). Memory on hard drives is used for storing data and the results of calculations, but
is not used for active calculations the way RAM is used. In order for calculations to be done
on data that are on the hard drive array, it must first be read into a node’s RAM. After the calculations
are done, the results are typically written back out onto the hard drive array.</p>
<p>On a typical cluster, there are usually several different “portions” of the hard drive
array attached to every <em>node</em>. One part of the array holds <em>home directories</em> of the
different users. Each user typically has a limited amount of storage in their home directory, but
the data in these home directories is usually safe or protected, meaning you can put a file there
and expect that it will be there next week, month, or year, etc. It is also likely that the home directories are
backed up (but check the docs for your cluster to confirm this!). Since the space in home directories is limited,
you typically will not put large data sets in your home directory for long-term storage, but you will
store scripts and programs, and such items as cloned GitHub repositories there. Another
type of storage that exists on the hard drive array is called <em>persistent long-term storage</em>. This type of storage
is purchased for use by research groups to store large quantities of
data on the cluster for long periods of time. As discussed in the last chapter, the rise of cloud-based storage solutions,
like Google Drive, offering unlimited storage to institutional users, makes persistent long-term storage less
important (and less cost-effective) for many research groups. Finally, the third type of
storage in the hard drive array is called <em>scratch storage</em>. There are usually fairly
light limits (if any at all) to how much data can be placed in scratch storage, but there
will typically be Draconian <em>time limits</em> placed on your scratch storage. For example, on the
Hoffman2 cluster at UCLA, you are granted 2 Tb of <code>scratch</code> storage, but any files that
have sat unmodified in <code>scratch</code> for more than 14 days will be deleted (and if space is tight, the
system administrators may delete things from <code>scratch</code> in far fewer then 14 days.) Check
your local cluster documentation for information about time and space limits on <code>scratch</code>.</p>
<p>On many clusters, scratch space is also configured to be very fast for input and output
(for example, on many systems, the scratch storage will be composed of solid state drives rather
than spinning hard disks). On jobs that require that a lot of data be accessed from the
drive or written to it (this includes most operations on BAM files), significant decreases
in overall running time can be seen by using fast storage. Finally, scratch space exists on a cluster
expressly as <em>the place</em> to put data and outputs on the hard drive <em>when running jobs</em>. For all these
reasons, when you run jobs, you will always want to read and write data from and to <code>scratch</code>. We
will talk more about the specifics of doing so, but for now, you should be developing a generic picture
of your cluster computing workflow that looks like:</p>
<ol style="list-style-type: decimal">
<li>Download big data files from the cloud to <code>scratch</code></li>
<li>Run analyses on those data files, writing output back to <code>scratch</code>.</li>
<li>When done, copy, to the cloud, any products that you wish to keep.</li>
<li>Remove data and outputs left on <code>scratch</code>.</li>
</ol>
<p>As is the case with virtually all modern desktop or laptop computers, within each node,
there are multiple (typically between 16 and 48) <em>cores</em>, which are the computer chip units that actually
do the computations within a node. A <em>serial job</em> is one that just runs on a single core
within a <em>node</em>, while a <em>parallel job</em> might run on multiple cores, at the same time, within
a single node. In such a parallel job, each core has access to the same data within the
node’s RAM (a “shared-memory,” parallel job). The <em>core</em>
is the fundamental unit of computing machinery that gets allocated to perform jobs in an HPCC.</p>
<p>Most of the nodes in a cluster are there to hold the cores which are the computational workhorses,
slogging through calculations for the HPCC’s myriad users. However, some nodes
are more appropriate to certain types of computations than others (for example, some
might have lots of memory for doing genome assembly, while others will have
big, hurkin’, graphical processing units to be used for GPU calculations). Or, some nodes
might be available preferentially for different users than for others. For these
reasons, nodes are grouped into different collections. Depending on the system you are
using, these collections of different nodes are called, either, <em>partitions</em> or <em>queues</em>.
The world of SLURM uses the term <em>partitions</em> for these collections, and we will adopt
that language as well, using <em>queue</em> to refer to the line of jobs that are waiting to start
on an HPCC; however we warn the reader that other job schedulers (like the Univa Grid Engine)
use the term “queues” to refer to collections of nodes.</p>
<p>On every cluster, however, there will be one to several nodes that are reserved not for
doing computation, but for allowing users to access the cluster. These are called the
<em>login</em> nodes or the <em>head</em> nodes. These nodes are <em>solely</em> for logging in, light editing of
scripts, minor manipulation of directories, and scheduling and managing jobs. They are absolutely <em>not</em>
for doing major computations. For example, you should never login to the head node and immediately
start using it, in an interactive <code>bash</code> session to, say, sort BAM files or run <code>bwa mem</code>
to do alignments. Running commands that require a lot of computation, or a lot of
input and output from the disk, on the login nodes is an egregious
<em>faux pas</em> of cluster computing. Doing so can
negatively impact the ability of other users to login or otherwise get their work done, and it
might lead the system administrators to disable your account. Therefore,
never do it! All of your hardcore computation on a cluster <em>has</em> to be done on
a <em>compute node</em>. We will show how to do that shortly, but first we will talk about why.</p>
</div>
<div id="cluster-computing-and-the-job-scheduler" class="section level2">
<h2><span class="header-section-number">8.2</span> Cluster computing and the job scheduler</h2>
<p>When you do work, or stream a video, or surf the web on your laptop computer, there are numerous
different computer processes running, to make sure that your computer keeps working and doing what it
is supposed to be doing. In the case of your laptop, the operating system, itself, orchestrates
all these different processes, making sure that each one is given some compute time on your
laptop’s processors in order to get its work done. Your laptop’s operating system has a fair bit
of flexibility in how it allocates resources to these different processes: it has multiple
cores to assign different processes to, <em>and</em> it
allows multiple processes to run on a single core, alternating between these different processes over
different <em>cycles</em> of the central processing unit. Things work differently
on a shared resource like an HPCC. The main, interesting problem of cluster computing is basically
this: lots of people want to use the cluster to run big jobs, but the cluster does not
run like a single computer.</p>
<p>The cluster is not happy to give lots of different jobs
from lots of different users a chance to all run on the same core, sharing time by dividing up cycles.
When a user wants to use the computational resources of an HPCC,
she cannot just start a job and be confident that it will launch immediately and be granted
at least a few CPU cycles every now and again. Rather, on an HPCC, every <em>job</em> that is submitted
by a user will be assigned to a dedicated core (or several cores, if requested, and granted)
with a dedicated amount of memory.
If a core is not available for use, the job “gets in line” (into the “queue”, so to speak)
where it sits and waits (doing nothing) until a core
(with sufficient associated resources, like RAM memory) becomes available. When such a core
becomes available in the cluster, the job gets launched on it. All of this is orchestrated by the job scheduler,
of which SLURM is an example.</p>
<p>In this computing model, a job, once it is launched, ties up the core and the memory
that has been allocated to it until the job is finished. While that job is running, no one
else’s jobs or processes can run on the core or share the RAM memory that was allocated to the job.
For this reason, the job scheduler, needs to know, <em>ahead of time</em> how long each job might run and
what resources will be required during that time. A simple contrived example illustrates things easily:
imagine that Joe and Cheryl each have 1000 separate jobs to run. Each of Cheryl’s jobs involves running
a machine-learning algorithm to identify seabirds in high-resolution, aerial images of the ocean, and
takes only about 20 minutes running on a single core. Each of Joe’s jobs, on the other hand, involves mapping billions of
sequencing reads, a task which requires about 36 hours when run on a single core.
If their cluster has only 640 cores, and Joe submits his jobs first,
then, if the job scheduler were naive, it might put all of his jobs in line first, requiring some 50 or 60 hours
before the first of Cheryl’s jobs even runs. This would be a huge buzz kill for Cheryl.
However, if Cheryl and Joe both have to provide estimates to the scheduler of how
long their jobs will run, the scheduler can make more equitable decisions, starting a few of Joe’s jobs, but retaining
many more cores for Cheryl’s jobs, each of which runs much faster.</p>
<p>Thus, when you want to run any jobs on a cluster, you must provide an estimate of the resources
that the job will require. The three main axes upon which these resources are measured are:</p>
<ol style="list-style-type: decimal">
<li>The number of cores the job will require.</li>
<li>The maximum amount of RAM (memory) the job will require.</li>
<li>The amount of time for which the job will run.</li>
</ol>
<p>Requests for large amounts of resources for long periods of time generally take longer to start.
There are two main reasons for this: either 1) the scheduler does not want to launch too many
long-duration, high-memory jobs because it anticipates other users will want to use resources
down the road and no single user should tie up the compute resources
for too long; or 2) there are so many jobs running on myriad nodes and cores, that only
infrequently do nodes with sufficient numbers of cores and RAM come available to start
the new jobs.</p>
<p>The second reason is a particular bane of new cluster users who unwittingly request more resources
than actually exist (i.e. 52 cores, when no single node has more than 32; or 50 Gb of RAM when no single
node has more than 48 Gb). Unfortunately (or, perhaps comically, if you have a sick sense of
humor), most job schedulers will not notify you of this sort of transgression.
Rather, your job will just sit in line waiting to be launched, but it never will be, because sufficient
resources never become available!</p>
<p>It is worth noting that regardless of whether reason 1 or reason 2 is the dominant cause influencing
how long it takes to start a job, asking for fewer resources for less time will generally allow your
jobs to start faster. Particularly because of reason #2, however, breaking your jobs down (if possible) into
small chunks that will run relatively quickly on a single core with low RAM needs can render many more
opportunities for your jobs to start, letting you tap into resources that are not often fully utilized
in a cluster. Since I started working on large cluster in which it took a long time to start a job
that required all or most of the cores on a single node, but in which there were many nodes
harboring a few cores that were not being used, I tend to endorse this approach…</p>
<p>Since the requested resources for a job play such a large role in wait times for jobs to start,
you might wonder why people don’t intentionally underestimate the resources they request for their
jobs. The answer is simple: the job scheduler is a highly efficient and completely dispassionate
manager. If you requested 2 hours for your job, but your job has not finished in that amount of time,
the job scheduler will waste no time hemming and hawing or having an emotional struggle with itself
about whether it should stop your job. No, at 2 hours and 0.2 seconds it WILL kill your job, regardless
of whether it is just about to finish, or not. Similarly, if you requested 4 Gb or RAM, but five hours into
your job, the program you are running ends up using 5 Gb or RAM to store a large chunk of data, your
job WILL be killed, immediately.</p>
<p>Thus, it is best to be able to accurately estimate the time and resources a job will
require. You always want to request more time and resources than your job will
actually need, but not too much more. A large part of getting good at computing
in a shared cluster resource is gaining experience in predicting how long different jobs will
run, and how much RAM they will require. Later we will describe how the records of your
jobs, stored by the job scheduler, can be accessed and organized to aid in predicting
the resource demand of future jobs.</p>
</div>
<div id="learning-about-the-resources-on-your-hpcc" class="section level2">
<h2><span class="header-section-number">8.3</span> Learning about the resources on your HPCC</h2>
<p>Most computing clusters have a web page that describes the configuration of the system
and the resources available on it. However, you can use various SLURM commands to
learn about those resources as well, and also to gain information about which
of the resources are in use and how many users are waiting to use them.</p>
<p>Requests for resources, and for <em>information</em> about the computing cluster, are made to the job scheduler
using a few basic commands. As said, we will focus on the commands available in a SLURM-based system.
In a later section, (after a discussion of installing software that you might need on your cluster)
we will more fully cover the commands used to <em>launch</em>, <em>schedule</em>, and manage jobs. Here
we will first explore the SLURM commands that
you can use to “get to know” your cluster.</p>
<p>All SLURM commands begin with an <code>s</code>, and all SLURM systems support the <code>sinfo</code> command
that gives you information about the cluster’s nodes and their status (whether they are currently
running jobs or not.) On your cluster, <code>man sinfo</code> will tell you about this command.
On the <code>Sedna</code> cluster, which just got installed and therefore does not
have many users, we see:</p>
<div class="sourceCode" id="cb234"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb234-1" data-line-number="1"><span class="ex">%</span> sinfo</a>
<a class="sourceLine" id="cb234-2" data-line-number="2"><span class="ex">PARTITION</span> AVAIL  TIMELIMIT  NODES  STATE NODELIST</a>
<a class="sourceLine" id="cb234-3" data-line-number="3"><span class="ex">nodes*</span>       up   infinite     28   idle node[01-28]</a>
<a class="sourceLine" id="cb234-4" data-line-number="4"><span class="ex">himem</span>        up   infinite      1  alloc himem01</a>
<a class="sourceLine" id="cb234-5" data-line-number="5"><span class="ex">himem</span>        up   infinite      2   idle himem[02-03]</a></code></pre></div>
<p>which tells us that there are two <em>partitions</em> (collection of nodes)
named <code>nodes</code> and <code>himem</code>. The <code>himem</code> partition has one node which is
currently allocated to a job (<code>STATE</code> = <code>alloc</code>), and two more that
are not currently allocated to jobs. The <code>himem</code> partition holds machines
with lots of RAM memory for tasks like sequence assembly (and, when I ran that
command, one of the nodes in <code>himem</code> was busy assembling the genome of
some interesting sea creature [need to ask Krista again what the hell it was…].
It also has 28 compute nodes in the
<code>nodes</code> partition that are free. This is a very small cluster.</p>
<p>If you do the same command on SUMMIT you get many more lines of output. There
are many more different partitions, and there is a lot of information about how many
nodes are in each:</p>
<div class="sourceCode" id="cb235"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb235-1" data-line-number="1"><span class="ex">%</span> sinfo</a>
<a class="sourceLine" id="cb235-2" data-line-number="2"><span class="ex">PARTITION</span>        AVAIL  TIMELIMIT  NODES  STATE NODELIST</a>
<a class="sourceLine" id="cb235-3" data-line-number="3"><span class="ex">shas*</span>               up 1-00:00:00      2 drain* shas[0136-0137]</a>
<a class="sourceLine" id="cb235-4" data-line-number="4"><span class="ex">shas*</span>               up 1-00:00:00      2  down* shas[0404,0506]</a>
<a class="sourceLine" id="cb235-5" data-line-number="5"><span class="ex">shas*</span>               up 1-00:00:00      1  drain shas0101</a>
<a class="sourceLine" id="cb235-6" data-line-number="6"><span class="ex">shas*</span>               up 1-00:00:00      3   resv shas[0102,0521,0853]</a>
<a class="sourceLine" id="cb235-7" data-line-number="7"><span class="ex">shas*</span>               up 1-00:00:00     74    mix shas[0125,0130,0133,0138,0141,0149,0156,0158,0218,0222,0229,0236-0237,0240-0241,0243,0246,0255,0303,0311,0314,0322,0335,0337,0341,0343,0351,0357,0402,0411,0413,0415-0416,0418,0423,0428,0432-0433,0435-0436,0440,0452,0455-0456,0459,0501,0504,0514,0522-0524,0526-0527,0556,0608,0611-0613,0615-0616,0631,0637,0801,0810-0811,0815,0834-0835,0850,0855,0907-0909,0921]</a>
<a class="sourceLine" id="cb235-8" data-line-number="8"><span class="ex">shas*</span>               up 1-00:00:00    359  alloc shas[0103-0124,0126-0129,0131-0132,0134-0135,0139-0140,0142-0148,0150-0155,0157,0159-0160,0201-0217,0219-0221,0223-0225,0227-0228,0230-0235,0238-0239,0242,0247-0254,0256-0260,0301-0302,0304-0310,0312-0313,0315-0321,0323-0334,0336,0338-0340,0342,0344-0350,0352-0356,0358-0360,0401,0405-0410,0412,0414,0417,0419-0422,0424-0427,0429-0431,0434,0437-0439,0442-0451,0453-0454,0457-0458,0460,0502-0503,0505,0507-0513,0515-0520,0525,0528-0555,0560,0605-0607,0609,0614,0617-0630,0632-0636,0638-0650,0652-0664,0802-0809,0812-0814,0816-0833,0836-0849,0851-0852,0854,0856-0860,0901-0906,0910-0913,0915-0920,0922-0932]</a>
<a class="sourceLine" id="cb235-9" data-line-number="9"><span class="ex">shas*</span>               up 1-00:00:00     11   idle shas[0226,0244-0245,0403,0441,0557-0559,0610,0651,0914]</a>
<a class="sourceLine" id="cb235-10" data-line-number="10"><span class="ex">shas-testing</span>        up   infinite      2 drain* shas[0136-0137]</a>
<a class="sourceLine" id="cb235-11" data-line-number="11"><span class="ex">shas-testing</span>        up   infinite      2  down* shas[0404,0506]</a>
<a class="sourceLine" id="cb235-12" data-line-number="12"><span class="ex">shas-testing</span>        up   infinite      1  drain shas0101</a>
<a class="sourceLine" id="cb235-13" data-line-number="13"><span class="ex">shas-testing</span>        up   infinite      3   resv shas[0102,0521,0853]</a>
<a class="sourceLine" id="cb235-14" data-line-number="14"><span class="ex">shas-testing</span>        up   infinite     74    mix shas[0125,0130,0133,0138,0141,0149,0156,0158,0218,0222,0229,0236-0237,0240-0241,0243,0246,0255,0303,0311,0314,0322,0335,0337,0341,0343,0351,0357,0402,0411,0413,0415-0416,0418,0423,0428,0432-0433,0435-0436,0440,0452,0455-0456,0459,0501,0504,0514,0522-0524,0526-0527,0556,0608,0611-0613,0615-0616,0631,0637,0801,0810-0811,0815,0834-0835,0850,0855,0907-0909,0921]</a>
<a class="sourceLine" id="cb235-15" data-line-number="15"><span class="ex">shas-testing</span>        up   infinite    359  alloc shas[0103-0124,0126-0129,0131-0132,0134-0135,0139-0140,0142-0148,0150-0155,0157,0159-0160,0201-0217,0219-0221,0223-0225,0227-0228,0230-0235,0238-0239,0242,0247-0254,0256-0260,0301-0302,0304-0310,0312-0313,0315-0321,0323-0334,0336,0338-0340,0342,0344-0350,0352-0356,0358-0360,0401,0405-0410,0412,0414,0417,0419-0422,0424-0427,0429-0431,0434,0437-0439,0442-0451,0453-0454,0457-0458,0460,0502-0503,0505,0507-0513,0515-0520,0525,0528-0555,0560,0605-0607,0609,0614,0617-0630,0632-0636,0638-0650,0652-0664,0802-0809,0812-0814,0816-0833,0836-0849,0851-0852,0854,0856-0860,0901-0906,0910-0913,0915-0920,0922-0932]</a>
<a class="sourceLine" id="cb235-16" data-line-number="16"><span class="ex">shas-testing</span>        up   infinite     11   idle shas[0226,0244-0245,0403,0441,0557-0559,0610,0651,0914]</a>
<a class="sourceLine" id="cb235-17" data-line-number="17"><span class="ex">shas-interactive</span>    up   infinite      2 drain* shas[0136-0137]</a>
<a class="sourceLine" id="cb235-18" data-line-number="18"><span class="ex">shas-interactive</span>    up   infinite      2  down* shas[0404,0506]</a>
<a class="sourceLine" id="cb235-19" data-line-number="19"><span class="ex">shas-interactive</span>    up   infinite      1  drain shas0101</a>
<a class="sourceLine" id="cb235-20" data-line-number="20"><span class="ex">shas-interactive</span>    up   infinite      3   resv shas[0102,0521,0853]</a>
<a class="sourceLine" id="cb235-21" data-line-number="21"><span class="ex">shas-interactive</span>    up   infinite     74    mix shas[0125,0130,0133,0138,0141,0149,0156,0158,0218,0222,0229,0236-0237,0240-0241,0243,0246,0255,0303,0311,0314,0322,0335,0337,0341,0343,0351,0357,0402,0411,0413,0415-0416,0418,0423,0428,0432-0433,0435-0436,0440,0452,0455-0456,0459,0501,0504,0514,0522-0524,0526-0527,0556,0608,0611-0613,0615-0616,0631,0637,0801,0810-0811,0815,0834-0835,0850,0855,0907-0909,0921]</a>
<a class="sourceLine" id="cb235-22" data-line-number="22"><span class="ex">shas-interactive</span>    up   infinite    359  alloc shas[0103-0124,0126-0129,0131-0132,0134-0135,0139-0140,0142-0148,0150-0155,0157,0159-0160,0201-0217,0219-0221,0223-0225,0227-0228,0230-0235,0238-0239,0242,0247-0254,0256-0260,0301-0302,0304-0310,0312-0313,0315-0321,0323-0334,0336,0338-0340,0342,0344-0350,0352-0356,0358-0360,0401,0405-0410,0412,0414,0417,0419-0422,0424-0427,0429-0431,0434,0437-0439,0442-0451,0453-0454,0457-0458,0460,0502-0503,0505,0507-0513,0515-0520,0525,0528-0555,0560,0605-0607,0609,0614,0617-0630,0632-0636,0638-0650,0652-0664,0802-0809,0812-0814,0816-0833,0836-0849,0851-0852,0854,0856-0860,0901-0906,0910-0913,0915-0920,0922-0932]</a>
<a class="sourceLine" id="cb235-23" data-line-number="23"><span class="ex">shas-interactive</span>    up   infinite     11   idle shas[0226,0244-0245,0403,0441,0557-0559,0610,0651,0914]</a>
<a class="sourceLine" id="cb235-24" data-line-number="24"><span class="ex">sgpu</span>                up 1-00:00:00      1   resv sgpu0501</a>
<a class="sourceLine" id="cb235-25" data-line-number="25"><span class="ex">sgpu</span>                up 1-00:00:00      1  alloc sgpu0502</a>
<a class="sourceLine" id="cb235-26" data-line-number="26"><span class="ex">sgpu</span>                up 1-00:00:00      9   idle sgpu[0101-0102,0201-0202,0301-0302,0401-0402,0801]</a>
<a class="sourceLine" id="cb235-27" data-line-number="27"><span class="ex">sgpu-testing</span>        up   infinite      1   resv sgpu0501</a>
<a class="sourceLine" id="cb235-28" data-line-number="28"><span class="ex">sgpu-testing</span>        up   infinite      1  alloc sgpu0502</a>
<a class="sourceLine" id="cb235-29" data-line-number="29"><span class="ex">sgpu-testing</span>        up   infinite      9   idle sgpu[0101-0102,0201-0202,0301-0302,0401-0402,0801]</a>
<a class="sourceLine" id="cb235-30" data-line-number="30"><span class="ex">sknl</span>                up 1-00:00:00      1  drain sknl0710</a>
<a class="sourceLine" id="cb235-31" data-line-number="31"><span class="ex">sknl</span>                up 1-00:00:00      1   resv sknl0706</a>
<a class="sourceLine" id="cb235-32" data-line-number="32"><span class="ex">sknl</span>                up 1-00:00:00     18  alloc sknl[0701-0705,0707-0709,0711-0720]</a>
<a class="sourceLine" id="cb235-33" data-line-number="33"><span class="ex">sknl-testing</span>        up   infinite      1  drain sknl0710</a>
<a class="sourceLine" id="cb235-34" data-line-number="34"><span class="ex">sknl-testing</span>        up   infinite      1   resv sknl0706</a>
<a class="sourceLine" id="cb235-35" data-line-number="35"><span class="ex">sknl-testing</span>        up   infinite     18  alloc sknl[0701-0705,0707-0709,0711-0720]</a>
<a class="sourceLine" id="cb235-36" data-line-number="36"><span class="ex">smem</span>                up 7-00:00:00      1   drng smem0201</a>
<a class="sourceLine" id="cb235-37" data-line-number="37"><span class="ex">smem</span>                up 7-00:00:00      4  alloc smem[0101,0301,0401,0501]</a>
<a class="sourceLine" id="cb235-38" data-line-number="38"><span class="ex">ssky</span>                up 1-00:00:00      1   drng ssky0944</a>
<a class="sourceLine" id="cb235-39" data-line-number="39"><span class="ex">ssky</span>                up 1-00:00:00      1    mix ssky0952</a>
<a class="sourceLine" id="cb235-40" data-line-number="40"><span class="ex">ssky</span>                up 1-00:00:00      3  alloc ssky[0942-0943,0951]</a>
<a class="sourceLine" id="cb235-41" data-line-number="41"><span class="ex">ssky-preemptable</span>    up 1-00:00:00      1   drng ssky0944</a>
<a class="sourceLine" id="cb235-42" data-line-number="42"><span class="ex">ssky-preemptable</span>    up 1-00:00:00      1    mix ssky0952</a>
<a class="sourceLine" id="cb235-43" data-line-number="43"><span class="ex">ssky-preemptable</span>    up 1-00:00:00      9  alloc ssky[0933-0934,0937-0940,0942-0943,0951]</a>
<a class="sourceLine" id="cb235-44" data-line-number="44"><span class="ex">ssky-preemptable</span>    up 1-00:00:00      9   idle ssky[0935-0936,0941,0945-0950]</a>
<a class="sourceLine" id="cb235-45" data-line-number="45"><span class="ex">ssky-ucb-aos</span>        up 7-00:00:00      6  alloc ssky[0933-0934,0937-0940]</a>
<a class="sourceLine" id="cb235-46" data-line-number="46"><span class="ex">ssky-ucb-aos</span>        up 7-00:00:00      3   idle ssky[0935-0936,0941]</a>
<a class="sourceLine" id="cb235-47" data-line-number="47"><span class="ex">ssky-csu-mbp</span>        up 7-00:00:00      1   idle ssky0945</a>
<a class="sourceLine" id="cb235-48" data-line-number="48"><span class="ex">ssky-csu-asb</span>        up 7-00:00:00      1   idle ssky0946</a>
<a class="sourceLine" id="cb235-49" data-line-number="49"><span class="ex">ssky-csu-rsp</span>        up 7-00:00:00      4   idle ssky[0947-0950]</a></code></pre></div>
<p>Yikes! That is a lot of info. The numbers that you see on the ends of the lines there
are node numbers.</p>
<p>If you wanted to see information about each node in the <code>shas</code> partition, you could
print long information for each node like this:</p>
<div class="sourceCode" id="cb236"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb236-1" data-line-number="1"><span class="ex">%</span> sinfo -l -N</a></code></pre></div>
<p>On Summit, that creates almost 1500 lines of output. Some nodes are listed several times
because they belong to different partitions. To look at results for just the standard compute
parition (<code>shas</code>: 380 nodes with Intel Xeon Haswell processors) you can use</p>
<div class="sourceCode" id="cb237"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb237-1" data-line-number="1"><span class="ex">%</span> sinfo -l -n -p shas</a></code></pre></div>
<p>Try that command. (Or try a similar command with an appropriate partition name on your
own cluster.) If you want to see explicitly how many cores are available vs allocated
on each node, how much total memory each node has, and how much of that total
memory is free, in that partition you can do:</p>
<div class="sourceCode" id="cb238"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb238-1" data-line-number="1"><span class="ex">%</span> sinfo -N -p shas -O nodelist,cpusstate,memory,allocmem,freemem</a></code></pre></div>
<p>The top part of the output from that command on SUMMIT looks like:</p>
<pre><code>NODELIST            CPUS(A/I/O/T)       MEMORY              ALLOCMEM            FREE_MEM            
shas0101            0/0/24/24           116368              0                   125156              
shas0102            0/24/0/24           116368              0                   112325              
shas0103            24/0/0/24           116368              116352              87228               
shas0104            24/0/0/24           116368              116352              80769  </code></pre>
<p>This says <code>shas0101</code> has 24 CPUs that are <em>Out</em> (not functional at this point).
<code>shas0102</code>, on the other hand, has 24 CPUs that are <em>Idle</em>, while <code>shas0103</code> has
24 CPUs that are allocated, and so forth. (In our parlance, here,
CPU is being used to mean “core”). All of the nodes have
116 Gb of memory total. Most of them have about that much memory allocated to
the jobs they are running.</p>
<p>We can throw down some <code>awk</code> to count the total number of available cores
(and the total number of all the cores):</p>
<div class="sourceCode" id="cb240"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb240-1" data-line-number="1"><span class="ex">%</span> sinfo -N -p shas -O  nodelist,cpusstate <span class="kw">|</span> <span class="fu">awk</span> -F<span class="st">&quot;/&quot;</span> <span class="st">&#39;{avail+=$2; tots+=$4} END {print &quot;Out of&quot;, tots, &quot;cores, there are&quot;, avail, &quot;available&quot;}&#39;</span> </a>
<a class="sourceLine" id="cb240-2" data-line-number="2"><span class="ex">Out</span> of 10848 cores, there are 331 available</a></code></pre></div>
<p>That could explain why it can be hard to get time on the supercomputer: at this time, only about 3% of the cores in the system
are idle, waiting for jobs to go on them.</p>
<p>If you want to see how many jobs are in line, waiting to be launched, you can use the
<code>squeue</code> command. Simply issuing the command <code>squeue</code> will give a (typically very long)
list of all jobs that are either currently running or are in the queue,
waiting to be launched. This is worth doing in order to see how many
different people are using (or waiting to use) the resources.</p>
<p>If we run the command on Sedna, we see that (like we saw before)
only one job is currently running:</p>
<div class="sourceCode" id="cb241"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb241-1" data-line-number="1"><span class="ex">%</span> squeue</a>
<a class="sourceLine" id="cb241-2" data-line-number="2">             <span class="ex">JOBID</span> PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)</a>
<a class="sourceLine" id="cb241-3" data-line-number="3">               <span class="ex">160</span>     himem MaSuRCA_   ggoetz  R 34-15:56:16      1 himem01</a></code></pre></div>
<p>Holy cow! Looking at the TIME column you can see that the genome assembly job
has been running for over 34 days! (The time format is DAYS-Hours:Minutes:Seconds).</p>
<p>The output can be filtered to just PENDING
or RUNNING jobs using the <code>-t</code> option. So,</p>
<div class="sourceCode" id="cb242"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb242-1" data-line-number="1"><span class="ex">squeue</span> -t PENDING</a></code></pre></div>
<p>lists all jobs waiting to be launched. If you wanted to see jobs that a particular
user (like yourself) has running or pending, you can use the <code>-u</code> option. For example,</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb243-1" data-line-number="1"><span class="ex">squeue</span> -u eriq@colostate.edu</a></code></pre></div>
<p>In fact that is such a worthwhile command that it is worth creating an alias
for it by adding a line such as the following to your <code>~/.bashrc</code>:</p>
<div class="sourceCode" id="cb244"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb244-1" data-line-number="1"><span class="bu">alias</span> myjobs=<span class="st">&#39;squeue -u eriq@colostate.edu&#39;</span></a></code></pre></div>
<p>Then, typing <code>myjobs</code> shows all of the jobs that you have running, or waiting to be launched
on the cluster.</p>
</div>
<div id="installing-software-on-an-hpcc" class="section level2">
<h2><span class="header-section-number">8.4</span> Installing Software on an HPCC</h2>
<p>In order to do anything useful on a cluster you will need to have software programs
for analyzing data. As we have seen, a lot of the nuts and bolts of writing command
lines uses utilities that are found on every Unix computer. However almost always
your bioinformatic analyses will require programs or software that
do not come “standard” with Unix. For example, the specialized programs for
sequence assembly and alignment will have to be <em>installed</em> on the cluster
in order for you to be able to use them.</p>
<p>It turns out that installing software on a Unix machine (or cluster) has not
always been a particularly easy thing to do for a number or reasons. First, for
a long time, Unix software was largely distributed in the form of <em>source code</em>: the
actual programming code (text) written by the developers that describes the actions
that a program takes. Such computer code cannot be run directly, it first must be
<em>compiled</em> into a <em>binary</em> or <em>executable</em> program. Doing this can be a challenging
process. First, computer code compilation can be very time consuming (if you use R
on Linux, and install all your packages from CRAN—which requires compilation—you
will know that!). Secondly, vexing errors and failures can occur when the
compiler or the computer architecture is in conflict with the program code. (I have lost
entire days trying to solve compiling problems). On top of that, in order to run, most
programs do not operate in a standalone fashion; rather, while a program is running, it
typically depends on computer code and routines that must be stored in separate <em>libraries</em> on your
Unix computer. These libraries are known as program <em>dependencies</em>. So, installing a program
requires not just installing the program itself, but also ensuring that the program’s dependencies
are installed <em>and</em> that the program knows where they are installed. As if that were not enough,
the dependencies of some programs can conflict (be incompatible) with the dependencies of other
programs, and particular versions of a program might require particular versions of
the dependencies. Additionally, some versions of some programs might not work with particular
versions (types of chips) of some computer systems. Finally, most systems for installing
software that were in place on Unix
machines a decade ago required that whoever was installing software have <em>administrative privileges</em>
on the computer. On an HPCC, none of the typical users have administrative privileges which are,
as you might guess, reserved for the system administrators.</p>
<p>For all the reasons above, installing software on an HPCC used to be a harrowing
affair: you either had to be fluent in compilers and libraries to do it yourself in your
home directory or you had to beg your system administrator. (Though our cluster computing
sysadmins at NMFS are wonderful, that is not always the case…<a href="https://www.wiw.org/~chris/archive/dilbert/">see Dilbert</a>).
On HPCC’s the system administrators have to contend with requests from multiple users
for different software and different versions. They solve this (somewhat headachey)
problem by installing software into separate “compartments” that allow different software
and versions to be maintained on the system without all of it being accessible at once.
Doing so, they create <em>modules</em> of software. This is discussed in a few more sections.</p>
<p>Today, however, a large group of motivated people have created a software management
system called Miniconda that tries to solve many of the problems encountered in maintaining
software on a computer system. First, Miniconda maintains a huge repository of
programs that are already <em>pre-compiled</em> for a number of different chip architectures, so that
programs can usually be installed without the time-consuming compiling process. Second, the repository
maintains critical information on the dependencies for each software program, and about
conflicts and incompatibilities between different versions of programs, architectures and
dependencies. Third, the Miniconda system is built from the ground up to make it easy to maintain
separate software <em>environments</em> on your system. These different environments have different
software programs or different versions of different software programs. Such an approach
was originally used so developers could use a single computer to test any new code
they had written in a number of different
computing environments; however, it has become an incredibly valuable tool for ensuring
that your analyses are reproducible: you can give people not just the data and scripts that you
used for the analysis, but also the computing/software environment (with all the same
software versions) that you used for the analysis. And, finally, all of this
can be done with Miniconda without have administrative privileges. Effectively,
Miniconda manages all these software programs and dependencies <em>within your home directory</em>.</p>
<div id="miniconda" class="section level3">
<h3><span class="header-section-number">8.4.1</span> Miniconda</h3>
<p>We will first walk you through a few steps with Miniconda to install some
bioinformatic software into an environment on your cluster. After that we will discuss more about
the underlying philosophy of Miniconda, and how it is operating.</p>
<div id="installing-or-updating-miniconda" class="section level4">
<h4><span class="header-section-number">8.4.1.1</span> Installing or updating Miniconda</h4>
<ol style="list-style-type: decimal">
<li><p>To do installation of software, you probably should not be on SUMMIT’s login nodes.
They offer “compile nodes” that should be suitable for installing with Miniconda. So, do this:</p>
<div class="sourceCode" id="cb245"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb245-1" data-line-number="1"><span class="fu">ssh</span> scompile</a></code></pre></div></li>
<li><p>If you are on Hummingbird, be sure to get a <code>bash</code> shell before doing anything else, by typing <code>bash</code>
(if you have not already set <code>bash</code> as your default shell (see the previous chapter)).</p></li>
<li><p>First, check if you have Miniconda. Update it if you do, and install it if you don’t:</p>
<div class="sourceCode" id="cb246"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb246-1" data-line-number="1"><span class="co"># just type conda at the command line:</span></a>
<a class="sourceLine" id="cb246-2" data-line-number="2"><span class="ex">conda</span></a></code></pre></div>
<p>If you see some help information, then you already have Miniconda (or Anaconda) and you
should merely update it with:</p>
<div class="sourceCode" id="cb247"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb247-1" data-line-number="1"><span class="ex">conda</span> update conda</a></code></pre></div>
<p>If you get an error telling you that your computer does not know about a command, <code>conda</code>, then
you do not have Miniconda and you must install it. You do that by downloading the Miniconda package
with <code>wget</code> and then running the Miniconda installer, like this:</p>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb248-1" data-line-number="1"><span class="co"># start in your home directory and do the following:</span></a>
<a class="sourceLine" id="cb248-2" data-line-number="2">mkdir conda_install</a>
<a class="sourceLine" id="cb248-3" data-line-number="3">cd conda_install<span class="op">/</span></a>
<a class="sourceLine" id="cb248-4" data-line-number="4">wget https<span class="op">:</span><span class="er">//</span>repo.anaconda.com<span class="op">/</span>miniconda<span class="op">/</span>Miniconda3<span class="op">-</span>latest<span class="op">-</span>Linux<span class="op">-</span>x86_<span class="fl">64.</span>sh</a>
<a class="sourceLine" id="cb248-5" data-line-number="5">chmod u<span class="op">+</span>x Miniconda3<span class="op">-</span>latest<span class="op">-</span>Linux<span class="op">-</span>x86_<span class="fl">64.</span>sh </a>
<a class="sourceLine" id="cb248-6" data-line-number="6">.<span class="op">/</span>Miniconda3<span class="op">-</span>latest<span class="op">-</span>Linux<span class="op">-</span>x86_<span class="fl">64.</span>sh </a></code></pre></div>
<p>That launches the Miniconda installer (it is a shell script). Follow the prompts and agree
to the license and to the default install location and to initialize conda.
At the end, it tells you to log out of your shell and log back in for changes to take
effect. It turns out that it suffices to do <code>cd ~; source .bash_profile</code></p></li>
</ol>
<p>Once you complete the above, your command prompt should have been changed to something that looks like:</p>
<div class="sourceCode" id="cb249"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb249-1" data-line-number="1"><span class="kw">(</span><span class="ex">base</span><span class="kw">)</span> [<span class="ex">~</span>]<span class="ex">--%</span></a></code></pre></div>
<p>The <code>(base)</code> is telling you that you are in Miniconda’s <em>base</em> environment. Typically you want to keep the
<em>base</em> environment clean of installed software, so we will install software into a new environment.</p>
</div>
<div id="installing-software-into-a-bioinformatics-environment" class="section level4">
<h4><span class="header-section-number">8.4.1.2</span> Installing software into a bioinformatics environment</h4>
<p>If everything went according to plan above, then we are ready to use Miniconda to
install some software for bioinformatics. We will install a few programs that we will
use extensively in the next few weeks: <code>bwa</code>, <code>samtools</code>, and <code>bcftools</code>. We will
install these programs into a conda environment that we will name <code>bioinf</code> (short
for “bioinformatics”). It takes just a single command:</p>
<div class="sourceCode" id="cb250"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb250-1" data-line-number="1"><span class="ex">conda</span> create -n bioinf -c bioconda bwa samtools bcftools</a></code></pre></div>
<p>That should only take a few seconds.</p>
<p>To test that we got the programs we must <em>activate</em> the <code>bioinf</code> environment, and then issue
the commands, <code>bwa</code>, <code>samtools</code>, and <code>bcftools</code>. Each of those should spit back some help
information. If so, that means they are installed correctly! It looks like this:</p>
<div class="sourceCode" id="cb251"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb251-1" data-line-number="1"><span class="ex">conda</span> activate bioinf</a></code></pre></div>
<p>After that you should get a command prompt that starts with <code>(bioinf)</code>, telling you that the
active conda environment is <code>bioinf</code>. Now, try these commands:</p>
<div class="sourceCode" id="cb252"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb252-1" data-line-number="1"><span class="ex">bwa</span></a>
<a class="sourceLine" id="cb252-2" data-line-number="2"><span class="ex">samtools</span></a>
<a class="sourceLine" id="cb252-3" data-line-number="3"><span class="ex">bcftools</span></a></code></pre></div>
</div>
<div id="uninstalling-miniconda-and-its-associated-environments" class="section level4">
<h4><span class="header-section-number">8.4.1.3</span> Uninstalling Miniconda and its associated environments</h4>
<p>It may become necessary at some point to uninstall Miniconda. One important case of this
is if you end up overflowing your home directory with conda-installed software. In this case,
unless you have installed numerous, complex environments, the simplest thing to do is
to “uninstall” Miniconda, reinstall it in a location with fewer hard-drive space constraints,
and then simply recreate the environments you need, as you did originally.</p>
<p>This is actually quite germane to SUMMIT users. The size quota on home directories on SUMMIT is
only 2 Gb, so you can easily fill up your home directory by installing a few conda environments.
To check how much of the hard
drive space allocated to you is in use on SUMMIT, use the <code>curc-quota</code> command. (Check the documentation
for how to check space on other HPCCs, but note that Hummingbird users get 1 TB on their home
directories). Instead of
using your home directory to house your Miniconda software, on SUMMIT you can put it in your
<code>projects</code> storage area. Each user gets more storage (like 250 Gb) in a directory
called <code>/projects/username</code> where <code>username</code> is replaced by your SUMMIT username,
for example: <code>/projects/eriq@colostate.edu</code></p>
<p>To “uninstall” Miniconda, you first must delete the <code>miniconda3</code> directory in your
home directory (if that is where it got installed to). This can take a while. It is done with:</p>
<div class="sourceCode" id="cb253"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb253-1" data-line-number="1"><span class="fu">rm</span> -rf ~/miniconda3</a></code></pre></div>
<p>Then you have to delete the lines between <code># &gt;&gt;&gt;</code> and <code># &lt;&lt;&lt;</code>, wherever they occur in your <code>~/.bashrc</code> and <code>~/bash_profile</code>
files, i.e., you will have to remove all of the lines that look
something like thius:</p>
<div class="sourceCode" id="cb254"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb254-1" data-line-number="1"><span class="co"># &gt;&gt;&gt; conda initialize &gt;&gt;&gt;</span></a>
<a class="sourceLine" id="cb254-2" data-line-number="2"><span class="co"># !! Contents within this block are managed by &#39;conda init&#39; !!</span></a>
<a class="sourceLine" id="cb254-3" data-line-number="3"><span class="va">__conda_setup=</span><span class="st">&quot;</span><span class="va">$(</span><span class="st">&#39;/Users/eriq/miniconda3/bin/conda&#39;</span> <span class="st">&#39;shell.bash&#39;</span> <span class="st">&#39;hook&#39;</span> <span class="op">2&gt;</span> <span class="ex">/dev/null</span><span class="va">)</span><span class="st">&quot;</span></a>
<a class="sourceLine" id="cb254-4" data-line-number="4"><span class="kw">if</span><span class="bu"> [</span> <span class="va">$?</span> <span class="ot">-eq</span> 0<span class="bu"> ]</span>; <span class="kw">then</span></a>
<a class="sourceLine" id="cb254-5" data-line-number="5">    <span class="bu">eval</span> <span class="st">&quot;</span><span class="va">$__conda_setup</span><span class="st">&quot;</span></a>
<a class="sourceLine" id="cb254-6" data-line-number="6"><span class="kw">else</span></a>
<a class="sourceLine" id="cb254-7" data-line-number="7">    <span class="kw">if</span><span class="bu"> [</span> <span class="ot">-f</span> <span class="st">&quot;/Users/eriq/miniconda3/etc/profile.d/conda.sh&quot;</span><span class="bu"> ]</span>; <span class="kw">then</span></a>
<a class="sourceLine" id="cb254-8" data-line-number="8">        <span class="bu">.</span> <span class="st">&quot;/Users/eriq/miniconda3/etc/profile.d/conda.sh&quot;</span></a>
<a class="sourceLine" id="cb254-9" data-line-number="9">    <span class="kw">else</span></a>
<a class="sourceLine" id="cb254-10" data-line-number="10">        <span class="bu">export</span> <span class="va">PATH=</span><span class="st">&quot;/Users/eriq/miniconda3/bin:</span><span class="va">$PATH</span><span class="st">&quot;</span></a>
<a class="sourceLine" id="cb254-11" data-line-number="11">    <span class="kw">fi</span></a>
<a class="sourceLine" id="cb254-12" data-line-number="12"><span class="kw">fi</span></a>
<a class="sourceLine" id="cb254-13" data-line-number="13"><span class="bu">unset</span> <span class="va">__conda_setup</span></a>
<a class="sourceLine" id="cb254-14" data-line-number="14"><span class="co"># &lt;&lt;&lt; conda initialize &lt;&lt;&lt;</span></a></code></pre></div>
<p>After all those conda lines are removed from your <code>~/.bashrc</code> and <code>~/bash_profile</code>, logging out
and logging back in, you should be free from conda and ready to reinstall it in
a different location.</p>
<p>To reinstall miniconda in a different location, just follow the
installation instructions above, but when you are running the
<code>./Miniconda3-latest-Linux-x86_64.sh</code> script, instead of choosing the default
install location, use a location in your project directory. For example, for me, that is:
<code>/projects/eriq@colostate.edu/miniconda3</code>.</p>
<p>Then, recreate the <code>bioinf</code> environment described above.</p>
<p>If you are having fun making environments and you think that you might like to use R on
the cluster, then you might want to make an environment with some bioinformatics software
that also has
the latest version of R on miniconda installed. At the time of writing that was R 3.6.1. So,
do:</p>
<div class="sourceCode" id="cb255"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb255-1" data-line-number="1"><span class="ex">conda</span> create -n binfr -c bioconda bwa samtools bcftools r-base=3.6.1 r-essentials</a></code></pre></div>
<p>That makes an environment called <code>binfr</code> (which turns out to also be <strong>way</strong> easier to type that <code>bioinfr</code>).
The <code>r-essentials</code> in the above command line is the name for a collection of 200 commonly used R packages (including
the <code>tidyverse</code>). This procedure takes a little while, but it is still far less painful than using the
version of R that is installed on SUMMIT with the <code>modules</code> packages, and then trying to build the tidyverse
from source with <code>install.packages()</code>.</p>
</div>
<div id="what-is-miniconda-doing" class="section level4">
<h4><span class="header-section-number">8.4.1.4</span> What is Miniconda doing?</h4>
<p>This is a good question. We won’t go deeply into the specifics, but will skim the
surface of a few topics that can help you understand what Miniconda is doing.</p>
<p>First, Miniconda is downloading programs and their dependencies into the <code>miniconda3</code>
directory. Based on the lists of dependencies and conflicts for each program that is being
installed, it makes a sort of “equation,” which it can “solve” to find the versions of
different programs and libraries that can be installed and which should “play nicely with
one another (and with your specific computer architecture.” While it is solving this
“equation” it is doing so while also doing its best
to optimize features of the programs (like using the latest versions, if possible).
Solving this “equation” is an example of a Boolean Satisfiability problem, which is a known
class of difficult (time-consuming) problems. If you are requesting a lot of programs, and
especially if you do not constrain your request (by demanding a certain version of
the program) then “solving” the request may take a long time. However, when installing
just a few bioinformatics programs it is unlikely to every take too terribly long.</p>
<p>Once miniconda has decided on which versions of which programs and dependencies to install,
it downloads them and then places them into the requested environment (or the active environment
if no environment is specifically requested). If program is installed into an environment, then you
can access that program by activating the environment (i.e. <code>conda activate bioinf</code>). Importantly,
if you don’t activate the environment, you won’t be able to use the programs installed there.
We will see later in writing bioinformatic scripts, you will always have to explicitly
activate a desired conda environment when you run a script on a compute node through the job
scheduler.</p>
<p>The way that Miniconda delivers programs in an environment is by storing all the programs
in a special environment directory (within the <code>miniconda3/envs</code> directory), and then, when
the environment is activated, all that is really happening is that <code>conda</code> is manipulating your
PATH variable (and likely a few others) to include directories with the environment’s directory
within the <code>miniconda3/envs</code> directory. An easy way to see this is simply by
inspecting your path variable while in differnt environments. Here we compare the PATH
variable in the <code>base</code> environment, versus in the <code>bioinf</code> environment, versus in the
<code>binfr</code> environment:</p>
<div class="sourceCode" id="cb256"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb256-1" data-line-number="1"><span class="kw">(</span><span class="ex">base</span><span class="kw">)</span> [<span class="ex">~</span>]<span class="ex">--%</span> echo <span class="va">$PATH</span></a>
<a class="sourceLine" id="cb256-2" data-line-number="2"><span class="ex">/projects/eriq@colostate.edu/miniconda3</span>/bin:<span class="ex">/projects/eriq@colostate.edu/miniconda3</span>/condabin:<span class="ex">/usr/local</span>/bin:/bin:<span class="ex">/usr</span>/bin:<span class="ex">/usr</span>/bin:<span class="ex">/usr/local</span>/sbin:<span class="ex">/usr</span>/sbin:<span class="ex">/home/eriq@colostate.edu</span>/bin:<span class="ex">/home/eriq@colostate.edu/bin</span></a>
<a class="sourceLine" id="cb256-3" data-line-number="3"><span class="kw">(</span><span class="ex">base</span><span class="kw">)</span> [<span class="ex">~</span>]<span class="ex">--%</span> conda activate bioinf</a>
<a class="sourceLine" id="cb256-4" data-line-number="4"><span class="kw">(</span><span class="ex">bioinf</span><span class="kw">)</span> [<span class="ex">~</span>]<span class="ex">--%</span> echo <span class="va">$PATH</span></a>
<a class="sourceLine" id="cb256-5" data-line-number="5"><span class="ex">/projects/eriq@colostate.edu/miniconda3/envs/bioinf</span>/bin:<span class="ex">/projects/eriq@colostate.edu/miniconda3</span>/condabin:<span class="ex">/usr/local</span>/bin:/bin:<span class="ex">/usr</span>/bin:<span class="ex">/usr</span>/bin:<span class="ex">/usr/local</span>/sbin:<span class="ex">/usr</span>/sbin:<span class="ex">/home/eriq@colostate.edu</span>/bin:<span class="ex">/home/eriq@colostate.edu/bin</span></a>
<a class="sourceLine" id="cb256-6" data-line-number="6"><span class="kw">(</span><span class="ex">bioinf</span><span class="kw">)</span> [<span class="ex">~</span>]<span class="ex">--%</span> conda activate binfr</a>
<a class="sourceLine" id="cb256-7" data-line-number="7"><span class="kw">(</span><span class="ex">binfr</span><span class="kw">)</span> [<span class="ex">~</span>]<span class="ex">--%</span> echo <span class="va">$PATH</span></a>
<a class="sourceLine" id="cb256-8" data-line-number="8"><span class="ex">/projects/eriq@colostate.edu/miniconda3/envs/binfr</span>/bin:<span class="ex">/projects/eriq@colostate.edu/miniconda3</span>/condabin:<span class="ex">/usr/local</span>/bin:/bin:<span class="ex">/usr</span>/bin:<span class="ex">/usr</span>/bin:<span class="ex">/usr/local</span>/sbin:<span class="ex">/usr</span>/sbin:<span class="ex">/home/eriq@colostate.edu</span>/bin:<span class="ex">/home/eriq@colostate.edu/bin</span></a></code></pre></div>
</div>
<div id="what-programs-are-available-on-minconda" class="section level4">
<h4><span class="header-section-number">8.4.1.5</span> What programs are available on Minconda?</h4>
<p>There are quite a few programs for multiple platforms. If you are wondering
whether a particular program is available from Miniconda, the easiest first
step is to Google it. For example, search for <code>miniconda bowtie</code>.</p>
<p>You can also search from the command line using <code>conda search</code>. Note that most
bioinformatics programs you will be interested in are available on a conda
<em>channel</em> called <code>bioconda</code>. You probably saw the <code>-c bioconda</code> option
applied to the <code>conda create</code> commands above. That options tells conda to search
the Bioconda channel for programs and packages.</p>
<p>Here, you can try searching for a couple of packages that you might
end up using to analyze genomic data:</p>
<div class="sourceCode" id="cb257"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb257-1" data-line-number="1"><span class="ex">conda</span> search -c bioconda plink</a>
<a class="sourceLine" id="cb257-2" data-line-number="2"></a>
<a class="sourceLine" id="cb257-3" data-line-number="3"><span class="co"># and next:</span></a>
<a class="sourceLine" id="cb257-4" data-line-number="4"></a>
<a class="sourceLine" id="cb257-5" data-line-number="5"><span class="ex">conda</span> search -c bioconda angsd</a></code></pre></div>
</div>
<div id="can-i-add-more-programs-to-an-environment" class="section level4">
<h4><span class="header-section-number">8.4.1.6</span> Can I add more programs to an environment?</h4>
<p>This is a worthwhile question. Imagine that we have been happily working in our <code>bioinf</code> conda environment
for a few months. We have finished all our tasks with <code>bwa</code>, <code>samtools</code>, and <code>bcftools</code>, but perhaps now we
want to analyze some of the data with <code>angsd</code> or <code>plink</code>. Can we add those programs to our
<code>bioinf</code> environment? The short answer is “Yes!”. The steps are easy.</p>
<ol style="list-style-type: decimal">
<li>Activate the environment you wish to add the programs to (i.e. <code>conda activate bioinf</code> for example).</li>
<li><p>Then use <code>conda install</code>. For example to install specific versions of <code>plink</code> and <code>angsd</code> that we
saw above while searching for those packages we might do:</p>
<div class="sourceCode" id="cb258"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb258-1" data-line-number="1"><span class="ex">conda</span> install -c bioconda plink=1.90b6.12 angsd=0.931</a></code></pre></div></li>
</ol>
<p>Now, the longer answer is “Yes, but…” The big “but” there occurs because if different
programs require the same dependencies, but rely on different versions of the dependencies,
installing programs over different commands can cause miniconda to not identify some
incompatibilities between program dependencies. A germane example occurs if you first install
<code>samtools</code> into an environment, and then, after that, you install <code>bcftools</code>, like this:</p>
<div class="sourceCode" id="cb259"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb259-1" data-line-number="1"><span class="ex">conda</span> create -n samtools-first        # create an empty environment</a>
<a class="sourceLine" id="cb259-2" data-line-number="2"><span class="ex">conda</span> activate samtools-first         # activate the environment</a>
<a class="sourceLine" id="cb259-3" data-line-number="3"><span class="ex">conda</span> install -c bioconda samtools    # install samtools</a>
<a class="sourceLine" id="cb259-4" data-line-number="4"><span class="ex">conda</span> install -c bioconda bcftools    # install bcftools</a>
<a class="sourceLine" id="cb259-5" data-line-number="5"><span class="ex">bcftools</span>                              # try running bcftools</a></code></pre></div>
<p>When you try running the last line, <code>bcftools</code> barfs on you like so:</p>
<div class="sourceCode" id="cb260"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb260-1" data-line-number="1"><span class="ex">bcftools</span>: error while loading shared libraries: libcrypto.so.1.0.0: cannot open shared object file: No such file or directory</a></code></pre></div>
<p>So, often, installing extra programs does not create problems, but it can. If you find yourself battling
errors from conda-installed programs, see if you can correct that by creating a new environment and installing all the programs
you want at the same time, in one fell swoop, using <code>conda create</code>, as in:</p>
<div class="sourceCode" id="cb261"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb261-1" data-line-number="1"><span class="ex">conda</span> create -n binfr -c bioconda bwa samtools bcftools r-base=3.6.1 r-essentials</a></code></pre></div>
</div>
<div id="exporting-environments" class="section level4">
<h4><span class="header-section-number">8.4.1.7</span> Exporting environments</h4>
<p>In our introduction to Miniconda, we mentioned that it is a great boon to
reproducibility. Clearly, your analyses will be more reproducible if it is
easier for others to install software to repeat your analyses. However, Miniconda
takes that one step further, allowing you to generate a list of
the specific versions of
all software and dependencies in a conda environment. This list is a complete
record of your environment, and, supplied to conda, it is a specification of
exactly how to recreate that environment.</p>
<p>The process of creating such a list is called <em>exporting</em> the conda
environment. Here we demonstrate its use by exporting the <code>bioinf</code>
environment from SUMMIT to a simple text file. Then we use that text file
to recreate the environment on my laptop.</p>
<div class="sourceCode" id="cb262"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb262-1" data-line-number="1"><span class="co"># on summit:</span></a>
<a class="sourceLine" id="cb262-2" data-line-number="2"><span class="ex">conda</span> activate bioinf          # activate the environment</a>
<a class="sourceLine" id="cb262-3" data-line-number="3"><span class="ex">conda</span> env export               # export the environment</a></code></pre></div>
<p>The last command above just sends the exported environment
to stdout, looking like this:</p>
<pre><code>name: bioinf
channels:
  - bioconda
  - defaults
dependencies:
  - _libgcc_mutex=0.1=main
  - bcftools=1.9=ha228f0b_4
  - bwa=0.7.17=hed695b0_7
  - bzip2=1.0.8=h7b6447c_0
  - ca-certificates=2020.1.1=0
  - curl=7.68.0=hbc83047_0
  - htslib=1.9=ha228f0b_7
  - krb5=1.17.1=h173b8e3_0
  - libcurl=7.68.0=h20c2e04_0
  - libdeflate=1.0=h14c3975_1
  - libedit=3.1.20181209=hc058e9b_0
  - libgcc-ng=9.1.0=hdf63c60_0
  - libssh2=1.8.2=h1ba5d50_0
  - libstdcxx-ng=9.1.0=hdf63c60_0
  - ncurses=6.1=he6710b0_1
  - openssl=1.1.1d=h7b6447c_4
  - perl=5.26.2=h14c3975_0
  - samtools=1.9=h10a08f8_12
  - tk=8.6.8=hbc83047_0
  - xz=5.2.4=h14c3975_4
  - zlib=1.2.11=h7b6447c_3
prefix: /projects/eriq@colostate.edu/miniconda3/envs/bioinf</code></pre>
<p>The format of this information is YAML (Yet Another Markup Language),
(which we saw in the headers of RMarkdown documents, too).</p>
<p>If we stored that output in a file:</p>
<div class="sourceCode" id="cb264"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb264-1" data-line-number="1"><span class="ex">conda</span> env export <span class="op">&gt;</span> bioinf.yml</a></code></pre></div>
<p>And then copied that file to another computer, then we can recreate
the environment on that other computer with:</p>
<div class="sourceCode" id="cb265"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb265-1" data-line-number="1"><span class="ex">conda</span> env create -f bioinf.yml</a></code></pre></div>
<p>That should work fine if the new computer is of the same architecture (i.e., both
are Linux computers, or both are Macs). However, the specific build numbers
referenced in the YAML (i.e. things like the <code>h7b6447c_3</code> part of the program name)
can create problems when installing on other architectures. In that case, we must
export without the build names:</p>
<div class="sourceCode" id="cb266"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb266-1" data-line-number="1"><span class="ex">conda</span> env export --no-builds <span class="op">&gt;</span> bioinf.yml</a></code></pre></div>
<p>Even that might fail if the dependencies differ on different architectures,
in which case you can export just the list of the actual programs that
you reqested be installed, by using the <code>--from-history</code> option. For example:</p>
<div class="sourceCode" id="cb267"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb267-1" data-line-number="1"><span class="ex">%</span> conda env export --from-history</a>
<a class="sourceLine" id="cb267-2" data-line-number="2"><span class="ex">name</span>: bioinf</a>
<a class="sourceLine" id="cb267-3" data-line-number="3"><span class="ex">channels</span>:</a>
<a class="sourceLine" id="cb267-4" data-line-number="4">  <span class="ex">-</span> defaults</a>
<a class="sourceLine" id="cb267-5" data-line-number="5"><span class="ex">dependencies</span>:</a>
<a class="sourceLine" id="cb267-6" data-line-number="6">  <span class="ex">-</span> bwa</a>
<a class="sourceLine" id="cb267-7" data-line-number="7">  <span class="ex">-</span> bcftools</a>
<a class="sourceLine" id="cb267-8" data-line-number="8">  <span class="ex">-</span> samtools</a>
<a class="sourceLine" id="cb267-9" data-line-number="9"><span class="ex">prefix</span>: /projects/eriq@colostate.edu/miniconda3/envs/bioinf</a></code></pre></div>
<p>Though, even that fails, cuz it doesn’t list bioconda in there.</p>
</div>
</div>
<div id="modules" class="section level3">
<h3><span class="header-section-number">8.4.2</span> Modules</h3>
<p>This is if your sys admin has made it easy.</p>
<p>The big point I must make here is that these don’t work like conda environments.
Environments are “all-or-nothing,” whereas modules are “cumulative” and can be
layered atop one another. (If doing so creates conflicts, then the sysadmins have to
figure that out…perhaps this is one reason that modules will often carry older (if not
completely antiquated) versions of software.)</p>
</div>
</div>
<div id="getting-compute-resources-allocated-to-your-jobs-on-an-hpcc" class="section level2">
<h2><span class="header-section-number">8.5</span> Getting compute resources allocated to your jobs on an HPCC</h2>
<div id="interactive-sessions" class="section level3">
<h3><span class="header-section-number">8.5.1</span> Interactive sessions</h3>
<p>Although most of your serious computing on the cluster will occur in <em>batch</em> jobs
that get submitted to the scheduler and run without any further interaction from you, the user,
you will begin our foray into the SLURM job scheduling commands with getting an interactive
shell on a compute node. “Why?” you might ask? Well, before you launch any newly-written
script as a <em>batch</em> job
on the cluster, you really should run through the code inside that script and ensure that
things work as they should. Since your script likely does some heavy computing, you can’t do this on the
head nodes of the cluster.</p>
<p>The standard way to get a shell on a compute node of a cluster is to request it with:</p>
<div class="sourceCode" id="cb268"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb268-1" data-line-number="1"><span class="ex">srun</span> --pty /bin/bash</a></code></pre></div>
<p>This tells slurm to run <code>bash</code> (which is typically found at <code>/bin/bash</code> on all Unix system) within
a pseodo-terminal (the <code>--pty</code> part). If you run this command you will be given a bash shell on a compute
node (usually). Check which node you are on, by following up with:</p>
<div class="sourceCode" id="cb269"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb269-1" data-line-number="1"><span class="fu">hostname</span></a></code></pre></div>
<p>Now, on summit, there is a different way of getting an interactive shell that is recommended. The
sysadmins have written their own script to do it, called <code>sinteractive</code>.
So, you could do:</p>
<div class="sourceCode" id="cb270"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb270-1" data-line-number="1"><span class="ex">sinteractive</span></a></code></pre></div>
<p>Now, try this.</p>
<ol style="list-style-type: decimal">
<li><p>If you are not already in a tmux session, start a new <code>tmux</code> session called <code>inclass</code> on the login node.</p>
<div class="sourceCode" id="cb271"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb271-1" data-line-number="1"><span class="ex">tmux</span> new -s inclass</a></code></pre></div></li>
<li><p>Get a bash shell on a compute node</p>
<div class="sourceCode" id="cb272"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb272-1" data-line-number="1"><span class="ex">sinteractive</span></a>
<a class="sourceLine" id="cb272-2" data-line-number="2"><span class="co"># or</span></a>
<a class="sourceLine" id="cb272-3" data-line-number="3"><span class="ex">srun</span> --pty /bin/bash</a></code></pre></div></li>
<li><p>Check what host you are on with <code>hostname</code></p>
<div class="sourceCode" id="cb273"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb273-1" data-line-number="1"><span class="fu">hostname</span></a></code></pre></div>
Do not proceed if you are on a login node still.</li>
<li><p>Use your alias <code>myjobs</code> (if you have created it) to see what jobs you have running.</p>
<div class="sourceCode" id="cb274"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb274-1" data-line-number="1"><span class="ex">myjobs</span></a>
<a class="sourceLine" id="cb274-2" data-line-number="2"></a>
<a class="sourceLine" id="cb274-3" data-line-number="3"><span class="co"># or, if you don&#39;t have that aliased:</span></a>
<a class="sourceLine" id="cb274-4" data-line-number="4"><span class="ex">squeue</span> -u your_username</a></code></pre></div></li>
<li><p>Navigate to your scratch directory. On SUMMIT there is a symbolic link called <code>scratch</code> in your home directory.</p>
<div class="sourceCode" id="cb275"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb275-1" data-line-number="1"><span class="bu">cd</span> scratch</a></code></pre></div></li>
<li>Make sure you have followed the link in your email to the shared google drive folder I sent you last night.</li>
<li><p>Now, use rclone to list the files in the folder I shared with you:</p>
<div class="sourceCode" id="cb276"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb276-1" data-line-number="1"><span class="ex">rclone</span> ls --drive-shared-with-me gdrive-pers:CSU-con-gen-2020</a></code></pre></div>
Note that <code>gdrive-pers</code> above is the name of my rclone configuration associated with the gmail account where
I got an email (from myself at my CSU account) about the shared folder. You will have to change <code>gdrive-pers</code>
to be appropriate to your config, depending on how you set it up. Note also how you access folders shared with
you using <code>--drive-shared-with-me</code>.</li>
<li><p>After you get the above command to work, you should see:</p>
<div class="sourceCode" id="cb277"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb277-1" data-line-number="1"><span class="ex">416265040</span> tiny-genomic-data.zip</a>
<a class="sourceLine" id="cb277-2" data-line-number="2"><span class="ex">179506752</span> big-fastq-play.zip</a></code></pre></div></li>
<li><p>Use rclone to download <code>tiny-genomic-data.zip</code> (about 416 MB):</p>
<div class="sourceCode" id="cb278"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb278-1" data-line-number="1"><span class="ex">rclone</span> copy -P  --drive-shared-with-me gdrive-pers:CSU-con-gen-2020/tiny-genomic-data.zip ./</a></code></pre></div></li>
<li><p>Unzip the file you just downloaded:</p>
<div class="sourceCode" id="cb279"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb279-1" data-line-number="1"><span class="fu">unzip</span> tiny-genomic-data.zip</a></code></pre></div></li>
<li><p>Change directories into <code>tiny-genomic-data</code></p>
<div class="sourceCode" id="cb280"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb280-1" data-line-number="1"><span class="fu">unzip</span> tiny-genomic-data.zip</a></code></pre></div></li>
<li><p>Download the Chinook salmon genome into a directory called <code>genome</code></p>
<div class="sourceCode" id="cb281"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb281-1" data-line-number="1"> <span class="fu">mkdir</span> genome</a>
<a class="sourceLine" id="cb281-2" data-line-number="2"> <span class="bu">cd</span> genome</a>
<a class="sourceLine" id="cb281-3" data-line-number="3"> <span class="fu">wget</span> ftp://ftp.ncbi.nlm.nih.gov/genomes/genbank/vertebrate_other/Oncorhynchus_tshawytscha/all_assembly_versions/GCA_002872995.1_Otsh_v1.0/GCA_002872995.1_Otsh_v1.0_genomic.fna.gz</a></code></pre></div>
That dude is 760 Mb, but it goes really quickly.</li>
<li><p>Activate your <code>bioinf</code> conda environment to be able to use <code>bwa</code>:</p>
<div class="sourceCode" id="cb282"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb282-1" data-line-number="1"><span class="ex">conda</span> activate bioinf</a></code></pre></div></li>
<li><p>Index the genome with <code>bwa</code> in order to align reads to it:</p>
<div class="sourceCode" id="cb283"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb283-1" data-line-number="1"><span class="ex">bwa</span> index GCA_002872995.1_Otsh_v1.0_genomic.fna.gz </a></code></pre></div>
This will take about 10 to 15 minutes. During that time, do this:
<ul>
<li>If you are doing this within a <code>tmux</code> session, create a new tmux window (<code>cntrl-b c</code>). Note that this shell opens on the login node (the one that you started your tmux session on.)</li>
<li>Rename this new tmux window “browse” (<code>cntrl-b ,</code>)</li>
<li>Navigate to <code>~/scratch/cd tiny-genomic-data/genome/</code> and list the files there. These new files
are being created by bwa.</li>
<li>Switch back to your original tmux window.</li>
</ul></li>
<li>If <code>bwa</code> is still running, ask me some questions. Or maybe I will tell you more about how conda works…</li>
<li><p>If <code>bwa</code> is still running after that, ask me about <em>nested tmux sessions</em>.</p></li>
</ol>
<p>Here is some nice stuff for summarizing all the information from the different runs from the chinook-wgs project:</p>
<div class="sourceCode" id="cb284"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb284-1" data-line-number="1"><span class="ex">qacct</span> -o eriq -b 09271925 -j ml <span class="kw">|</span> <span class="ex">tidy-qacct</span></a></code></pre></div>
<p>Explain scratch space and how clusters are configured with respect to storage, etc.</p>
<p>Strategies—break names up with consistent characters:</p>
<ul>
<li>dashes within population names</li>
<li>underscores for different groups of chromosomes</li>
<li>periods for catenating pairs of pops</li>
</ul>
<p>etc. Basically, it just makes it much easier to split things up
when the time comes.</p>
</div>
</div>
<div id="the-queue-slurmsgeuge" class="section level2">
<h2><span class="header-section-number">8.6</span> The Queue (SLURM/SGE/UGE)</h2>
</div>
<div id="modules-package" class="section level2">
<h2><span class="header-section-number">8.7</span> Modules package</h2>
</div>
<div id="compiling-programs-without-admin-privileges" class="section level2">
<h2><span class="header-section-number">8.8</span> Compiling programs without admin privileges</h2>
<p>Inevitably you will want to use a piece of software that is not available as
a module or is not otherwise installed on they system.</p>
<p>Typically these software programs have a frightful web of dependencies.</p>
<p>Unix/Linux distros typically maintain all these dependencies as libraries or packages
that can be installed using a <code>rpm</code> or <code>yum</code>. However, the simple “plug-and-play” approach
to using these programs requires have administrator privileges so that the software can
be installed in one of the (typically protected) paths in the root (like <code>/usr/bin</code>).</p>
<p>But, you can use these programs to install packages into your home directory. Once you have done
that, you need to let your system know where to look for these packages when it needs them
(i.e., when running a program or <em>linking</em> to it whilst compiling up a program that uses it
as a dependency.</p>
<p>Hoffman2 runs CentOS. Turns out that CentOS uses <code>yum</code> as a package manager.</p>
<p>Let’s see if we can install llvm using yum.</p>
<div class="sourceCode" id="cb285"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb285-1" data-line-number="1"><span class="ex">yum</span> search all llvm <span class="co"># &lt;- this got me to devtoolset-7-all.x86_64 : Package shipping all available toolsets.</span></a>
<a class="sourceLine" id="cb285-2" data-line-number="2"></a>
<a class="sourceLine" id="cb285-3" data-line-number="3"><span class="co"># a little web searching made it look like llvm-toolset-7-5.0.1-4.el7.x86_64.rpm or devtoolset-7-llvm-7.0-5.el7.x86_64.rpm</span></a>
<a class="sourceLine" id="cb285-4" data-line-number="4"><span class="co"># might be what we want.  The first is a dependency of the second...</span></a>
<a class="sourceLine" id="cb285-5" data-line-number="5"><span class="fu">mkdir</span> ~/centos</a></code></pre></div>
<p>Was using instructions at <a href="https://stackoverflow.com/questions/36651091/how-to-install-packages-in-linux-centos-without-root-user-with-automatic-depen">https://stackoverflow.com/questions/36651091/how-to-install-packages-in-linux-centos-without-root-user-with-automatic-depen</a></p>
<p>Couldn’t get yum downloader to download any packages. The whole thing looked like it was going to
be a mess, so I thought I would try with miniconda.</p>
<p>I installed miniconda (python 2.7 version) into <code>/u/nobackup/kruegg/eriq/programs/miniconda/</code> and then did this:</p>
<div class="sourceCode" id="cb286"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb286-1" data-line-number="1"><span class="co"># probably could have listed them all at once, but wanted to watch them go </span></a>
<a class="sourceLine" id="cb286-2" data-line-number="2"><span class="co"># one at a time...</span></a>
<a class="sourceLine" id="cb286-3" data-line-number="3"><span class="ex">conda</span> install numpy</a>
<a class="sourceLine" id="cb286-4" data-line-number="4"><span class="ex">conda</span> install scipy</a>
<a class="sourceLine" id="cb286-5" data-line-number="5"><span class="ex">conda</span> install pandas</a>
<a class="sourceLine" id="cb286-6" data-line-number="6"><span class="ex">conda</span> install numba</a>
<a class="sourceLine" id="cb286-7" data-line-number="7"></a>
<a class="sourceLine" id="cb286-8" data-line-number="8"><span class="co"># those all ran great.</span></a>
<a class="sourceLine" id="cb286-9" data-line-number="9"></a>
<a class="sourceLine" id="cb286-10" data-line-number="10"><span class="ex">conda</span> install pysnptools</a>
<a class="sourceLine" id="cb286-11" data-line-number="11"></a>
<a class="sourceLine" id="cb286-12" data-line-number="12"><span class="co"># that one didn&#39;t find a match, but I found on the web that I should try:</span></a>
<a class="sourceLine" id="cb286-13" data-line-number="13"><span class="ex">conda</span> install -c bioconda pysnptools </a>
<a class="sourceLine" id="cb286-14" data-line-number="14"></a>
<a class="sourceLine" id="cb286-15" data-line-number="15"><span class="co"># that worked!</span></a></code></pre></div>
<p>Also we want to touch briefly on LD_PATH (linking failures—and note that libraries are often
named libxxx.a) and CPATH (for failure to find xxxx.h), etc.</p>
</div>
<div id="job-arrays" class="section level2">
<h2><span class="header-section-number">8.9</span> Job arrays</h2>
<p>Quick note: Redefine IFS to break on TABs so you can have full commands in there.
This is super useful for parsing job-array COMMLINES files.</p>
<pre><code>IFS=$&#39;\t\n&#39;; BOP=($(echo boing | awk &#39;{printf(&quot;first\tsecond\tthird that is long\tfourth\n&quot;);}&#39;)); IFS=$&#39; \t\n&#39;;</code></pre>
<p>Definitely mention the <code>eval</code> keyword in bash for when you want to print
command lines with redirects.</p>
<p>Show the routine for it, and develop a good approach to efficiently
orchestrating redos. If you know the taskIDs of the ones that failed
then it is pretty easy to write an awk script that picks out the
commands and puts them in a new file. Actually, it is probably
better to just cycle over the numbers and use the -t option
to launch each. Then there is now changing the job-ids file.</p>
<p>In fact, I am starting to think that the -t option is better than
putting it into the file.</p>
<p>Question: if you give something on the command line, does that override
the directive in the header of the file? If so, then you don’t even
need to change the file. Note that using the qsub command line options
instead of the directives really opens up a lot of possibilities for
writing useful scripts that are flexible.</p>
<p>Also use short names for the jobs and have a system for naming the
redos (append numbers so you know which round it is, too)
possibly base the name on the ways things failed the first time. Like,
<code>fsttf1</code> = “Fst run for things that failed due to time limits, 1”. Or
structure things so that redos can just be done by invoking it with -t
and the jobid.</p>
</div>
<div id="writing-stdout-and-stderr-to-files" class="section level2">
<h2><span class="header-section-number">8.10</span> Writing stdout and stderr to files</h2>
<p>This is always good to do. Note that <code>stdbuf</code> is super useful here so that
things don’t get buffered super long. (PCAngsd doesn’t seem to write antyhing till
the end…)</p>
</div>
<div id="breaking-stuff-down" class="section level2">
<h2><span class="header-section-number">8.11</span> Breaking stuff down</h2>
<p>It is probably worth talking about how problems can be broken down into
smaller ones. Maybe give an example, and then say that we will be talking about
this for every step of the way in bioinformatic pipelines.</p>
<p>One thing to note—sometimes processes go awry for one reason or another.
When things are in smaller chunks it is not such a huge investment to
re-run it. (Unlike stuff that runs for two weeks before you realize that
it ain’t working right).</p>

</div>
</div>



            </section>

          </div>
        </div>
      </div>
<a href="working-on-remote-servers.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="introduction-to-reproducible-research.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook/js/app.min.js"></script>
<script src="libs/gitbook/js/lunr.js"></script>
<script src="libs/gitbook/js/clipboard.min.js"></script>
<script src="libs/gitbook/js/plugin-search.js"></script>
<script src="libs/gitbook/js/plugin-sharing.js"></script>
<script src="libs/gitbook/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook/js/plugin-bookdown.js"></script>
<script src="libs/gitbook/js/jquery.highlight.js"></script>
<script src="libs/gitbook/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/eriqande/eca-bioinf-handbook/edit/master/1.035-hpc.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/eriqande/eca-bioinf-handbook/commits/master/1.035-hpc.Rmd",
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["eca-bioinf-handbook.pdf", "eca-bioinf-handbook.epub"],
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
