<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 Rstudio and Project-centered Organization | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics</title>
  <meta name="description" content="A book example for a Chapman &amp; Hall book." />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 Rstudio and Project-centered Organization | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A book example for a Chapman &amp; Hall book." />
  <meta name="github-repo" content="yihui/bookdown-crc" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Rstudio and Project-centered Organization | Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics" />
  
  <meta name="twitter:description" content="A book example for a Chapman &amp; Hall book." />
  

<meta name="author" content="Eric C. Anderson" />


<meta name="date" content="2022-02-17" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="introduction-to-reproducible-research.html"/>
<link rel="next" href="version-control.html"/>
<script src="libs/header-attrs/header-attrs.js"></script>
<script src="libs/jquery/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook/css/style.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ECA's Bioinformatics Handbook</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="erics-notes-of-what-he-might-do.html"><a href="erics-notes-of-what-he-might-do.html"><i class="fa fa-check"></i><b>2</b> Eric’s Notes of what he might do</a>
<ul>
<li class="chapter" data-level="2.1" data-path="erics-notes-of-what-he-might-do.html"><a href="erics-notes-of-what-he-might-do.html#table-of-topics"><i class="fa fa-check"></i><b>2.1</b> Table of topics</a></li>
</ul></li>
<li class="part"><span><b>I Part I: Essential Computing Skills</b></span></li>
<li class="chapter" data-level="3" data-path="overview-of-essential-computing-skills.html"><a href="overview-of-essential-computing-skills.html"><i class="fa fa-check"></i><b>3</b> Overview of Essential Computing Skills</a></li>
<li class="chapter" data-level="4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html"><i class="fa fa-check"></i><b>4</b> Essential Unix/Linux Terminal Knowledge</a>
<ul>
<li class="chapter" data-level="4.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#getting-a-bash-shell-on-your-system"><i class="fa fa-check"></i><b>4.1</b> Getting a bash shell on your system</a></li>
<li class="chapter" data-level="4.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#navigating-the-unix-filesystem"><i class="fa fa-check"></i><b>4.2</b> Navigating the Unix filesystem</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#changing-the-working-directory-with-cd"><i class="fa fa-check"></i><b>4.2.1</b> Changing the working directory with <code>cd</code></a></li>
<li class="chapter" data-level="4.2.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#comm-prompt"><i class="fa fa-check"></i><b>4.2.2</b> Updating your command prompt</a></li>
<li class="chapter" data-level="4.2.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#tab-completion-for-paths"><i class="fa fa-check"></i><b>4.2.3</b> TAB-completion for paths</a></li>
<li class="chapter" data-level="4.2.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#listing-the-contents-of-a-directory-with-ls"><i class="fa fa-check"></i><b>4.2.4</b> Listing the contents of a directory with <code>ls</code></a></li>
<li class="chapter" data-level="4.2.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#globbing"><i class="fa fa-check"></i><b>4.2.5</b> Globbing</a></li>
<li class="chapter" data-level="4.2.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#what-makes-a-good-file-name"><i class="fa fa-check"></i><b>4.2.6</b> What makes a good file-name?</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#the-anatomy-of-a-unix-command"><i class="fa fa-check"></i><b>4.3</b> The anatomy of a Unix command</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#anatomy-command"><i class="fa fa-check"></i><b>4.3.1</b> The <code>command</code></a></li>
<li class="chapter" data-level="4.3.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#the-options"><i class="fa fa-check"></i><b>4.3.2</b> The <em>options</em></a></li>
<li class="chapter" data-level="4.3.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#arguments"><i class="fa fa-check"></i><b>4.3.3</b> Arguments</a></li>
<li class="chapter" data-level="4.3.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#getting-information-about-unix-commands"><i class="fa fa-check"></i><b>4.3.4</b> Getting information about Unix commands</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#handling-manipulating-and-viewing-files-and-streams"><i class="fa fa-check"></i><b>4.4</b> Handling, Manipulating, and Viewing files and streams</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#creating-new-directories"><i class="fa fa-check"></i><b>4.4.1</b> Creating new directories</a></li>
<li class="chapter" data-level="4.4.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#fundamental-file-handling-commands"><i class="fa fa-check"></i><b>4.4.2</b> Fundamental file-handling commands</a></li>
<li class="chapter" data-level="4.4.3" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#viewing-files"><i class="fa fa-check"></i><b>4.4.3</b> “Viewing” Files</a></li>
<li class="chapter" data-level="4.4.4" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#redirecting-standard-output-and"><i class="fa fa-check"></i><b>4.4.4</b> Redirecting standard output: <code>&gt;</code> and <code>&gt;&gt;</code></a></li>
<li class="chapter" data-level="4.4.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#stdin-and"><i class="fa fa-check"></i><b>4.4.5</b> stdin, <code>&lt;</code> and <code>|</code></a></li>
<li class="chapter" data-level="4.4.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#stderr"><i class="fa fa-check"></i><b>4.4.6</b> stderr</a></li>
<li class="chapter" data-level="4.4.7" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#symbolic-links"><i class="fa fa-check"></i><b>4.4.7</b> Symbolic links</a></li>
<li class="chapter" data-level="4.4.8" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#file-perm"><i class="fa fa-check"></i><b>4.4.8</b> File Permissions</a></li>
<li class="chapter" data-level="4.4.9" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#editing-text-files-at-the-terminal"><i class="fa fa-check"></i><b>4.4.9</b> Editing text files at the terminal</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#unix-env"><i class="fa fa-check"></i><b>4.5</b> Customizing your Environment</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#appearances-matter"><i class="fa fa-check"></i><b>4.5.1</b> Appearances matter</a></li>
<li class="chapter" data-level="4.5.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#where-are-my-programscommands-at"><i class="fa fa-check"></i><b>4.5.2</b> Where are my programs/commands at?!</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#a-few-more-important-keystrokes"><i class="fa fa-check"></i><b>4.6</b> A Few More Important Keystrokes</a></li>
<li class="chapter" data-level="4.7" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#a-short-list-of-additional-useful-commands."><i class="fa fa-check"></i><b>4.7</b> A short list of additional useful commands.</a></li>
<li class="chapter" data-level="4.8" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#two-important-computing-concepts"><i class="fa fa-check"></i><b>4.8</b> Two important computing concepts</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#compression"><i class="fa fa-check"></i><b>4.8.1</b> Compression</a></li>
<li class="chapter" data-level="4.8.2" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#hashing"><i class="fa fa-check"></i><b>4.8.2</b> Hashing</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="essential-unixlinux-terminal-knowledge.html"><a href="essential-unixlinux-terminal-knowledge.html#unix-quick-study-guide"><i class="fa fa-check"></i><b>4.9</b> Unix: Quick Study Guide</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="shell-programming.html"><a href="shell-programming.html"><i class="fa fa-check"></i><b>5</b> Shell programming</a>
<ul>
<li class="chapter" data-level="5.1" data-path="shell-programming.html"><a href="shell-programming.html#an-example-script"><i class="fa fa-check"></i><b>5.1</b> An example script</a></li>
<li class="chapter" data-level="5.2" data-path="shell-programming.html"><a href="shell-programming.html#the-structure-of-a-bash-script"><i class="fa fa-check"></i><b>5.2</b> The Structure of a Bash Script</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="shell-programming.html"><a href="shell-programming.html#paragraph-before-shebang"><i class="fa fa-check"></i><b>5.2.1</b> A bit more on <code>;</code> and <code>&amp;</code></a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="shell-programming.html"><a href="shell-programming.html#variables"><i class="fa fa-check"></i><b>5.3</b> Variables</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="shell-programming.html"><a href="shell-programming.html#assigning-values-to-variables"><i class="fa fa-check"></i><b>5.3.1</b> Assigning values to variables</a></li>
<li class="chapter" data-level="5.3.2" data-path="shell-programming.html"><a href="shell-programming.html#accessing-values-from-variables"><i class="fa fa-check"></i><b>5.3.2</b> Accessing values from variables</a></li>
<li class="chapter" data-level="5.3.3" data-path="shell-programming.html"><a href="shell-programming.html#what-does-the-shell-do-with-the-value-substituted-for-a-variable"><i class="fa fa-check"></i><b>5.3.3</b> What does the shell do with the value substituted for a variable?</a></li>
<li class="chapter" data-level="5.3.4" data-path="shell-programming.html"><a href="shell-programming.html#quotes-and-var-subs"><i class="fa fa-check"></i><b>5.3.4</b> Double and Single Quotation Marks and Variable Substitution</a></li>
<li class="chapter" data-level="5.3.5" data-path="shell-programming.html"><a href="shell-programming.html#one-useful-fancy-variable-substitution-method"><i class="fa fa-check"></i><b>5.3.5</b> One useful, fancy, variable-substitution method</a></li>
<li class="chapter" data-level="5.3.6" data-path="shell-programming.html"><a href="shell-programming.html#bash-arithmetic"><i class="fa fa-check"></i><b>5.3.6</b> Integer Arithmetic with Shell Variables</a></li>
<li class="chapter" data-level="5.3.7" data-path="shell-programming.html"><a href="shell-programming.html#variable-arrays"><i class="fa fa-check"></i><b>5.3.7</b> Variable arrays</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="shell-programming.html"><a href="shell-programming.html#evaluate-a-command-and-substitute-the-result-on-the-command-line"><i class="fa fa-check"></i><b>5.4</b> Evaluate a command and substitute the result on the command line</a></li>
<li class="chapter" data-level="5.5" data-path="shell-programming.html"><a href="shell-programming.html#groupingcollecting-output-from-multiple-commands-commands-and-commands"><i class="fa fa-check"></i><b>5.5</b> Grouping/Collecting output from multiple commands: <code>(commands)</code> and <code>{ commands; }</code></a></li>
<li class="chapter" data-level="5.6" data-path="shell-programming.html"><a href="shell-programming.html#exit-status"><i class="fa fa-check"></i><b>5.6</b> Exit Status</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="shell-programming.html"><a href="shell-programming.html#combinations-of-exit-statuses"><i class="fa fa-check"></i><b>5.6.1</b> Combinations of exit statuses</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="shell-programming.html"><a href="shell-programming.html#unix-for-loops"><i class="fa fa-check"></i><b>5.7</b> Loops and repetition</a></li>
<li class="chapter" data-level="5.8" data-path="shell-programming.html"><a href="shell-programming.html#more-conditional-evaluation-if-then-else-and-friends"><i class="fa fa-check"></i><b>5.8</b> More Conditional Evaluation: <code>if</code>, <code>then</code>, <code>else</code>, and friends</a></li>
<li class="chapter" data-level="5.9" data-path="shell-programming.html"><a href="shell-programming.html#finallypositional-parameters"><i class="fa fa-check"></i><b>5.9</b> Finally…positional parameters</a></li>
<li class="chapter" data-level="5.10" data-path="shell-programming.html"><a href="shell-programming.html#basename-and-dirname-two-useful-little-utilities"><i class="fa fa-check"></i><b>5.10</b> <code>basename</code> and <code>dirname</code> two useful little utilities</a></li>
<li class="chapter" data-level="5.11" data-path="shell-programming.html"><a href="shell-programming.html#bash-functions"><i class="fa fa-check"></i><b>5.11</b> <code>bash</code> functions</a></li>
<li class="chapter" data-level="5.12" data-path="shell-programming.html"><a href="shell-programming.html#reading-files-line-by-line"><i class="fa fa-check"></i><b>5.12</b> reading files line by line</a></li>
<li class="chapter" data-level="5.13" data-path="shell-programming.html"><a href="shell-programming.html#further-reading"><i class="fa fa-check"></i><b>5.13</b> Further reading</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html"><i class="fa fa-check"></i><b>6</b> Sed, awk, and regular expressions</a>
<ul>
<li class="chapter" data-level="6.1" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#awk"><i class="fa fa-check"></i><b>6.1</b> awk</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#line-cycling-tests-and-actions"><i class="fa fa-check"></i><b>6.1.1</b> Line-cycling, tests and actions</a></li>
<li class="chapter" data-level="6.1.2" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#column-splitting-fields--f-nf-print-ofs-and-begin"><i class="fa fa-check"></i><b>6.1.2</b> Column splitting, fields, <code>-F</code>, <code>$</code>, <code>NF</code>, <code>print</code>, <code>OFS</code> and <code>BEGIN</code></a></li>
<li class="chapter" data-level="6.1.3" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#a-brief-introduction-to-regular-expressions"><i class="fa fa-check"></i><b>6.1.3</b> A brief introduction to regular expressions</a></li>
<li class="chapter" data-level="6.1.4" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#a-variety-of-tests"><i class="fa fa-check"></i><b>6.1.4</b> A variety of tests</a></li>
<li class="chapter" data-level="6.1.5" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#code-in-the-action-blocks"><i class="fa fa-check"></i><b>6.1.5</b> Code in the action blocks</a></li>
<li class="chapter" data-level="6.1.6" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#using-awk-to-assign-to-shell-variables"><i class="fa fa-check"></i><b>6.1.6</b> Using <code>awk</code> to assign to shell variables</a></li>
<li class="chapter" data-level="6.1.7" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#passing-variables-into-awk-with--v"><i class="fa fa-check"></i><b>6.1.7</b> Passing Variables into <code>awk</code> with <code>-v</code></a></li>
<li class="chapter" data-level="6.1.8" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#writing-awk-scripts-in-files"><i class="fa fa-check"></i><b>6.1.8</b> Writing awk scripts in files</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="sed-awk-and-regular-expressions.html"><a href="sed-awk-and-regular-expressions.html#sed"><i class="fa fa-check"></i><b>6.2</b> sed</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html"><i class="fa fa-check"></i><b>7</b> Working on remote servers</a>
<ul>
<li class="chapter" data-level="7.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#accessing-remote-computers"><i class="fa fa-check"></i><b>7.1</b> Accessing remote computers</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#windows"><i class="fa fa-check"></i><b>7.1.1</b> Windows</a></li>
<li class="chapter" data-level="7.1.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#hummingbird"><i class="fa fa-check"></i><b>7.1.2</b> Hummingbird</a></li>
<li class="chapter" data-level="7.1.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#summit"><i class="fa fa-check"></i><b>7.1.3</b> Summit</a></li>
<li class="chapter" data-level="7.1.4" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#sedna"><i class="fa fa-check"></i><b>7.1.4</b> Sedna</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#transferring-files-to-remote-computers"><i class="fa fa-check"></i><b>7.2</b> Transferring files to remote computers</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#sftp-and-several-systems-that-use-it"><i class="fa fa-check"></i><b>7.2.1</b> <code>sftp</code> and several systems that use it</a></li>
<li class="chapter" data-level="7.2.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#git"><i class="fa fa-check"></i><b>7.2.2</b> git</a></li>
<li class="chapter" data-level="7.2.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#globus"><i class="fa fa-check"></i><b>7.2.3</b> Globus</a></li>
<li class="chapter" data-level="7.2.4" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#interfacing-with-the-cloud"><i class="fa fa-check"></i><b>7.2.4</b> Interfacing with “The Cloud”</a></li>
<li class="chapter" data-level="7.2.5" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#get-seqs"><i class="fa fa-check"></i><b>7.2.5</b> Getting files from a sequencing center</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#tmux"><i class="fa fa-check"></i><b>7.3</b> <code>tmux</code>: the terminal multiplexer</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#an-analogy-for-how-tmux-works"><i class="fa fa-check"></i><b>7.3.1</b> An analogy for how <code>tmux</code> works</a></li>
<li class="chapter" data-level="7.3.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#first-steps-with-tmux"><i class="fa fa-check"></i><b>7.3.2</b> First steps with <code>tmux</code></a></li>
<li class="chapter" data-level="7.3.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#further-steps-with-tmux"><i class="fa fa-check"></i><b>7.3.3</b> Further steps with <code>tmux</code></a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#tmux-for-mac-users"><i class="fa fa-check"></i><b>7.4</b> <code>tmux</code> for Mac users</a></li>
<li class="chapter" data-level="7.5" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#installing-software-on-an-hpcc"><i class="fa fa-check"></i><b>7.5</b> Installing Software on an HPCC</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#modules"><i class="fa fa-check"></i><b>7.5.1</b> Modules</a></li>
<li class="chapter" data-level="7.5.2" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#miniconda"><i class="fa fa-check"></i><b>7.5.2</b> Miniconda</a></li>
<li class="chapter" data-level="7.5.3" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#installing-java-programs"><i class="fa fa-check"></i><b>7.5.3</b> Installing Java Programs</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#vim-its-time-to-get-serious-with-text-editing"><i class="fa fa-check"></i><b>7.6</b> <code>vim</code>: it’s time to get serious with text editing</a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="working-on-remote-servers.html"><a href="working-on-remote-servers.html#using-neovim-and-nvim-r-and-tmux-to-use-r-well-on-the-cluster"><i class="fa fa-check"></i><b>7.6.1</b> Using neovim and Nvim-R and tmux to use R well on the cluster</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="chap-HPCC.html"><a href="chap-HPCC.html"><i class="fa fa-check"></i><b>8</b> High Performance Computing Clusters (HPCC’s)</a>
<ul>
<li class="chapter" data-level="8.1" data-path="chap-HPCC.html"><a href="chap-HPCC.html#an-oversimplified-but-useful-view-of-a-computing-cluster"><i class="fa fa-check"></i><b>8.1</b> An oversimplified, but useful, view of a computing cluster</a></li>
<li class="chapter" data-level="8.2" data-path="chap-HPCC.html"><a href="chap-HPCC.html#cluster-computing-and-the-job-scheduler"><i class="fa fa-check"></i><b>8.2</b> Cluster computing and the job scheduler</a></li>
<li class="chapter" data-level="8.3" data-path="chap-HPCC.html"><a href="chap-HPCC.html#slurm-info"><i class="fa fa-check"></i><b>8.3</b> Learning about the resources on your HPCC</a></li>
<li class="chapter" data-level="8.4" data-path="chap-HPCC.html"><a href="chap-HPCC.html#getting-compute-resources-allocated-to-your-jobs-on-an-hpcc"><i class="fa fa-check"></i><b>8.4</b> Getting compute resources allocated to your jobs on an HPCC</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="chap-HPCC.html"><a href="chap-HPCC.html#interactive-sessions"><i class="fa fa-check"></i><b>8.4.1</b> Interactive sessions</a></li>
<li class="chapter" data-level="8.4.2" data-path="chap-HPCC.html"><a href="chap-HPCC.html#slurm-batch"><i class="fa fa-check"></i><b>8.4.2</b> Batch jobs</a></li>
<li class="chapter" data-level="8.4.3" data-path="chap-HPCC.html"><a href="chap-HPCC.html#slurm-job-arrays"><i class="fa fa-check"></i><b>8.4.3</b> SLURM Job Arrays</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="chap-HPCC.html"><a href="chap-HPCC.html#prepation-interlude-an-in-class-exercise-to-make-sure-everything-is-configured-correctly"><i class="fa fa-check"></i><b>8.5</b> PREPATION INTERLUDE: An in-class exercise to make sure everything is configured correctly</a></li>
<li class="chapter" data-level="8.6" data-path="chap-HPCC.html"><a href="chap-HPCC.html#more-boneyard"><i class="fa fa-check"></i><b>8.6</b> More Boneyard…</a></li>
<li class="chapter" data-level="8.7" data-path="chap-HPCC.html"><a href="chap-HPCC.html#the-queue-slurmsgeuge"><i class="fa fa-check"></i><b>8.7</b> The Queue (SLURM/SGE/UGE)</a></li>
<li class="chapter" data-level="8.8" data-path="chap-HPCC.html"><a href="chap-HPCC.html#modules-package"><i class="fa fa-check"></i><b>8.8</b> Modules package</a></li>
<li class="chapter" data-level="8.9" data-path="chap-HPCC.html"><a href="chap-HPCC.html#compiling-programs-without-admin-privileges"><i class="fa fa-check"></i><b>8.9</b> Compiling programs without admin privileges</a></li>
<li class="chapter" data-level="8.10" data-path="chap-HPCC.html"><a href="chap-HPCC.html#job-arrays"><i class="fa fa-check"></i><b>8.10</b> Job arrays</a></li>
<li class="chapter" data-level="8.11" data-path="chap-HPCC.html"><a href="chap-HPCC.html#writing-stdout-and-stderr-to-files"><i class="fa fa-check"></i><b>8.11</b> Writing stdout and stderr to files</a></li>
<li class="chapter" data-level="8.12" data-path="chap-HPCC.html"><a href="chap-HPCC.html#breaking-stuff-down"><i class="fa fa-check"></i><b>8.12</b> Breaking stuff down</a></li>
</ul></li>
<li class="part"><span><b>II Part II: Reproducible Research Strategies</b></span></li>
<li class="chapter" data-level="9" data-path="introduction-to-reproducible-research.html"><a href="introduction-to-reproducible-research.html"><i class="fa fa-check"></i><b>9</b> Introduction to Reproducible Research</a></li>
<li class="chapter" data-level="10" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html"><i class="fa fa-check"></i><b>10</b> Rstudio and Project-centered Organization</a>
<ul>
<li class="chapter" data-level="10.1" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#organizing-big-projects"><i class="fa fa-check"></i><b>10.1</b> Organizing big projects</a></li>
<li class="chapter" data-level="10.2" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#using-rstudio-in-workflows-with-remote-computers-and-hpccs"><i class="fa fa-check"></i><b>10.2</b> Using RStudio in workflows with remote computers and HPCCs</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#keeping-an-rstudio-project-in-sync-with-github"><i class="fa fa-check"></i><b>10.2.1</b> Keeping an RStudio project “in sync” with GitHub</a></li>
<li class="chapter" data-level="10.2.2" data-path="rstudio-and-project-centered-organization.html"><a href="rstudio-and-project-centered-organization.html#evaluating-scripts-line-by-line-on-a-remote-machine-from-within-rstudio"><i class="fa fa-check"></i><b>10.2.2</b> Evaluating scripts line by line on a remote machine from within RStudio</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="version-control.html"><a href="version-control.html"><i class="fa fa-check"></i><b>11</b> Version control</a>
<ul>
<li class="chapter" data-level="11.1" data-path="version-control.html"><a href="version-control.html#why-use-version-control"><i class="fa fa-check"></i><b>11.1</b> Why use version control?</a></li>
<li class="chapter" data-level="11.2" data-path="version-control.html"><a href="version-control.html#git-workings"><i class="fa fa-check"></i><b>11.2</b> How git works</a></li>
<li class="chapter" data-level="11.3" data-path="version-control.html"><a href="version-control.html#git-workflow-patterns"><i class="fa fa-check"></i><b>11.3</b> git workflow patterns</a></li>
<li class="chapter" data-level="11.4" data-path="version-control.html"><a href="version-control.html#using-git-with-rstudio"><i class="fa fa-check"></i><b>11.4</b> using git with Rstudio</a></li>
<li class="chapter" data-level="11.5" data-path="version-control.html"><a href="version-control.html#git-on-the-command-line"><i class="fa fa-check"></i><b>11.5</b> git on the command line</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="a-fast-furious-overview-of-the-tidyverse.html"><a href="a-fast-furious-overview-of-the-tidyverse.html"><i class="fa fa-check"></i><b>12</b> A fast, furious overview of the tidyverse</a></li>
<li class="chapter" data-level="13" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html"><i class="fa fa-check"></i><b>13</b> Authoring reproducibly with Rmarkdown</a>
<ul>
<li class="chapter" data-level="13.1" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#notebooks"><i class="fa fa-check"></i><b>13.1</b> Notebooks</a></li>
<li class="chapter" data-level="13.2" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#references"><i class="fa fa-check"></i><b>13.2</b> References</a>
<ul>
<li class="chapter" data-level="13.2.1" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#zotero-and-rmarkdown"><i class="fa fa-check"></i><b>13.2.1</b> Zotero and Rmarkdown</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#bookdown"><i class="fa fa-check"></i><b>13.3</b> Bookdown</a></li>
<li class="chapter" data-level="13.4" data-path="authoring-reproducibly-with-rmarkdown.html"><a href="authoring-reproducibly-with-rmarkdown.html#google-docs"><i class="fa fa-check"></i><b>13.4</b> Google Docs</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html"><i class="fa fa-check"></i><b>14</b> Managing Workflows with Snakemake</a>
<ul>
<li class="chapter" data-level="14.1" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html#a-simple-snakemake-example"><i class="fa fa-check"></i><b>14.1</b> A Simple Snakemake Example</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html#snakefile-1-a-simple-rule-to-trimmomcawesome-the-files-from-001"><i class="fa fa-check"></i><b>14.1.1</b> Snakefile #1: A simple rule to TrimmoMcAwesome the files from 001</a></li>
<li class="chapter" data-level="14.1.2" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html#snakefile-2-a-single-trimmomcawesome-rule-using-wildcards"><i class="fa fa-check"></i><b>14.1.2</b> Snakefile #2: A single TrimmoMcAwesome rule using wildcards</a></li>
<li class="chapter" data-level="14.1.3" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html#snakefile-3-add-trim_stupendous-in-there"><i class="fa fa-check"></i><b>14.1.3</b> Snakefile #3: Add <code>trim_stupendous</code> in there</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html#ugly-complex-file-namesa-job-for-snakemake-input-functions"><i class="fa fa-check"></i><b>14.2</b> Ugly, complex file names…a job for Snakemake input functions!</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html#playing-with-pandas"><i class="fa fa-check"></i><b>14.2.1</b> Playing with Pandas</a></li>
<li class="chapter" data-level="14.2.2" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html#a-quick-sidebar-spoofing-a-snakemake-input-function"><i class="fa fa-check"></i><b>14.2.2</b> A Quick Sidebar: Spoofing a Snakemake input function</a></li>
<li class="chapter" data-level="14.2.3" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html#snakefile-4-an-input-function-for-trim_awesome"><i class="fa fa-check"></i><b>14.2.3</b> Snakefile #4: an input function for trim_awesome</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html#input-functions-from-python-dictionaries"><i class="fa fa-check"></i><b>14.3</b> Input functions from Python Dictionaries</a></li>
<li class="chapter" data-level="14.4" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html#generating-lists-of-files-with-expand"><i class="fa fa-check"></i><b>14.4</b> Generating lists of files with expand()</a></li>
<li class="chapter" data-level="14.5" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html#a-python-primer-for-using-snakemake"><i class="fa fa-check"></i><b>14.5</b> A Python Primer for Using Snakemake</a></li>
<li class="chapter" data-level="14.6" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html#using-snakemake-on-a-computing-cluster"><i class="fa fa-check"></i><b>14.6</b> Using Snakemake on a Computing Cluster</a>
<ul>
<li class="chapter" data-level="14.6.1" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html#installing-and-configuring-the-snakemake-slurm-profile"><i class="fa fa-check"></i><b>14.6.1</b> Installing and Configuring the Snakemake SLURM profile</a></li>
<li class="chapter" data-level="14.6.2" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html#playing-with-our-new-slurm-profile"><i class="fa fa-check"></i><b>14.6.2</b> Playing with our new SLURM profile</a></li>
<li class="chapter" data-level="14.6.3" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html#managing-cluster-resources-with-snakemake"><i class="fa fa-check"></i><b>14.6.3</b> Managing cluster resources with Snakemake</a></li>
<li class="chapter" data-level="14.6.4" data-path="managing-workflows-with-snakemake.html"><a href="managing-workflows-with-snakemake.html#grouping-short-jobs-to-run-on-a-single-node"><i class="fa fa-check"></i><b>14.6.4</b> Grouping short jobs to run on a single node</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Part III: Bioinformatic Analyses</b></span></li>
<li class="chapter" data-level="15" data-path="overview-of-bioinformatic-analyses.html"><a href="overview-of-bioinformatic-analyses.html"><i class="fa fa-check"></i><b>15</b> Overview of Bioinformatic Analyses</a></li>
<li class="chapter" data-level="16" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html"><i class="fa fa-check"></i><b>16</b> DNA Sequences and Sequencing</a>
<ul>
<li class="chapter" data-level="16.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#dna-stuff"><i class="fa fa-check"></i><b>16.1</b> DNA Stuff</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#dna-replication-with-dna-polymerase"><i class="fa fa-check"></i><b>16.1.1</b> DNA Replication with DNA Polymerase</a></li>
<li class="chapter" data-level="16.1.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#the-importance-of-the-3-hydroxyl"><i class="fa fa-check"></i><b>16.1.2</b> The importance of the 3’ hydroxyl…</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#sanger-sequencing"><i class="fa fa-check"></i><b>16.2</b> Sanger sequencing</a></li>
<li class="chapter" data-level="16.3" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#illumina-sequencing-by-synthesis"><i class="fa fa-check"></i><b>16.3</b> Illumina Sequencing by Synthesis</a></li>
<li class="chapter" data-level="16.4" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#library-prep-protocols"><i class="fa fa-check"></i><b>16.4</b> Library Prep Protocols</a>
<ul>
<li class="chapter" data-level="16.4.1" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#wgs"><i class="fa fa-check"></i><b>16.4.1</b> WGS</a></li>
<li class="chapter" data-level="16.4.2" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#rad-seq-methods"><i class="fa fa-check"></i><b>16.4.2</b> RAD-Seq methods</a></li>
<li class="chapter" data-level="16.4.3" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#amplicon-sequencing"><i class="fa fa-check"></i><b>16.4.3</b> Amplicon Sequencing</a></li>
<li class="chapter" data-level="16.4.4" data-path="dna-sequences-and-sequencing.html"><a href="dna-sequences-and-sequencing.html#capture-arrays-rapture-etc."><i class="fa fa-check"></i><b>16.4.4</b> Capture arrays, RAPTURE, etc.</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html"><i class="fa fa-check"></i><b>17</b> Bioinformatic file formats</a>
<ul>
<li class="chapter" data-level="17.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sequences"><i class="fa fa-check"></i><b>17.1</b> Sequences</a></li>
<li class="chapter" data-level="17.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#fastq"><i class="fa fa-check"></i><b>17.2</b> FASTQ</a>
<ul>
<li class="chapter" data-level="17.2.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#illumina-ids"><i class="fa fa-check"></i><b>17.2.1</b> Line 1: Illumina identifier lines</a></li>
<li class="chapter" data-level="17.2.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#bqscores"><i class="fa fa-check"></i><b>17.2.2</b> Line 4: Base quality scores</a></li>
<li class="chapter" data-level="17.2.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#a-fastq-tidyverse-interlude"><i class="fa fa-check"></i><b>17.2.3</b> A FASTQ ‘tidyverse’ Interlude</a></li>
<li class="chapter" data-level="17.2.4" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#comparing-read-1-to-read-2"><i class="fa fa-check"></i><b>17.2.4</b> Comparing read 1 to read 2</a></li>
</ul></li>
<li class="chapter" data-level="17.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#fasta"><i class="fa fa-check"></i><b>17.3</b> FASTA</a>
<ul>
<li class="chapter" data-level="17.3.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#genomic-ranges"><i class="fa fa-check"></i><b>17.3.1</b> Genomic ranges</a></li>
<li class="chapter" data-level="17.3.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#extracting-genomic-ranges-from-a-fasta-file"><i class="fa fa-check"></i><b>17.3.2</b> Extracting genomic ranges from a FASTA file</a></li>
<li class="chapter" data-level="17.3.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#downloading-reference-genomes-from-ncbi"><i class="fa fa-check"></i><b>17.3.3</b> Downloading reference genomes from NCBI</a></li>
</ul></li>
<li class="chapter" data-level="17.4" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#alignments"><i class="fa fa-check"></i><b>17.4</b> Alignments</a>
<ul>
<li class="chapter" data-level="17.4.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#how-might-i-align-to-thee-let-me-count-the-ways"><i class="fa fa-check"></i><b>17.4.1</b> How might I align to thee? Let me count the ways…</a></li>
<li class="chapter" data-level="17.4.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#play-with-simple-alignments"><i class="fa fa-check"></i><b>17.4.2</b> Play with simple alignments</a></li>
<li class="chapter" data-level="17.4.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sam-flags"><i class="fa fa-check"></i><b>17.4.3</b> SAM Flags</a></li>
<li class="chapter" data-level="17.4.4" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#cigar"><i class="fa fa-check"></i><b>17.4.4</b> The CIGAR string</a></li>
<li class="chapter" data-level="17.4.5" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#the-seq-and-qual-columns"><i class="fa fa-check"></i><b>17.4.5</b> The SEQ and QUAL columns</a></li>
<li class="chapter" data-level="17.4.6" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sam-headers"><i class="fa fa-check"></i><b>17.4.6</b> SAM File Headers</a></li>
<li class="chapter" data-level="17.4.7" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#the-bam-format"><i class="fa fa-check"></i><b>17.4.7</b> The BAM format</a></li>
<li class="chapter" data-level="17.4.8" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#quick-self-study"><i class="fa fa-check"></i><b>17.4.8</b> Quick self study</a></li>
</ul></li>
<li class="chapter" data-level="17.5" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#variants"><i class="fa fa-check"></i><b>17.5</b> Variants</a>
<ul>
<li class="chapter" data-level="17.5.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#vcf-format-the-body"><i class="fa fa-check"></i><b>17.5.1</b> VCF Format – The Body</a></li>
<li class="chapter" data-level="17.5.2" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#vcf-format-the-header"><i class="fa fa-check"></i><b>17.5.2</b> VCF Format – The Header</a></li>
<li class="chapter" data-level="17.5.3" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#boneyard"><i class="fa fa-check"></i><b>17.5.3</b> Boneyard</a></li>
</ul></li>
<li class="chapter" data-level="17.6" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#segments"><i class="fa fa-check"></i><b>17.6</b> Segments</a></li>
<li class="chapter" data-level="17.7" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#conversionextractions-between-different-formats"><i class="fa fa-check"></i><b>17.7</b> Conversion/Extractions between different formats</a></li>
<li class="chapter" data-level="17.8" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#visualization-of-genomic-data"><i class="fa fa-check"></i><b>17.8</b> Visualization of Genomic Data</a>
<ul>
<li class="chapter" data-level="17.8.1" data-path="bioinformatic-file-formats.html"><a href="bioinformatic-file-formats.html#sample-data"><i class="fa fa-check"></i><b>17.8.1</b> Sample Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="18" data-path="genome-assembly.html"><a href="genome-assembly.html"><i class="fa fa-check"></i><b>18</b> Genome Assembly</a></li>
<li class="chapter" data-level="19" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><i class="fa fa-check"></i><b>19</b> Alignment of sequence data to a reference genome (and associated steps)</a>
<ul>
<li class="chapter" data-level="19.1" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#read-journey"><i class="fa fa-check"></i><b>19.1</b> The Journey of each DNA Fragment from Organism to Sequencing Read</a></li>
<li class="chapter" data-level="19.2" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#read-groups"><i class="fa fa-check"></i><b>19.2</b> Read Groups</a></li>
<li class="chapter" data-level="19.3" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#aligning-reads-with-bwa"><i class="fa fa-check"></i><b>19.3</b> Aligning reads with <code>bwa</code></a>
<ul>
<li class="chapter" data-level="19.3.1" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#indexing-the-genome-for-alignment"><i class="fa fa-check"></i><b>19.3.1</b> Indexing the genome for alignment</a></li>
<li class="chapter" data-level="19.3.2" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#mapping-reads-with-bwa-mem"><i class="fa fa-check"></i><b>19.3.2</b> Mapping reads with <code>bwa mem</code></a></li>
<li class="chapter" data-level="19.3.3" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#hold-it-right-there-buddy-what-about-the-read-groups"><i class="fa fa-check"></i><b>19.3.3</b> Hold it Right There, Buddy! What about the Read Groups?</a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#samtools"><i class="fa fa-check"></i><b>19.4</b> Processing alignment output with <code>samtools</code></a>
<ul>
<li class="chapter" data-level="19.4.1" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#samtools-subcommands"><i class="fa fa-check"></i><b>19.4.1</b> <code>samtools</code> subcommands</a></li>
</ul></li>
<li class="chapter" data-level="19.5" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#boneyard-below-here"><i class="fa fa-check"></i><b>19.5</b> BONEYARD BELOW HERE</a></li>
<li class="chapter" data-level="19.6" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#preprocess"><i class="fa fa-check"></i><b>19.6</b> Preprocess ?</a></li>
<li class="chapter" data-level="19.7" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#quick-notes-to-self-on-chaining-things"><i class="fa fa-check"></i><b>19.7</b> Quick notes to self on chaining things:</a></li>
<li class="chapter" data-level="19.8" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#merging-bam-files"><i class="fa fa-check"></i><b>19.8</b> Merging BAM files</a></li>
<li class="chapter" data-level="19.9" data-path="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html"><a href="alignment-of-sequence-data-to-a-reference-genome-and-associated-steps.html#divide-and-conquer-strategies"><i class="fa fa-check"></i><b>19.9</b> Divide and Conquer Strategies</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="variant-calling.html"><a href="variant-calling.html"><i class="fa fa-check"></i><b>20</b> Variant calling</a>
<ul>
<li class="chapter" data-level="20.1" data-path="variant-calling.html"><a href="variant-calling.html#genotype-likelihoods"><i class="fa fa-check"></i><b>20.1</b> Genotype Likelihoods</a>
<ul>
<li class="chapter" data-level="20.1.1" data-path="variant-calling.html"><a href="variant-calling.html#basic-sketch-of-genotype-likelihood-calculations"><i class="fa fa-check"></i><b>20.1.1</b> Basic Sketch of Genotype Likelihood Calculations</a></li>
<li class="chapter" data-level="20.1.2" data-path="variant-calling.html"><a href="variant-calling.html#specifics-of-different-genotype-likelihoods"><i class="fa fa-check"></i><b>20.1.2</b> Specifics of different genotype likelihoods</a></li>
<li class="chapter" data-level="20.1.3" data-path="variant-calling.html"><a href="variant-calling.html#computing-genotype-likelihoods-with-three-different-softwares"><i class="fa fa-check"></i><b>20.1.3</b> Computing genotype likelihoods with three different softwares</a></li>
<li class="chapter" data-level="20.1.4" data-path="variant-calling.html"><a href="variant-calling.html#a-directed-acyclic-graph-for-genotype-likelihoods"><i class="fa fa-check"></i><b>20.1.4</b> A Directed Acyclic Graph For Genotype Likelihoods</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="21" data-path="boneyard-1.html"><a href="boneyard-1.html"><i class="fa fa-check"></i><b>21</b> Boneyard</a></li>
<li class="chapter" data-level="22" data-path="basic-handling-of-vcf-files.html"><a href="basic-handling-of-vcf-files.html"><i class="fa fa-check"></i><b>22</b> Basic Handling of VCF files</a>
<ul>
<li class="chapter" data-level="22.1" data-path="basic-handling-of-vcf-files.html"><a href="basic-handling-of-vcf-files.html#bcftools"><i class="fa fa-check"></i><b>22.1</b> bcftools</a>
<ul>
<li class="chapter" data-level="22.1.1" data-path="basic-handling-of-vcf-files.html"><a href="basic-handling-of-vcf-files.html#tell-me-about-my-vcf-file"><i class="fa fa-check"></i><b>22.1.1</b> Tell me about my VCF file!</a></li>
<li class="chapter" data-level="22.1.2" data-path="basic-handling-of-vcf-files.html"><a href="basic-handling-of-vcf-files.html#get-fragmentsparts-of-my-vcf-file"><i class="fa fa-check"></i><b>22.1.2</b> Get fragments/parts of my VCF file</a></li>
<li class="chapter" data-level="22.1.3" data-path="basic-handling-of-vcf-files.html"><a href="basic-handling-of-vcf-files.html#combine-vcf-files-in-various-ways"><i class="fa fa-check"></i><b>22.1.3</b> Combine VCF files in various ways</a></li>
<li class="chapter" data-level="22.1.4" data-path="basic-handling-of-vcf-files.html"><a href="basic-handling-of-vcf-files.html#filter-out-variants-for-a-variety-of-reasons"><i class="fa fa-check"></i><b>22.1.4</b> Filter out variants for a variety of reasons</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="23" data-path="bioinformatics-for-rad-seq-data-with-and-without-a-reference-genome.html"><a href="bioinformatics-for-rad-seq-data-with-and-without-a-reference-genome.html"><i class="fa fa-check"></i><b>23</b> Bioinformatics for RAD seq data with and without a reference genome</a></li>
<li class="chapter" data-level="24" data-path="processing-amplicon-sequencing-data.html"><a href="processing-amplicon-sequencing-data.html"><i class="fa fa-check"></i><b>24</b> Processing amplicon sequencing data</a></li>
<li class="chapter" data-level="25" data-path="genome-annotation.html"><a href="genome-annotation.html"><i class="fa fa-check"></i><b>25</b> Genome Annotation</a></li>
<li class="chapter" data-level="26" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html"><i class="fa fa-check"></i><b>26</b> Whole genome alignment strategies</a>
<ul>
<li class="chapter" data-level="26.1" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#mapping-of-scaffolds-to-a-closely-related-genome"><i class="fa fa-check"></i><b>26.1</b> Mapping of scaffolds to a closely related genome</a></li>
<li class="chapter" data-level="26.2" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#obtaining-ancestral-states-from-an-outgroup-genome"><i class="fa fa-check"></i><b>26.2</b> Obtaining Ancestral States from an Outgroup Genome</a>
<ul>
<li class="chapter" data-level="26.2.1" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#using-lastz-to-align-coho-to-the-chinook-genome"><i class="fa fa-check"></i><b>26.2.1</b> Using LASTZ to align coho to the chinook genome</a></li>
<li class="chapter" data-level="26.2.2" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#try-on-the-chinook-chromosomes"><i class="fa fa-check"></i><b>26.2.2</b> Try on the chinook chromosomes</a></li>
<li class="chapter" data-level="26.2.3" data-path="whole-genome-alignment-strategies.html"><a href="whole-genome-alignment-strategies.html#explore-the-other-parameters-more"><i class="fa fa-check"></i><b>26.2.3</b> Explore the other parameters more</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="27" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html"><i class="fa fa-check"></i><b>27</b> An example full workflow from sequences to variants</a>
<ul>
<li class="chapter" data-level="27.1" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#a-word-on-organizing-workflows"><i class="fa fa-check"></i><b>27.1</b> A word on organizing workflows</a></li>
<li class="chapter" data-level="27.2" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step-1a.-quality-checking-the-reads"><i class="fa fa-check"></i><b>27.2</b> Step 1a. Quality Checking the Reads</a></li>
<li class="chapter" data-level="27.3" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step-1b.-trimming-the-reads"><i class="fa fa-check"></i><b>27.3</b> Step 1b. Trimming the Reads</a></li>
<li class="chapter" data-level="27.4" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step-2.-mapping-each-unit-to-the-genome"><i class="fa fa-check"></i><b>27.4</b> Step 2. Mapping each unit to the genome</a></li>
<li class="chapter" data-level="27.5" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step-3.-merging-bams-from-the-same-sample-library"><i class="fa fa-check"></i><b>27.5</b> Step 3. Merging bams from the same sample + library</a></li>
<li class="chapter" data-level="27.6" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step-4.-marking-duplicates"><i class="fa fa-check"></i><b>27.6</b> Step 4. Marking duplicates</a></li>
<li class="chapter" data-level="27.7" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step-5.-merging-dup-marked-bams-from-the-same-sample"><i class="fa fa-check"></i><b>27.7</b> Step 5. Merging dup-marked bams from the same sample</a></li>
<li class="chapter" data-level="27.8" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step-6.-creating-a-gvcf-file-for-each-sample"><i class="fa fa-check"></i><b>27.8</b> Step 6. Creating a gVCF file for each sample</a></li>
<li class="chapter" data-level="27.9" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step-7.-importing-the-gvcf-files-into-a-genomicsdatabase"><i class="fa fa-check"></i><b>27.9</b> Step 7. Importing the gVCF files into a GenomicsDatabase</a></li>
<li class="chapter" data-level="27.10" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step8.-calling-variants-and-creating-vcf-files-from-the-genomicsdatabase"><i class="fa fa-check"></i><b>27.10</b> Step8. Calling variants and creating VCF files from the GenomicsDatabase</a></li>
<li class="chapter" data-level="27.11" data-path="an-example-full-workflow-from-sequences-to-variants.html"><a href="an-example-full-workflow-from-sequences-to-variants.html#step9.-filtering-variants"><i class="fa fa-check"></i><b>27.11</b> Step9. Filtering variants</a></li>
</ul></li>
<li class="part"><span><b>IV Part IV: Analysis of Big Variant Data</b></span></li>
<li class="chapter" data-level="28" data-path="bioinformatic-analysis-on-variant-data.html"><a href="bioinformatic-analysis-on-variant-data.html"><i class="fa fa-check"></i><b>28</b> Bioinformatic analysis on variant data</a></li>
<li class="part"><span><b>V Part V: Population Genomics</b></span></li>
<li class="chapter" data-level="29" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html"><i class="fa fa-check"></i><b>29</b> Topics in pop gen</a>
<ul>
<li class="chapter" data-level="29.1" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#coalescent"><i class="fa fa-check"></i><b>29.1</b> Coalescent</a></li>
<li class="chapter" data-level="29.2" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#measures-of-genetic-diversity-and-such"><i class="fa fa-check"></i><b>29.2</b> Measures of genetic diversity and such</a></li>
<li class="chapter" data-level="29.3" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#demographic-inference-with-partial-a-partial-i-and-moments"><i class="fa fa-check"></i><b>29.3</b> Demographic inference with <span class="math inline">\(\partial a \partial i\)</span> and <em>moments</em></a></li>
<li class="chapter" data-level="29.4" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#balls-in-boxes"><i class="fa fa-check"></i><b>29.4</b> Balls in Boxes</a></li>
<li class="chapter" data-level="29.5" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#some-landscape-genetics"><i class="fa fa-check"></i><b>29.5</b> Some landscape genetics</a></li>
<li class="chapter" data-level="29.6" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#relationship-inference"><i class="fa fa-check"></i><b>29.6</b> Relationship Inference</a></li>
<li class="chapter" data-level="29.7" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#tests-for-selection"><i class="fa fa-check"></i><b>29.7</b> Tests for Selection</a></li>
<li class="chapter" data-level="29.8" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#multivariate-associations-gea-etc."><i class="fa fa-check"></i><b>29.8</b> Multivariate Associations, GEA, etc.</a></li>
<li class="chapter" data-level="29.9" data-path="topics-in-pop-gen.html"><a href="topics-in-pop-gen.html#estimating-heritability-in-the-wild"><i class="fa fa-check"></i><b>29.9</b> Estimating heritability in the wild</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references-1.html"><a href="references-1.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://bookdown.org" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Practical Computing and Bioinformatics for Conservation and Evolutionary Genomics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="rstudio-and-project-centered-organization" class="section level1" number="10">
<h1><span class="header-section-number">Chapter 10</span> Rstudio and Project-centered Organization</h1>
<p>Somewhere talk about here::here(): <a href="https://github.com/jennybc/here_here" class="uri">https://github.com/jennybc/here_here</a></p>
<div id="organizing-big-projects" class="section level2" number="10.1">
<h2><span class="header-section-number">10.1</span> Organizing big projects</h2>
<p>By “big” I mean something like the chinook project, or your typical thing this is
a chapter in a dissertation or a paper.</p>
<p>I think it is useful for number things in order on a three-digit system, and
at the top of each make directories <code>outputs</code> and <code>intermediates</code>, like this:</p>
<div class="sourceCode" id="cb374"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb374-1"><a href="rstudio-and-project-centered-organization.html#cb374-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">file.path</span>(<span class="fu">c</span>(<span class="st">&quot;outputs&quot;</span>, <span class="st">&quot;intermediates&quot;</span>), <span class="st">&quot;203&quot;</span>), <span class="at">recursive =</span> <span class="cn">TRUE</span>, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>I had previously used two variables <code>output_dir</code> and <code>interm_dir</code> to specify these in each
notebook, but now I think it would be better to just hardwire those, for a few reasons:</p>
<ul>
<li>Sometimes you are working on two notebooks at once in the same environment and you
don’t want to get confused about where things should get written.</li>
<li>You can’t use those variables in shell blocks of code, where you will just have
to write the paths out anyway.</li>
<li>Hard-wiring the paths forces you to think about the fact that once you establish
the name for something, you should not change it, ever.</li>
<li>Hard-wiring the paths makes it easy to identify access to those different files. In particular
you can write an R script that rips through all the lines of code in the Rmds (and R files) in your
project and records all the instances of writing and reading of files from the outputs and intermediates
directories. If you do this, you can make a pretty cool dependency graph so that you can visualize
what you need to keep to clean things up for a final reproducible project. <em>Note: I should
write a little R package that can analyze such dependencies in a project. Unless there is
already something like that. (Note that these are not package dependencies, but, rather, internal
project dependencies. Note that if one is consistent with using readr functions it would be
pretty easy to find all those instances of <code>read_*</code> and <code>write_*</code> and that makes it clear
why standardized syntax like that is so useful.</em> Hey! Notice that this type of analysis
would be made simple if we just focused on dependencies between different Rmds. That is
probably the level we want to keep it at as well. Ideally you can make a graph of all files
that are output from one Rmd and read into another. That would be a fun graph to make of the
Chinook project.</li>
<li>Note. You should keep 900-999 as 100 slots for Rnotebooks for the final reproducible project
to go with a publication. So, you can pare down all the previous notebooks and things.</li>
<li>Hey! Sometimes you are going to want to write or read files that have been auto-produced.
For example, if you are cycling over chromosomes, you might have output files that start
something like: <code>outputs/302/chromo_output_</code>. So, when generating those names,
make sure that the full prefix is in there, and has a trailing underscore. Then
you can still find it with a regex search, and also recognize it as a
signifying a class of output files.</li>
</ul>
</div>
<div id="using-rstudio-in-workflows-with-remote-computers-and-hpccs" class="section level2" number="10.2">
<h2><span class="header-section-number">10.2</span> Using RStudio in workflows with remote computers and HPCCs</h2>
<p>To this point in the text/course we have spent a good deal of time
learning about bash shell scripts and familiarizing ourselves with
operating on the command line while logged into remote computers, like those
of the HPCC administered by our university or agency. For the most part,
we have not been terribly concerned about integrating RStudio into these
workflows on remote computers; however, if you are familiar with RStudio
as an IDE for working with R on your local computer, and you have grown
comfortable with its interface: source code editor windows, file browser pane,
streamlined GUI for doing git version control, etc., then you are in a good
position to incorporate RStudio into your workflows on remote computers.
By leveraging your familiarity with RStudio, and merely learning a few new
ways of using it, you can use RStudio to great effect for testing bash
scripts line-by-line in order to make sure that they will run, as expected,
on your HPCC. It also works well for keeping the scripts and source codes
of your projects synchronized between your HPCC and your local laptop, providing
some backup and redundancy to your projects.</p>
<p>To be sure, RStudio was not originally developed to provide a solution for
shell-script-intensive projects on HPCCs. And, there will be a few hardcore
Unix holdouts that believe that the best way to “do Unix” is to learn a
purely text-based editing system (like <code>vim</code> or <code>emacs</code>, and stick with that).
But, if you are already comfortable with the GUI-integrated development environment of
RStudio, you don’t have a lot to lose from trying to incorporate it into your
remote computer workflows. Surely, the long-term persistence of other GUI
frameworks for working with remote computers (like X-windows) argues that there
is some value in letting people work in the sort of GUI environments they might
be used to.</p>
<p>It should be noted that RStudio does produce a truly wondrous product
called RStudio Server that allows a remote computer to run an RStudio
session and to serve it up to a client (that would be like you or me on a laptop)
via a web browser. This provides a nearly seamless RStudio experience: your
web browser window literally becomes RStudio itself and it hardly feels like
you are working on a remote machine. Such an approach, however, is not
great in most HPCC environments for a number of reasons:</p>
<ol style="list-style-type: decimal">
<li>Setting up the RStudio Server requires administrator access to the remote
machine, which you almost certainly do not have on your HPCC.</li>
<li>Most system administrators are not keen to set up an RStudio Server on the
HPCC systems they administer.</li>
<li>RStudio Server would not be the best platform for doing Unix-based
bioinformatic workflows. You still need to use your computer’s job scheduler.</li>
<li>If many people are going to be using the RStudio Server, the paid, licensed
version is required, that costs some $’s.</li>
</ol>
<p>So, what I will describe here can be thought of as a sort of “poor-man’s”
RStudio Server—a way to use RStudio to do source code editing of shell and R
scripts on your laptop, but send those
lines of code, one-by-one, to the server to test them and make sure that they work. Then,
once you are convinced that you have all the kinks worked out of that code, you
can easily send your edits as script files to your HPCC server to do “production
runs” of your bioinformatic workflows or analysis projects. In this chapter we will
cover three useful aspects of this type of workflow:</p>
<ol style="list-style-type: decimal">
<li>Using GitHub to transfer changes to an RStudio project from your laptop to
your server (and vice versa).</li>
<li>Running shell code line by line by “sending it” for evaluation to the RStudio
terminal window.</li>
<li>Running R code on a remote R session by sending it via the terminal window.</li>
</ol>
<div id="keeping-an-rstudio-project-in-sync-with-github" class="section level3" number="10.2.1">
<h3><span class="header-section-number">10.2.1</span> Keeping an RStudio project “in sync” with GitHub</h3>
<p>The ways of using RStudio to interact with an HPCC here all involve maintaining
two “copies” of an RStudio Project: one copy on your laptop (the local copy) and
one copy on the remote server (the remote copy). The remote copy of the project
is not one that you ever “open RStudio” on, but it maintains the directory
structure of your local RStudio project (it is kind of like a mirror of your
local RStudio project).</p>
<p>Typically, you will write code
(shell scripts and things) on your local copy of the project, using RStudio’s
intuitive GUI source code editory window. Then you can test that code by
sending lines of code from the source code editor window on the local RStudio project
to a command line interpreter on the remote machine. When you are done editing code
on the local machine, and testing it on the remote machine, then, you are ready
to copy those scripts (and any edits you made on them while testing the code) to your
remote machine in order to really run them (using SLURM <code>sbatch</code>, for example).
You can use git and GitHub to make it easy to transfer files and changes from your local
RStudio project to your remote project.</p>
<p>A few principles that make things simpler when working on a project that exists
on both a remote server and on your laptop:</p>
<ol style="list-style-type: decimal">
<li>Don’t keep large data files under git version control. You should only
version-control (and synchronize, using git, between your laptop and the server)
your source-code files. git and GitHub are not there for you to backup your huge
FASTQ files or large output files like BAMs and VCFs. You should always
use your <code>.gitignore</code> file to avoid accidentally committing large files to git.</li>
<li>Any time when you have two separate git repositories that you are pushing and
pulling to GitHub, you want to avoid situations in which you have changed the same
file in both of them. Doing so leads to <em>merge conflicts</em>. It is best to exclusively
edit files on your laptop for some time. Then commit them, push them to GitHub,
pull them down to the repository on your remote server and use them. If you end up
editing any files in the repository on the remote server, you will want to commit those
and push them to GitHub and pull them down to your laptop, before editing more from
your laptop.</li>
</ol>
<div id="an-exercise-with-chr-32-bioinformatics" class="section level4" number="10.2.1.1">
<h4><span class="header-section-number">10.2.1.1</span> An exercise with chr-32-bioinformatics</h4>
<p>If you followed the instructions from last week, you should have <code>git clone</code>d your
repository so that you have a directory on your HPCC that is named
<code>chr-32-bioinformatics-githubname</code> (where githubname is your GitHub handle), and if you
list the contents, it should look like:</p>
<div class="sourceCode" id="cb375"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb375-1"><a href="rstudio-and-project-centered-organization.html#cb375-1" aria-hidden="true" tabindex="-1"></a><span class="kw">(</span><span class="ex">base</span><span class="kw">)</span> <span class="ex">[chr-32-bioinformatics-eriqande]--%</span> ls</span>
<span id="cb375-2"><a href="rstudio-and-project-centered-organization.html#cb375-2" aria-hidden="true" tabindex="-1"></a><span class="ex">chinook-all-prefixes-for-samples.tsv</span>  fastqs                 obtain-and-prep-data.sh</span>
<span id="cb375-3"><a href="rstudio-and-project-centered-organization.html#cb375-3" aria-hidden="true" tabindex="-1"></a><span class="ex">chinook-fastq-meta-data.tsv</span>           genome                 README-01.txt</span>
<span id="cb375-4"><a href="rstudio-and-project-centered-organization.html#cb375-4" aria-hidden="true" tabindex="-1"></a><span class="ex">chr-32-bioinformatics.Rproj</span>           map-N-files-from-K.sh  run-single-job.sh</span></code></pre></div>
<p>The directories <code>fastqs</code> and <code>genome</code> should hold the FASTQ files and the
bwa-indexed genome that you downloaded using <code>rclone</code>.</p>
<div id="push-any-changed-scripts-to-github" class="section level5" number="10.2.1.1.1">
<h5><span class="header-section-number">10.2.1.1.1</span> Push any changed scripts to GitHub</h5>
<p>In the process of downloading those files, you might have made changes to
<code>obtain-and-prep-data.sh</code>. You can see if you changed anything in the repository
by doing</p>
<div class="sourceCode" id="cb376"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb376-1"><a href="rstudio-and-project-centered-organization.html#cb376-1" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> status</span></code></pre></div>
<p>If it shows files that have been modified. You can commit them.<br />
<em>Be sure not to commit large files, like the data files and the indexed genome!!</em></p>
<div class="sourceCode" id="cb377"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb377-1"><a href="rstudio-and-project-centered-organization.html#cb377-1" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> add file-name1 file-name2  <span class="co"># replace file-name with whatever scripts you have modified</span></span>
<span id="cb377-2"><a href="rstudio-and-project-centered-organization.html#cb377-2" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> commit</span>
<span id="cb377-3"><a href="rstudio-and-project-centered-organization.html#cb377-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb377-4"><a href="rstudio-and-project-centered-organization.html#cb377-4" aria-hidden="true" tabindex="-1"></a><span class="co"># then you write a commit message and save that file</span></span></code></pre></div>
<p>Once you commit that, you can check with <code>git status</code> to make sure the repository
is <em>clean</em>.</p>
<p>Then, push those changes back to GitHub:</p>
<div class="sourceCode" id="cb378"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb378-1"><a href="rstudio-and-project-centered-organization.html#cb378-1" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> push origin master</span></code></pre></div>
</div>
<div id="make-sure-you-have-a-local-copy-of-your-repository" class="section level5" number="10.2.1.1.2">
<h5><span class="header-section-number">10.2.1.1.2</span> Make sure you have a local copy of your repository</h5>
<p>Now, if you have not already done it, get a copy of that repository on
your local machine. You can do this by using a line like:</p>
<div class="sourceCode" id="cb379"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb379-1"><a href="rstudio-and-project-centered-organization.html#cb379-1" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> clone https://eriqande@github.com/CSU-con-gen-bioinformatics-2020/chr-32-bioinformatics-eriqande</span></code></pre></div>
<p>on your laptop terminal (replacing <code>eriqande</code>—in two places above!—with your correct GitHub username). Or by using the RStudio’s “New Project-&gt;From Version Control” option
and opening a git repository with a URL like this:</p>
<div class="sourceCode" id="cb380"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb380-1"><a href="rstudio-and-project-centered-organization.html#cb380-1" aria-hidden="true" tabindex="-1"></a><span class="ex">https://eriqande@github.com/CSU-con-gen-bioinformatics-2020/chr-32-bioinformatics-eriqande</span></span></code></pre></div>
<p>again, replacing <code>eriqande</code> with your GitHub name.</p>
<p>That gives you your <em>local copy</em> of the repository.</p>
<p>Notice that this local copy does not have the <code>fastqs</code> or <code>genome</code> directory
inside it. That is OK. We don’t want to run code locally anyway.</p>
<p>In the next section we see how to use this local RStudio repository
to start an RStudio session to send command lines to the remote machine.</p>
</div>
</div>
</div>
<div id="evaluating-scripts-line-by-line-on-a-remote-machine-from-within-rstudio" class="section level3" number="10.2.2">
<h3><span class="header-section-number">10.2.2</span> Evaluating scripts line by line on a remote machine from within RStudio</h3>
<p>Continuing within our <em>local</em> RStudio project, which you should have opened (so
that RStudio is running in that project…)</p>
<div id="open-a-terminal-window-to-send-shell-code-to-your-remote-machine" class="section level4" number="10.2.2.1">
<h4><span class="header-section-number">10.2.2.1</span> Open a Terminal window to send shell code to your remote machine</h4>
<p>We want a unix shell on the remote machine where our remote repository is.</p>
<ul>
<li>Go to the Terminal tab in your local RStudio. That will give you a Unix shell
on your laptop. Use that shell to use <code>ssh</code> to login to your remote computer.
For example:</li>
</ul>
<div class="sourceCode" id="cb381"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb381-1"><a href="rstudio-and-project-centered-organization.html#cb381-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ssh</span> username@remote.server.edu</span></code></pre></div>
<p>Huge note! RStudio doesn’t use the same <code>ssh</code> program that ships with git bash on
Windows. So, if you are using a windows machine, if using ssh in the RStudio terminal
window gives you errors, you must explicitly use the version of ssh that ships with git
bash. Try giving the full path to ssh like this:</p>
<div class="sourceCode" id="cb382"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb382-1"><a href="rstudio-and-project-centered-organization.html#cb382-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/usr/bin/ssh</span> username@remote.server.edu</span></code></pre></div>
<p>For details about that, see <a href="https://github.com/rstudio/rstudio/issues/4623">https://github.com/rstudio/rstudio/issues/4623</a> I hope this fixes issues with Windows. I no longer have a windows machine
to test things on.</p>
<p>Once you have logged on to your remote machine, navigate (using <code>cd</code>) to the
directory that holds your remote RStudio project. For example, on my system that
is:</p>
<pre><code>/home/eriq@colostate.edu/scratch/COURSE_STUFF/chr-32-bioinformatics-eriqande</code></pre>
</div>
<div id="optional-start-a-tmux-session" class="section level4" number="10.2.2.2">
<h4><span class="header-section-number">10.2.2.2</span> Optional: start a tmux session</h4>
<p>The RStudio terminal is god-awful slow if a lot of text is streaming
through it. I thought that this might be ameliorated by starting a tmux session,
but it is not. Nonetheless, as we have seen previously, it can be good
to work within a tmux session for many reasons, not the least of which is that
it is easy to get back to it. (Also, if RStudio bombs while you are working in the
terminal, it will kill your shell session, so it is nice to be able to get back to
that same session using tmux.)
So, if you are comfortable with tmux, create a new tmux session with something like:</p>
<div class="sourceCode" id="cb384"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb384-1"><a href="rstudio-and-project-centered-organization.html#cb384-1" aria-hidden="true" tabindex="-1"></a><span class="ex">tmux</span> new <span class="at">-s</span> chr-32</span></code></pre></div>
</div>
<div id="not-optional-get-an-interactive-shell-on-a-compute-node" class="section level4" number="10.2.2.3">
<h4><span class="header-section-number">10.2.2.3</span> Not optional: Get an interactive shell on a compute node</h4>
<p>We don’t want to be doing computing on the login node. Also, we want to
test all the code on a compute node…</p>
<p>On summit:</p>
<div class="sourceCode" id="cb385"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb385-1"><a href="rstudio-and-project-centered-organization.html#cb385-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sinteractive</span></span></code></pre></div>
<p>Note, you might have to do this a couple of times. Their sinteractive script
seems a little buggy. In fact, early this morning it was completely failing,
but I was able to get an interactive session using what I use on Sedna:</p>
<div class="sourceCode" id="cb386"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb386-1"><a href="rstudio-and-project-centered-organization.html#cb386-1" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--pty</span> /bin/bash</span></code></pre></div>
<p>On Hummingbird: y’all that work on Hummingbird have said the system has some vagaries…</p>
</div>
<div id="open-map-n-files-from-k.sh-on-your-local-rstudio" class="section level4" number="10.2.2.4">
<h4><span class="header-section-number">10.2.2.4</span> Open <code>map-N-files-from-K.sh</code> on your local RStudio</h4>
<p>Do this from the file browser in your local RStudio. It will open up a source code editor window.</p>
</div>
<div id="evaluate-lines-of-code-in-map-n-files-from-k.sh-on-the-remote-server" class="section level4" number="10.2.2.5">
<h4><span class="header-section-number">10.2.2.5</span> Evaluate lines of code in <code>map-N-files-from-K.sh</code> on the remote server</h4>
<p>First, lets evaluate lines 36 and 37 in <code>map-N-files-from-K.sh</code>. Do this
by putting your cursor anywhere on line 36 (which says <code>source ~/.bashrc</code>)
and hit Option-CMD-Return (on a Mac) or Alt-cntrl-Return (on Windows)</p>
<p>That keystroke sends the whole command line where the cursor is to the
active Terminal window. <em>And</em> and places the cursor at the next line.</p>
<p>So you can evaluate that too…Option-Cmd-Return (Alt-cntrl-Return) and
Boom! You have evaluated the line <code>conda activate bioinf</code>.</p>
<p>This is much like CMD-Return (or cntrl-Return on Windows) to evaluate code in an
R script in the RStudio R console.</p>
<p>(Note, if you are editing a file with <code>.sh</code> extension, then, even CMD-Return
will send code to the Terminal, not to the R console. But, it is good idea to get
into the habit of using CMD-Option-Return (Alt-cntrl-Return) when you want to
send code to a remote terminal.)</p>
<p>That last line:</p>
<div class="sourceCode" id="cb387"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb387-1"><a href="rstudio-and-project-centered-organization.html#cb387-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate bioinf</span></code></pre></div>
<p>was important because we will be using bwa and samtools.
You should have these
in your conda bioinf environment.</p>
<p>Once that is done we are ready to start testing lines of code
from the scripts in our RStudio project, <em>on the remote server</em>.</p>
<p>At any rate, you should feel the power in your hands at this point.
This framework lets you step through code line by line and evaluate
each line to make sure it works, or, if you are reviewing someone else’s
scripts, or trying to understand what they do, you can work through them
this way.</p>
<p>At this point, we are ready to start chugging through these command lines.</p>
</div>
<div id="a-word-on-variables-etc." class="section level4" number="10.2.2.6">
<h4><span class="header-section-number">10.2.2.6</span> A word on variables, etc.</h4>
<p>If you are going to be running through shell code line by line, you
must be congnizant of the shell variables that are defined in the code.</p>
<p>You typically can’t just start from the middle of the script and begin
evaluating code. You have to start from the top of the script to make sure
that, by the time you get further down into the script, any variables that
may have been defined in the intervening lines and get <em>used</em> in the script are,
indeed, properly defined.</p>
<p>So, always start from the top.</p>
<p>In our case, after the usage block (which we don’t hassle with since we
are not running the script as a standalone script, but rather line by line
from <em>inside</em> it!) two variables are defined immediately:</p>
<div class="sourceCode" id="cb388"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb388-1"><a href="rstudio-and-project-centered-organization.html#cb388-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Now, get the first file index and the last file index</span></span>
<span id="cb388-2"><a href="rstudio-and-project-centered-organization.html#cb388-2" aria-hidden="true" tabindex="-1"></a><span class="va">START=$2</span></span>
<span id="cb388-3"><a href="rstudio-and-project-centered-organization.html#cb388-3" aria-hidden="true" tabindex="-1"></a><span class="va">STOP=$(($START</span> <span class="op">+</span> <span class="va">$1</span> <span class="op">-</span> <span class="dv">1</span><span class="va">))</span></span></code></pre></div>
<p>These basically tell us the indexes of the files that will be processed in the
<code>for</code> loop coming up. i.e., if START=1 and STOP=8, that means that, in the loop
below this point in the script, files 1 through 8 will be processed.</p>
<p>We can set these now if we want. Let’s set</p>
<div class="sourceCode" id="cb389"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb389-1"><a href="rstudio-and-project-centered-organization.html#cb389-1" aria-hidden="true" tabindex="-1"></a><span class="va">START=</span>1</span>
<span id="cb389-2"><a href="rstudio-and-project-centered-organization.html#cb389-2" aria-hidden="true" tabindex="-1"></a><span class="va">STOP=</span>8</span></code></pre></div>
<p>You can do that by just typing those on the command line.</p>
<p>Once that is done, we can evaluate (Option-CMD-Return) the following line
that starts:</p>
<div class="sourceCode" id="cb390"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb390-1"><a href="rstudio-and-project-centered-organization.html#cb390-1" aria-hidden="true" tabindex="-1"></a><span class="va">SM_TAGS=</span>...</span></code></pre></div>
<p>Then, see what the result is, with:</p>
<div class="sourceCode" id="cb391"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb391-1"><a href="rstudio-and-project-centered-organization.html#cb391-1" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="va">$SM_TAGS</span></span></code></pre></div>
<p>Those are the fundamentals of evaluating code line by line: evaluate the line. If
it assigned a variable, print the value of the variable (<code>echo</code>) to see what its value
is set to. Repeat…</p>
</div>
<div id="testing-things-in-for-loops" class="section level4" number="10.2.2.7">
<h4><span class="header-section-number">10.2.2.7</span> Testing things in <code>for</code> loops</h4>
<p>To test command lines, between the <code>do</code> and the <code>done</code> of a bash <code>for</code> loop, we need
to evaluate those lines. Recall that the body of the for loop will likely be evaluated
many times (that code gets looped over…) but we will want to test it just once, at
least. To prepare for this, we must set the variables that the <code>for</code> loop is cycling over.</p>
<p>When we see:</p>
<div class="sourceCode" id="cb392"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb392-1"><a href="rstudio-and-project-centered-organization.html#cb392-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span><span class="kw">((</span><span class="va">Idx</span><span class="op">=</span><span class="va">$START</span><span class="kw">;</span> <span class="va">Idx</span><span class="op">&lt;=</span><span class="va">$STOP</span><span class="kw">;</span> <span class="va">Idx</span><span class="op">++</span><span class="kw">));</span> <span class="cf">do</span></span></code></pre></div>
<p>It means that the <code>for</code> loop is cycling over different values assigned to the
<code>Idx</code> variable. That is, the first time through, <code>Idx</code> is set to the value of
START (which is 1, in our case). The next time through, <code>Idx</code> is set to the
value of 2, and so forth.</p>
<p>So, we can run through the first loop of this <code>for</code> loop code faithfully, by setting</p>
<div class="sourceCode" id="cb393"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb393-1"><a href="rstudio-and-project-centered-organization.html#cb393-1" aria-hidden="true" tabindex="-1"></a><span class="va">Idx=</span>1</span></code></pre></div>
<p>Do that, and then do <code>echo $Idx</code> to print the value of <code>Idx</code> to make sure you got it set
correctly. Then evaluate the next line.</p>
</div>
<div id="multi-line-commands" class="section level4" number="10.2.2.8">
<h4><span class="header-section-number">10.2.2.8</span> Multi-line commands</h4>
<p>Soon, we get to a mutli-line command that looks like:</p>
<div class="sourceCode" id="cb394"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb394-1"><a href="rstudio-and-project-centered-organization.html#cb394-1" aria-hidden="true" tabindex="-1"></a><span class="va">ASSIGNMENTS=$(</span><span class="fu">awk</span> <span class="at">-v</span> LINE=<span class="va">$Idx</span> <span class="st">&#39;</span></span>
<span id="cb394-2"><a href="rstudio-and-project-centered-organization.html#cb394-2" aria-hidden="true" tabindex="-1"></a><span class="st">    $1 == &quot;index&quot; {for(i=1; i&lt;=NF; i++) vars[i]=$i; next}</span></span>
<span id="cb394-3"><a href="rstudio-and-project-centered-organization.html#cb394-3" aria-hidden="true" tabindex="-1"></a><span class="st">    $1 == LINE {for(i=1; i&lt;=NF; i++) printf(&quot;%s=%s; &quot;, vars[i], $i)}</span></span>
<span id="cb394-4"><a href="rstudio-and-project-centered-organization.html#cb394-4" aria-hidden="true" tabindex="-1"></a><span class="st">  &#39;</span> chinook-fastq-meta-data.tsv<span class="va">)</span></span></code></pre></div>
<p>You can evaluate that line-by-line, by starting with your cursor
anywhere on the line that says <code>ASSIGNMENTS</code> and then hitting
Option-CMD-Return four times.</p>
<p><em>Or</em> you can highlight all the text in all four of those command lines
and then hit Option-CMD-Return once. Note that you have to highlight
<em>all</em> of the text you want to evaluate. i.e. what gets evaluated is exactly—and
only—the text that is highlighted.</p>
</div>
<div id="continue" class="section level4" number="10.2.2.9">
<h4><span class="header-section-number">10.2.2.9</span> Continue…</h4>
<p>The remaining lines until you get to the <code>bwa</code> command are mostly
just variable assignments. Evaluate them all and afterward
check the results by <code>echo</code>ing the values of a few variables.</p>
</div>
<div id="the-bwa-command" class="section level4" number="10.2.2.10">
<h4><span class="header-section-number">10.2.2.10</span> The bwa command</h4>
<p>There is a lot going on in line 92 to 97 (multiple commands strung or piped
together with <code>&amp;&amp;</code> and <code>|</code>). Highlight all those lines and then evaluate them
(Opt-Command-Return).</p>
<p>It takes about 30 seconds to run.</p>
<p>Question: What file do you think is the final output of all of those commands?
It is stored in a variable; what is the name of that variable? What is its value?</p>
<p>Note that by this point we have gotten to the end of the
<code>for</code> loop.</p>
</div>
<div id="evaluating-the-entire-for-loop-from-start1-to-stop8" class="section level4" number="10.2.2.11">
<h4><span class="header-section-number">10.2.2.11</span> Evaluating the entire <code>for</code> loop from START=1 to STOP=8</h4>
<p>Check to make sure that you have START and STOP set correctly:</p>
<div class="sourceCode" id="cb395"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb395-1"><a href="rstudio-and-project-centered-organization.html#cb395-1" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="va">$START</span> <span class="va">$STOP</span></span></code></pre></div>
<p>If you get <code>1 8</code> as the response, then you are good to go. If not, then
do:</p>
<div class="sourceCode" id="cb396"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb396-1"><a href="rstudio-and-project-centered-organization.html#cb396-1" aria-hidden="true" tabindex="-1"></a><span class="va">START=</span>1</span>
<span id="cb396-2"><a href="rstudio-and-project-centered-organization.html#cb396-2" aria-hidden="true" tabindex="-1"></a><span class="va">STOP=</span>8</span></code></pre></div>
<p>And re-evaluate the <code>SM_TAGS=</code> line (line 48) to make sure that variable
is set.</p>
<p>Now, to evaluate the whole for loop you can highlight everything from
the <code>for...</code> line to the <code>done</code> line (inclusive) and hit Option-CMD-Return.</p>
<p>This will take about 2 minutes to run through.</p>
</div>
<div id="check-the-stderr-output-from-all-the-tools" class="section level4" number="10.2.2.12">
<h4><span class="header-section-number">10.2.2.12</span> Check the stderr output from all the tools</h4>
<p>Look at the value of the variables that hold the paths
to the stderr files:</p>
<div class="sourceCode" id="cb397"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb397-1"><a href="rstudio-and-project-centered-organization.html#cb397-1" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="va">$BWA_STDERR</span></span>
<span id="cb397-2"><a href="rstudio-and-project-centered-organization.html#cb397-2" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="va">$SAM_VIEW_STDERR</span></span>
<span id="cb397-3"><a href="rstudio-and-project-centered-organization.html#cb397-3" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="va">$SAM_FIX_STDERR</span></span>
<span id="cb397-4"><a href="rstudio-and-project-centered-organization.html#cb397-4" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="va">$SAM_SORT_STDERR</span></span></code></pre></div>
<p>The answers you get are those that were set on the last time time
the computer ran through the loop. But it shows that they are
all in the <code>stderr</code> directory.</p>
<p>We can check to see if anything went wrong by looking that the top few lines
of all of those files that were produced (for all the 8 different times
through the loop), with:</p>
<div class="sourceCode" id="cb398"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb398-1"><a href="rstudio-and-project-centered-organization.html#cb398-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> stderr/<span class="pp">*</span> <span class="kw">|</span> <span class="fu">less</span></span></code></pre></div>
<p>That pipes the output through the <code>less</code> pager/viewer. Space bar makes it go
down a “page” <code>b</code> makes it go <em>back</em> a “page” and <code>q</code> is what you use
to get out of the <code>less</code> viewer.</p>
<p>You should have seen progress messages from bwa, and then a bunch of empty files…</p>
</div>
<div id="now-you-are-ready-to-run-the-next-block-of-code" class="section level4" number="10.2.2.13">
<h4><span class="header-section-number">10.2.2.13</span> Now, you are ready to run the next block of code</h4>
<p>Starting on line 106 you can run through the next bit.</p>
<p>Check the value of SM_TAGS:</p>
<div class="sourceCode" id="cb399"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb399-1"><a href="rstudio-and-project-centered-organization.html#cb399-1" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="va">$SM_TAGS</span></span></code></pre></div>
<p>It should be a single value: <code>CH_plate1_A01</code>.</p>
<p>So we can set <code>SM</code> (the value being cycled over) to that:</p>
<div class="sourceCode" id="cb400"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb400-1"><a href="rstudio-and-project-centered-organization.html#cb400-1" aria-hidden="true" tabindex="-1"></a><span class="va">SM=</span>CH_plate1_A01</span></code></pre></div>
<p>And then run through the lines of code within that last for loop to
understand what they are doing.</p>
<p>At the end of it, you should be able to do</p>
<div class="sourceCode" id="cb401"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb401-1"><a href="rstudio-and-project-centered-organization.html#cb401-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> view <span class="va">$MKDUP_OUTPUT</span> <span class="kw">|</span> <span class="fu">head</span> </span></code></pre></div>
<p>to see a few lines of SAM file.</p>
<p>If that all worked. Then it means you should be all set
up to run <code>run-single-job.sh</code> using <code>sbatch</code>. Once you have
successfully done that you are very nearly done with the homework.</p>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction-to-reproducible-research.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="version-control.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook/js/app.min.js"></script>
<script src="libs/gitbook/js/clipboard.min.js"></script>
<script src="libs/gitbook/js/plugin-search.js"></script>
<script src="libs/gitbook/js/plugin-sharing.js"></script>
<script src="libs/gitbook/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook/js/plugin-bookdown.js"></script>
<script src="libs/gitbook/js/jquery.highlight.js"></script>
<script src="libs/gitbook/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/eriqande/eca-bioinf-handbook/edit/master/2.01-rstudio-prj-org.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/eriqande/eca-bioinf-handbook/commits/master/2.01-rstudio-prj-org.Rmd",
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["eca-bioinf-handbook.pdf", "eca-bioinf-handbook.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
